---
title: "Energetic Profile-Based Protein Comparison: A New and Fast Approach for Structural, and Evolutionary analysis"
author: "Peyman & Mehdi"
date: "2024-02-19"
output: 
  html_document:
    fig_caption: yes
fontsize: 11pt
geometry: margin=1in
graphics: yes
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,warning=FALSE, fig.pos= "h")
```

# Library   
```{r libraries, include=FALSE}
#library(kableExtra)
options(knitr.table.format = "latex")
library(readr)
library(data.table)
library(MASS)
library(tidyverse)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(ggpointdensity)
library(ggpubr)
library(igraph)
library(fpc)
library(grid)
library(gridExtra)
library(reshape2)
library(visNetwork)
library(pheatmap)
library(umap)
library(Rtsne)
library(viridis)
library(viridisLite)
library(bio3d)
library(caret)
library(e1071)
library(randomForest)
library(seqinr)
library(geometry)
library(ape)
library(colorspace)
library(dendextend)
library(dynamicTreeCut)
library(caret)
library(ggtree)
library(phangorn)
library(tanggle)
library(stringr)
library(protr)
library(proxy)
library(class)
library(readxl)
library(ggh4x)
```

# Custom distance

```{r}
#  denoiseEnergy<-function(x,n){
#    # x is a vectore with 210 dimentional
#    # n is the number of eigenvalues to reconstruct
#    m <- matrix(0,20,20)
#    m[lower.tri(m,diag = T)]<-x
#    m<-m+t(m)
#    diag(m)<-diag(m)/2
#    s<-svd(m)
#    s$d[(n+1):20] <- 0
#    m <- s$u %*% diag(s$d)  %*% t(s$v)
#    x <- m[lower.tri(m,diag = T)]
#    return(x)
#  }
# -----------------------------------------
manhat <- function(X, Y){
  #x <- denoiseEnergy(X,1); y <- denoiseEnergy(Y,1)
  x<-X; y<-Y
  xy <- abs(x - y)
  xy <- xy[xy != 0]
  if(length(xy)==0){
    d <- 0
  }else{
    d <- sum(xy)/length(xy)
  }
  return(d)
}
```

# Parse Scope 2.08
This chunk parse the astral 95/40 data downloaded form Scope 2.08 and organize it into a suitable data-frame with columns "scopeID","pdbID" ,"Position", "class","fold","superfamily","family","seq". Since we already made astral_95.rds, we do not need to run the following two chunks (ASTRAL95 and ASTRAL40).

## ASTRAL95

```{r}
astrals_95 <- read_table("Data/csv/astral-scopedom-seqres-gd-sel-gs-bib-95-2.08.fa", col_names = FALSE)
astrals_95_seq <- readFASTA("Data/csv/astral-scopedom-seqres-gd-sel-gs-bib-95-2.08.fa")

rows_keep <- grep(">", astrals_95$X1)
astrals_95 <- astrals_95[rows_keep,c(1,2,3)]
colnames(astrals_95) <- c("scopeID","ID_scope","Position")
astral_95 <- data.frame(
    scopeID = character(dim(astrals_95)[1]),
    pdbID = character(dim(astrals_95)[1]),
    Position = character(dim(astrals_95)[1]),
    seq = t(data.frame(astrals_95_seq)))
row.names(astral_95) <- c(1:dim(astral_95)[1])
astral_95$scopeID <- substr(astrals_95$scopeID, start = 2, stop = nchar(astrals_95$scopeID))
astral_95$pdbID <- substr(astrals_95$scopeID, start = 3, stop = 6)
#----------Split scopeID into class, fold, super and family---------
split_values <- strsplit(astrals_95$ID_scope, "\\.")
matrix_values <- matrix("", nrow = nrow(astrals_95), ncol = 4)
for (i in 1:4) {
  matrix_values[, i] <- sapply(split_values, function(x) paste(x[1:i], collapse = "."))
}
df_split <- as.data.frame(matrix_values)
colnames(df_split) <- c("class", "fold", "superfamily", "family")
#-------------------------------------------------------------------
# ------------split Position, start and end--------------------------
astral_95$Position <- gsub("[()]", "", astrals_95$Position)
#-----------------------------------------------------------------
astral_95 <- cbind(astral_95,df_split)
astral_95 <- astral_95[,c("scopeID","pdbID" ,"Position", "class","fold","superfamily","family","seq")]
#astral_95$seq <- toupper(astral_95$seq)
#saveRDS(astral_95, file = "Data/rds/astral_95.rds")
```

## ASTRAL40

```{r}
astrals_40 <- read_table("Data/csv/astral-scopedom-seqres-gd-sel-gs-bib-40-2.08.fa", col_names = FALSE)
astrals_40_seq <- readFASTA("Data/csv/astral-scopedom-seqres-gd-sel-gs-bib-40-2.08.fa")

rows_keep <- grep(">", astrals_40$X1)
astrals_40 <- astrals_40[rows_keep,c(1,2,3)]
colnames(astrals_40) <- c("pdbID","scopeID","chainID")
astral_40 <- data.frame(
  scope = character(dim(astrals_40)[1]),
    pdbID = character(dim(astrals_40)[1]),
    chainID = character(dim(astrals_40)[1]),
    position = integer(dim(astrals_40)[1]),
    start = integer(dim(astrals_40)[1]),
    end = integer(dim(astrals_40)[1]),
  seq = t(data.frame(astrals_40_seq)))
astral_40$scope <- substr(astrals_40$pdbID, start = 2, stop = nchar(astrals_40$pdbID))
astral_40$pdbID <- substr(astrals_40$pdbID, start = 3, stop = 6)
#----------Split scopeID into class, fold, super and family---------
split_values <- strsplit(astrals_40$scopeID, "\\.")
matrix_values <- matrix("", nrow = nrow(astrals_40), ncol = 4)
for (i in 1:4) {
  matrix_values[, i] <- sapply(split_values, function(x) paste(x[1:i], collapse = "."))
}
df_split <- as.data.frame(matrix_values)
colnames(df_split) <- c("class", "fold", "superfamily", "family")
#-------------------------------------------------------------------
# ------------split chainID, start and end--------------------------
astral_40$chainID <- gsub("[()]", "", astrals_40$chainID)
astral_40$position <- substr(astral_40$chainID, start = regexpr(":", astral_40$chainID) + 1, stop = nchar(astral_40$chainID))
astral_40$chainID <- substr(astral_40$chainID,start = 1, stop = 1)
astral_40$start <- NA
astral_40$end <- NA
split_values <- strsplit(astral_40$position, "-")
for (i in seq_along(split_values)) {
  if (length(split_values[[i]]) == 2) {
    astral_40$start[i] <- as.numeric(split_values[[i]][1])
    astral_40$end[i] <- as.numeric(split_values[[i]][2])
  }
}
#-----------------------------------------------------------------
astral_40 <- cbind(astral_40,df_split)
#astral_40$length <- apply(data.frame(astral_40$seq),1,nchar)
#saveRDS(astral_40, file = "Data/rds/astral_40.rds")
```


# Energy Calculation

Sequence_Energy_Estimator
Energy Calculation prerequisite


The following chunk retrieves the predicted pairwise energy values from the training dataset for the computation of the energy profile based on sequence.

```{r}
energy_dell_dunbrack <- read.csv("Data/csv/energy.csv",header = FALSE,sep ="" )
energy_dell_dunbrack <- energy_dell_dunbrack[,5:33]
source("Functions/amacid2num.R")
source("Functions/atomtype2num167.R")
source("Functions/Energy_PC210.R")
source("Functions/NetworkFrame.R")
source("Functions/inedx_HelixSheet.R")
source("Functions/split_position.R")
# ------Pairwise Amino Acid names-------------
AA <- c("PHE","LEU","ILE","VAL","TRP","TYR","MET","CYS","HIS","THR",
        "ARG","ALA","ASN","GLN","PRO","SER","ASP","GLY","LYS","GLU")
letters_list <- aa321(AA)
pair_inter <- character(0)
for (i in 1:20) {
  for (j in i:20) {
    pair_inter <- c(pair_inter, paste0(letters_list[i], letters_list[j]))
  }
} 
# -----------------------------------
# -----  Sequence Energy Estimator (predictor)
aaenergy <- read.csv("Data/csv/Pij.csv", header = F,sep = ";")
aaenergy <- (aaenergy + t(aaenergy))/2
colnames(aaenergy) = rownames(aaenergy)<- letters_list
#aa210 <- aaenergy[lower.tri(aaenergy,diag = T)]
```

## ASTRAL95/40

This chunk calculate the profile of energy for ASTRAL95/40 based on sequence and structure.

```{r}
# -------- At first, run the 'Energy Calculation prerequisite' chunk.
# ----------------------------------------------------
astral <- readRDS("Data/rds/astral_95.rds")  # astral_40.rds 
Nprotein <- nrow(astral)
num_columns <- 210
E210.astral <- data.frame(pdbID = astral$pdbID,
                            scopeID= astral$scopeID,
                            length = rep(NA,Nprotein),
                            position = astral$Position,
                            class = astral$class,
                            fold = astral$fold,
                            superfamily = astral$superfamily,
                            family = astral$family,
                            matrix(NA, ncol = num_columns, nrow = length(astral$pdbID)))
colnames(E210.astral)[9:218] = pair_inter
E210.astral_SS <- E210.seq_astral <- E210.seqPDB_astral <- E210.astral 
for (NP in 1:Nprotein) {
  if (NP %% 50 == 0) {print(NP)}
  chain_resno <- split_position(astral$Position[NP])
  pdbname <- astral$pdbID[NP]
  tryCatch({
    pdb0 <- read.pdb(pdbname,verbose = FALSE)
    pdblist = list()
    for (ind in 1:dim(chain_resno)[1]){
      sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=chain_resno$chain[ind])
      sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
      sele4 <- atom.select(pdb0, "noh")
      sele <- combine.select(sele1, sele2,sele4, operator="AND")
      if (!(chain_resno$start[ind]=="NA") & !(chain_resno$end[ind])=="NA"){
         st <- chain_resno$start[ind]
         en <- chain_resno$end[ind]
         sele3 <- atom.select(pdb0,  resno=st:en, verbose=FALSE)
         sele <- combine.select(sele1, sele2,sele3,sele4, operator="AND")
      }
      pdblist[[ind]] <- trim.pdb(pdb0, sele,verbose = FALSE)
    }
    new_pdb = pdblist[[1]]
    if (dim(chain_resno)[1] > 1) {
      for (ind in 2:dim(chain_resno)[1]){
      new_pdb <- cat.pdb(new_pdb, pdblist[[ind]], rechain=FALSE)
      }
    }
    pdbX <- new_pdb$atom
    E210.astral$length[NP] = length(unique(pdbX$resno))
    E210.astral_SS$length[NP] = length(unique(pdbX$resno))
    E210.seqPDB_astral$length[NP] = length(unique(pdbX$resno))
    Net_Frame <- NetworkFrame(PDB = new_pdb)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210.astral[NP,9:218] <- energy210
    # ****************keep res helix & sheet************************
    indhs <- inedx_HelixSheet(PDB = new_pdb)
    Net_Frame_hs <- Net_Frame[Net_Frame$resnoif %in% indhs & Net_Frame$resnojf %in% indhs,]
    if(dim(Net_Frame_hs)[1]!=0){
        Net_Frame <- Net_Frame_hs
    }
    energy210_ss <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210.astral_SS[NP,9:218] <- energy210_ss
    #-----------sequence enregy based on PDB--------
    pdbXca <- pdbX[pdbX$elety == "CA",]
    seq = aa321(pdbXca$resid)
    E210.seqPDB_astral$length[NP] = length(seq)
    freq = c()
    for (j in 1:20){
        freq[j] <-  length(grep(tolower(letters_list[j]),tolower(seq)))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
    E210.seqPDB_astral[NP,9:218] <- aa210_p
#-----------sequence enregy based on astral--------
    seq = astral$seq[NP]
    seq = gsub("X", "", seq)
    E210.seq_astral$length[NP] = nchar(seq)
    seq = unlist(strsplit(seq,""))
    freq = c()
    for (j in 1:20){
        freq[j] <-  length(grep(letters_list[j],tolower(seq)))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
    E210.seq_astral[NP,9:218] <- aa210_p
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}
# -----------------------------------------------------------
# ----------- astral 95 ------
#saveRDS(E210.astral,file = "Data/rds/E210.astral_95.rds")
#saveRDS(E210.astral_SS,file = "Data/rds/E210.astral_95_SS.rds")
#saveRDS(E210.seq_astral,file = "Data/rds/E210.seq_astral_95.rds")
#saveRDS(E210.seqPDB_astral,file = "Data/rds/E210.seqPDB_astral_95.rds")
# -----------------------------------------------------------
# ----------- astral 40 ------
#saveRDS(E210.astral,file = "Data/rds/E210.astral_40.rds")
#saveRDS(E210.astral_SS,file = "Data/rds/E210.astral_40_SS.rds")
#saveRDS(E210.seq_astral,file = "Data/rds/E210.seq_astral_40.rds")
#saveRDS(E210.seqPDB_astral,file = "Data/rds/E210.seqPDB_astral_40.rds")
```


## Ferritin-like Superfamily

```{r}
ferritin <- read.csv("Data/csv/Ferritin_Like.csv",sep = ",")
energy_dell_dunbrack <- read.csv("Data/csv/energy.csv",header = FALSE,sep ="" )
energy_dell_dunbrack <- energy_dell_dunbrack[,5:33]
source("Functions/amacid2num.R")
source("Functions/atomtype2num167.R")
source("Functions/Energy_PC210.R")
source("Functions/NetworkFrame.R")
source("Functions/inedx_HelixSheet.R")
Nprotein <- nrow(ferritin)
# ------Pairwise Amino Acid names-------------
AA <- c("PHE","LEU","ILE","VAL","TRP","TYR","MET","CYS","HIS","THR",
        "ARG","ALA","ASN","GLN","PRO","SER","ASP","GLY","LYS","GLU")
letters_list <- aa321(AA)
pair_inter <- character(0)
for (i in 1:20) {
  for (j in i:20) {
    pair_inter <- c(pair_inter, paste0(letters_list[i], letters_list[j]))
  }
} 
#----------------------------------------------
num_columns <- 210
E210_ferritin <- data.frame(pdbID = ferritin$pdbID,
                            length = rep(NA,Nprotein),
                            group = ferritin$group,
                            matrix(NA, ncol = num_columns, nrow = length(ferritin$pdbID)))
colnames(E210_ferritin) = c("pdbID","length","group",pair_inter)
E210_ferritin_SS <- E210_ferritin
for (NP in 1:Nprotein) {
  print(NP)
  pdbname <- ferritin$pdbID[NP]
  E210_ferritin_SS$pdbID[NP] <- pdbname
  ch <- ferritin$chainID[NP]
  E210_ferritin$pdbID[NP] <- paste0(pdbname,ch)
  E210_ferritin_SS$pdbID[NP] <- paste0(pdbname,ch)
  tryCatch({
    pdb0 <- read.pdb(pdbname)
    sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
    sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
    sele4 <- atom.select(pdb0, "noh")
    sele <- combine.select(sele1, sele2,sele4, operator="AND")
    if (!is.na(ferritin$start[NP])){
      st <- ferritin$start[NP]
      en <- ferritin$end[NP]
      sele3 <- atom.select(pdb0,  resno=st:en, verbose=TRUE)
      sele <- combine.select(sele1, sele2,sele3,sele4, operator="AND")
    }
    pdbX <- pdb0$atom[sele$atom,]
    E210_ferritin$length[NP] = length(unique(pdbX$resno))
    E210_ferritin_SS$length[NP] = length(unique(pdbX$resno))
    Net_Frame <- NetworkFrame(PDB = pdb0, AtomSelect = sele)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210_ferritin[NP,4:213] <- energy210
    # ****************keep res helix & sheet************************
    indhs <- inedx_HelixSheet(PDB = pdb0)
    Net_Frame_hs <- Net_Frame[Net_Frame$resnoif %in% indhs & Net_Frame$resnojf %in% indhs,]
    if(dim(Net_Frame_hs)[1]!=0){
        Net_Frame <- Net_Frame_hs
    }
    energy210_ss <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210_ferritin_SS[NP,4:213] <- energy210_ss
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}
#saveRDS(E210_ferritin,file = "Data/rds/E210_ferritin.rds")
#saveRDS(E210_ferritin_SS,file = "Data/rds/E210_ferritin_SS.rds")
```


## Covid_Spike200

```{r}
# -------- At first, run the 'Energy Calculation prerequisite' chunk.
# ----------------------------------------------------
spike <- read.csv("Data/csv/spike200.csv", sep=";")
Nprotein <- nrow(spike)
#----------------------------------------------
num_columns <- 210
E210_spike200 <- data.frame(pdbID = spike$pdbID,
                            length = rep(NA,Nprotein),
                            virus = spike$VIRUS,
                            conformation = spike$Conformation,
                            matrix(NA,ncol=num_columns,nrow=Nprotein))

colnames(E210_spike200) = c("pdbID","length","virus", "conformation",pair_inter)
E210.seq_spike200 <- E210_spike200_SS <- E210_spike200

for (NP in 1:Nprotein) {
  print(NP)
  ch <- substr(spike$pdbID[NP],5,5)
  pdbname <- substr(spike$pdbID[NP],1,4)
  E210_spike200$pdbID[NP] <- spike$pdbID[NP]
  E210_spike200_SS$pdbID[NP] <- spike$pdbID[NP]
  E210.seq_spike200$pdbID[NP] <- spike$pdbID[NP]
  tryCatch({
    pdb0 <- read.pdb(tolower(pdbname))
    sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
    sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
    sele4 <- atom.select(pdb0, "noh")
    sele <- combine.select(sele1, sele2,sele4, operator="AND")
    pdbX <- pdb0$atom[sele$atom,]
    pdbXca <- pdbX[pdbX$elety == "CA",]
    seq <- aa321(pdbXca$resid)
    E210.seq_spike200$length[NP] <- length(unique(pdbX$resno))
    freq <- c()
    for (j in 1:20){
        freq[j] <-  length(grep(tolower(letters_list[j]),tolower(seq)))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p <- pair_es[lower.tri(pair_es,diag = T)]
    E210.seq_spike200[NP,5:214] <- aa210_p
    E210_spike200$length[NP] <- length(unique(pdbX$resno))
    E210_spike200_SS$length[NP] <- length(unique(pdbX$resno))
    # -------------------
    new_pdb <- trim.pdb(pdb0, sele,verbose = FALSE)
    Net_Frame <- NetworkFrame(PDB = new_pdb)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210_spike200[NP,5:214] <- energy210
    # ****************keep res helix & sheet************************
    indhs <- inedx_HelixSheet(PDB = new_pdb)
    Net_Frame_hs <- Net_Frame[Net_Frame$resnoif %in% indhs & Net_Frame$resnojf %in% indhs,]
    if(dim(Net_Frame_hs)[1]!=0){
        Net_Frame <- Net_Frame_hs
    }
    energy210_ss <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210_spike200_SS[NP,5:214] <- energy210_ss
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}

#saveRDS(E210_spike200,file = "Data/rds/E210_spike200.rds")
#saveRDS(E210_spike200_SS,file = "Data/rds/E210_spike200_SS.rds")
#saveRDS(E210.seq_spike200,file = "Data/rds/E210.seq_spike200.rds")
```

## DTI preprocess
This chunk extracts AA sequences from the drug-target
```{r}
library(rentrez)
get_protids_from_geneids <- function(gene_ids) {
  # Get the protein elink.
  protein_elink = rentrez::entrez_link(dbfrom = "gene",
                                        id = gene_ids,
                                        db = "protein")
  # Get the protein id (refseq database)
  protein_ids = protein_elink$links$gene_protein_refseq
  return(protein_ids)
}

make_aa_fasta <- function(prot_ids) { #, nameFile
  # Make a multi-fasta file with each protein id
  sapply(prot_ids, function(x) {
    protein_esummary = rentrez::entrez_summary(db = "protein", id = x)
    protein_fasta = rentrez::entrez_fetch(db = "protein",
                                          id = protein_esummary$uid,
                                          rettype = "fasta")
    return(protein_fasta)
  } )
}

seq2df<-function(seq){
  x<-strsplit(seq,'\n')
  head<-x[[1]][1]
  seq<-paste(x[[1]][-c(1,10)],collapse = '')
  return(c(head,seq))
}

# ------------------------------
dti <- read_excel('Data/csv/41467_2019_9186_MOESM4_ESM_dti_barabasi.xlsx')
colnames(dti)<-c('drugID','entrezID')
dti$entrezID<-as.character(dti$entrezID)

sab<-readRDS('Data/rds/sAB.rds')
colnames(sab)[1:2]<-c('d1','d2')

d65<-unique(c(sab$d1,sab$d2))
df_seq<-c()
for(i in 1:65){
  df<-dti[dti$drugID%in%d65[i],]
  df2<-c()
  for (j in 1:nrow(df)) {
    print(paste0(i,', ',j,'/',nrow(df)))
    prot_ids <- get_protids_from_geneids(df$entrezID[j])
    z<-make_aa_fasta(prot_ids)
    zz<-data.frame(prot=df$entrezID[j],t(sapply(z, seq2df)))
    df2<-rbind(df2,zz)
  }
  df2<-data.frame(drug=d65[i],df2)
  df_seq<-rbind(df_seq,df2)
}

colnames(df_seq)[3:4]<-c('Isoform','seq_aa')
df_seq$length <- nchar(df_seq$seq)
#saveRDS(df_seq,'Data/rds/DTI_df_seq.rds')
```

## DTI Energy

```{r}
# -------- At first, run the 'Energy Calculation prerequisite' chunk.
# ----------------------------------------------------
df_seq <- readRDS('Data/rds/DTI_df_seq.rds')

E210.seq_dti <- data.frame(df_seq,matrix(NA,nrow(df_seq),210))
colnames(E210.seq_dti)[-c(1:5)] <- pair_inter
for (NP in 1:nrow(df_seq)) {
  if(NP%%10==0)print(NP)
  seq = unlist(strsplit(unlist(E210.seq_dti$seq_aa[NP]),''))
    freq = c()
    for (j in 1:20){
        freq[j] <-  length(grep(tolower(letters_list[j]),tolower(seq)))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
    E210.seq_dti[NP,-c(1:5)] <- aa210_p
}
#saveRDS(E210.seq_dti,'Data/rds/E210.seq_dti.rds')
```

## DTI drug similarity 
This Chunk calculates the similarity between drugs based on Eq 6 in the paper.
```{r}
# -------- At first, run the 'Prerequisite of Fig and dis' chunk.
# ----------------------------------------------------
sab <- readRDS('Data/rds/sAB.rds')
colnames(sab)[1:2] <- c('d1','d2')
E210.seq_dti<-readRDS('Data/rds/E210.seq_dti.rds')

df0 <- E210.seq_dti 
df <- cbind(df0[,1:5],df0[,-c(1:5)]/df0$length) 

df2 <- aggregate(.~drug+entrezID,df[,-c(3:5)],mean)
drugs<-unique(df2$drug)
#########################
my_sab<-matrix(0,65,65)
for (i in 1:65) {
  print(i)
  for (j in 1:65) {
    if(i>=j){
      x<-drugs[i]
      y<-drugs[j]
      zx<-df2[df2$drug%in%x,]
      zy<-df2[df2$drug%in%y,]
      xnames<-paste0(zx$drug,'_',zx$entrezID)
      ynames<-paste0(zy$drug,'_',zy$entrezID)
      z<-rbind(zx,zy)
      dis<-as.matrix(dist(z[,-c(1:4)],method = manhat))
      colnames(dis)<-rownames(dis)<-paste0(z$drug,'_',z$entrezID)
      diag(dis)<-NA
      dis[is.nan(dis)] <- 0
      disx<-dis[xnames,xnames]
      dxx<-mean(diag(as.matrix(disx[apply(disx,2,which.min),])))
      disy<-dis[ynames,ynames]
      dyy<-mean(diag(as.matrix(disy[apply(disy,2,which.min),])))
      disxy<-dis[xnames,ynames]
      disyx<-dis[ynames,xnames]
      dxy<-mean(c(diag(as.matrix(disxy[apply(disxy,2,which.min),])),
                  diag(as.matrix(disyx[apply(disyx,2,which.min),]))))
      my_sab[i,j]<-dxy-((dxx+dyy)/2)
      my_sab[j,i]<-my_sab[i,j]
    }
  }
}
diag(my_sab)<-NA
colnames(my_sab)<-rownames(my_sab)<-drugs
# saveRDS(my_sab,'Data/rds/my_sab.rds')
```

## Bacteriocin Tm-Vec

```{r}
# -------- At first, run the 'Energy Calculation prerequisite' chunk.
# ----------------------------------------------------
# ----------- the csv file is from TM-Vec guys.
tm <- read.csv('Data/csv/tm-vec.csv') 
nprotein <- nrow(tm)

for (i in 1:nprotein){
  if(i%%100==0)print(i)
  # ---- seq1
  seq <- unlist(strsplit(tolower(tm$Seq_x[i]), ""))
  freq <- c()
  for (j in 1:20){
  freq[j] <-  length(grep(tolower(letters_list[j]),seq))
  }
  freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
  en_fr <- aaenergy * freq_matrix
  en_fr <- en_fr/length(seq)
  pair_es <- diag(freq) %*% as.matrix(en_fr)
  aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
  CEP_x <- c(length(seq),aa210_p)
  # ---- seq2
  seq <- unlist(strsplit(tolower(tm$Seq_y[i]), ""))
  freq = c()
  for (j in 1:20){
  freq[j] <-  length(grep(tolower(letters_list[j]),seq))
  }
  freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
  en_fr <- aaenergy * freq_matrix
  en_fr <- en_fr/length(seq)
  pair_es <- diag(freq) %*% as.matrix(en_fr)
  aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
  CEP_y <- c(length(seq),aa210_p)
  # ----------
  tm$CEP_dis[i] <- manhat(CEP_x[-1],CEP_y[-1])
}

tm$CEP_dis <- (tm$CEP_dis-min(tm$CEP_dis))/(max(tm$CEP_dis)-min(tm$CEP_dis))
tm$CEP_dis <- 1-(tm$CEP_dis) 

tm$status <- if_else(tm$class_x==tm$class_y,'same class','different class')
tm$status2 <- if_else(tm$Subclass_x==tm$Subclass_y&tm$class_x=='Class_1'&tm$class_y=='Class_1','same class(sub class1)',NA)
# saveRDS(tm,'Data/rds/bacteriocin_tm-vec.rds')
# ---------------------------------------------------------
# ------- ENERGY
# ---------------------------------------
tm_unique <- tm[!duplicated(tm$ID_x),]
nprotein <- nrow(tm_unique)
tm_unique <- data.frame(tm_unique[,c(1,7,8,9,14)], 
                        length=nchar(tm_unique$Seq_x),
                        matrix(NA,nrow=nprotein,ncol=210))
colnames(tm_unique)[-c(1:6)] <- pair_inter

for (i in 1:nprotein){
  print(i)
  seq = unlist(strsplit(tolower(tm_unique$Seq_x[i]), ""))
  freq = c()
  for (j in 1:20){
  freq[j] <-  length(grep(tolower(letters_list[j]),seq))
  }
freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
en_fr <- aaenergy * freq_matrix
en_fr <- en_fr/length(seq)
pair_es <- diag(freq) %*% as.matrix(en_fr)
aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
tm_unique[i,7:216] <- aa210_p
}
# saveRDS(tm_unique,'Data/rds/E210.seq_bacteriocin.rds')
```

## CT_Ho

```{r}
# -------- At first, run the 'Energy Calculation prerequisite' chunk.
# ----------------------------------------------------
df <- read.csv('Data/csv/CT_Ho_cathID_filtered.csv')
df$chain <- substr(df$pdbID,5,5)
df$pdbID <- substr(df$pdbID,1,4)

E210.CT_Ho <- data.frame(pdbID = df$pdbID,chainID = df$chain,length = NA,
                         start = df$start,end = df$end,
                         cathID = df$cathID,
                         class = df$class,
                         architecture = df$architecture,
                         topology = df$topology,
                         homologous = df$homologous,
                         matrix(NA, ncol = 210, nrow = nrow(df)))

colnames(E210.CT_Ho)[-c(1:10)] <- pair_inter
E210.seq_CT_Ho <- E210.CT_Ho

for(NP in 1:nrow(E210.CT_Ho)){
  print(NP)
  pdbname <- E210.CT_Ho$pdbID[NP]
  ch <- E210.CT_Ho$chainID[NP]
  tryCatch({
    pdb0 <- read.pdb(pdbname)
      sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
      sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
      sele4 <- atom.select(pdb0, "noh")
      sele <- combine.select(sele1, sele2,sele4, operator="AND")
      if (!is.na(E210.CT_Ho$start[NP])){
        pdbX <- pdb0$atom[sele$atom,]
        shift_pos <- min(pdbX$resno)-1
         st <- E210.CT_Ho$start[NP]+shift_pos
         en <- E210.CT_Ho$end[NP]+shift_pos
         sele3 <- atom.select(pdb0,  resno=st:en, verbose=TRUE)
         sele <- combine.select(sele, sele3, operator="AND")
      }
      new_pdb <- trim.pdb(pdb0, sele,verbose = FALSE)
      pdbX <- new_pdb$atom
      E210.CT_Ho$length[NP] = length(unique(pdbX$resno))
      Net_Frame <- NetworkFrame(PDB = new_pdb)
      energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
      E210.CT_Ho[NP,11:220] <- energy210
      # ****************seq************************
      pdbXca <- pdbX[pdbX$elety == "CA",]
      seq <- tolower(aa321(pdbXca$resid))
      E210.seq_CT_Ho$length[NP] = length(seq)
      freq = c()
      for (j in 1:20){
          freq[j] <- length(grep(tolower(letters_list[j]),seq))
      }
      freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
      en_fr <- aaenergy * freq_matrix
      en_fr <- en_fr/length(seq)
      pair_es <- diag(freq) %*% as.matrix(en_fr)
      aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
      E210.seq_CT_Ho[NP,11:220] <- aa210_p
      }, error = function(e) {
       cat("Error downloading file:", e$message, "\n")
        }
  )
}
# saveRDS(E210.CT_Ho,'Data/rds/E210.CT_Ho.rds')
# saveRDS(E210.seq_CT_Ho,'Data/rds/E210.seq_CT_Ho.rds')
```

## five superfamily

```{r}
# -------- At first, run the 'Energy Calculation prerequisite' chunk.
# ----------------------------------------------------
df5 <- readRDS('Data/rds/fiveSF.rds')
Nprotein <- nrow(df5)
#----------------------------------------------
num_columns <- 210
E210.seq_df5 <- data.frame(df5,
                       matrix(NA,ncol=num_columns,nrow=Nprotein))

for (NP in 1:Nprotein) {
  print(NP)
  chain_resno <- split_position(df5$position[NP])
  pdbname <- df5$pdbID[NP]
  tryCatch({
    pdb0 <- read.pdb(pdbname,verbose = FALSE)
    pdblist = list()
    for (ind in 1:dim(chain_resno)[1]){
      sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=chain_resno$chain[ind])
      sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
      sele4 <- atom.select(pdb0, "noh")
      sele <- combine.select(sele1, sele2,sele4, operator="AND")
      if (!(chain_resno$start[ind]=="NA") & !(chain_resno$end[ind])=="NA"){
         st <- chain_resno$start[ind]
         en <- chain_resno$end[ind]
         sele3 <- atom.select(pdb0,  resno=st:en, verbose=FALSE)
         sele <- combine.select(sele1, sele2,sele3,sele4, operator="AND")
      }
      pdblist[[ind]] <- trim.pdb(pdb0, sele,verbose = FALSE)
    }
    new_pdb = pdblist[[1]]
    if (dim(chain_resno)[1] > 1) {
      for (ind in 2:dim(chain_resno)[1]){
      new_pdb <- cat.pdb(new_pdb, pdblist[[ind]], rechain=FALSE)
      }
    }
    pdbX <- new_pdb$atom
    #-----------sequence enregy based on PDB--------
    pdbXca <- pdbX[pdbX$elety == "CA",]
    seq <- aa321(pdbXca$resid)
    freq = c()
    for (j in 1:20){
        freq[j] <-  length(grep(tolower(letters_list[j]),tolower(seq)))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
    E210.seq_df5[NP,9:218] <- aa210_p
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}

# saveRDS(E210.seq_df5,file = "Data/rds/E210.seq_df5.rds")
```







