---
title: "Protein Similarity Based on Energy Comparison"
output: 
  pdf_document:
    fig_caption: yes
fontsize: 11pt
geometry: margin=1in
graphics: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,warning=FALSE, fig.pos= "h")
```

# Library   
```{r libraries,mythem,distance, include=FALSE}
library(kableExtra)
options(knitr.table.format = "latex")
library(readr)
library(data.table)
library(MASS)
library(tidyverse)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(ggpointdensity)
library(ggpubr)
library(igraph)
library(fpc)
library(grid)
library(gridExtra)
library(reshape2)
library(visNetwork)
library(pheatmap)
library(umap)
library(Rtsne)
library(viridis)
library(viridisLite)
library(bio3d)
library(caret)
library(e1071)
library(randomForest)
library(seqinr)
library(geometry)
library(ape)
library(colorspace)
library(dendextend)
library(dynamicTreeCut)
library(caret)
library(ggtree)
library(phangorn)
library(tanggle)
library(stringr)
library(protr)
library(proxy)
library(class)
library(readxl)
library(ggh4x)

mythem <- theme_bw(base_line_size = .1)+
  theme(axis.text = element_text(size=15),
        axis.title = element_text(size = 20),
        legend.text = element_text(size = 12),
        legend.title = element_text(size=15),
        strip.text = element_text(size = 15),
        title = element_text(size = 20),
        panel.spacing = unit(2, "lines")
  )
# ***********
manhat <- function(X, Y){
  x = X; y = Y
  xy = abs(x - y)
  xy = xy[xy != 0]
  if(length(xy)==0){
    d <- 0
  }else{
    d = sum(xy)/length(xy)
  }
  return(d)
}


```

# Parse Scope 2.08
This chunk parse the astral 95/40 data downloaded form Scope 2.08 and organize it into a suitable data-frame with columns "scopeID","pdbID" ,"Position", "class","fold","superfamily","family","seq". Since we already made astral_95.rds, we do not need to run the following two chunks (ASTRAL95 and ASTRAL40).

## ASTRAL95

```{r}
astrals_95 <- read_table("Data/csv/astral-scopedom-seqres-gd-sel-gs-bib-95-2.08.fa", col_names = FALSE)
astrals_95_seq <- readFASTA("Data/csv/astral-scopedom-seqres-gd-sel-gs-bib-95-2.08.fa")

rows_keep <- grep(">", astrals_95$X1)
astrals_95 <- astrals_95[rows_keep,c(1,2,3)]
colnames(astrals_95) <- c("scopeID","ID_scope","Position")
astral_95 <- data.frame(
    scopeID = character(dim(astrals_95)[1]),
    pdbID = character(dim(astrals_95)[1]),
    Position = character(dim(astrals_95)[1]),
    seq = t(data.frame(astrals_95_seq)))
row.names(astral_95) <- c(1:dim(astral_95)[1])
astral_95$scopeID <- substr(astrals_95$scopeID, start = 2, stop = nchar(astrals_95$scopeID))
astral_95$pdbID <- substr(astrals_95$scopeID, start = 3, stop = 6)
#----------Split scopeID into class, fold, super and family---------
split_values <- strsplit(astrals_95$ID_scope, "\\.")
matrix_values <- matrix("", nrow = nrow(astrals_95), ncol = 4)
for (i in 1:4) {
  matrix_values[, i] <- sapply(split_values, function(x) paste(x[1:i], collapse = "."))
}
df_split <- as.data.frame(matrix_values)
colnames(df_split) <- c("class", "fold", "superfamily", "family")
#-------------------------------------------------------------------
# ------------split Position, start and end--------------------------
astral_95$Position <- gsub("[()]", "", astrals_95$Position)
#-----------------------------------------------------------------
astral_95 <- cbind(astral_95,df_split)
astral_95 <- astral_95[,c("scopeID","pdbID" ,"Position", "class","fold","superfamily","family","seq")]
astral_95$seq <- toupper(astral_95$seq)
#saveRDS(astral_95, file = "Data/rds/astral_95.rds")
```

## ASTRAL40

```{r}
astrals_40 <- read_table("Data/csv/astral-scopedom-seqres-gd-sel-gs-bib-40-2.08.fa", col_names = FALSE)
astrals_40_seq <- readFASTA("Data/csv/astral-scopedom-seqres-gd-sel-gs-bib-40-2.08.fa")

rows_keep <- grep(">", astrals_40$X1)
astrals_40 <- astrals_40[rows_keep,c(1,2,3)]
colnames(astrals_40) <- c("pdbID","scopeID","chainID")
astral_40 <- data.frame(
  scope = character(dim(astrals_40)[1]),
    pdbID = character(dim(astrals_40)[1]),
    chainID = character(dim(astrals_40)[1]),
    position = integer(dim(astrals_40)[1]),
    start = integer(dim(astrals_40)[1]),
    end = integer(dim(astrals_40)[1]),
  seq = t(data.frame(astrals_40_seq)))
astral_40$scope <- substr(astrals_40$pdbID, start = 2, stop = nchar(astrals_40$pdbID))
astral_40$pdbID <- substr(astrals_40$pdbID, start = 3, stop = 6)
#----------Split scopeID into class, fold, super and family---------
split_values <- strsplit(astrals_40$scopeID, "\\.")
matrix_values <- matrix("", nrow = nrow(astrals_40), ncol = 4)
for (i in 1:4) {
  matrix_values[, i] <- sapply(split_values, function(x) paste(x[1:i], collapse = "."))
}
df_split <- as.data.frame(matrix_values)
colnames(df_split) <- c("class", "fold", "superfamily", "family")
#-------------------------------------------------------------------
# ------------split chainID, start and end--------------------------
astral_40$chainID <- gsub("[()]", "", astrals_40$chainID)
astral_40$position <- substr(astral_40$chainID, start = regexpr(":", astral_40$chainID) + 1, stop = nchar(astral_40$chainID))
astral_40$chainID <- substr(astral_40$chainID,start = 1, stop = 1)
astral_40$start <- NA
astral_40$end <- NA
split_values <- strsplit(astral_40$position, "-")
for (i in seq_along(split_values)) {
  if (length(split_values[[i]]) == 2) {
    astral_40$start[i] <- as.numeric(split_values[[i]][1])
    astral_40$end[i] <- as.numeric(split_values[[i]][2])
  }
}
#-----------------------------------------------------------------
astral_40 <- cbind(astral_40,df_split)
astral_40$length <- apply(data.frame(astral_40$seq),1,nchar)
#saveRDS(astral_40, file = "Data/rds/astral_40.rds")

```



# Energy Calculation


## Sequence_Energy_Estimator

The following chunk retrieves the predicted pairwise energy values from the training dataset for the computation of the energy profile based on sequence.

```{r}
energy_dell_dunbrack <- read.csv("Data/csv/energy.csv",header = FALSE,sep ="" )
energy_dell_dunbrack <- energy_dell_dunbrack[,5:33]
source("Functions/amacid2num.R")
source("Functions/atomtype2num167.R")
source("Functions/Energy_PC210.R")
source("Functions/NetworkFrame.R")
source("Functions/inedx_HelixSheet.R")
source("Functions/split_position.R")
# ------Pairwise Amino Acid names-------------
AA <- c("PHE","LEU","ILE","VAL","TRP","TYR","MET","CYS","HIS","THR",
        "ARG","ALA","ASN","GLN","PRO","SER","ASP","GLY","LYS","GLU")
letters_list <- aa321(AA)
pair_inter <- character(0)
for (i in 1:20) {
  for (j in i:20) {
    pair_inter <- c(pair_inter, paste0(letters_list[i], letters_list[j]))
  }
} 
aaenergy <- read.csv("Data/csv/Pij.csv", header = F,sep = ";")
aaenergy <- (aaenergy + t(aaenergy))/2
colnames(aaenergy) = rownames(aaenergy)<- letters_list
aa210 <- aaenergy[lower.tri(aaenergy,diag = T)]

```

## ASTRAL95/40

This chunk calculate the profile of energy for ASTRAL95/40 based on sequence and structure.

```{r}
astral95 <- readRDS("Data/rds/astral_95.rds")
Nprotein <- nrow(astral95)
num_columns <- 210
E210.astral95 <- data.frame(pdbID = astral95$pdbID,
                            scopeID= astral95$scopeID,
                            length = rep(NA,Nprotein),
                            position = astral95$Position,
                            class = astral95$class,
                            fold = astral95$fold,
                            superfamily = astral95$superfamily,
                            family = astral95$family,
                            matrix(NA, ncol = num_columns, nrow = length(astral95$pdbID)))
colnames(E210.astral95)[9:218] = pair_inter
E210.astral95_SS <- E210.seq_astral95 <- E210.seqPDB_astral95 <- E210.astral95 
for (NP in 1:Nprotein) {
#  if (NP %% 50 == 0) {
   print(NP)
#  }
  chain_resno <- split_position(astral95$Position[NP])
  pdbname <- astral95$pdbID[NP]
  tryCatch({
    pdb0 <- read.pdb(pdbname,verbose = FALSE)
    pdblist = list()
    for (ind in 1:dim(chain_resno)[1]){
      sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=chain_resno$chain[ind])
      sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
      sele4 <- atom.select(pdb0, "noh")
      sele <- combine.select(sele1, sele2,sele4, operator="AND")
      if (!(chain_resno$start[ind]=="NA") & !(chain_resno$end[ind])=="NA"){
         st <- chain_resno$start[ind]
         en <- chain_resno$end[ind]
         sele3 <- atom.select(pdb0,  resno=st:en, verbose=FALSE)
         sele <- combine.select(sele1, sele2,sele3,sele4, operator="AND")
      }
      pdblist[[ind]] <- trim.pdb(pdb0, sele,verbose = FALSE)
    }
    new_pdb = pdblist[[1]]
    if (dim(chain_resno)[1] > 1) {
      for (ind in 2:dim(chain_resno)[1]){
      new_pdb <- cat.pdb(new_pdb, pdblist[[ind]], rechain=FALSE)
      }
    }
    pdbX <- new_pdb$atom
    E210.astral95$length[NP] = length(unique(pdbX$resno))
    E210.astral95_SS$length[NP] = length(unique(pdbX$resno))
    E210.seqPDB_astral95$length[NP] = length(unique(pdbX$resno))
    Net_Frame <- NetworkFrame(PDB = new_pdb)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210.astral95[NP,9:218] <- energy210
    # ****************keep res helix & sheet************************
    indhs <- inedx_HelixSheet(PDB = new_pdb)
    Net_Frame_hs <- Net_Frame[Net_Frame$resnoif %in% indhs & Net_Frame$resnojf %in% indhs,]
    if(dim(Net_Frame_hs)[1]!=0){
        Net_Frame <- Net_Frame_hs
    }
    energy210_ss <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210.astral95_SS[NP,9:218] <- energy210_ss
    #-----------sequence enregy based on PDB--------
    pdbXca <- pdbX[pdbX$elety == "CA",]
    seq = aa321(pdbXca$resid)
    E210.seqPDB_astral95$length[NP] = length(seq)
    freq = c()
    for (j in 1:20){
        freq[j] <-  length(grep(letters_list[j],tolower(seq)))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
    E210.seqPDB_astral95[NP,9:218] <- aa210_p
#-----------sequence enregy based on astral--------
    seq = astral95$seq[NP]
    seq = gsub("X", "", seq)
    E210.seq_astral95$length[NP] = nchar(seq)
    seq = unlist(strsplit(seq,""))
    freq = c()
    for (j in 1:20){
        freq[j] <-  length(grep(letters_list[j],tolower(seq)))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
    E210.seq_astral95[NP,9:218] <- aa210_p
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}
saveRDS(E210.astral95,file = "Data/rds/E210.astral_95.rds")
saveRDS(E210.astral95_SS,file = "Data/rds/E210.astral_95_SS.rds")
saveRDS(E210.seq_astral95,file = "Data/rds/E210.seq_astral_95.rds")
saveRDS(E210.seqPDB_astral95,file = "Data/rds/E210.seqPDB_astral_95.rds")
```


## Ferritin-like Superfamily

```{r}
ferritin <- read.csv("Data/csv/Ferritin_Like.csv",sep = ";")
energy_dell_dunbrack <- read.csv("Data/csv/energy.csv",header = FALSE,sep ="" )
energy_dell_dunbrack <- energy_dell_dunbrack[,5:33]
source("Functions/amacid2num.R")
source("Functions/atomtype2num167.R")
source("Functions/Energy_PC210.R")
source("Functions/NetworkFrame.R")
source("Functions/inedx_HelixSheet.R")
Nprotein <- nrow(ferritin)
# ------Pairwise Amino Acid names-------------
AA <- c("PHE","LEU","ILE","VAL","TRP","TYR","MET","CYS","HIS","THR",
        "ARG","ALA","ASN","GLN","PRO","SER","ASP","GLY","LYS","GLU")
letters_list <- aa321(AA)
pair_inter <- character(0)
for (i in 1:20) {
  for (j in i:20) {
    pair_inter <- c(pair_inter, paste0(letters_list[i], letters_list[j]))
  }
} 
#----------------------------------------------
num_columns <- 210
E210_ferritin <- data.frame(pdbID = ferritin$pdbID,length = rep(NA,Nprotein),group = ferritin$group,
                                         matrix(NA, ncol = num_columns, nrow = length(ferritin$pdbID)))
colnames(E210_ferritin) = c("pdbID","length","group",pair_inter)
E210_ferritin_SS <- E210_ferritin
for (NP in 1:Nprotein) {
  print(NP)
  pdbname <- ferritin$pdbID[NP]
  E210_ferritin_SS$pdbID[NP] <- pdbname
  ch <- ferritin$chainID[NP]
  E210_ferritin$pdbID[NP] <- paste0(pdbname,ch)
  E210_ferritin_SS$pdbID[NP] <- paste0(pdbname,ch)
  tryCatch({
    pdb0 <- read.pdb(pdbname)
    sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
    sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
    sele4 <- atom.select(pdb0, "noh")
    sele <- combine.select(sele1, sele2,sele4, operator="AND")
    if (!is.na(ferritin$start[NP])){
      st <- ferritin$start[NP]
      en <- ferritin$end[NP]
      sele3 <- atom.select(pdb0,  resno=st:en, verbose=TRUE)
      sele <- combine.select(sele1, sele2,sele3,sele4, operator="AND")
    }
    pdbX <- pdb0$atom[sele$atom,]
    E210_ferritin$length[NP] = length(unique(pdbX$resno))
    E210_ferritin_SS$length[NP] = length(unique(pdbX$resno))
    Net_Frame <- NetworkFrame(PDB = pdb0, AtomSelect = sele)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210_ferritin[NP,4:213] <- energy210
    # ****************keep res helix & sheet************************
    indhs <- inedx_HelixSheet(PDB = pdb0)
    Net_Frame_hs <- Net_Frame[Net_Frame$resnoif %in% indhs & Net_Frame$resnojf %in% indhs,]
    if(dim(Net_Frame_hs)[1]!=0){
        Net_Frame <- Net_Frame_hs
    }
    energy210_ss <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210_ferritin_SS[NP,4:213] <- energy210_ss
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}
saveRDS(E210_ferritin,file = "Data/rds/E210_ferritin.rds")
saveRDS(E210_ferritin_SS,file = "Data/rds/E210_ferritin_SS.rds")
```


## Covid_Spike200

```{r}
#spike <- read.delim("Data/csv/spike_structures.tsv")
spike <- read.csv("Data/csv/spike200.csv", sep=";")
energy_dell_dunbrack <- read.csv("Data/csv/energy.csv",header = FALSE,sep ="" )
energy_dell_dunbrack <- energy_dell_dunbrack[,5:33]
source("Functions/amacid2num.R")
source("Functions/atomtype2num167.R")
source("Functions/Energy_PC210.R")
source("Functions/NetworkFrame.R")
source("Functions/inedx_HelixSheet.R")
aaenergy <- read.csv("Data/csv/Pij.csv", header = F,sep = ";")
aaenergy <- (aaenergy + t(aaenergy))/2
colnames(aaenergy) = rownames(aaenergy)<- 
  c("F","L","I","V","W","Y","M","C","H","T","R","A","N","Q","P","S","D","G","K","E")
#row.names(aaenergy) <- aa321(aaenergy[,1])
#aaenergy <- aaenergy[,-1]
#row.names(aaenergy) <- colnames(aaenergy)
letters_list <- tolower(row.names(aaenergy))
aa210<-aaenergy[lower.tri(aaenergy,diag = T)]

Nprotein <- nrow(spike)
# ------Pairwise Amino Acid names-------------
AA <- c("PHE","LEU","ILE","VAL","TRP","TYR","MET","CYS","HIS","THR",
        "ARG","ALA","ASN","GLN","PRO","SER","ASP","GLY","LYS","GLU")
letters_list <- aa321(AA)
pair_inter <- character(0)
for (i in 1:20) {
  for (j in i:20) {
    pair_inter <- c(pair_inter, paste0(letters_list[i], letters_list[j]))
  }
} 
#----------------------------------------------
num_columns <- 210
E210_spike200 <- data.frame(pdbID = spike$pdbID,length = rep(NA,Nprotein),virus = spike$VIRUS,
                         #description = spike$DESCRIPTION,
                         conformation = spike$Conformation,
                         matrix(NA, ncol = num_columns, nrow = length(spike$pdbID)))
colnames(E210_spike200) = c("pdbID","length","virus", "conformation",pair_inter)
E210_spike200_SS <- E210_spike200
E210.seq_spike200 <- data.frame(pdbID = spike$pdbID,length = rep(NA,Nprotein),
                             virus = spike$VIRUS,
                             conformation = spike$Conformation,
                                         matrix(NA, ncol = num_columns, 
                                                nrow = length(spike$pdbID)))

for (NP in 1:Nprotein) {
  print(NP)
  ch <- substr(spike$pdbID[NP],5,5)
  pdbname <- substr(spike$pdbID[NP],1,4)
  E210_spike200$pdbID[NP] <- spike$pdbID[NP]
  E210_spike200_SS$pdbID[NP] <- spike$pdbID[NP]
  E210.seq_spike200$pdbID[NP] <- spike$pdbID[NP]
  tryCatch({
    pdb0 <- read.pdb(tolower(pdbname))
    sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
    sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
    sele4 <- atom.select(pdb0, "noh")
    sele <- combine.select(sele1, sele2,sele4, operator="AND")
    pdbX <- pdb0$atom[sele$atom,]
    pdbXca <- pdbX[pdbX$elety == "CA",]
    seq = aa321(pdbXca$resid)
    E210.seq_spike200$length[NP] = length(unique(pdbX$resno))
    freq = c()
    for (j in 1:20){
        freq[j] <-  length(grep(letters_list[j],seq))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
    E210.seq_spike200[NP,5:214] <- aa210_p
    E210_spike200$length[NP] = length(unique(pdbX$resno))
    E210_spike200_SS$length[NP] = length(unique(pdbX$resno))
    Net_Frame <- NetworkFrame(PDB = pdb0, AtomSelect = sele)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210_spike200[NP,5:214] <- energy210
    # ****************keep res helix & sheet************************
    indhs <- inedx_HelixSheet(PDB = pdb0)
    Net_Frame_hs <- Net_Frame[Net_Frame$resnoif %in% indhs & Net_Frame$resnojf %in% indhs,]
    if(dim(Net_Frame_hs)[1]!=0){
        Net_Frame <- Net_Frame_hs
    }
    energy210_ss <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210_spike200_SS[NP,5:214] <- energy210_ss
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}

saveRDS(E210_spike200,file = "Data/rds/E210_spike200.rds")
saveRDS(E210_spike200_SS,file = "Data/rds/E210_spike200_SS.rds")
saveRDS(E210.seq_spike200,file = "Data/rds/E210.seq_spike200.rds")

```

## Ferritin-like_sequence

```{r}
ferritin <- read.csv("Data/csv/Ferritin_Like.csv",sep = ";")
energy_dell_dunbrack <- read.csv("Data/csv/energy.csv",header = FALSE,sep ="" )
energy_dell_dunbrack <- energy_dell_dunbrack[,5:33]
source("Functions/amacid2num.R")
source("Functions/atomtype2num167.R")
source("Functions/Energy_PC210.R")
Nprotein <- nrow(ferritin)
# ------Pairwise Amino Acid names-------------
AA <- c("PHE","LEU","ILE","VAL","TRP","TYR","MET","CYS","HIS","THR",
        "ARG","ALA","ASN","GLN","PRO","SER","ASP","GLY","LYS","GLU")
letters_list <- aa321(AA)
pair_inter <- character(0)
for (i in 1:20) {
  for (j in i:20) {
    pair_inter <- c(pair_inter, paste0(letters_list[i], letters_list[j]))
  }
} 
#----------------------------------------------
num_columns <- 210
E210_seq_ferritin <- data.frame(pdbID = ferritin$pdbID,length = rep(NA,Nprotein),group = ferritin$group,
                                         matrix(NA, ncol = num_columns, nrow = length(ferritin$pdbID)))
colnames(E210_seq_ferritin) = c("pdbID","length","group",pair_inter)
for (NP in 1:Nprotein) {
  print(NP)
  pdbname <- ferritin$pdbID[NP]
  ch <- ferritin$chainID[NP]
  E210_seq_ferritin$pdbID[NP] <- paste0(pdbname,ch)
  tryCatch({
    pdb0 <- read.pdb(pdbname)
    sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
    sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
    sele <- combine.select(sele1, sele2, operator="AND")
    if (!is.na(ferritin$start[NP])){
      st <- ferritin$start[NP]
      en <- ferritin$end[NP]
      sele3 <- atom.select(pdb0,  resno=st:en, verbose=TRUE)
      sele <- combine.select(sele1, sele2,sele3, operator="AND")
    }
    pdbX <- pdb0$atom[sele$atom,]
    pdbXca <- pdbX[pdbX$elety == "CA",]
    seq = aa321(pdbXca$resid)
    E210_seq_ferritin$length[NP] = length(unique(pdbX$resno))
    freq = c()
    for (j in 1:20){
        freq[j] <-  length(grep(letters_list[j],seq))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
    E210_seq_ferritin[NP,4:213] <- aa210_p
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}
saveRDS(E210_seq_ferritin,file = "Data/rds/E210_seq_ferritin.rds")
```

# Tree construction

## Ferritin-like Phylogenetic tree 

## Phylogenetic Tree_Ferritin1

```{r}
E210_ferritin <- readRDS("Data/rds/E210_ferritin.rds")
df <- E210_ferritin
df$group[41:45]<-'Mn_cataleses'

df$family[df$group%in%c('Bac','Fer','Dps')]<-'a.25.1.1'
df$family[!df$group%in%c('Bac','Fer','Dps')]<-'a.25.1.2'
df$family[41:45]<-'Mn_cataleses'

df <- df[with(df, order(df$family)), ]
#df <- df[with(df, order(df$group)), ]
row.names(df) <- df[,1]
rn <- df[,1]
dis <- proxy::dist(df[,-c(1,2,3,214)], method = manhat)
dis[is.nan(dis)]<-0
tree_data <- ape::bionj(1 -dis)
grp <- list(Bacteri     = rn[1:5],
            Dps  = rn[6:20],
            Ferritins  = rn[21:31],
            BMM_alpha = rn[32:33],
            BMM_beta  = rn[34:36],
            Fatty_acid  = rn[37:40],
            RNRR2  = rn[41:48],
            Mn_cataleses = rn[49:53])

p_data <- ggtree(tree_data, layout = 'circular',
                 branch.length='none',size = 1.5) 

t = groupOTU(p_data, grp, "groups") + aes(color=groups) +
  theme(legend.position="right",
        legend.title = element_blank(),
        legend.text = element_text(size = 15, colour = "black"))+
  geom_tiplab2(aes(label=label,angle=angle),
               hjust=-.1,size=5,fontface="bold",align=TRUE)+
  scale_color_manual(breaks = c('Bacteri', 'Dps', 'Ferritins',
                                'BMM_alpha','BMM_beta','Fatty_acid',
                                'RNRR2','Mn_cataleses'),
                     values = c('#F8766D','#00BE67','#00A9FF',
                                '#CD9600','#7CAE00','#00BFC4',
                               '#FF61CC','#C77CFF','gray'))

#------------------tamam--------
df[,-c(1,2,3)] = df[,-c(1,2,3)]
df$total = rowSums(df[,-c(1,2,3)])
ggplot(df,aes(group,total))+geom_boxplot()
df$family[df$group %in% c("Bac","Fer","Dps","Manganese cataleses")] = "a.25.1.1"
df$family[! df$group %in% c("Bac","Fer","Dps")] = "a.25.1.2"
t.test(df$total[df$family == "a.25.1.0"],df$total[df$family == "a.25.1.2"],alternative = "greater")
wilcox.test(df$total[df$family == "a.25.1.0"],df$total[df$family == "a.25.1.2"],alternative = "greater")
ggplot(df,aes(family,total))+geom_boxplot()

dis <- dist(df[,-c(1,2,3)], method="manhattan")
dis <- sqrt(dis)
#------------------

#----------------------------------------
tot<-data.frame(total=df$total)
rownames(tot)<-rownames(df)
dis <- dist(tot)



x = df[,-c(1:3)]
x  <- as.matrix(x) 
hc <- x %>% dist(method = 'manhattan') %>% hclust(method = "ward.D")
dend <- hc %>% as.dendrogram

labcol <- c( "blue", "green","red","black","yellow","pink","magenta","purple")
ord_sml <- df[,][order.dendrogram(dend),3]
group_cols <- labcol[as.factor(ord_sml)]

dend2 <- dend %>%
  #color_branches(k=3, col = c("red","blue","green3")) %>%
  color_labels(col =   group_cols) %>%
  set("labels_cex", 0.5) 
plot(dend2, ylab = "Height")
legend("topright", legend = as.character(unique(ord_sml)),
       pch = 15, pt.cex = 2.3, cex = 1.4, bty = 'n',
       inset = c(-.15, -.1), # place outside
       col = unique(group_cols),horiz = FALSE,xpd = TRUE)







nnet <- neighborNet(dis)
plot(nnet)
ggsplitnet(nnet) + geom_tiplab2(aes(label=label),
                                hjust=-.1,size=5,fontface="bold")



tree_data <- ape::bionj(dis)
grp <- list(Bacteri     = rn[1:5],
            BMM_alpha = rn[6:7],
            BMM_beta  = rn[8:10],
            Dps  = rn[11:25],
            Fatty_acid  = rn[26:29],
            Ferritins  = rn[30:40],
            NAA  = rn[41:45],
            RNRR2  = rn[46:53])

p_data <- ggtree(tree_data, layout = 'circular',
                 branch.length='none',size = 1.5) 

t = groupOTU(p_data, grp, "groups") + aes(color=groups) +
  theme(legend.position="right",
        legend.title = element_blank(),
        legend.text = element_text(size = 15, colour = "black"))+
  geom_tiplab2(aes(label=label,angle=angle),
               hjust=-.1,size=5,fontface="bold",align=TRUE)




df0 <- E210_ferritin
df0 <- df0[with(df0, order(df0$group)), ]
pc <- prcomp(df0[,-c(1,2,3)],scale. = F,center = T) 
Rot_E_abcd <- as.matrix(df0[,-c(1,2,3)]) %*% pc$rotation[,1:3] 
df <- data.frame(class = df0$group
                 ,Rot_E_abcd) 
rn <- df0[,1]
row.names(df) <- rn




# ------------- UMAP ------------------------
ump.E_PC_abcd <- umap(E_PC_abcd[,-c(1)], n_neighbors = 10,
            min_dist = 0.001, random_state = 123)
#-------- UMAMP Visualization ---------------
umap.df.E_PC_abcd <- data.frame(
  sf = E_PC_abcd$class,
  UMAP1 = ump.E_PC_abcd$layout[, 1],
  UMAP2 = ump.E_PC_abcd$layout[, 2])
um.E_PC_abcd <- ggplot(umap.df.E_PC_abcd,aes(UMAP1, UMAP2, color= sf))+
  geom_point(alpha=.7) 




```

## Phylogenetic Tree_Ferritin

```{r}
E210_seq_ferritin <- readRDS("Data/rds/E210_ferritin.rds")

df <- E210_seq_ferritin
df <- df[with(df, order(df$group)), ]
row.names(df) <- df[,1]
# PCA
#pc <- prcomp(df[,-seq(1:3)],scale. = F,center = T) 
#Rot_ferr <- as.matrix(df[,-seq(1:3)]) %*% pc$rotation[,1:5]
#dis <- dist(Rot_ferr, method="manhattan")
#dis <- log2(dis)
#------------------
rn <- df[,1]
#--------custom_distance--------------------
dis <- as.matrix(dist(df[,-c(1,3)], method = manhat))
#--------------------------------------------
dis <- sqrt(dis)
#------------------
tree_data <- ape::bionj(dis)
grp <- list(Bacteri     = rn[1:5],
            BMM_alpha = rn[6:7],
            BMM_beta  = rn[8:10],
            Dps  = rn[11:25],
            Fatty_acid  = rn[26:29],
            Ferritins  = rn[30:40],
            NAA  = rn[41:45],
            RNRR2  = rn[46:53])

p_data <- ggtree(tree_data, layout = 'circular',
                 branch.length='none',size = 1.5) 

t = groupOTU(p_data, grp, "groups") + aes(color=groups) +
  theme(legend.position="right",
        legend.title = element_blank(),
        legend.text = element_text(size = 15, colour = "black"))+
  geom_tiplab2(aes(label=label,angle=angle),
               hjust=-.1,size=5,fontface="bold",align=TRUE)
```

## Covid

```{r}
E210.seq_spike200 <- readRDS("Data/rds/E210.seq_spike200.rds")
E210_spike200 <- readRDS("Data/rds/E210_spike200.rds")
#-----------close
df = E210_spike200
df = df[!is.na(df$length),]
df = df[df$conformation == "Close",]

rownames(df) = df$pdbID
x = df[,-c(1,3:4)]/df$length
x  <- as.matrix(x) 
hc <- x %>% proxy::dist(method = manhat)
hc[is.nan(hc)] <- 0
hc <- hc%>% hclust(method = "ward.D2")
dend <- hc %>% as.dendrogram

labcol <- c( "blue", "green","red")
ord_sml <- df[,][order.dendrogram(dend),3]
group_cols <- labcol[as.factor(ord_sml)]

dend2 <- dend %>%
  #color_branches(k=3, col = c("red","blue","green3")) %>%
  color_labels(col =   group_cols) %>%
  set("labels_cex", 0.5) 
ggplot(df, aes(virus, tot)) + geom_boxplot()
df$tot<-rowSums(df[,-c(1:4)])
plot(dend2, ylab = "Height")
legend("topright", legend = as.character(unique(ord_sml)),
       pch = 15, pt.cex = 2.3, cex = 1.4, bty = 'n',
       inset = c(-.15, -.1), # place outside
       col = unique(group_cols),horiz = FALSE,xpd = TRUE)
#---------------tamam----------------------
df$total<-rowSums(df[,-c(1:4)]/df$length)
ggplot(df,aes(virus,total))+geom_boxplot()

#----------------------open
df = left_join(spike200,E210_spike,by = "pdbID")
df = df[!is.na(df$length),]
df = df[df$Conformation == "Open",]
rownames(df) = df$pdbID
len = as.numeric(df$length)
  x = df[,-c(1:7)]/len
x  <- as.matrix(x) 
hc <- x %>% dist(method = 'manhattan') %>% hclust(method = "ward.D2")
dend <- hc %>% as.dendrogram

labcol <- c( "blue", "green","red")
ord_sml <- df[,][order.dendrogram(dend),2]
group_cols <- labcol[as.factor(ord_sml)]

dend2 <- dend %>%
  #color_branches(k=3, col = c("red","blue","green3")) %>%
  color_labels(col =   group_cols) %>%
  set("labels_cex", 0.5) 
plot(dend2, ylab = "Height")
legend("topright", legend = as.character(unique(ord_sml)),
       pch = 15, pt.cex = 2.3, cex = 1.4, bty = 'n',
       inset = c(-.15, -.1), # place outside
       col = unique(group_cols),horiz = FALSE,xpd = TRUE)
```


## UMAP_manhat_norm

```{r}
df <- readRDS("Data/rds/E210.astral40.rds")
dfs <- df[df$class %in% c("a","b","c","d")& df$length > 100 & !is.na(df$FF),]
df1<-dfs[,-c(1:2,4:5,7:8)]
df1[,-1] <- df1[,-1]
#df1$total = rowSums(df1[,-c(1,2)])
df1<-data.frame(class=substr(df1$fold,1,1),df1)
#df1 = df1[c(sample(2743:5697,500),sample(2642,500),sample(5698:10159,500)),]
#df1 <- df1 %>% sample_frac(1)
#df1 <- df1[ !is.na(df1$FF),]

#df1 = df1 %>% group_by(fold) %>% 
#  dplyr::summarise(across(colnames(df1)[-c(1,2)], mean))
#df1<-data.frame(class=substr(df1$fold,1,1),df1)
dis <- as.matrix(proxy::dist(df1[,-c(1,3)], method = manhat_norm_diff))
ump <- umap(d = dis,
            n_neighbors = 5,
            min_dist = 0.1,
            input="dist")
ump_df <- data.frame(df1[,1:2],
                     UMAP1 = ump$layout[, 1],
                     UMAP2 = ump$layout[, 2])
ggplot(ump_df,aes(UMAP1, UMAP2, color= class)) +
   geom_point(alpha=.7) + 
  ylim(-15,15) +
  xlim(-10,20) +
   mythem

```

## UMAP at class level
```{r}
df <- readRDS("data/rds/E210.astral40.rds")
df = E210.astral_95
df = E210.aspartaze_SS
df = E210.seq_astral40
sel1<-c('a.39.1','a.45.1')
sel1<-c('a.29.2','a.29.3')
sel1<-c('a.102.1','a.116.1')

sel<-c('b.82.1','b.82.3')    #  'b.82.2'

sel1<-c('c.1.10','c.1.8')
sel1<-c('c.1.10','b.82.1')

dfs <- df[df$superfamily %in% sel1,]

dfs <- df[df$class %in% c("a","b")& df$length > 100 & !is.na(df$FF),]
df1<-dfs[,-c(1:5,7:8)]
df1[,-1] <- df1[,-1]/dfs$length
df1<-data.frame(class=substr(df1$fold,1,1),df1)
#df1$total = rowSums(df1[,-c(1,2)])
ggplot(df1,aes(class,total))+geom_boxplot()
df1 <- df1 %>% sample_frac(1)
df2 <- df1[c(1:2000),]

df1<-df1 %>% 
  group_by(fold) %>%
  dplyr::summarise(across(colnames(df1)[-1], mean))


#pc <- prcomp(df1,scale. = F,center = T) 
Rot_E_abcd <- as.matrix(df1[,-c(1:2)]) #%*% pc$rotation[,1:5] 
E_PC_abcd <- data.frame(class = df1$class,Rot_E_abcd) 
# ------------- UMAP ------------------------
ump.E_PC_abcd <- umap(E_PC_abcd[,-c(1)], n_neighbors = 20,
            min_dist = 0.001, random_state = 123)
#-----------customized UMAP-----------
tm_score <- function(x, y){
    d <- (sum(1/(1 + (x - y))))
  return(d)
}
tm_score2 <- function(x, y){
  xy<-rep(0,210)
  for(i in 1:210){
    if(x[i]<y[i]){
      xy[i]<-x[i]-y[i]
    }else{
      xy[i]<-y[i]-x[i]
    }
  }
  d <- (sum(1/(1 + xy)))
  return(d)
}
tm_score3 <- function(x, y){
  if (sum(x) > sum(y)){
    d <- (sum(1/(1 + (x - y))))
  }
  else{
    d <- (sum(1/(1 + ( -x + y))))
  }
    
  return(d)
}


dis <- as.matrix(dist(df1[,-c(1,2)], method = tm_score3))
table(df1$class)
a<-dis[1:1610,1:1610]
a<-a[lower.tri(a)]
b<-dis[1611:3770,1611:3770]
b<-b[lower.tri(b)]
ab<-dis[1:1610,1611:3770]
ab<-as.vector(ab)
df_dis<-data.frame(gr=rep(c('a','b','ab'),c(length(a),length(b),length(ab))),
                   dis=c(a,b,ab))
ggplot(df_dis,aes(gr,dis))+geom_boxplot()+ylim(-100,100)

umap_result <- umap(d = dis,
                    n_neighbors = 30,
                    min_dist = 0.01)
umap.df.E_PC_abcd <- data.frame(
  df1[,1:2],
  UMAP1 = umap_result$layout[, 1],
  UMAP2 = umap_result$layout[, 2])

 ggplot(umap.df.E_PC_abcd,aes(UMAP1, UMAP2, color= class))+
  geom_point(alpha=.7,show.legend = F) + 
  mythem

#-------- UMAMP Visualization ---------------
umap.df.E_PC_abcd <- data.frame(
  df1[,1:2],
  UMAP1 = ump.E_PC_abcd$layout[, 1],
  UMAP2 = ump.E_PC_abcd$layout[, 2])

 ggplot(umap.df.E_PC_abcd,aes(UMAP1, UMAP2, color= class))+
  geom_point(alpha=.7) +
  mythem

 
 um.E_PC_abcd <- ggplot(umap.df.E_PC_abcd,aes(UMAP1, UMAP2, color= sf))+
  geom_point(alpha=.7) +
  mythem
#-------------PCA 
 pc <- prcomp(df1,scale. = F,center = T)
pc.df <- data.frame(group=df$family, pc$x)
g.pc <- ggplot(pc.df, aes(x=PC1, y=PC2, col=group))+
  geom_point(size=2)+
  theme_bw(base_line_size = .3)+
  #stat_ellipse(aes(fill = group), geom="polygon",type = "norm",alpha=0.2)+
  mythem
 
```

## Tree

```{r}
# Best result with SS and pc = 100, ward.D
df = E210.aspartaze_SS
pc <- prcomp(df[,-seq(1:10)],scale. = F,center = T) 
Rot_E_abcd <- as.matrix(df[,-seq(1:10)]) %*% pc$rotation[,1:20] 
E_PC_abcd <- data.frame(class = df$class,Rot_E_abcd) 
E_PC_abcd <- umap.df.E_PC_abcd
x  <- E_PC_abcd[,-1] %>% as.matrix
rownames(x) <- df$pdbID
hc <- x %>% dist(method = 'manhattan') %>% hclust(method = "ward.D2")
 dfx = data.frame(x)

dend <- hc %>% as.dendrogram

labcol <- c( "blue", "red")
ord_sml <- E_PC_abcd[,][order.dendrogram(dend),1]
group_cols <- labcol[as.factor(ord_sml)]

dend2 <- dend %>%
  #color_branches(k=3, col = c("red","blue","green3")) %>%
  color_labels(col =   group_cols) %>%
  set("labels_cex", 0.5) 
plot(dend2, ylab = "Height")



df <- E210.aspartaze_SS
df <- df[with(df, order(df$family)), ]
row.names(df) <- paste0(df[,1],df[,2])
# PCA
#pc <- prcomp(df[,-seq(1:3)],scale. = F,center = T) 
#Rot_ferr <- as.matrix(df[,-seq(1:3)]) %*% pc$rotation[,1:5]
#dis <- dist(Rot_ferr, method="manhattan")
#dis <- log2(dis)
#------------------
rn <- paste0(df[,1],df[,2])
dis <- dist(df[,-c(1:10)], method="manhattan")
dis <- sqrt(dis)
#------------------
tree_data <- ape::bionj(dis)
grp <- list(A0     = rn[1:16],
            A1 = rn[17:49],
            A2  = rn[50:53])

p_data <- ggtree(tree_data, layout = 'circular',
                 branch.length='none',size = 1.5) 

t = groupOTU(p_data, grp, "groups") + aes(color=groups) +
  theme(legend.position="right",
        legend.title = element_blank(),
        legend.text = element_text(size = 15, colour = "black"))+
  geom_tiplab2(aes(label=label,angle=angle),
               hjust=-.1,size=5,fontface="bold",align=TRUE)

```



## UMAP covid

```{r}
E210_spike200 <- readRDS("Data/rds/E210_spike200.rds")
E210_spike200_SS <- readRDS("Data/rds/E210_spike200_SS.rds")
E210.seq_spike200 <- readRDS("Data/rds/E210.seq_spike200.rds")

#spike200 <- read.csv("Data/csv/spike200.csv", sep = ";")
#spike200$chain = substr(spike200$pdbID,5,5)
#spike200$pdbID = paste0(tolower(substr(spike200$pdbID,1,4)),spike200$chain)
#df = left_join(spike200,E210_spike,by = "pdbID")
df <- E210.seq_spike200
df = df[!is.na(df$length),]
df0 = df[,-seq(1:4)]/df$length
#df = df[df$length>100 & df$description == " Spike glycoprotein",]
pc <- prcomp(df0,scale. = F,center = T) 
#%*% pc$rotation[,1:10] 
Rot_E_abcd <- as.matrix(df0) 
E_PC_abcd <- data.frame(class = df$virus,conf = df$conformation,Rot_E_abcd) 
# ------------- UMAP ------------------------
ump.E_PC_abcd <- umap(E_PC_abcd[,-c(1,2)], n_neighbors = 40,
            min_dist = 0.1, random_state = 123)
#-------- UMAMP Visualization ---------------
umap.df.E_PC_abcd <- data.frame(shape = E_PC_abcd$conf,
  sf = E_PC_abcd$class,
  UMAP1 = ump.E_PC_abcd$layout[, 1],
  UMAP2 = ump.E_PC_abcd$layout[, 2])
um.E_PC_abcd <- ggplot(umap.df.E_PC_abcd,aes(UMAP1, UMAP2, color= sf, shape = shape))+
  geom_point(alpha=.7) +
  mythem

```



# Sequence Level

## Astral40

```{r}
astral <- read.fasta('Data/csv/astral-scopedom-seqres-gd-sel-gs-bib-40-2.08.fa')
aaenergy <- read.csv("Data/csv/Pij.csv", header = F,sep = ";")
astral40 <- readRDS("Data/rds/astral_40.rds")

aaenergy <- (aaenergy + t(aaenergy))/2
colnames(aaenergy) = rownames(aaenergy)<- 
  c("F","L","I","V","W","Y","M","C","H","T","R","A","N","Q","P","S","D","G","K","E")
#row.names(aaenergy) <- aa321(aaenergy[,1])
#aaenergy <- aaenergy[,-1]
#row.names(aaenergy) <- colnames(aaenergy)
letters_list <- tolower(row.names(aaenergy))
nprotein <- length(astral)
aa210<-aaenergy[lower.tri(aaenergy,diag = T)]
num_columns <- 210
E210.seq_astral40 <- data.frame(pdbID = astral40$pdbID,length = rep(NA,nprotein),
                                         matrix(NA, ncol = num_columns, 
                                                nrow = length(astral40$pdbID)))
for (i in 1:nprotein){
  print(i)
  #seq = unlist(strsplit(tolower(DTI_df__seq$seq[i]),""))
  seq = astral[[i]]
  freq = c()
  for (j in 1:20){
    freq[j] <-  length(grep(letters_list[j],seq))
  }
  freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
  en_fr <- aaenergy * freq_matrix
  en_fr <- en_fr/length(seq)
  pair_es <- diag(freq) %*% as.matrix(en_fr)
  aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
  E210.seq_astral40[i,3:212] <- aa210_p
  E210.seq_astral40[i,2] <- length(seq)
}

E210.seq_astral40 <- cbind(astral40,E210.seq_astral40[,-c(1)])
saveRDS(E210.seq_astral40,file = "Data/rds/E210.seq_astral40.rds")
# ---------------------------------structure---------------------------------
E210.astral40 <- readRDS("Data/rds/E210.astral40.rds")
df = E210.astral40
df <- df[df$length > 100 & !is.na(df$length) & df$class %in% c("a","b"),]
df1 <- df[,-seq(1:10)]/(df$length)
pc <- prcomp(df1,scale. = F,center = T) 
Rot_E_abcd <- as.matrix(df1) 
E_PC_abcd <- data.frame(class = df$class,Rot_E_abcd) 
# ------------- UMAP ------------------------
ump.E_PC_abcd <- umap(E_PC_abcd[,-c(1)], n_neighbors = 10,metric = "manhattan",
            min_dist = .0001, random_state = 123)
#-------- UMAMP Visualization ---------------
umap.df.E_PC_abcd <- data.frame(
  sf = E_PC_abcd$class,
  UMAP1 = ump.E_PC_abcd$layout[, 1],
  UMAP2 = ump.E_PC_abcd$layout[, 2])
um.E_PC_abcd <- ggplot(umap.df.E_PC_abcd,aes(UMAP1, UMAP2, color= sf))+
  geom_point(alpha=.7) +
  mythem
#------------tSNE--------------------------
E_PC_abcd<-E_PC_abcd[!duplicated(E_PC_abcd[,-1]),]
tsne<-Rtsne(E_PC_abcd[,-c(1)],pca=F,dims=2,perplexity=100,
            verbose=T, max_iter=1000,theta=0,
            learning_rate=300,normalize=F)

tsne.df <- data.frame(
  group = E_PC_abcd$class,
  tSNE1=tsne$Y[,1],
  tSNE2=tsne$Y[,2]
  )

g.tsne <- ggplot(tsne.df,aes(tSNE1, tSNE2, color= group))+
  geom_point(size=1.5,alpha=.7)+
  mythem

# ---------------------------------Sequence---------------------------------
e_seq_str <- data.frame(eseq = rowSums(E210.seq_astral40[,-c(1:10)]),
                     estr = rowSums(E210.astral40[,-c(1:6)]))

e_seq_str <- cbind(E210.seq_astral40[,c(1:10)],e_seq_str)
cor.test(e_seq_str$eseq,e_seq_str$estr)
# ---------------------------------Sequence95---------------------------------

e_seq_str <- data.frame(eseq = rowSums(E210.seq_astral95[,-c(1:10)]),
                     estr = rowSums(E210.astral95[,-c(1:9)]))

e_seq_str <- cbind(E210.seq_astral95[,c(1:10)],e_seq_str)
e_seq_str1 <- e_seq_str[-grep(',',e_seq_str$seq),]
ggplot(e_seq_str1,aes(estr,eseq,col=class))+geom_jitter()
cor.test(e_seq_str1$eseq,e_seq_str1$estr)
```




## Drug_target

```{r}
dti_barabasi <- read_excel("Data/csv/dti_barabasi.xlsx")
colnames(dti_barabasi) <- c("drug","entrez")
#dti_barabasi <- dti_barabasi[1:100,]
dti_barabasi$entrez <- as.character(dti_barabasi$entrez)
library(rentrez)
get_protids_from_geneids <- function(gene_ids) {
  
  # Get the protein elink.
  protein_elink <- rentrez::entrez_link(dbfrom = "gene", id = gene_ids, db = "protein")
  
  # Get the protein id (refseq database)
  protein_ids <- protein_elink$links$gene_protein_refseq
  protein_ids
}

make_aa_fasta <- function(prot_ids, nameFile) {
  # Make a multi-fasta file with each protein id
  sapply(prot_ids, function(x) {
    protein_esummary <- rentrez::entrez_summary(db = "protein", id = x)
    protein_fasta <- rentrez::entrez_fetch(db = "protein", id = protein_esummary$uid, rettype = "fasta")
    # save amino acid sequences into a FASTA file ("nameFile"")
    write(protein_fasta, file = paste(nameFile, ".fasta", sep = ""), append = TRUE)
    #return(protein_fasta)
  } )
}
#-------------------------------------
dti_seq <- readRDS("Data/rds/DTI_df_seq.rds")
sAB <- readRDS("Data/rds/sAB.rds")

aaenergy <- read.csv("Data/csv/Pij.csv", header = F,sep = ";")
aaenergy <- (aaenergy + t(aaenergy))/2
colnames(aaenergy) = rownames(aaenergy)<- 
  c("F","L","I","V","W","Y","M","C","H","T","R","A","N","Q","P","S","D","G","K","E")
letters_list <- tolower(row.names(aaenergy))
nprotein <- nrow(dti_seq)
aa210<-aaenergy[lower.tri(aaenergy,diag = T)]
num_columns <- 210
E210.seq_dti <- data.frame(pdbID = dti_seq$drug,length = rep(NA,nprotein),
                                         matrix(NA, ncol = num_columns, 
                                                nrow = nrow(dti_seq)))
for (i in 1:nprotein){
  print(i)
  #seq = unlist(strsplit(tolower(DTI_df__seq$seq[i]),""))
  seq = unlist(strsplit(tolower(dti_seq$seq[i]), ""))
  freq = c()
  for (j in 1:20){
    freq[j] <-  length(grep(letters_list[j],seq))
  }
  freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
  en_fr <- aaenergy * freq_matrix
  en_fr <- en_fr/length(seq)
  pair_es <- diag(freq) %*% as.matrix(en_fr)
  aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
  E210.seq_dti[i,3:212] <- aa210_p
  E210.seq_dti[i,2] <- length(seq)
}
df0 = E210.seq_dti
df = cbind(drug = df[,1],df[,-c(1:2)]/df$length)
df <- df %>%
  group_by(drug) %>%
 dplyr::summarise(across(X1:X210, mean))

x  <- as.matrix(df[,-c(1)]) 
row.names(x) = df$drug
dist_matrix <- as.matrix(dist(x, method = "manhattan"))
dist_df <- as.data.frame(as.table(dist_matrix))
colnames(dist_df) <- c("d1", "d2", "distance")
dist_df <- dist_df[dist_df$d1 != dist_df$d2, ]

```

## Bacteriocin

```{r}
bacteriocin <- readRDS("Data/rds/bacteriocin.rds")
aaenergy <- read.csv("Data/csv/Pij.csv", header = F,sep = ";")
aaenergy <- (aaenergy + t(aaenergy))/2
colnames(aaenergy) = rownames(aaenergy)<-c("F","L","I","V","W","Y","M","C","H","T","R","A","N","Q","P","S","D","G","K","E")
letters_list <- tolower(row.names(aaenergy))
nprotein <- nrow(bacteriocin)
aa210<-aaenergy[lower.tri(aaenergy,diag = T)]
num_columns <- 210
E210.seq_bacteriocin <- data.frame(class = bacteriocin$class,seq=bacteriocin$seq,
                                   length = rep(NA,nprotein),
                                   matrix(NA, ncol = num_columns,nrow = nrow(bacteriocin)))
for (i in 1:nprotein){
  print(i)
  seq = unlist(strsplit(tolower(E210.seq_bacteriocin$seq[i]), ""))
  freq = c()
  for (j in 1:20){
  freq[j] <-  length(grep(letters_list[j],seq))
  }
freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
en_fr <- aaenergy * freq_matrix
en_fr <- en_fr/length(seq)
pair_es <- diag(freq) %*% as.matrix(en_fr)
aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
E210.seq_bacteriocin[i,4:213] <- aa210_p
E210.seq_bacteriocin[i,3] <- length(seq)
}

saveRDS(E210.seq_bacteriocin,'Data/rds/E210.seq_bacteriocin.rds')
E210.seq_bacteriocin = readRDS("Data/rds/E210.seq_bacteriocin.rds")
df = E210.seq_bacteriocin
df1 <- df[,-seq(1:1)]

dis <- as.matrix(proxy::dist(df1, method = manhat))
ump <- umap(d = dis,
            n_neighbors = 40,
            min_dist = 0.01,
            input="dist")

ump_df <- data.frame(df[,1:2],
                     UMAP1 = ump$layout[, 1],
                     UMAP2 = ump$layout[, 2])
ggplot(ump_df,aes(UMAP1, UMAP2, color= class)) +
   geom_point(alpha=.7)



Rot_E_abcd <- as.matrix(df1)
E_PC_abcd <- data.frame(class = df$class,Rot_E_abcd)
# ------------- UMAP ------------------------
ump.E_PC_abcd <- umap(E_PC_abcd[,-c(1)], n_neighbors = 25,metric = "manhattan",
                      min_dist = 0.1, random_state = 123)
#-------- UMAMP Visualization ---------------
umap.df.E_PC_abcd <- data.frame(
sf = E_PC_abcd$class,
UMAP1 = ump.E_PC_abcd$layout[, 1],
UMAP2 = ump.E_PC_abcd$layout[, 2])
ggplot(umap.df.E_PC_abcd,aes(UMAP1, UMAP2, color= sf))+
geom_point(alpha=.7) +
mythem

#----------tree
df = E210.seq_bacteriocin
len = as.numeric(df$length)
x = df[,-c(1:2)]/df$length
x  <- as.matrix(x)
hc <- x %>% dist(method = 'manhattan') %>% hclust(method = "ward.D2")
dend <- hc %>% as.dendrogram
labcol <- c( "blue", "green","red")
ord_sml <- df[,][order.dendrogram(dend),1]
group_cols <- labcol[as.factor(ord_sml)]
dend2 <- dend %>%
#color_branches(k=3, col = c("red","blue","green3")) %>%
color_labels(col =   group_cols) %>%
set("labels_cex", 0.5)
plot(dend2, ylab = "Height")
legend("topright", legend = as.character(unique(ord_sml)),
pch = 15, pt.cex = 2.3, cex = 1.4, bty = 'n',
inset = c(-.15, -.1), # place outside
col = unique(group_cols),horiz = FALSE,xpd = TRUE)
#------------boxplot
df = E210.seq_bacteriocin
x = df[,-c(1:1)]
hc <- as.matrix(x %>% dist(method = manhat_norm_maxl))
class1 = hc[1:499,1:499]
class1=class1[lower.tri(class1)]
class2 = hc[500:728,500:728]
class2=class2[lower.tri(class2)]
class3 = hc[729:821,729:821]
class3=class3[lower.tri(class3)]
class12 = hc[500:728,1:499]
class12 = hc[500:728,1:499]
class12=class12[lower.tri(class12)]
class13 = hc[729:821,1:499]
class13=class13[lower.tri(class13)]
class23 = hc[729:821,500:728]
class23=class23[lower.tri(class23)]
df<-data.frame(comp=c(rep('same',sum(length(class1),length(class2),length(class3))),
rep('dif',sum(length(class12),length(class13),length(class23)))),
dis=c(class1,class2,class3,class12,class13,class23))
ggplot(df,aes(comp,dis))+geom_boxplot()
df$scale_dis<-1/(1+df$dis)


```
## Remote homology

```{r}
df <- readRDS("data/rds/E210.seq_astral40.rds")
df = df[!is.na(df$FL),]
fa1 = "a.25.1.0"
fa2 = "a.25.1.2"
df_fa = df[df$family %in% c(fa1,fa2),]
dis_fa <- as.matrix(df_fa[,-c(1:8)] %>% dist(method = 'manhattan'))
table(df_fa$family)
# *****
class1 = dis_fa[1:15,1:15]
class1=class1[lower.tri(class1)]
class2 = dis_fa[16:26,16:26]
class2=class2[lower.tri(class2)]
class12 = dis_fa[1:15,16:26]
class12=as.vector(class12)

df_family<-data.frame(gr=rep(c('within','between'),
                             c(length(class1)+length(class2),length(class12))),
                      distance=c(class1,class2,class12))

g1<-ggplot(df_family,aes(gr,distance))+geom_boxplot()+ggtitle('a.25.1.0, a.25.1.2')
# *****************
sf1<-'a.29.2'
sf2<-'a.29.3'
df_sf = df[df$superfamily %in% c(sf1,sf2),]
dis_sf <- as.matrix(df_sf[,-c(1:8)] %>% dist(method = 'manhattan'))
table(df_sf$superfamily)
# *****
class1 = dis_sf[1:31,1:31]
class1=class1[lower.tri(class1)]
class2 = dis_sf[32:50,32:50]
class2=class2[lower.tri(class2)]
class12 = dis_sf[1:31,32:50]
class12=as.vector(class12)

df_superfamily<-data.frame(gr=rep(c('within','between'),
                             c(length(class1)+length(class2),length(class12))),
                      distance=c(class1,class2,class12))

g2<-ggplot(df_superfamily,aes(gr,distance))+geom_boxplot()+ggtitle('a.29.2, a.29.3')
ggarrange(g1,g2)
```

## fold
```{r}
df <- readRDS("Data/rds/E210.astral40.rds")
df <- df[!is.na(df$FL) & df$class %in% c('a','b'), ]

dis <- as.matrix(dist(df[,-c(1:2,4:8)],method = manhat))
dis <- data.frame(fold=df$fold,superf = df$superfamily,length = df$length, dis)

n = nrow(df)
lengths = (df$length)
fold = df$fold
super = df$superfamily
class = df$class
distances = as.matrix(dis[,-c(1:3)])
protein_data <- data.frame(
  Protein1 = rep(lengths, each = n),
  Protein2 = rep(lengths, times = n),
  fold1 = rep(fold, each = n),
  fold2 = rep(fold, times = n),
  super1 = rep(super, each = n),
  super2 = rep(super, times = n),
  class1 = rep(class, each = n),
  class2 = rep(class, times = n),
  Distance = as.vector(distances),
  LengthDifference = abs(rep(lengths, each = n) - rep(lengths, times = n))
)


protein_data = protein_data[protein_data$Distance !=0,]
protein_data = protein_data[protein_data$class1 !="b" & protein_data$class2 !="b",]
different_proteins <- protein_data[protein_data$super1 != protein_data$super2 
                                   & protein_data$fold1 == protein_data$fold2,]
different_proteins$group = "Diff"

#pairs_random_differ <- different_proteins %>%
#  group_by(LengthDifference) %>%
#  sample_n(2) %>%
#  ungroup()
#lm(formula = Distance ~ LengthDifference, data = pairs_random_differ)
same_proteins <- protein_data[protein_data$super1 == protein_data$super2 ,]
same_proteins$group = "same"

#pairs_random_same <- same_proteins %>%
#  group_by(LengthDifference) %>%
#  sample_n(2) %>%
#  ungroup()
#lm(formula = Distance ~ LengthDifference, data = pairs_random_same)

df_diff_same = rbind(same_proteins,different_proteins)
ggplot(data = df_diff_same,aes(group,Distance,color = group)) +
  geom_boxplot()

diff_fold <- different_proteins[sample(nrow(different_proteins),
                                       size = 4000, replace = FALSE), ]
lm(formula = Distance ~ LengthDifference, data = diff_fold)
plot(diff_fold$LengthDifference,diff_fold$Distance)

same_fold <- protein_data[protein_data$fold1 == protein_data$fold2,]
same_fold <- same_fold[sample(nrow(same_fold), size = 4000, replace = FALSE), ]
lm(formula = Distance ~ (LengthDifference+1), data = same_fold)
plot(same_fold$LengthDifference,same_fold$Distance)


df_diff200 = different_proteins[different_proteins$LengthDifference<100,]
df_same200 = same_proteins[same_proteins$LengthDifference<100,]

dff200 = rbind(df_diff200,df_same200)
dff200 = dff200[dff200$Distance!=0,]
dff200$disnorm = dff200$Distance/(6.66 + (dff200$LengthDifference *- 0.036))
ggplot(dff200,aes(group,Distance)) + geom_boxplot()


gr<-unique(dis$superf)
between<-c() 
within<-c()
for (i in 1:length(gr)) {  
  print(i)   
  x <- gr[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+3)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+3)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}

median(within,na.rm = T)
median(between,na.rm = T)

bet<-between[sample(length(between),10000)]
wit<-within[sample(length(within),10000)]
t.test(wit,bet)

df_fold <- data.frame(gr=rep(c('within','between'),c(length(wit),length(bet))),
                    distance=c(wit,bet))
ggplot(df_fold,aes(gr,distance,fill=gr))+geom_boxplot(alpha=.3)


```

## cusomize distance

```{r}
library(umap)
data <- matrix(runif(100), ncol = 2)
custom_distance_matrix <- dist(data, method = "euclidean")
umap_result <- umap(d = as.matrix(custom_distance_matrix))
plot(umap_result$layout, 
     main = "UMAP with Custom Distance Matrix",
     n_neighbors = 10,
     min_dist = 0.1, 
     random_state = 123)

```


# CT_Ho

```{r}
df <- read.csv('Data/csv/CT_Ho_cathID_filtered.csv')
df$chain <- substr(df$pdbID,5,5)
df$pdbID <- substr(df$pdbID,1,4)

E210.CT_Ho <- data.frame(pdbID = df$pdbID,chainID = df$chain,length = NA,
                            start = df$start,end = df$end,
                         cathID = df$cathID,
                                        class = df$class,
                                         architecture = df$architecture,
                                         topology = df$topology,
                                         homologous = df$homologous,
                                         matrix(NA, ncol = 210, nrow = nrow(df)))
colnames(E210.CT_Ho)[-c(1:10)] <- pair_inter
E210.seq_CT_Ho <- E210.CT_Ho

for(NP in 1:nrow(E210.CT_Ho)){
  print(NP)
  pdbname <- E210.CT_Ho$pdbID[NP]
  ch <- E210.CT_Ho$chainID[NP]
  tryCatch({
    pdb0 <- read.pdb(pdbname)
      sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
      sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
      sele4 <- atom.select(pdb0, "noh")
      sele <- combine.select(sele1, sele2,sele4, operator="AND")
      if (!is.na(E210.CT_Ho$start[NP])){
        pdbX <- pdb0$atom[sele$atom,]
        shift_pos <- min(pdbX$resno)-1
         st <- E210.CT_Ho$start[NP]+shift_pos
         en <- E210.CT_Ho$end[NP]+shift_pos
         sele3 <- atom.select(pdb0,  resno=st:en, verbose=TRUE)
         sele <- combine.select(sele, sele3, operator="AND")
      }
    pdbX <- pdb0$atom[sele$atom,]
    E210.CT_Ho$length[NP] = length(unique(pdbX$resno))
    Net_Frame <- NetworkFrame(PDB = pdb0,sele)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210.CT_Ho[NP,11:220] <- energy210
    # ****************seq************************
    pdbXca <- pdbX[pdbX$elety == "CA",]
    seq = aa321(pdbXca$resid)
    E210.seq_CT_Ho$length[NP] = length(seq)
    freq = c()
    for (j in 1:20){
        freq[j] <-  length(grep(letters_list[j],seq))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
    E210.seq_CT_Ho[NP,11:220] <- aa210_p
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}
saveRDS(E210.CT_Ho,'Data/rds/E210.CT_Ho.rds')
saveRDS(E210.seq_CT_Ho,'Data/rds/E210.seq_CT_Ho.rds')

###########################
df <- readRDS('Data/rds/E210.CT_Ho.rds')
df <- df[!is.na(df$FF),]
twoSF <- unique(df$cathID)

dis <- as.matrix(dist(df[,c(3,11:220)], method = manhat))
diag(dis) <- NA

im <- apply(dis,1, FUN = which.min)
res <- data.frame(sorce=df$cathID, target=df$cathID[im])
#res$pred <- ifelse(res[,1]==res[,2],"1","0")
## ----------
cm <- matrix(0,2,2)
colnames(cm) <- rownames(cm) <- twoSF
for (sf in twoSF) {
  x <- res[res$sorce%in%sf,]
  z <- table(x$target)
  cm[sf,names(z)]<-z
}
round(sum(diag(cm))/sum(cm),2)
# --------------
ump <- umap(d = dis,
            n_neighbors = 50,
            min_dist = 0.1,
            input="dist")
ump_df <- data.frame(class=df$cathID,
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])
ggplot(ump_df,aes(UMAP1, UMAP2, color= class)) +
    geom_point(alpha=.7) + 
  mythem
```

# five superfamily

```{r}
# -------- 5 sf
win <- 'a.4.5'
imu <- 'b.1.1'
ph <- 'b.55.1'
ub <- 'd.15.1'
ntf <- 'd.17.4'

fiveSF <- c(win,imu,ph,ub,ntf)
#############
df5 <- readRDS('Data/rds/E210.seq_df5.rds')
df5 <- df5[!is.na(df5$FF),]
###############
dis <- as.matrix(dist(df5[,c(3,9:218)],method = manhat))
diag(dis) <- NA

im <- apply(dis,1, FUN = which.min)
res <- data.frame(sorce=df5$superfamily, target=df5$superfamily[im])
#res$pred <- ifelse(res[,1]==res[,2],"1","0")
cm <- matrix(0,5,5)
colnames(cm) <- rownames(cm) <- fiveSF
for (sf in fiveSF) {
  x <- res[res$sorce%in%sf,]
  z <- table(x$target)
  cm[sf,names(z)]<-z
}
sum(diag(cm))/sum(cm)
Precision <- diag(cm)/colSums(cm)
Recall <- diag(cm)/rowSums(cm)
F1 <- (2 * Precision * Recall) /(Precision + Recall)
round(F1,2)
# ------------------------------------
trainControl <- trainControl(method="repeatedcv", number=10, repeats=3)
metric <- "Accuracy"

df5$superfamily<-factor(df5$superfamily)
validationIndex <- createDataPartition(df5$superfamily, p=0.70, list=FALSE)

train <- df5[validationIndex,c(7,9:218)] # 70% of data to training
test <- df5[-validationIndex,c(7,9:218)] 

grid <- expand.grid(.k=seq(1,1,by=1))
fit.knn <- train(superfamily~., data=train, method="knn", 
                 metric=metric, tuneGrid=grid, trControl=trainControl)
knn.k2 <- fit.knn$bestTune # keep this optimal k for testing with stand alone knn() function in next section
print(fit.knn)

prediction <- predict(fit.knn, newdata = test)
cf <- confusionMatrix(prediction, test$superfamily)
print(cf)
cf$byClass[,'F1']
#############
kfolds <- createFolds(df5$superfamily, k = 10)
res<-c()
for (i in 1:10) {
  print(i)
  x <- kfolds[[i]]
  tr <- df5[-x,c(7,9:218)]
  te <- df5[x,c(7,9:218)]
  #rf <- tuneRF(tr[,-1],tr$superfamily, stepFactor=1.5, improve=1e-5,plot = T,trace = T, doBest = T)
  rf = randomForest(superfamily~.,data = tr)
  p2 <- predict(rf, te)
  cm2 <- confusionMatrix(p2, te$superfamily)
  Accuracy <- c(cm2$overall['Accuracy'], F1=cm2$byClass[,'F1'])
  res<-rbind(res,Accuracy)
}
res <- rbind(res,colMeans(res))
res <- round(res,2)
rownames(res) <- c(paste0("Fold_",1:length(kfolds)),'Average')
colnames(res)[-1] <- gsub('.Class','',colnames(res)[-1])
```

# Tm-Vec

```{r}
tm <- read.csv('Data/csv/tm-vec.csv')
nprotein <- nrow(tm)

for (i in 1:nprotein){
  if(i%%100==0)print(i)
  # ---- seq1
  seq = unlist(strsplit(tolower(tm$Seq_x[i]), ""))
  freq = c()
  for (j in 1:20){
  freq[j] <-  length(grep(letters_list[j],seq))
  }
  freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
  en_fr <- aaenergy * freq_matrix
  en_fr <- en_fr/length(seq)
  pair_es <- diag(freq) %*% as.matrix(en_fr)
  aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
  CEP_x <- c(length(seq),aa210_p)
  # ---- seq2
  seq = unlist(strsplit(tolower(tm$Seq_y[i]), ""))
  freq = c()
  for (j in 1:20){
  freq[j] <-  length(grep(letters_list[j],seq))
  }
  freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
  en_fr <- aaenergy * freq_matrix
  en_fr <- en_fr/length(seq)
  pair_es <- diag(freq) %*% as.matrix(en_fr)
  aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
  CEP_y <- c(length(seq),aa210_p)
  # ----------
  tm$CEP_dis[i] <- manhat(CEP_x,CEP_y)
  tm$CEP_dis_normLen[i] <- manhat(CEP_x/CEP_x[1],CEP_y/CEP_y[1])
}

tm$CEP_dis_MaxMin <- (tm$CEP_dis-min(tm$CEP_dis))/(max(tm$CEP_dis)-min(tm$CEP_dis))
tm$CEP_dis_normLen_MaxMin <- (tm$CEP_dis_normLen-min(tm$CEP_dis_normLen))/
  (max(tm$CEP_dis_normLen)-min(tm$CEP_dis_normLen))
tm$CEP_dis_MaxMin1 <- 1-(tm$CEP_dis_MaxMin)
tm$CEP_dis_normLen_MaxMin1 <- 1-(tm$CEP_dis_normLen_MaxMin)
tm$status <- if_else(tm$class_x==tm$class_y,'same class','different class')
tm$status2 <- if_else(tm$Subclass_x==tm$Subclass_y&tm$class_x=='Class_1'&tm$class_y=='Class_1','same class(sub class1)',NA)

saveRDS(tm,'Data/rds/bacteriocin_tm-vec.RDS')
# -----------------------------------------
# ----------- FIG8
tm <- readRDS('Data/rds/bacteriocin_tm-vec.RDS')
tm$CEP_dis_MaxMin1 <- tm$CEP_dis_MaxMin1-.36
tm2 <- tm[!is.na(tm$status2),-28]
colnames(tm2)[28]<-'status'
tm2 <- rbind(tm[,-29],tm2)



tm2 <- tm2[,c(5,20,21,6,26,28)]
colnames(tm2)[-6] <- c('Alphafold2','Omegafold','ESMfold','TM_vec','CEP')
tm2 <- melt(tm2,variable.name = 'method',value.name = 'score')
tm2$method <- factor(tm2$method,levels = c('TM_vec','Alphafold2','Omegafold','ESMfold','CEP'))
p1<-ggplot(tm2, aes(status,score,fill=method))+
  #geom_violin(trim = TRUE,
  #            alpha = 0.3)+
  geom_boxplot(outlier.shape = NA,
               color='black',
               size=rep(c(rep(.3,4),1),3),
               alpha=rep(c(rep(.4,4),1),3),
               width=.4
               #position = position_dodge(width = .9)
               )+
  coord_cartesian(ylim = c(0.1, 1))+
  xlab('')+
    geom_hline(yintercept = .5, linetype="dashed", 
             color = "red", size=.7)+
  mythem+
  #facet_wrap(~status)+
  theme(legend.title = element_blank(),
        axis.text.y = element_text(size=13),
        axis.text.x = element_text(color = 'black',size = 15, face = 'bold'),
        axis.title.y = element_text(size = 20,vjust = 2, face = 'bold'),
        legend.text = element_text(size = 15))+
  scale_fill_discrete(labels=c('TM_vec','Alphafold2',
                               'Omegafold','ESMfold',
                               expression(bold("CEP"))))

# ---------------------------------------
tm_unique <- tm[!duplicated(tm$ID_x),]
nprotein <- nrow(tm_unique)
tm_unique <- data.frame(tm_unique[,c(1,7,8,9,14)], 
                        length=nchar(tm_unique$Seq_x),
                        matrix(NA,nrow=nprotein,ncol=210))
colnames(tm_unique)[-c(1:6)] <- pair_inter

for (i in 1:nprotein){
  print(i)
  seq = unlist(strsplit(tolower(tm_unique$Seq_x[i]), ""))
  freq = c()
  for (j in 1:20){
  freq[j] <-  length(grep(tolower(letters_list[j]),seq))
  }
freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
en_fr <- aaenergy * freq_matrix
en_fr <- en_fr/length(seq)
pair_es <- diag(freq) %*% as.matrix(en_fr)
aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
tm_unique[i,7:216] <- aa210_p
}

df <- tm_unique[,-c(1:5)]
dis <- as.matrix(proxy::dist(df[,-1], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 100,
            min_dist = 0.1,
            input="dist")

ump_df <- data.frame(class=tm_unique$class_x,
                     subclass=tm_unique$Subclass_x,
                     UMAP1 = ump$layout[, 1],
                     UMAP2 = ump$layout[, 2])

p2<-ggplot(ump_df,
       aes(UMAP1, UMAP2, color= class)) +
  geom_point(alpha=.7, size=2) + 
  mythem+
  theme(legend.title = element_blank(),
        legend.position = 'bottom',
        axis.title = element_text(size = 15),
        axis.title.y = element_text(size = 15,vjust = 5),
        legend.text = element_text(size = 15))+
  scale_color_manual(values = c('green','magenta','blue'))


p1 <- ggarrange(NULL,NULL,
          NULL,p1,ncol = 2,nrow = 2,
          labels = c('','B'),
          heights = c(0.1,1),
          widths = c(.1,1),
          font.label = list(size = 50),
          hjust = 0,vjust = 1.2)
p2 <- ggarrange(NULL,NULL,
          NULL,p2,ncol = 2,nrow = 2,
          labels = c('','A'),
          heights = c(0.1,1),
          widths = c(.1,1),
          font.label = list(size = 50),
          hjust = .5,vjust = 1.2)

ggarrange(p2,NULL,p1,ncol = 3,nrow = 1,
          widths = c(.5,.05,1))
# save FIG8.pdf 15x8
# ------------------
################### total ###################
df2<-df
df2$total<-rowSums(df2[,-c(1)])
my_comparisons<-list(c('Class_1','Class_2'),
                     c('Class_2','Class_3'),
                     c('Class_1','Class_3'))

df2$total<-df2$total/1000
df2 <- data.frame(class=tm_unique$class_x,df2)
df2 $class <- factor(df2$class, levels = c('Class_1','Class_2','Class_3'))
p3<-ggboxplot(df2, x='class',y='total', 
          fill = 'class',alpha=.5,outlier.shape = NA)+
  stat_summary(fun.y=mean, geom="point", shape=8,
               size=2, color="red", fill="red")+xlab('')+
  theme_bw(base_line_size = .2)+mythem+ylab('Total energy')+
  theme(legend.position = 'none',
        axis.text.x = element_text(size=13),
        axis.text.y = element_text(size=13),
        axis.title.y = element_text(size=20))+
  ylim(-6.5,2)+
  scale_fill_viridis(discrete = TRUE, alpha=0.6)+
  stat_compare_means(comparisons = my_comparisons, method = 't.test',
                     label.y = c(-1,-.5,0),
                     step.increase = .03,tip.length = 0.005)+
  scale_fill_manual(values = c('green','magenta','blue'))

p23<-ggarrange(p2,NULL,p3,nrow = 1,
          widths = c(1, 0.1,1),    
          labels = c('A', "",'B'),
          font.label = list(size = 30),
          hjust = -.5,vjust = 1.2)

ggarrange(p23,NULL,p1,ncol = 1,
          heights = c(1, 0.2, 1),
          labels = c('', "C"),
          font.label = list(size = 30),
          hjust = -.5,vjust = 1.2)


```

# Figures

```{r}
# astral 95
df_seq <- readRDS("Data/rds/E210.seq_astral_95.rds") 
df_seq = df_seq[!is.na(df_seq$FL),] 

df_str <- readRDS("Data/rds/E210.astral_95.rds") 
df_str = df_str[!is.na(df_str$FL),] 
m<-match(df_str$scopeID,df_seq$scopeID)
df_str <- df_str[-which(is.na(m)),]
# **** total energy 95
df_tot95 <- data.frame(data='Astral 95', class=df_str$class,
                       length = df_seq$length,
                       seq=rowSums(df_seq[,-c(1:8)]),
                       str=rowSums(df_str[,-c(1:8)]))
# **** distance 95
dis_seq <- dist(df_seq[,-c(1:8)],method = manhat)
dis_str <- dist(df_str[,-c(1:8)],method = manhat)
idx <- sample(length(dis_str),100000)
df_dis95 <- data.frame(data='Astral 95',
                       dis_str=as.vector(dis_str)[idx],
                       dis_seq=as.vector(dis_seq)[idx])
saveRDS(df_dis95,'manuscript/fig_data/dis_95_manhat.rds')
# ************************************************************
# astral 40
df_seq <- readRDS("Data/rds/E210.seq_astral40.rds") 
df_seq = df_seq[!is.na(df_seq$FL),] 

df_str <- readRDS("Data/rds/E210.astral40.rds") 
df_str = df_str[!is.na(df_str$FL),] 
m<-match(df_str$scopeID,df_seq$scopeID)
df_str <- df_str[-which(is.na(m)),]
# **** total energy 40
df_tot40 <- data.frame(data='Astral 40', class=df_str$class,
                       length=df_seq$length,
                       seq=rowSums(df_seq[,-c(1:8)]),
                       str=rowSums(df_str[,-c(1:8)]))
# **** distance 40
dis_seq <- dist(df_seq[,-c(1:8)],method = manhat)
dis_str <- dist(df_str[,-c(1:8)],method = manhat)
idx <- sample(length(dis_str),100000)
df_dis40 <- data.frame(data='Astral 40',
                       dis_str=as.vector(dis_str)[idx],
                       dis_seq=as.vector(dis_seq)[idx])
saveRDS(df_dis40,'manuscript/fig_data/dis_40_manhat.rds')
# **************************************
df_tot <- rbind(df_tot40, df_tot95)
saveRDS(df_tot, 'manuscript/fig_data/df_total_energy.rds')
```

## Fig1

```{r}
# ************** total energy 40 and 95
df_tot <- readRDS('manuscript/fig_data/df_total_energy.rds')
df_tot$data <- ifelse(df_tot$data=='Astral 40','ASTRAL40','ASTRAL95')
AB <- ggplot(df_tot, aes(seq/2000,str/1000))+
       geom_pointdensity(adjust = .1,alpha=.3,show.legend = T)+
       geom_smooth(method = 'glm',col='red')+
       stat_cor(method = "pearson", size=5, color = "red", geom = "label") +
       scale_color_viridis()+facet_wrap(vars(data))+
       theme_bw(base_line_size = .3)+
       xlab('Total energy (sequence)')+ylab('Total energy (structure)')+
  mythem
# ************** distance 40 and 95
df_dis40<-readRDS('manuscript/fig_data/dis_40_manhat.rds')
df_dis95<-readRDS('manuscript/fig_data/dis_95_manhat.rds')
df_dis <- rbind(df_dis40, df_dis95)
df_dis$data <- ifelse(df_dis$data=='Astral 40','ASTRAL40','ASTRAL95')

CD <- ggplot(df_dis, aes(dis_seq,dis_str))+
       geom_pointdensity(adjust = .1,alpha=.3,show.legend = T)+
       geom_smooth(method = 'glm',col='red')+
       stat_cor(method = "pearson", size=5, color = "red", geom = "label") +
       scale_color_viridis()+facet_wrap(vars(data))+
       theme_bw(base_line_size = .3)+
       xlab('Distance (CPE)')+ylab('Distance (SPE)')+
       mythem

ggarrange(NULL,
    AB, NULL,CD, ncol = 1,
    heights = c(0.09, 1, 0.17, 1),
    labels = c('',"A", "", "B"),
    font.label = list(size = 30),    # color = "red"
    hjust = -.5,vjust = 0
    )

# save FIG1.pdf 10x9
```

## Fig2

```{r}
df_tot <- readRDS('manuscript/fig_data/df_total_energy.rds')
df_tot <- df_tot[df_tot$class%in%c('a','b','c','d'),]
df_tot$class[df_tot$class=='a']<-'All_alpha'
df_tot$class[df_tot$class=='b']<-'All_beta'
df_tot$class[df_tot$class=='c']<-'alpha/beta'
df_tot$class[df_tot$class=='d']<-'alpha+beta'
df_tot$data <- ifelse(df_tot$data=='Astral 40','ASTRAL40','ASTRAL95')

AB <- ggplot(df_tot,aes(class,str/length,fill=class))+
  geom_boxplot(outlier.alpha = .01)+
  facet_wrap(vars(data))+
  xlab('')+ylab('Total energy (structure)')+
  coord_cartesian(ylim = c(-15, -4))+
  mythem+theme(legend.title = element_blank(),
               axis.text.x = element_blank())

CD <- ggplot(df_tot,aes(class,seq/(2*length),fill=class))+
  geom_boxplot(outlier.alpha = .1)+
  facet_wrap(vars(data))+
  xlab('')+ylab('Total energy (sequence)')+
  coord_cartesian(ylim = c(-15, -6))+
  mythem+theme(legend.title = element_blank(),
               axis.text.x = element_blank())

#  ------------- add pValue
#  df_tot2 <- df_tot
#  df_tot2$str <- df_tot2$str/df_tot2$length
#  df_tot2$seq <- df_tot2$seq/(2*df_tot2$length)
#  
#  my_comparisons <- list(c('All_alpha','All_beta'),
#                         c('All_alpha','alpha/beta'),
#                         c('All_alpha','alpha+beta'),
#                         c('All_beta','alpha/beta'),
#                         c('All_beta','alpha+beta'),
#                         c('alpha/beta','alpha+beta'))
#  
#  AB <- ggboxplot(df_tot2, x='class',y='str', 
#                 fill = 'class',alpha=.8,outlier.shape = NA)+
#    theme_bw(base_line_size = .1)+facet_wrap(vars(data))+
#    xlab('')+ylab('Total energy (structure)')+
#    coord_cartesian(ylim = c(-14, -2))+
#    theme(axis.title = element_text(size=20),
#          axis.text.y = element_text(size=15),
#          axis.text.x = element_blank(),
#          strip.text = element_text(size = 20),
#          legend.title = element_blank(),
#          legend.text = element_text(size = 20),
#          panel.spacing = unit(2, "lines"))+
#    stat_compare_means(comparisons = my_comparisons,
#                       method = 't.test',
#                       size=7,label = "p.signif",
#                       label.y = c(-7.3,-6.8,-6.3,-5.8,-5.3,-4.8),
#                       tip.length = 0,vjust = .8,bracket.size = .1)
#  
#  
#  CD <- ggboxplot(df_tot2, x='class',y='seq', 
#            fill = 'class',alpha=.8,outlier.shape = NA)+
#    theme_bw(base_line_size = .1)+facet_wrap(vars(data))+
#    xlab('')+ylab('Total energy (structure)')+
#    coord_cartesian(ylim = c(-14, -4))+
#    theme(axis.title = element_text(size=20),
#          axis.text.y = element_text(size=15),
#          axis.text.x = element_blank(),
#          strip.text = element_text(size = 20),
#          legend.title = element_blank(),
#          legend.text = element_text(size = 20),
#          panel.spacing = unit(2, "lines"))+
#    stat_compare_means(comparisons = my_comparisons,
#                       method = 't.test',
#                       size=7,label = "p.signif",
#                       label.y = c(-7.4,-6.9,-6.4,-5.9,-5.4,-4.9),
#                       tip.length = 0,vjust = .8,bracket.size = .1)
#  

ggarrange(NULL,
    AB, NULL,CD, ncol = 1,
    heights = c(0.09, 1, 0.17, 1),
    labels = c('',"A", "", "B"),
    common.legend = T, legend = 'bottom',
    font.label = list(size = 30),    # color = "red"
    hjust = -.5,vjust = 0
    )
# save FIG2.pdf 10x9
```

## Fig5EF

```{r}
# ******** a.29.3
# ****** structure40
df <- readRDS("Data/rds/E210.astral40.rds") 
df = df[!is.na(df$FL),] 
idx<- df$superfamily%in% 'a.29.3'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$family[idx], dis)
fa<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(fa)) {  
  print(i)   
  x <- fa[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
df_family_str <- data.frame(data='ASTRAL40-superfamily a.29.3',type='SPE',
                            gr=rep(c('within','between'),
                                   c(length(within),length(between))),
                            distance=c(within,between)) 
# ****** sequence40
df <- readRDS("Data/rds/E210.seq_astral40.rds") 
df = df[!is.na(df$FL),] 
idx<- df$superfamily%in% 'a.29.3'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$family[idx], dis)
fa<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(fa)) {  
  print(i)   
  x <- fa[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
df_family_seq <- data.frame(data='ASTRAL40-superfamily a.29.3',type='CPE',
                            gr=rep(c('within','between'),
                                   c(length(within),length(between))),
                            distance=c(within,between)) 
# **********
df_family <- rbind(df_family_seq, df_family_str)

E <- ggboxplot(df_family, x='gr',y='distance', 
          fill = 'gr',alpha=.5,outlier.shape = NA)+
  stat_summary(fun.y=mean, geom="point", shape=8,
               size=2, color="red", fill="red")+xlab('')+
  mythem+
  theme(legend.position = 'none',
        axis.text.x = element_blank())+
  scale_fill_viridis(discrete = TRUE, alpha=0.6)+
  stat_compare_means(method = 't.test',size=3,label.y = 18,
                     label.x = 1.4,
                     aes(label = paste0("p = ", ..p.format..)))+
  facet_nested(~ data + type)+
  coord_cartesian(ylim = c(0, 20))
mylegend <- get_legend(E+
                       theme(legend.position = 'bottom',
                             legend.title = element_blank(),
                             legend.text = element_text(size = 20)))
# ******************************************
# ****** structure95
df <- readRDS("Data/rds/E210.astral_95.rds") 
df = df[!is.na(df$FL),] 
idx<- df$superfamily%in% 'a.29.3'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$family[idx], dis)
fa<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(fa)) {  
  print(i)   
  x <- fa[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
df_family_str <- data.frame(data='ASTRAL95-superfamily a.29.3',type='SPE',
                            gr=rep(c('within','between'),
                                   c(length(within),length(between))),
                            distance=c(within,between)) 
# ****** sequence95
df <- readRDS("Data/rds/E210.seq_astral_95.rds") 
df = df[!is.na(df$FL),] 
idx<- df$superfamily%in% 'a.29.3'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$family[idx], dis)
fa<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(fa)) {  
  print(i)   
  x <- fa[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
df_family_seq <- data.frame(data='ASTRAL95-superfamily a.29.3',type='CPE',
                            gr=rep(c('within','between'),
                                   c(length(within),length(between))),
                            distance=c(within,between)) 
# **********
df_family <- rbind(df_family_seq, df_family_str)

FF <- ggboxplot(df_family, x='gr',y='distance', 
          fill = 'gr',alpha=.5,outlier.shape = NA)+
  stat_summary(fun.y=mean, geom="point", shape=8,
               size=2, color="red", fill="red")+xlab('')+
  mythem+
  theme(legend.position = 'none',
    axis.text.x = element_blank())+
  scale_fill_viridis(discrete = TRUE, alpha=0.6)+
  stat_compare_means(method = 't.test',size=3,
                     label.y = 18,
                     label.x = 1.4,
                     aes(label = paste0("p = ", ..p.format..)))+
  facet_nested(~ data + type)+
  coord_cartesian(ylim = c(0, 20))
# ********
E<-ggarrange(
  NULL,E, ncol = 1,
  heights = c(0.07, 1),
  labels = c('', "E"),
  font.label = list(size = 30), # color = "red"
  hjust = -.5,vjust = 0
)
FF<-ggarrange(
  NULL,FF, ncol = 1,
  heights = c(0.07, 1),
  labels = c('', "F"),
  font.label = list(size = 30), # color = "red"
  hjust = -.5,vjust = 0
)
EF<-ggarrange(
    E, NULL,FF, nrow = 1,
    widths = c(1, 0.3, 1),
    labels = c("", "", "")
    )

```

## Fig5CD

```{r}
# ******** a.29
# ****** structure40
df <- readRDS("Data/rds/E210.astral40.rds") 
df = df[!is.na(df$FL),] 
idx<- df$fold%in% 'a.29'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$superfamily[idx], dis)
sf<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(sf)) {  
  print(i)   
  x <- sf[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
df_superfamily_str <- data.frame(data='ASTRAL40-fold a.29',type='SPE',
                            gr=rep(c('within','between'),
                                   c(length(within),length(between))),
                            distance=c(within,between)) 
# ****** sequence40
df <- readRDS("Data/rds/E210.seq_astral40.rds") 
df = df[!is.na(df$FL),] 
idx<- df$fold%in% 'a.29'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$superfamily[idx], dis)
sf<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(sf)) {  
  print(i)   
  x <- sf[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
df_superfamily_seq <- data.frame(data='ASTRAL40-fold a.29',type='CPE',
                            gr=rep(c('within','between'),
                                   c(length(within),length(between))),
                            distance=c(within,between)) 
# **********
df_superfamily <- rbind(df_superfamily_seq, df_superfamily_str)

C <- ggboxplot(df_superfamily, x='gr',y='distance', 
          fill = 'gr',alpha=.5,outlier.shape = NA)+
  stat_summary(fun.y=mean, geom="point", shape=8,
               size=2, color="red", fill="red")+xlab('')+
  mythem+
  theme(legend.position = 'none',
        axis.text.x = element_blank())+
  scale_fill_viridis(discrete = TRUE, alpha=0.6)+
  stat_compare_means(comparisons = my_comparisons, method = 't.test',
                     size=3,label.y = 17,
                     label.x = 1.4,bracket.size = NA,tip.length = NA)+
  facet_nested(~ data + type)+
  coord_cartesian(ylim = c(0, 20))
# ******************************************
# ****** structure95
df <- readRDS("Data/rds/E210.astral_95.rds") 
df = df[!is.na(df$FL),] 
idx<- df$fold%in% 'a.29'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$superfamily[idx], dis)
sf<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(sf)) {  
  print(i)   
  x <- sf[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
df_superfamily_str <- data.frame(data='ASTRAL95-fold a.29',type='SPE',
                            gr=rep(c('within','between'),
                                   c(length(within),length(between))),
                            distance=c(within,between)) 
# ****** sequence95
df <- readRDS("Data/rds/E210.seq_astral_95.rds") 
df = df[!is.na(df$FL),] 
idx<- df$fold%in% 'a.29'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$superfamily[idx], dis)
sf <- unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(sf)) {  
  print(i)   
  x <- sf[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
df_superfamily_seq <- data.frame(data='ASTRAL95-fold a.29',type='CPE',
                            gr=rep(c('within','between'),
                                   c(length(within),length(between))),
                            distance=c(within,between)) 
# **********
df_superfamily <- rbind(df_superfamily_seq, df_superfamily_str)

D <- ggboxplot(df_superfamily, x='gr',y='distance', 
          fill = 'gr',alpha=.5,outlier.shape = NA)+
  stat_summary(fun.y=mean, geom="point", shape=8,
               size=2, color="red", fill="red")+xlab('')+
  mythem+
  theme(legend.position = 'none',
        axis.text.x = element_blank())+
  scale_fill_viridis(discrete = TRUE, alpha=0.6)+
  stat_compare_means(comparisons = my_comparisons, method = 't.test',size=3,
                     label.x = 1.4,label.y = 17,bracket.size = NA,tip.length = NA)+
  facet_nested(~ data + type)+
  coord_cartesian(ylim = c(0, 20))
# ********
C<-ggarrange(
  NULL,C, ncol = 1,
  heights = c(0.07, 1),
  labels = c('', "C"),
  font.label = list(size = 30), # color = "red"
  hjust = -.5,vjust = 0
)
D<-ggarrange(
  NULL,D, ncol = 1,
  heights = c(0.07, 1),
  labels = c('', "D"),
  font.label = list(size = 30), # color = "red"
  hjust = -.5,vjust = 0
)
CD<-ggarrange(
    C, NULL,D, nrow = 1,
    widths = c(1, 0.3, 1),
    labels = c("", "", "")
    )
```

## Fig5AB

```{r}
my_comparisons<-list(c('within','between'))
# ******** a
# ****** structure40
df <- readRDS("Data/rds/E210.astral40.rds") 
df = df[!is.na(df$FL),] 
idx<- df$class%in% 'a'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$fold[idx], dis)
fo<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(fo)) {  
  print(i)   
  x <- fo[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
wit <- within[sample(length(within),1000)] 
bet <- between[sample(length(between),1000)] 
df_fold_str <- data.frame(data='ASTRAL40-class alpha',type='SPE',
                            gr=rep(c('within','between'),c(1000,1000)),
                            distance=c(wit,bet)) 
# ****** sequence40
df <- readRDS("Data/rds/E210.seq_astral40.rds") 
df = df[!is.na(df$FL),] 
idx<- df$class%in% 'a'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$fold[idx], dis)
fo <- unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(fo)) {  
  print(i)   
  x <- fo[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
wit <- within[sample(length(within),1000)] 
bet <- between[sample(length(between),1000)] 
df_fold_seq <- data.frame(data='ASTRAL40-class alpha',type='CPE',
                            gr=rep(c('within','between'),c(1000,1000)),
                            distance=c(wit,bet))
# **********
df_fold <- rbind(df_fold_seq, df_fold_str)

A <- ggboxplot(df_fold, x='gr',y='distance', 
          fill = 'gr',alpha=.5,outlier.shape = NA)+
  stat_summary(fun.y=mean, geom="point", shape=8,
               size=2, color="red", fill="red")+xlab('')+
  mythem+
  theme(legend.position = 'none',
        axis.text.x = element_blank())+
  scale_fill_viridis(discrete = TRUE, alpha=0.6)+
  stat_compare_means(comparisons = my_comparisons, method = 't.test',size=3,
                     label.y = 30,bracket.size = NA,tip.length = NA)+
  facet_nested(~ data + type)+
  coord_cartesian(ylim = c(0, 40))
# ******************************************
# ****** structure95
df <- readRDS("Data/rds/E210.astral_95.rds") 
df = df[!is.na(df$FL),] 
idx<- df$class%in% 'a'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$fold[idx], dis)
fo<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(fo)) {  
  print(i)   
  x <- fo[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
wit <- within[sample(length(within),1000)]
bet <- between[sample(length(between),1000)] 
df_fold_str <- data.frame(data='ASTRAL95-class alpha',type='SPE',
                            gr=rep(c('within','between'),c(1000,1000)),
                            distance=c(wit,bet))
# ****** sequence95
df <- readRDS("Data/rds/E210.seq_astral_95.rds") 
df = df[!is.na(df$FL),] 
idx<- df$class%in% 'a'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$fold[idx], dis)
fo<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(fo)) {  
  print(i)   
  x <- fo[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
wit <- within[sample(length(within),1000)] 
bet <- between[sample(length(between),1000)] 
df_fold_seq <- data.frame(data='ASTRAL95-class alpha',type='CPE',
                            gr=rep(c('within','between'),c(1000,1000)),
                            distance=c(wit,bet)) 
# **********
df_fold <- rbind(df_fold_seq, df_fold_str)

B <- ggboxplot(df_fold, x='gr',y='distance', 
          fill = 'gr',alpha=.5,outlier.shape = NA)+
  stat_summary(fun.y=mean, geom="point", shape=8,
               size=2, color="red", fill="red")+xlab('')+
  mythem+
  theme(legend.position = 'none',
        axis.text.x = element_blank())+
  scale_fill_viridis(discrete = TRUE, alpha=0.6)+
  stat_compare_means(comparisons = my_comparisons, method = 't.test',size=3,
                     label.y = 30, bracket.size = NA,tip.length = NA)+
  facet_nested(~ data + type)+
  coord_cartesian(ylim = c(0, 40))
# ********
A<-ggarrange(
  NULL,A, ncol = 1,
  heights = c(0.07, 1),
  labels = c('', "A"),
  font.label = list(size = 30), # color = "red"
  hjust = -.5,vjust = 0
)
B<-ggarrange(
  NULL,B, ncol = 1,
  heights = c(0.07, 1),
  labels = c('', "B"),
  font.label = list(size = 30), # color = "red"
  hjust = -.5,vjust = 0
)
AB<-ggarrange(
    A, NULL,B, nrow = 1,
    widths = c(1, 0.3, 1),
    labels = c("", "", "")
    )
```

##Fig5All

```{r}
f5all <- ggarrange(NULL,AB,CD,EF,ncol = 1,
          heights = c(0.1,1,1,1))
cowplot::plot_grid(f5all, mylegend, ncol = 1, rel_heights = c(1,.03))
# save pdf 10x9
```

## Fig3

```{r}
# ******** str40
df <- readRDS("Data/rds/E210.astral40.rds")
df <- df[df$class %in% c("a","b")& df$length > 100 & !is.na(df$FF),]
df <- df[,c(5,3,9:218)]
dis <- as.matrix(proxy::dist(df[,-1]/df$length, method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_str40 <- data.frame(data='Astral 40',type='Structure',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])
ump_df_str40$class <- ifelse(ump_df_str40$class=='a','All-alpha','All-beta')

# ******** seq40
df <- readRDS("Data/rds/E210.seq_astral40.rds")
df <- df[df$class %in% c("a","b")& df$length > 100 & !is.na(df$FF),]
df <- df[,c(5,3,9:218)]
dis <- as.matrix(proxy::dist(df[,-c(1)]/df$length, method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_seq40 <- data.frame(data='Astral 40',type='Sequence',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])
ump_df_seq40$class <- ifelse(ump_df_seq40$class=='a','All-alpha','All-beta')

# ******** str95
df <- readRDS("Data/rds/E210.astral_95.rds")
df <- df[df$class %in% c("a","b")& df$length > 100 & !is.na(df$FF),]
df <- df[,c(5,3,9:218)]
dis <- as.matrix(proxy::dist(df[,-c(1)]/df$length, method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_str95 <- data.frame(data='Astral 95',type='Structure',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])
ump_df_str95 <- ump_df_str95[ump_df_str95$UMAP1> -1.5 & ump_df_str95$UMAP1< 4.5,]
ump_df_str95 <- ump_df_str95[ump_df_str95$UMAP2> -1.5 & ump_df_str95$UMAP2< 6,]
ump_df_str95$class <- ifelse(ump_df_str95$class=='a','All-alpha','All-beta')

# ******** seq95
df <- readRDS("Data/rds/E210.seq_astral_95.rds")
df <- df[df$class %in% c("a","b")& df$length > 100 & !is.na(df$FF),]
df <- df[,c(5,3,9:218)]
dis <- as.matrix(proxy::dist(df[,-c(1)]/df$length, method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_seq95 <- data.frame(data='Astral 95',type='Sequence',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])
ump_df_seq95 <- ump_df_seq95[ump_df_seq95$UMAP1> -5 & ump_df_seq95$UMAP1< 1.5,]
ump_df_seq95 <- ump_df_seq95[ump_df_seq95$UMAP2> -2.5 & ump_df_seq95$UMAP2< 3,]
ump_df_seq95$class <- ifelse(ump_df_seq95$class=='a','All-alpha','All-beta')
# **** rbind
ump_df <- rbind(ump_df_str40,ump_df_seq40,
                ump_df_str95,ump_df_seq95)

save(ump_df_seq40,ump_df_seq95,ump_df_str40,ump_df_str95,file = 'manuscript/fig_data/UMPs.RData')
load('manuscript/fig_data/UMPs.RData')
ump_df_str40$type <- ifelse(ump_df_str40$type=='Structure','SPE','CPE')
ump_df_seq40$type <- ifelse(ump_df_seq40$type=='Structure','SPE','CPE')
ump_df_str95$type <- ifelse(ump_df_str95$type=='Structure','SPE','CPE')
ump_df_seq95$type <- ifelse(ump_df_seq95$type=='Structure','SPE','CPE')

ump_df_str40$data <- ump_df_seq40$data <- 'ASTRAL40'
ump_df_str95$data <- ump_df_seq95$data <- 'ASTRAL95'

p1<-ggplot(ump_df_str40,aes(UMAP1, UMAP2, color= class)) +
  geom_point(alpha=.7,size=.5) + 
  facet_grid(data ~ type)+xlab('')+ylab('')+
  mythem+scale_color_manual(values = c('green','magenta'))+
  theme(legend.title = element_blank(),
        legend.text = element_text(size = 18))+
  guides(color = guide_legend(override.aes = list(size = 5)))

p2<-ggplot(ump_df_seq40,aes(UMAP1, UMAP2, color= class)) +
  geom_point(alpha=.7,size=.5) + 
  facet_grid( ~ type)+xlab('')+
  #facet_grid(data ~ type)+xlab('')+ylab('')+
  mythem+scale_color_manual(values = c('green','magenta'))+
  theme(legend.title = element_blank(),
        legend.text = element_text(size = 18))+
  guides(color = guide_legend(override.aes = list(size = 5)))

p3<-ggplot(ump_df_str95,aes(UMAP1, UMAP2, color= class)) +
  geom_point(alpha=.7,size=.5) +
  facet_grid(rows = vars(data))+ylab('')+
  mythem+scale_color_manual(values = c('green','magenta'))+
  theme(legend.title = element_blank(),
        legend.text = element_text(size = 18))+
  guides(color = guide_legend(override.aes = list(size = 5)))


p4<-ggplot(ump_df_seq95,aes(UMAP1, UMAP2, color= class)) +
  geom_point(alpha=.7,size=.5) + 
  #facet_grid(rows = vars(data))+ylab('')+
  mythem+scale_color_manual(values = c('green','magenta'))+
  theme(legend.title = element_blank(),
        legend.text = element_text(size = 18))+
  guides(color = guide_legend(override.aes = list(size = 5)))

# ********
ggarrange(NULL,NULL,NULL,
          p2,NULL,p1,
          NULL,NULL,NULL,
          p4,NULL,p3,ncol = 3,nrow = 4,
          common.legend = T,
          labels = c('','','',
                     "A", "", "B",
                     '','','',
                     'C','','D'),
          font.label = list(size = 30),
          widths = c(1, 0.12, 1),
          heights = c(.1, 1, 0.12, 1),
          hjust = -.5,vjust = 0,legend = 'bottom')

# save FIG3.pdf 10x10
```

## Fig4A

```{r}
# ******** str40
df <- readRDS("Data/rds/E210.astral40.rds")
df <- df[df$fold%in%c('a.100','a.104'),]
df <- df[!is.na(df$FF),]
df <- df[,c(6,3,9:218)]
df$length <- as.numeric(df$length)
dis <- as.matrix(proxy::dist(df[,-1], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_str <- data.frame(data='Astral 40',type='SPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])

# ******** seq40
df <- readRDS("Data/rds/E210.seq_astral40.rds")
df <- df[df$fold%in%c('a.100','a.104'),]
df <- df[!is.na(df$FF),]
df <- df[,c(6,3,9:218)]
df$length <- as.numeric(df$length)
dis <- as.matrix(proxy::dist(df[,-1], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_seq <- data.frame(data='Astral 40',type='CPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])
z <- rbind(ump_df_str,ump_df_seq)
z$data <- 'ASTRAL40'
z$class <- ifelse(z$class=='a.100', 'a.100    ','a.104')

p1 <- ggplot(z,aes(UMAP1, UMAP2, color= class)) +
    geom_point(alpha=.7) + 
    facet_nested(~ data+type,independent = T,scales = T)+
  ylab('')+xlab('')+
  mythem+
  theme(axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 18))+
  guides(color = guide_legend(override.aes = list(size = 3)))
# ---------------------
# ******** str95
df <- readRDS("Data/rds/E210.astral_95.rds")
df <- df[df$fold%in%c('a.100','a.104'),]
df <- df[!is.na(df$FF),]
df <- df[,c(6,3,9:218)]
df$length <- as.numeric(df$length)
dis <- as.matrix(proxy::dist(df[,-1], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_str <- data.frame(data='Astral 95',type='SPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])

# ******** seq95
df <- readRDS("Data/rds/E210.seq_astral_95.rds")
df <- df[df$fold%in%c('a.100','a.104'),]
df <- df[!is.na(df$FF),]
df <- df[,c(6,3,9:218)]
df$length <- as.numeric(df$length)
dis <- as.matrix(proxy::dist(df[,-1], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_seq <- data.frame(data='Astral 95',type='CPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])
z <- rbind(ump_df_str,ump_df_seq)
z$data <- 'ASTRAL95'
z$class <- ifelse(z$class=='a.100', 'a.100    ','a.104')


p2 <- ggplot(z,aes(UMAP1, UMAP2, color= class)) +
    geom_point(alpha=.7) + 
    facet_nested(~ data+type,independent = T,scales = T)+
  ylab('')+xlab('')+
  mythem+
  theme(axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 18))+
  guides(color = guide_legend(override.aes = list(size = 3)))

A<-ggarrange(NULL,NULL,NULL,NULL,
             NULL,p1,NULL,p2,nrow = 2,ncol = 4,
          common.legend = T,legend = 'right',
          widths = c(.1,1, 0.0, 1),
          heights = c(.07,1),
          labels = c('','','','',
                     '',"A", "", ""),
  font.label = list(size = 30),
  hjust = .2,vjust = 0)

```

## Fig4B

```{r}
# ******** str40
df <- readRDS("Data/rds/E210.astral40.rds")
df <- df[df$superfamily%in%c('a.29.2','a.29.3'),]
df <- df[!is.na(df$FF),]
df <- df[,c(7,3,9:218)]
df$length <- as.numeric(df$length)
dis <- as.matrix(proxy::dist(df[,-1], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_str <- data.frame(data='Astral 40',type='SPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])

# ******** seq40
df <- readRDS("Data/rds/E210.seq_astral40.rds")
df <- df[df$superfamily%in%c('a.29.2','a.29.3'),]
df <- df[!is.na(df$FF),]
df <- df[,c(7,3,9:218)]
df$length <- as.numeric(df$length)
dis <- as.matrix(proxy::dist(df[,-1], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_seq <- data.frame(data='Astral 40',type='CPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])
z <- rbind(ump_df_str,ump_df_seq)
z$data <- 'ASTRAL40'
z$class <- ifelse(z$class=='a.29.2', 'a.29.2   ','a.29.3')

p1 <- ggplot(z,aes(UMAP1, UMAP2, color= class)) +
    geom_point(alpha=.7) + 
    facet_nested(~ data+type,independent = T,scales = T)+
  ylab('')+xlab('')+
  mythem+
  theme(axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 18))+
  guides(color = guide_legend(override.aes = list(size = 3)))
# ---------------------
# ******** str95
df <- readRDS("Data/rds/E210.astral_95.rds")
df <- df[df$superfamily%in%c('a.29.2','a.29.3'),]
df <- df[!is.na(df$FF),]
df <- df[,c(7,3,9:218)]
df$length <- as.numeric(df$length)
dis <- as.matrix(proxy::dist(df[,-1], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_str <- data.frame(data='Astral 95',type='SPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])

# ******** seq95
df <- readRDS("Data/rds/E210.seq_astral_95.rds")
df <- df[df$superfamily%in%c('a.29.2','a.29.3'),]
df <- df[!is.na(df$FF),]
df <- df[,c(7,3,9:218)]
df$length <- as.numeric(df$length)
dis <- as.matrix(proxy::dist(df[,-1], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_seq <- data.frame(data='Astral 95',type='CPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])
z <- rbind(ump_df_str,ump_df_seq)
z$data <- 'ASTRAL95'
z$class <- ifelse(z$class=='a.29.2', 'a.29.2   ','a.29.3')

p2 <- ggplot(z,aes(UMAP1, UMAP2, color= class)) +
    geom_point(alpha=.7) + 
    facet_nested(~ data+type,independent = T,scales = T)+
  ylab('')+xlab('')+
  mythem+
  theme(axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 18))+
  guides(color = guide_legend(override.aes = list(size = 3)))

B <- ggarrange(NULL,NULL,NULL,NULL,
             NULL,p1,NULL,p2,nrow = 2,ncol = 4,
          common.legend = T,legend = 'right',
          widths = c(.1,1, 0.0, 1),
          heights = c(.07,1),
          labels = c('','','','',
                     '',"B", "", ""),
  font.label = list(size = 30),
  hjust = .2,vjust = 0)


```

## Fig4C

```{r}
# ******** str40
df <- readRDS("Data/rds/E210.astral40.rds")
df <- df[df$family%in%c('a.25.1.0','a.25.1.2'),]
df <- df[!is.na(df$FF),]
df <- df[,c(8,3,9:218)]
df$length <- as.numeric(df$length)
dis <- as.matrix(proxy::dist(df[,-1], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 10,
            min_dist = 0.1,
            input="dist")
ump_df_str <- data.frame(data='Astral 40',type='SPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])

# ******** seq40
df <- readRDS("Data/rds/E210.seq_astral40.rds")
df <- df[df$family%in%c('a.25.1.0','a.25.1.2'),]
df <- df[!is.na(df$FF),]
df <- df[,c(8,3,9:218)]
df$length <- as.numeric(df$length)
dis <- as.matrix(proxy::dist(df[,-1], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 10,
            min_dist = 0.1,
            input="dist")
ump_df_seq <- data.frame(data='Astral 40',type='CPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])
z <- rbind(ump_df_str,ump_df_seq)
z$data <- 'ASTRAl40'

p1 <- ggplot(z,aes(UMAP1, UMAP2, color= class)) +
    geom_point(alpha=.7) + 
    facet_nested(~ data+type,independent = T,scales = T)+
  ylab('')+xlab('')+
  mythem+
  theme(axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 18))+
  guides(color = guide_legend(override.aes = list(size = 3)))
# ---------------------
# ******** str95
df <- readRDS("Data/rds/E210.astral_95.rds")
df <- df[df$family%in%c('a.25.1.0','a.25.1.2'),]
df <- df[!is.na(df$FF),]
df <- df[,c(8,3,9:218)]
df$length <- as.numeric(df$length)
dis <- as.matrix(proxy::dist(df[,-1], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 10,
            min_dist = 0.1,
            input="dist")
ump_df_str <- data.frame(data='Astral 95',type='SPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])

# ******** seq95
df <- readRDS("Data/rds/E210.seq_astral_95.rds")
df <- df[df$family%in%c('a.25.1.0','a.25.1.2'),]
df <- df[!is.na(df$FF),]
df <- df[,c(8,3,9:218)]
df$length <- as.numeric(df$length)
dis <- as.matrix(proxy::dist(df[,-1], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 10,
            min_dist = 0.1,
            input="dist")
ump_df_seq <- data.frame(data='Astral 95',type='CPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])
z <- rbind(ump_df_str,ump_df_seq)
z$data <- 'ASTRAL95'

p2 <- ggplot(z,aes(UMAP1, UMAP2, color= class)) +
    geom_point(alpha=.7) + 
    facet_nested(~ data+type,independent = T,scales = T)+
  ylab('')+xlab('')+
  mythem+
  theme(axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 18))+
  guides(color = guide_legend(override.aes = list(size = 3)))

C <- ggarrange(NULL,NULL,NULL,NULL,
             NULL,p1,NULL,p2,nrow = 2,ncol = 4,
          common.legend = T,legend = 'right',
          widths = c(.1,1, 0.0, 1),
          heights = c(.07,1),
          labels = c('','','','',
                     '',"C", "", ""),
  font.label = list(size = 30),
  hjust = .2,vjust = 0)
```

##Fig4ABC

```{r}
ggarrange(NULL,A,B,C,ncol = 1,
          heights = c(0.07,1,1,1))
# save FIG4.pdf 10x10
```

## Fig6

```{r}
E210.seq_spike200 <- readRDS("Data/rds/E210.seq_spike200.rds")
#-----------close
df = E210.seq_spike200
df = df[!is.na(df$length),]
df = df[df$conformation == "Close",]

rownames(df) = df$pdbID
x = df[,-c(1:4)]/df$length
x  <- as.matrix(x) 
hc <- x %>% proxy::dist(method = manhat)
hc <- hc%>% hclust(method = "ward.D2")
dend <- hc %>% as.dendrogram

#  labcol <- c( "blue", "green","red")
#  ord_sml <- df[,][order.dendrogram(dend),3]
#  group_cols <- labcol[as.factor(ord_sml)]
#  
#  dend2 <- dend %>%
#    color_labels(col =   group_cols) %>%
#    set("labels_cex", .8) 
#  
#  circlize_dendrogram(dend2)
#  par( mar= c(5,10,8,3) )
#  plot(dend2, #ylab="Height", 
#       main='Coronaviruse Spike Clustering Based On CPE')
#  title(ylab="Height", line=3, cex.lab=3)
#  legend("topright", legend = as.character(unique(ord_sml)),
#         pch = 15, pt.cex = 2.3, cex = 1.4, bty = 'n',
#         inset = c(-.5, -.5), y.intersp=0.15, x.intersp = .2,
#         col = unique(group_cols),horiz = FALSE,xpd = TRUE)
#  

df <- df[with(df, order(df$virus)), ]
row.names(df) <- df$pdbID
rn <- df$pdbID
tree_data <- as.phylo(dend)
grp <- list(MERS_Cov     = rn[1:32],
            SARS_Cov  = rn[33:63],
            SARS_Cov2  = rn[64:143])

p_data <- ggtree(tree_data, layout = 'circular',
                 #branch.length='none',
                 size = 1) 

t_CPE = groupOTU(p_data, grp, "groups") + aes(color=groups) +
  theme(legend.position="none",
        plot.title = element_text(size = 17, hjust = 0.5,vjust = 5))+
  geom_tiplab2(aes(label=label,angle=angle),
               hjust=-.1,size=5.5,fontface="bold",align=F)+
  scale_color_discrete(breaks=c('MERS_Cov', 'SARS_Cov', 'SARS_Cov2'),
                      type =  c('black',"blue","green","red" ))+
  ggtitle('Coronaviruse Spike Clustering Based On CPE')

# -----------------------
E210_spike200 <- readRDS("Data/rds/E210_spike200.rds")
#-----------close
df = E210_spike200
df = df[!is.na(df$length),]
df = df[df$conformation == "Close",]

rownames(df) = df$pdbID
x = df[,-c(1:4)]/df$length
x  <- as.matrix(x) 
hc <- x %>% proxy::dist(method = manhat)
hc <- hc%>% hclust(method = "ward.D2")
dend <- hc %>% as.dendrogram

df <- df[with(df, order(df$virus)), ]
row.names(df) <- df$pdbID
rn <- df$pdbID

tree_data <- as.phylo(dend)
grp <- list(MERS_Cov     = rn[1:32],
            SARS_Cov  = rn[33:63],
            SARS_Cov2  = rn[64:143])

p_data <- ggtree(tree_data, layout = 'circular',
                 #branch.length='none',
                 size = 1) 

t_SPE = groupOTU(p_data, grp, "groups") + aes(color=groups) +
  theme(legend.position="none",
        plot.title = element_text(size = 17, hjust = 0.5,vjust = 5))+
  geom_tiplab2(aes(label=label,angle=angle),
               hjust=-.1,size=5.5,fontface="bold",align=F)+
  scale_color_discrete(breaks=c('MERS_Cov', 'SARS_Cov', 'SARS_Cov2'),
                      type =  c('black',"blue","green","red"  ))+
  ggtitle('Coronaviruse Spike Clustering Based On SPE')

mylegend <- get_legend(t_SPE+
                       theme(legend.position = 'bottom',
                             legend.title = element_blank(),
                             legend.text = element_text(size = 20)))


AB<-ggarrange(NULL,
    t_CPE, NULL,t_SPE, ncol = 1,nrow = 4,
    heights = c(.15, 1,.1, 1),
    #widths = c(1,.1,1),
    labels = c('',"A", "", "B"),
    font.label = list(size = 55), # color = "red"
    hjust = -10,vjust = 0
    )
cowplot::plot_grid(AB, mylegend, ncol = 1, rel_heights = c(1,.1))
# save FIG6.pdf 13x8
```

## Fig8

```{r}
# ********************************************
# ------------------------------
# --------- go to the 'Tm-Vec' chunk
# ------------------------------
# ********************************************
df <- readRDS("Data/rds/E210.seq_bacteriocin.rds") 
df <- df[df$length>30,]
######## umap
dis <- as.matrix(proxy::dist(df[,-c(1,2)], method = manhat))
ump <- umap(d = dis, input='dist',
            n_neighbors = 70,
            min_dist = 0.5)
ump_df <- data.frame(df[,1:2],
                     UMAP1 = ump$layout[, 1],
                     UMAP2 = ump$layout[, 2])

p2 <- ggplot(ump_df,aes(UMAP1, UMAP2, color= class)) +
  geom_point(alpha=.7, size=1) + 
  mythem+
  theme(legend.title = element_blank())+
  scale_color_manual(values = c('green','magenta','blue'))
# ----------------------------------------------
 dis <- as.matrix(dist(df[,-c(1,2)],method = manhat))
 dis <- data.frame(gr=df$class, dis)
 cl<-unique(dis$gr)
 
 my_comparisons<-list(c('within','between'))
 between<-c() 
 within<-c()
 for (i in 1:length(cl)) {  
   print(i)   
   x <- cl[i]  
   idx <- which(dis$gr==x)
   y <- dis[idx,(idx+1)]
   within <- c(within, y[lower.tri(y)])
   x <- dis[idx,-c(1,idx+1)]   
   x <- as.vector(as.matrix(x))  
   x <- x[!is.na(x)]   
   between<-c(between,x) 
 }
 df_same_diff <- data.frame(gr=rep(c('within','between'),
                                c(length(within),length(between))),
                         distance=c(within,between)) 
 df_same_diff$dis_norm <- (df_same_diff$distance - min(df_same_diff$distance))/
   (max(df_same_diff$distance)-min(df_same_diff$distance))
 
 p1<-ggboxplot(df_same_diff, x='gr',y='distance', 
           fill = 'gr',alpha=.5,outlier.shape = NA)+
   stat_summary(fun.y=mean, geom="point", shape=8,
                size=2, color="red", fill="red")+xlab('')+
   theme_bw(base_line_size = .2)+mythem+
   theme(legend.position = 'none',
         axis.text.x = element_text(size=13),
         axis.text.y = element_text(size=13),
         axis.title.y = element_text(size=20))+
   scale_fill_viridis(discrete = TRUE, alpha=0.6)+
   coord_cartesian(ylim = c(0, 40))+
   stat_compare_means(comparisons = my_comparisons, method = 't.test',
                      label.y = 35,tip.length = 0.01)
 
################ total ###################
df2<-df
df2$total<-rowSums(df2[,-c(1,2)])
my_comparisons<-list(c('class1','class2'),
                     c('class2','class3'),
                     c('class1','class3'))

df2$total<-df2$total/1000
p3<-ggboxplot(df2, x='class',y='total', 
          fill = 'class',alpha=.5,outlier.shape = NA)+
  stat_summary(fun.y=mean, geom="point", shape=8,
               size=2, color="red", fill="red")+xlab('')+
  theme_bw(base_line_size = .2)+mythem+ylab('Total energy')+
  theme(legend.position = 'none',
        axis.text.x = element_text(size=13),
        axis.text.y = element_text(size=13),
        axis.title.y = element_text(size=20))+
  ylim(-6.5,2)+
  scale_fill_viridis(discrete = TRUE, alpha=0.6)+
  stat_compare_means(comparisons = my_comparisons, method = 't.test',
                     label.y = c(-1,-.5,0),
                     step.increase = .03,tip.length = 0.005)+
  scale_fill_manual(values = c('green','magenta','blue'))


p13<-ggarrange(p1,NULL,p3,nrow = 1,
          widths = c(1, 0.12, 1),
          labels = c("B", "", "C"),
    font.label = list(size = 30),
    hjust = -.5,vjust = 0)


 ggarrange(p2,NULL,p13,ncol = 1,
           heights = c(1, 0.1, 1),
           labels = c('A', ""),
           font.label = list(size = 30),
           hjust = -.5,vjust = 1.2)
```

##Fig7

```{r}
E210_ferritin <- readRDS("Data/rds/E210_ferritin.rds")
df <- E210_ferritin
df$group[41:45]<-'Rubrerthrin'

df$family[df$group%in%c('Bac','Fer','Dps')]<-'a.25.1.1'
df$family[!df$group%in%c('Bac','Fer','Dps')]<-'a.25.1.2'
df$family[41:45]<-'Rubrerthrin'

df <- df[with(df, order(df$family)), ]
row.names(df) <- df[,1]
rn <- df[,1]
dis <- proxy::dist(df[,-c(1:3,214)], method = manhat)
tree_data <- ape::bionj(dis)
grp <- list(Bacteri     = rn[1:5],
            Dps  = rn[6:20],
            Ferritins  = rn[21:31],
            BMM_alpha = rn[32:33],
            BMM_beta  = rn[34:36],
            Fatty_acid  = rn[37:40],
            RNRR2  = rn[41:48],
            Rubrerthrin = rn[49:53])

p_data <- ggtree(tree_data, layout = 'circular',
                 branch.length='none',size = 1.5) 

t = groupOTU(p_data, grp, "groups") + aes(color=groups) +
  theme(legend.position="right",
        legend.title = element_blank(),
        legend.text = element_text(size = 15, colour = "black"))+
  geom_tiplab2(aes(label=label,angle=angle),
               hjust=-.1,size=5,fontface="bold",align=TRUE)+
  scale_color_discrete(breaks=c('Bacteri', 'Dps', 'Ferritins',
                                'BMM_alpha','BMM_beta','Fatty_acid','RNRR2','Rubrerthrin'),
                      type =  c("#F8766D","#CD9600","#7CAE00","#00BE67","#00BFC4",
                                                           "#00A9FF","#FF61CC","#C77CFF" ))

                      
#g <- ggplot_build(t)
#cols<-unlist(unique(g$data[[1]]["colour"]))
nnet <- neighborNet(dis)
ggsplitnet(nnet) + geom_tiplab2(aes(label=label),
                                hjust=-.1,size=5,fontface="bold")
write.nexus.networx(nnet,'Data/rds/E210_ferritin_manhat.nxs')
###############
dis <- data.frame(gr=df$group, as.matrix(dis))
fa<-unique(dis$gr)

my_comparisons<-list(c('within','between'))
between<-c() 
within<-c()
for (i in 1:length(fa)) {  
  print(i)   
  x <- fa[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
df_family <- data.frame(gr=rep(c('within','between'),c(length(within),length(between))),
                        distance=c(within,between)) 

ggboxplot(df_family, x='gr',y='distance', 
          fill = 'gr',alpha=.5,outlier.shape = NA)+
  stat_summary(fun.y=mean, geom="point", shape=8,
               size=2, color="red", fill="red")+xlab('')+
  theme_bw(base_line_size = .2)+mythem+
  theme(axis.text.x = element_blank())+
  scale_fill_viridis(discrete = TRUE, alpha=0.6)+
  stat_compare_means(comparisons = my_comparisons, method = 't.test',size=10)

###########

df2<-df
df2$total<-rowSums(df2[,-c(1:3,214)])
df2$group<-factor(df2$group,levels = c('Bac','Dps','Fer','Rubrerthrin',
                                       'BMA','BMB','FAA','RNR'))
ggplot(df2, aes(x=group,y=total, fill=group))+
  geom_boxplot(size=.2,outlier.alpha = .5)+
  stat_summary(fun.y=mean, geom="point", shape=8,
               size=2, color="red", fill="red")+xlab('')+
  theme_bw(base_line_size = .2)+mythem+ylab('Total energy')+
  theme(legend.position = 'none',
        axis.text.x = element_blank())+
  scale_fill_manual(values = c("#F8766D" ,"#00BE67", "#00A9FF"  ,"#C77CFF",
                    "#CD9600", "#7CAE00" ,"#00BFC4" ,"#FF61CC"))

```

##Fig9

```{r}
library(rentrez)
get_protids_from_geneids <- function(gene_ids) {
  # Get the protein elink.
  protein_elink <- rentrez::entrez_link(dbfrom = "gene", id = gene_ids, db = "protein")
  # Get the protein id (refseq database)
  protein_ids <- protein_elink$links$gene_protein_refseq
  protein_ids
}

make_aa_fasta <- function(prot_ids) { #, nameFile
  # Make a multi-fasta file with each protein id
  sapply(prot_ids, function(x) {
    protein_esummary <- rentrez::entrez_summary(db = "protein", id = x)
    protein_fasta <- rentrez::entrez_fetch(db = "protein", id = protein_esummary$uid, rettype = "fasta")
    return(protein_fasta)
  } )
}

seq2df<-function(seq){
  x<-strsplit(seq,'\n')
  head<-x[[1]][1]
  seq<-paste(x[[1]][-c(1,10)],collapse = '')
  return(c(head,seq))
}


dti <- read_excel('Data/csv/41467_2019_9186_MOESM4_ESM.xlsx')
colnames(dti)<-c('drugID','entrezID')
dti$entrezID<-as.character(dti$entrezID)

sab<-readRDS('Data/rds/sAB.rds')
colnames(sab)[1:2]<-c('d1','d2')

d65<-unique(c(sab$d1,sab$d2))
df_seq<-c()
for(i in 1:65){
  df<-dti[dti$drugID%in%d65[i],]
  df2<-c()
  for (j in 1:nrow(df)) {
    print(paste0(i,', ',j,'/',nrow(df)))
    prot_ids <- get_protids_from_geneids(df$entrezID[j])
    z<-make_aa_fasta(prot_ids)
    zz<-data.frame(prot=df$entrezID[j],t(sapply(z, seq2df)))
    df2<-rbind(df2,zz)
  }
  df2<-data.frame(drug=d65[i],df2)
  df_seq<-rbind(df_seq,df2)
}

df_seq$length <- nchar(df_seq$seq_aa)
colnames(df_seq)[3:4]<-c('header_seq','seq')
# -------- Energy
letters_list <- toupper(letters_list)
E210.seq_dti <- data.frame(df_seq,matrix(NA,nrow(df_seq),210))
colnames(E210.seq_dti)[-c(1:5)] <- pair_inter
for (NP in 1:nrow(df_seq)) {
  if(NP%%10==0)print(NP)
  seq = unlist(strsplit(unlist(E210.seq_dti$seq_aa[NP]),''))
    freq = c()
    for (j in 1:20){
        freq[j] <-  length(grep(letters_list[j],seq))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
    E210.seq_dti[NP,-c(1:5)] <- aa210_p
}
saveRDS(E210.seq_dti,'Data/rds/E210.seq_dti.rds')
################################
sab <- readRDS('Data/rds/sAB.rds')
colnames(sab)[1:2] <- c('d1','d2')
E210.seq_dti<-readRDS('Data/rds/E210.seq_dti.rds')

df0 <- E210.seq_dti 
df <- cbind(df0[,1:5],df0[,-c(1:5)]/df0$length) 
#df <- df %>%   group_by(drug) %>%  dplyr::summarise(across(X1:X210, mean))
df2 <- aggregate(.~drug+entrezID,df[,-c(3:5)],mean)
drugs<-unique(df2$drug)
#########################
my_sab<-matrix(0,65,65)
for (i in 1:65) {
  print(i)
  for (j in 1:65) {
    if(i>=j){
      x<-drugs[i]
      y<-drugs[j]
      zx<-df2[df2$drug%in%x,]
      zy<-df2[df2$drug%in%y,]
      xnames<-paste0(zx$drug,'_',zx$entrezID)
      ynames<-paste0(zy$drug,'_',zy$entrezID)
      z<-rbind(zx,zy)
      dis<-as.matrix(dist(z[,-c(1:4)],method = manhat))
      colnames(dis)<-rownames(dis)<-paste0(z$drug,'_',z$entrezID)
      diag(dis)<-NA
      dis[is.nan(dis)] <- 0
      disx<-dis[xnames,xnames]
      dxx<-mean(diag(as.matrix(disx[apply(disx,2,which.min),])))
      disy<-dis[ynames,ynames]
      dyy<-mean(diag(as.matrix(disy[apply(disy,2,which.min),])))
      disxy<-dis[xnames,ynames]
      disyx<-dis[ynames,xnames]
      dxy<-mean(c(diag(as.matrix(disxy[apply(disxy,2,which.min),])),
                  diag(as.matrix(disyx[apply(disyx,2,which.min),]))))
      my_sab[i,j]<-dxy-((dxx+dyy)/2)
      my_sab[j,i]<-my_sab[i,j]
    }
  }
}
diag(my_sab)<-NA
colnames(my_sab)<-rownames(my_sab)<-drugs

my_sab <- as.data.frame(as.table(my_sab)) 
colnames(my_sab) <- c("d1", "d2", "my_sAB") 
my_sab <- my_sab[!is.na(my_sab$my_sAB), ]
##############
df3 <- left_join(sab,my_sab,c('d1','d2'))
saveRDS(df3,'Data/rds/df3_sAB.rds')


p1<-ggplot(df3,aes(sAB,my_sAB))+
  geom_point(alpha=.5)+geom_smooth(method = 'lm')+
  stat_cor(col='red',size=8)+mythem+
    theme(axis.title = element_text(size=20),
        axis.text.x = element_text(size = 13),
        axis.text.y = element_text(size = 13),)+
  xlab('sAB (PPI)')+ylab('sAB (CPE)')
# save FIG9.pdf 10x7

df3$sAB_bin <- cut(df3$sAB, breaks = 18, labels = paste0('Bin ',1:18))
mids <- round(hist(df3$sAB, breaks = 18, plot = FALSE)$mids,1)
mids <- data.frame(sAB_bin = paste0('Bin ',1:18),mids)
df3 <- left_join(df3,mids,by = 'sAB_bin')
df3$mids <- factor(df3$mids)

p2<-ggplot(df3,aes(mids,my_sAB))+
  geom_boxplot(alpha=.5)+geom_smooth(method = 'lm', aes(group=1))+
  stat_cor(col='red',size=8, aes(group=1))+mythem+
  #theme()+
  xlab('sAB (PPI)')+ylab('sAB (CPE)')


ggarrange(p1,p2,ncol = 1)
```