---
title: "Protein Similarity Based on Energy Comparison"
output: 
  pdf_document:
    fig_caption: yes
fontsize: 11pt
geometry: margin=1in
graphics: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,warning=FALSE, fig.pos= "h")
```

# Library   
```{r libraries and reading files, include=FALSE}
library(kableExtra)
options(knitr.table.format = "latex")
library(readr)
library(data.table)
library(MASS)
library(tidyverse)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(ggpointdensity)
library(ggpubr)
library(igraph)
library(fpc)
library(grid)
library(gridExtra)
library(reshape2)
library(visNetwork)
library(pheatmap)
library(umap)
library(Rtsne)
library(mixOmics)
library(viridis)
library(viridisLite)
library(fgsea)
library(bio3d)
library(caret)
library(e1071)
library(randomForest)
library(seqinr)
library(ShortRead)
library(geometry)
library(ape)
library(colorspace)
library(dendextend)
library(dynamicTreeCut)
library(caret)
library(ggtree)
library(phangorn)
library(tanggle)
library(stringr)
library(protr)
library(proxy)
mythem <- theme_bw(base_line_size = .2)+
  theme(axis.text = element_text(size=12, face = "bold"),
        axis.title = element_text(size = 15, face = "bold"),
        legend.text = element_text(size = 15, colour = "black"),
        legend.title = element_blank(),
        title = element_text(size = 15)
        )
# ************************************
# *********************************
manhat_norm <- function(X, Y){
  x = X[-1];y = Y[-1]
  xy = abs(x - y)
  xy = xy[xy != 0]
  d = sum(xy)/length(xy)
  Lmax = max(X[1], Y[1])
  d = d / (0.3 * (Lmax^0.6))
  return(d)
}
```

## Lenght_dependent

```{r}
pc25 <- read.csv("Data/csv/cullpdb_pc25.0_res0.0-1.5_noBrks_len40-10000_R0.25_Xray_d2023_10_23_chains3040", sep="")
pc25 <- pc25[nchar(pc25$PDBchain)<6,]
rownames(pc25) <- pc25$PDBchain
pairs <- data.frame(t(combn(pc25$PDBchain, 2)))
pairs$len1 <- pc25[pairs$X1,'len']
pairs$len2 <- pc25[pairs$X2,'len']
pairs$min_len <- apply(pairs[,3:4],1,min)
pairs$max_len <- apply(pairs[,3:4],1,max)
pairs$mean_len <- apply(pairs[,3:4],1,mean)

set.seed(123)
pairs_random <- pairs %>%
  group_by(max_len) %>%
  sample_n(1) %>%
  ungroup()

energy_dell_dunbrack <- read.csv("Data/csv/energy.csv",header = FALSE,sep ="" )
energy_dell_dunbrack <- energy_dell_dunbrack[,5:33]
source("Functions/amacid2num.R")
source("Functions/atomtype2num167.R")
source("Functions/Energy_PC210.R")
source("Functions/NetworkFrame.R")

pairs_random$dis <- 0
for(i in 179:nrow(pairs_random)){
  print(i)
  pdb1 <- read.cif(substr(pairs_random$X1[i],1,4))
  pdb2 <- read.cif(substr(pairs_random$X2[i],1,4))
  ch1 <- unique(substr(pairs_random$X1[i],5,5))
  ch2 <- unique(substr(pairs_random$X2[i],5,5))
  
  sele1 <- atom.select(pdb1,"protein",type = "ATOM", chain=ch1)
  sele2 <- atom.select(pdb1, "protein", elety=c("H","OXT"),inverse=TRUE)
  sele4 <- atom.select(pdb1, "noh")
  sele <- combine.select(sele1, sele2,sele4, operator="AND")
  pdb1 <- trim.pdb(pdb1, sele,verbose = FALSE)
  
  sele1 <- atom.select(pdb2,"protein",type = "ATOM", chain=ch2)
  sele2 <- atom.select(pdb2, "protein", elety=c("H","OXT"),inverse=TRUE)
  sele4 <- atom.select(pdb2, "noh")
  sele <- combine.select(sele1, sele2,sele4, operator="AND")
  pdb2 <- trim.pdb(pdb2, sele,verbose = FALSE)
  
  E210_2pdb <- data_frame(p1=rep(0,210),p2=rep(0,210))
  
  Net_Frame <- NetworkFrame(PDB = pdb1)
  energy210 <- Energy_PC210(Net_Frame ,energy_dell_dunbrack)
  E210_2pdb$p1 <- energy210
  
  Net_Frame <- NetworkFrame(PDB = pdb2)
  energy210 <- Energy_PC210(Net_Frame, energy_dell_dunbrack)
  E210_2pdb$p2 <- energy210
  
  df <- data.frame(t(E210_2pdb))
  pairs_random$dis[i] <- dist(df,method = manhat_norm)
}

lm(log10(dis)~log10(max_len),pairs_random)
ggplot(pairs_random[pairs_random$max_len>80,],aes(max_len,dis))+geom_point()+
    scale_y_log10()+scale_x_log10()+
  geom_smooth(method = 'lm')+
  stat_cor(col='red')
lm(log10(dis)~log10(max_len),pairs_random[pairs_random$max_len>80,])

#step <- 5
#sel_pairs_random <- pairs_random[pairs_random$max_len %in% seq(40, nrow(pairs_random), by = step), ]

pairs_random$adj <- apply(pairs_random[,c(6,8)],1,function(x){return(x[2]/(.30*x[1]^0.60))})
ggplot(pairs_random[pairs_random$max_len>80,],aes(max_len,adj))+geom_point()+
    scale_y_log10()+scale_x_log10()+
  geom_smooth(method = 'lm')+
  stat_cor(col='red')

```



## 1t56-3loc

```{r}
pdb0<-read.pdb('1t56')
################ select domain
ch<-'A'
pdb0 <- trim.pdb(pdb0,atom.select(pdb0, chain=ch))
s <- 22
e <- 94
infoatom <- pdb0$atom
infoatom <- infoatom[!duplicated(infoatom$resno),]
resno <- infoatom$resno[infoatom$resno%in%c(s:e)]
sele <- atom.select(pdb0,resno=resno)
pdb_1t56 <- trim.pdb(pdb0, inds=sele)
################ select domain
pdb0<-read.pdb('3loc')
ch<-'A'
pdb0 <- trim.pdb(pdb0,atom.select(pdb0, chain=ch))
s <- 11
e <- 85
infoatom <- pdb0$atom
infoatom <- infoatom[!duplicated(infoatom$resno),]
resno <- infoatom$resno[infoatom$resno%in%c(s:e)]
sele <- atom.select(pdb0,resno=resno)
pdb_3loc <- trim.pdb(pdb0, inds=sele)
#---------------------------------
E210_2pdb <- data_frame(d3locA1=rep(0,210),d1t56A1=rep(0,210))
E210_2pdb_SS <- data_frame(d3locA1=rep(0,210),d1t56A1=rep(0,210),)
E210_2pdb_seq <- data_frame(d3locA1=rep(0,210),d1t56A1=rep(0,210),)

new_pdb <- pdb_3loc
    pdbX <- new_pdb$atom
    Net_Frame <- NetworkFrame(PDB = new_pdb)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210_2pdb$d3locA1 <- energy210
    # ****************keep res helix & sheet************************
    indhs <- inedx_HelixSheet(PDB = new_pdb)
    Net_Frame_hs <- Net_Frame[Net_Frame$resnoif %in% indhs & Net_Frame$resnojf %in% indhs,]
    if(dim(Net_Frame_hs)[1]!=0){
        Net_Frame <- Net_Frame_hs
    }
    energy210_ss <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210_2pdb_SS$d3locA1 <- energy210_ss
    #-----------sequence enregy based on PDB--------
    pdbXca <- pdbX[pdbX$elety == "CA",]
    seq = aa321(pdbXca$resid)
    freq = c()
    for (j in 1:20){
        freq[j] <-  length(grep(letters_list[j],tolower(seq)))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
    E210_2pdb_seq$d3locA1 <- aa210_p
    
v210to20 <- function(x){
  mat<-matrix(0,20,20)
  mat[lower.tri(mat,diag = T)]<-x
  mat<-t(mat)
  mat[lower.tri(mat,diag = T)]<-x
  return(colSums(mat))
}

df<-data.frame(apply(E210_2pdb,2,v210to20))
cor.test(df$d3locA1,df$d1t56A1)
plot(df$d3locA1,df$d1t56A1)
```



## Scope 2.08
```{r}
astrals_95 <- read_table("Data/csv/astral-scopedom-seqres-gd-sel-gs-bib-95-2.08.fa", col_names = FALSE)
astrals_95_seq <- readFASTA("Data/csv/astral-scopedom-seqres-gd-sel-gs-bib-95-2.08.fa")

rows_keep <- grep(">", astrals_95$X1)
astrals_95 <- astrals_95[rows_keep,c(1,2,3)]
colnames(astrals_95) <- c("scopeID","ID_scope","Position")
astral_95 <- data.frame(
    scopeID = character(dim(astrals_95)[1]),
    pdbID = character(dim(astrals_95)[1]),
    Position = character(dim(astrals_95)[1]),
    seq = t(data.frame(astrals_95_seq)))
row.names(astral_95) <- c(1:dim(astral_95)[1])
astral_95$scopeID <- substr(astrals_95$scopeID, start = 2, stop = nchar(astrals_95$scopeID))
astral_95$pdbID <- substr(astrals_95$scopeID, start = 3, stop = 6)
#----------Split scopeID into class, fold, super and family---------
split_values <- strsplit(astrals_95$ID_scope, "\\.")
matrix_values <- matrix("", nrow = nrow(astrals_95), ncol = 4)
for (i in 1:4) {
  matrix_values[, i] <- sapply(split_values, function(x) paste(x[1:i], collapse = "."))
}
df_split <- as.data.frame(matrix_values)
colnames(df_split) <- c("class", "fold", "superfamily", "family")
#-------------------------------------------------------------------
# ------------split Position, start and end--------------------------
astral_95$Position <- gsub("[()]", "", astrals_95$Position)
#-----------------------------------------------------------------
astral_95 <- cbind(astral_95,df_split)
astral_95 <- astral_95[,c("scopeID","pdbID" ,"Position", "class","fold","superfamily","family","seq")]
astral_95$seq <- toupper(astral_95$seq)
saveRDS(astral_95, file = "Data/rds/astral_95.rds")
```

## DD

```{r}
scope_2.08 <- read.delim("Data/csv/dir.cla.scope.2.08-stable.txt", 
                         header=FALSE, comment.char="#")
scope_2.08 <- scope_2.08[,c(1:4)]
colnames(scope_2.08) <- c("scopeID","pdb","Position","ID_scope")
DD_seq <- readFASTA("Data/csv/DDtest.fasta")

dd_train <- read.csv("Data/csv/DDtest.fasta", header=FALSE, sep=";")
pdbid <- fold_name <- character(0)
for (line in dd_train$V1) {
  if (grepl("^FOLD", line)) {
    # Extract fold name
    current_fold <- line
    } else if (grepl("^>", line)) {
    # Extract pdbid and sequence
    parts <- strsplit(line, "\\s+")
    current_pdbid <- gsub(">", "", parts[[1]][1])  # Remove ">" symbol
    # Append to the vectors
    pdbid <- c(pdbid, current_pdbid)
    fold_name <- c(fold_name, current_fold)
  }
  }
df <- data.frame(pdbid, fold_name,scopeID=character(length(fold_name)),
                 seq = t(data.frame(DD_seq)))
for (i in 1:dim(df)[1]){
    df$scopeID[i] <- paste0("d", tolower(df$pdbid[i]))
    }

dx = scope_2.08[scope_2.08$scopeID %in% df$scopeID,]
DD <- merge(df,scope_2.08,"scopeID")
#----------Split scopeID into class, fold, super and family---------
split_values <- strsplit(DD$ID_scope, "\\.")
matrix_values <- matrix("", nrow = nrow(DD), ncol = 4)
for (i in 1:4) {
  matrix_values[, i] <- sapply(split_values, function(x) paste(x[1:i], collapse = "."))
  }
df_split <- as.data.frame(matrix_values)
colnames(df_split) <- c("class", "fold", "superfamily", "family")
#-------------------------------------------------------------------
DD$pdbID =DD$pdb
DD <- cbind(DD,df_split)
DD = DD[,c("scopeID","pdbID" ,"Position", "class","fold","superfamily","family","seq")]
saveRDS(DD,"Data/rds/DD_test.rds")
```


## EDD

```{r}
scope_2.08 <- read.delim("Data/csv/dir.cla.scope.2.08-stable.txt", 
                         header=FALSE, comment.char="#")
colnames(scope_2.08) <- c("scopeID","pdb","Position","ID_scope","description")
EDD <- read.csv("Data/csv/EDD.fasta", header=FALSE, sep=";")
EDD_seq <- readFASTA("Data/csv/EDD.fasta")

scopeID <- fold_name <- ID_scope <- Position <- character(0)
for (line in EDD$V1) {
  if (grepl("^FOLD", line)) {
    # Extract fold name
    current_fold <- line
  } else if (grepl("^>", line)) {
    # Extract scopeID and sequence
    res <- str_match(line, ">(\\S+)(?:_)?\\s(\\S+)\\s\\(([^)]+)\\)")[, 2:4]
    # Append to the vectors
    scopeID <- c(scopeID, res[1])
    ID_scope <- c(ID_scope,res[2])
    Position <- c(Position,res[3])
    fold_name <- c(fold_name, current_fold)
  }
}
EDD <- data.frame(scopeID, fold_name,ID_scope,Position, seq = t(data.frame(EDD_seq)))
#----------Split scopeID into class, fold, super and family---------
split_values <- strsplit(EDD$ID_scope, "\\.")
matrix_values <- matrix("", nrow = nrow(EDD), ncol = 4)
for (i in 1:4) {
  matrix_values[, i] <- sapply(split_values, function(x) paste(x[1:i], collapse = "."))
  }
df_split <- as.data.frame(matrix_values)
colnames(df_split) <- c("class", "fold", "superfamily", "family")
#-------------------------------------------------------------------
EDD$pdbID =substr(EDD$scopeID, start = 2, stop = 5)
EDD <- cbind(EDD,df_split)
EDD = EDD[,c("scopeID","pdbID" ,"Position", "class","fold","superfamily","family","seq")]
saveRDS(EDD,"Data/rds/EDD.rds")
```

## TG 

```{r}
scope_2.08 <- read.delim("Data/csv/dir.cla.scope.2.08-stable.txt", 
                         header=FALSE, comment.char="#")
colnames(scope_2.08) <- c("scopeID","pdb","Position","ID_scope","description")
TG0 <- read.csv("Data/csv/TG.fasta", header=FALSE, sep=";")
TG_seq <- readFASTA("Data/csv/TG.fasta")

scope <-  character(0)
for (line in TG0$V1) {
  if (grepl("^>", line)) {
    parts <- strsplit(line, "\\s+")
    scp <-  gsub("__", "a_", parts[[1]][2])
    scope <- c(scope,scp)
  }
}

TG_all <- data.frame(scopeID = scope, seq = t(data.frame(TG_seq)))
row.names(TG_all) <- scope

TG1 <- merge(TG_all,scope_2.08,"scopeID")
rem_tg <- TG_all[!(TG_all$scopeID %in% scope_2.08$scopeID),]
scope <-  character(0)
scope_or <- character(0)
for (i in 1:dim(rem_tg)[1]){
  sci <- rem_tg$scopeID[i]
  scope <- c(scope,gsub("_(1|2|3)$", "a\\1", sci))
  scope_or <- c(scope_or,sci)
}
TG_1 <- data.frame(scopeID = scope,seq = TG_all[TG_all$scopeID %in% scope_or,2])
TG2 <- merge(TG_1,scope_2.08,"scopeID")

rem_tg <- TG_1[! (TG_1$scopeID %in% scope_2.08$scopeID),]
scope <-  character(0)
scope_or <-  character(0)
for (i in 1:nrow(rem_tg)){
  sci <- rem_tg$scopeID[i]
  scope <- c(scope,gsub("a_$", "a1", sci))
  scope_or <- c(scope_or,sci)
}
TG_2 <- data.frame(scopeID = scope,seq = TG_all[TG_all$scopeID %in% scope_or,2])
TG3 <- merge(TG_2,scope_2.08,"scopeID")

rem_tg <- TG_2[! (TG_2$scopeID %in% scope_2.08$scopeID),]
  
TG <- rbind(TG1,TG2,TG3)
#----------Split scopeID into class, fold, super and family---------
split_values <- strsplit(TG$ID_scope, "\\.")
matrix_values <- matrix("", nrow = nrow(TG), ncol = 4)
for (i in 1:4) {
  matrix_values[, i] <- sapply(split_values, function(x) paste(x[1:i], collapse = "."))
  }
df_split <- as.data.frame(matrix_values)
colnames(df_split) <- c("class", "fold", "superfamily", "family")
#-------------------------------------------------------------------
TG$pdbID =substr(TG$scopeID, start = 2, stop = 5)
TG <- cbind(TG,df_split)
TG = TG[,c("scopeID","pdbID" ,"Position", "class","fold","superfamily","family","seq")]
saveRDS(TG,"Data/rds/TG.rds")
```



## Energy_Calculator
```{r}
energy_dell_dunbrack <- read.csv("Data/csv/energy.csv",header = FALSE,sep ="" )
energy_dell_dunbrack <- energy_dell_dunbrack[,5:33]
source("Functions/amacid2num.R")
source("Functions/atomtype2num167.R")
source("Functions/Energy_PC210.R")
source("Functions/NetworkFrame.R")
source("Functions/inedx_HelixSheet.R")
source("Functions/split_position.R")

astral95 <- readRDS("Data/rds/astral_95.rds")
Nprotein <- nrow(astral95)
# ------Pairwise Amino Acid names-------------
AA <- c("PHE","LEU","ILE","VAL","TRP","TYR","MET","CYS","HIS","THR",
        "ARG","ALA","ASN","GLN","PRO","SER","ASP","GLY","LYS","GLU")
letters_list <- aa321(AA)
pair_inter <- character(0)
for (i in 1:20) {
  for (j in i:20) {
    pair_inter <- c(pair_inter, paste0(letters_list[i], letters_list[j]))
  }
} 
aaenergy <- read.csv("Data/csv/Pij.csv", header = F,sep = ";")
aaenergy <- (aaenergy + t(aaenergy))/2
colnames(aaenergy) = rownames(aaenergy)<- 
  c("F","L","I","V","W","Y","M","C","H","T","R","A","N","Q","P","S","D","G","K","E")
letters_list <- tolower(row.names(aaenergy))
aa210<-aaenergy[lower.tri(aaenergy,diag = T)]
#----------------------------------------------
num_columns <- 210
E210.astral95 <- data.frame(pdbID = astral95$pdbID,
                            scopeID= astral95$scopeID,
                            length = rep(NA,Nprotein),
                            position = astral95$Position,
                            class = astral95$class,
                            fold = astral95$fold,
                            superfamily = astral95$superfamily,
                            family = astral95$family,
                            matrix(NA, ncol = num_columns, nrow = length(astral95$pdbID)))
colnames(E210.astral95)[9:218] = pair_inter
E210.astral95_SS <- E210.seq_astral95 <- E210.seqPDB_astral95 <- E210.astral95 
for (NP in 1:Nprotein) {
#  if (NP %% 50 == 0) {
   print(NP)
#  }
  chain_resno <- split_position(astral95$Position[NP])
  pdbname <- astral95$pdbID[NP]
  tryCatch({
    pdb0 <- read.pdb(pdbname,verbose = FALSE)
    pdblist = list()
    for (ind in 1:dim(chain_resno)[1]){
      sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=chain_resno$chain[ind])
      sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
      sele4 <- atom.select(pdb0, "noh")
      sele <- combine.select(sele1, sele2,sele4, operator="AND")
      if (!(chain_resno$start[ind]=="NA") & !(chain_resno$end[ind])=="NA"){
         st <- chain_resno$start[ind]
         en <- chain_resno$end[ind]
         sele3 <- atom.select(pdb0,  resno=st:en, verbose=FALSE)
         sele <- combine.select(sele1, sele2,sele3,sele4, operator="AND")
      }
      pdblist[[ind]] <- trim.pdb(pdb0, sele,verbose = FALSE)
    }
    new_pdb = pdblist[[1]]
    if (dim(chain_resno)[1] > 1) {
      for (ind in 2:dim(chain_resno)[1]){
      new_pdb <- cat.pdb(new_pdb, pdblist[[ind]], rechain=FALSE)
      }
    }
    pdbX <- new_pdb$atom
    E210.astral95$length[NP] = length(unique(pdbX$resno))
    E210.astral95_SS$length[NP] = length(unique(pdbX$resno))
    E210.seqPDB_astral95$length[NP] = length(unique(pdbX$resno))
    Net_Frame <- NetworkFrame(PDB = new_pdb)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210.astral95[NP,9:218] <- energy210
    # ****************keep res helix & sheet************************
    indhs <- inedx_HelixSheet(PDB = new_pdb)
    Net_Frame_hs <- Net_Frame[Net_Frame$resnoif %in% indhs & Net_Frame$resnojf %in% indhs,]
    if(dim(Net_Frame_hs)[1]!=0){
        Net_Frame <- Net_Frame_hs
    }
    energy210_ss <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210.astral95_SS[NP,9:218] <- energy210_ss
    #-----------sequence enregy based on PDB--------
    pdbXca <- pdbX[pdbX$elety == "CA",]
    seq = aa321(pdbXca$resid)
    E210.seqPDB_astral95$length[NP] = length(seq)
    freq = c()
    for (j in 1:20){
        freq[j] <-  length(grep(letters_list[j],tolower(seq)))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
    E210.seqPDB_astral95[NP,9:218] <- aa210_p
#-----------sequence enregy based on astral--------
    seq = astral95$seq[NP]
    seq = gsub("X", "", seq)
    E210.seq_astral95$length[NP] = nchar(seq)
    seq = unlist(strsplit(seq,""))
    freq = c()
    for (j in 1:20){
        freq[j] <-  length(grep(letters_list[j],tolower(seq)))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
    E210.seq_astral95[NP,9:218] <- aa210_p
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}
saveRDS(E210.astral95,file = "Data/rds/E210.astral_95.rds")
saveRDS(E210.astral95_SS,file = "Data/rds/E210.astral_95_SS.rds")
saveRDS(E210.seq_astral95,file = "Data/rds/E210.seq_astral_95.rds")
saveRDS(E210.seqPDB_astral95,file = "Data/rds/E210.seqPDB_astral_95.rds")
```

## UMAP_manhat_norm

```{r}
df <- readRDS("data/rds/E210.astral40.rds")
dfs <- df[df$class %in% c("a","b")& df$length > 100 & !is.na(df$FF),]
df1<-dfs[,-c(1:2,4:5,7:8)]
df1[,-1] <- df1[,-1]
#df1$total = rowSums(df1[,-c(1,2)])
df1<-data.frame(class=substr(df1$fold,1,1),df1)
#df1 = df1[c(sample(2743:5697,500),sample(2642,500),sample(5698:10159,500)),]
#df1 <- df1 %>% sample_frac(1)
#df1 <- df1[ !is.na(df1$FF),]

#df1 = df1 %>% group_by(fold) %>% 
#  dplyr::summarise(across(colnames(df1)[-c(1,2)], mean))
#df1<-data.frame(class=substr(df1$fold,1,1),df1)
dis <- as.matrix(proxy::dist(df1[,-c(1,3)], method = manhat_norm))
ump <- umap(d = dis,
            n_neighbors = 50,
            min_dist = 0.01,
            input="dist")
ump_df <- data.frame(df1[,1:2],
                     UMAP1 = ump$layout[, 1],
                     UMAP2 = ump$layout[, 2])
ggplot(ump_df,aes(UMAP1, UMAP2, color= class)) +
   geom_point(alpha=.7) + 
   mythem
#------------------tamam--------------------------
tm_score1 <- function(x, y){
  z= x - y
    d <- sum(1/(1 + z))
  return(d)
}
```

## UMAP at class level
```{r}
df <- readRDS("data/rds/E210.astral40.rds")
df = E210.astral_95
df = E210.aspartaze_SS
df = E210.seq_astral40
sel1<-c('a.39.1','a.45.1')
sel1<-c('a.29.2','a.29.3')
sel1<-c('a.102.1','a.116.1')

sel<-c('b.82.1','b.82.3')    #  'b.82.2'

sel1<-c('c.1.10','c.1.8')
sel1<-c('c.1.10','b.82.1')

dfs <- df[df$superfamily %in% sel1,]

dfs <- df[df$class %in% c("a","b")& df$length > 100 & !is.na(df$FF),]
df1<-dfs[,-c(1:5,7:8)]
df1[,-1] <- df1[,-1]/dfs$length
df1<-data.frame(class=substr(df1$fold,1,1),df1)
#df1$total = rowSums(df1[,-c(1,2)])
ggplot(df1,aes(class,total))+geom_boxplot()
df1 <- df1 %>% sample_frac(1)
df2 <- df1[c(1:2000),]

df1<-df1 %>% 
  group_by(fold) %>%
  dplyr::summarise(across(colnames(df1)[-1], mean))


#pc <- prcomp(df1,scale. = F,center = T) 
Rot_E_abcd <- as.matrix(df1[,-c(1:2)]) #%*% pc$rotation[,1:5] 
E_PC_abcd <- data.frame(class = df1$class,Rot_E_abcd) 
# ------------- UMAP ------------------------
ump.E_PC_abcd <- umap(E_PC_abcd[,-c(1)], n_neighbors = 20,
            min_dist = 0.001, random_state = 123)
#-----------customized UMAP-----------
tm_score <- function(x, y){
    d <- (sum(1/(1 + (x - y))))
  return(d)
}
tm_score2 <- function(x, y){
  xy<-rep(0,210)
  for(i in 1:210){
    if(x[i]<y[i]){
      xy[i]<-x[i]-y[i]
    }else{
      xy[i]<-y[i]-x[i]
    }
  }
  d <- (sum(1/(1 + xy)))
  return(d)
}
tm_score3 <- function(x, y){
  if (sum(x) > sum(y)){
    d <- (sum(1/(1 + (x - y))))
  }
  else{
    d <- (sum(1/(1 + ( -x + y))))
  }
    
  return(d)
}


dis <- as.matrix(dist(df1[,-c(1,2)], method = tm_score3))
table(df1$class)
a<-dis[1:1610,1:1610]
a<-a[lower.tri(a)]
b<-dis[1611:3770,1611:3770]
b<-b[lower.tri(b)]
ab<-dis[1:1610,1611:3770]
ab<-as.vector(ab)
df_dis<-data.frame(gr=rep(c('a','b','ab'),c(length(a),length(b),length(ab))),
                   dis=c(a,b,ab))
ggplot(df_dis,aes(gr,dis))+geom_boxplot()+ylim(-100,100)

umap_result <- umap(d = dis,
                    n_neighbors = 30,
                    min_dist = 0.01)
umap.df.E_PC_abcd <- data.frame(
  df1[,1:2],
  UMAP1 = umap_result$layout[, 1],
  UMAP2 = umap_result$layout[, 2])

 ggplot(umap.df.E_PC_abcd,aes(UMAP1, UMAP2, color= class))+
  geom_point(alpha=.7,show.legend = F) + 
  mythem

#-------- UMAMP Visualization ---------------
umap.df.E_PC_abcd <- data.frame(
  df1[,1:2],
  UMAP1 = ump.E_PC_abcd$layout[, 1],
  UMAP2 = ump.E_PC_abcd$layout[, 2])

 ggplot(umap.df.E_PC_abcd,aes(UMAP1, UMAP2, color= class))+
  geom_point(alpha=.7) +
  mythem

 
 um.E_PC_abcd <- ggplot(umap.df.E_PC_abcd,aes(UMAP1, UMAP2, color= sf))+
  geom_point(alpha=.7) +
  mythem
#-------------PCA 
 pc <- prcomp(df1,scale. = F,center = T)
pc.df <- data.frame(group=df$family, pc$x)
g.pc <- ggplot(pc.df, aes(x=PC1, y=PC2, col=group))+
  geom_point(size=2)+
  theme_bw(base_line_size = .3)+
  #stat_ellipse(aes(fill = group), geom="polygon",type = "norm",alpha=0.2)+
  mythem
 
```
## LDA

```{r}
library(MASS)
# Example data
X <- data.frame(Feature1 = c(1, 2, 3, 4, 5),
                Feature2 = c(2, 3, 4, 5, 6))
Y <- factor(c("ClassA", "ClassA", "ClassB", "ClassB", "ClassC"))
lda_model <- lda(X, Y)
new_data <- data.frame(Feature1 = c(2.5), Feature2 = c(4.5))
predictions <- predict(lda_model, newdata = new_data)
train_predictions <- predict(lda_model)
ind <- sample(2, nrow(df), replace = T, prob = c(0.8, 0.2))


df <- readRDS("data/rds/E210.astral40.rds")
df <- df[ !is.na(df$FF) & df$length > 100 & df$class %in% c("a","b"),c(colnames(df)[1:8],im$int[1:210])]
train <- df[ind==1,-c(1:4,6:8)]
test <- df[ind==2,-c(1:4,6:8)]
train[,-1] = train[,-1]/df$length
lda_model <- lda(class~.,train)
p1 <- predict(lda_model, train)$class
tab <- table(Predicted = p1, Actual = train$class)
tab
sum(diag(tab))/sum(tab)

p1 <- predict(lda_model, test)$class
tab <- table(Predicted = p1, Actual = test$class)
tab
sum(diag(tab))/sum(tab)
```



## Random forest

```{r}
set.seed(7)
df <- readRDS("data/rds/E210.astral40.rds")
df <- df[ !is.na(df$FF) & df$length > 100 & df$class %in% c("a","b"),]
df[,-c(1:8)] = df[,-c(1:8)]/df$length
df$total = rowSums(df[,-c(1:8)])

#. & df$class %in% c("a","b","c","d")
data<- cbind(class = as.factor(df$class),df[,-seq(1:8)])  # just need to change the class
table(data$class)
rf = randomForest(class~.,data = data, mtry = 14,ntree = 500, proximity = T)
varImpPlot(rf,
           sort = T,
           n.var = 10,
           main = "Top 10 - Variable Importance")

im<-data.frame(importance(rf))
im$int<-rownames(im)
im<-im[order(im$MeanDecreaseGini,decreasing = T),]
result_tune = c()
for (i in seq(100,1000,100)){
  print(i)
  bestmtry <- data.frame(tuneRF(data[,-1],data$class, stepFactor=2, improve=1e-7, ntree=i, mtryStart = 14))
  bestmtry$nTree<-i
  result_tune = rbind(result_tune,bestmtry)
}
rf = randomForest(class~.,data = data, mtry = 4,ntree = 400, proximity = T)
 p3 <- predict(rf, independent_test)
 confusionMatrix(p3, independent_test$class)
varImpPlot(rf,
           sort = T,
           n.var = 10,
           main = "Top 10 - Variable Importance")
 MDSplot(rf, train$class)
 # *****************  Applying k-Fold Cross Validation
kfolds <- createFolds(data$class, k = 10)
res<-c()
for (i in 1:10) {
  print(i)
  x <- kfolds[[i]]
  tr <- data[-x,]
  te <- data[x,]
  rf = randomForest(class~.,data = tr, mtry = 4,ntree = 400)
  p2 <- predict(rf, te)
  cm2 <- confusionMatrix(p2, te$class)
  Accuracy <- c(cm2$overall['Accuracy'], F1=cm2$byClass['F1'])
  res<-rbind(res,Accuracy)
}

hist(treesize(rf),
     main = "No. of Nodes for the Trees",
     col = "green")
varImpPlot(rf,
           sort = T,
           n.var = 10,
           main = "Top 10 - Variable Importance")
importance(rf)

MDSplot(rf, data$class)
MDSplot(bestmtry, test$class)

```

## Tree

```{r}
# Best result with SS and pc = 100, ward.D
df = E210.aspartaze_SS
pc <- prcomp(df[,-seq(1:10)],scale. = F,center = T) 
Rot_E_abcd <- as.matrix(df[,-seq(1:10)]) %*% pc$rotation[,1:20] 
E_PC_abcd <- data.frame(class = df$class,Rot_E_abcd) 
E_PC_abcd <- umap.df.E_PC_abcd
x  <- E_PC_abcd[,-1] %>% as.matrix
rownames(x) <- df$pdbID
hc <- x %>% dist(method = 'manhattan') %>% hclust(method = "ward.D2")
 dfx = data.frame(x)

dend <- hc %>% as.dendrogram

labcol <- c( "blue", "red")
ord_sml <- E_PC_abcd[,][order.dendrogram(dend),1]
group_cols <- labcol[as.factor(ord_sml)]

dend2 <- dend %>%
  #color_branches(k=3, col = c("red","blue","green3")) %>%
  color_labels(col =   group_cols) %>%
  set("labels_cex", 0.5) 
plot(dend2, ylab = "Height")



df <- E210.aspartaze_SS
df <- df[with(df, order(df$family)), ]
row.names(df) <- paste0(df[,1],df[,2])
# PCA
#pc <- prcomp(df[,-seq(1:3)],scale. = F,center = T) 
#Rot_ferr <- as.matrix(df[,-seq(1:3)]) %*% pc$rotation[,1:5]
#dis <- dist(Rot_ferr, method="manhattan")
#dis <- log2(dis)
#------------------
rn <- paste0(df[,1],df[,2])
dis <- dist(df[,-c(1:10)], method="manhattan")
dis <- sqrt(dis)
#------------------
tree_data <- ape::bionj(dis)
grp <- list(A0     = rn[1:16],
            A1 = rn[17:49],
            A2  = rn[50:53])

p_data <- ggtree(tree_data, layout = 'circular',
                 branch.length='none',size = 1.5) 

t = groupOTU(p_data, grp, "groups") + aes(color=groups) +
  theme(legend.position="right",
        legend.title = element_blank(),
        legend.text = element_text(size = 15, colour = "black"))+
  geom_tiplab2(aes(label=label,angle=angle),
               hjust=-.1,size=5,fontface="bold",align=TRUE)



```



## Energy_Calculator_95
```{r}
energy_dell_dunbrack <- read.csv("Data/csv/energy.csv",header = FALSE,sep ="" )
energy_dell_dunbrack <- energy_dell_dunbrack[,5:33]
source("Functions/amacid2num.R")
source("Functions/atomtype2num167.R")
source("Functions/Energy_PC210.R")
source("Functions/NetworkFrame.R")
source("Functions/inedx_HelixSheet.R")
astral95 <- readRDS("Data/rds/astral_95.rds")
Nprotein <- nrow(astral95)
# ------Pairwise Amino Acid names-------------
AA <- c("PHE","LEU","ILE","VAL","TRP","TYR","MET","CYS","HIS","THR",
        "ARG","ALA","ASN","GLN","PRO","SER","ASP","GLY","LYS","GLU")
letters_list <- aa321(AA)
pair_inter <- character(0)
for (i in 1:20) {
  for (j in i:20) {
    pair_inter <- c(pair_inter, paste0(letters_list[i], letters_list[j]))
  }
} 
#----------------------------------------------
num_columns <- 210
E210.astral95 <- data.frame(pdbID = astral95$pdbID,chainID = astral95$chainID,length = rep(NA,Nprotein),
                            start = astral95$start,end = astral95$end,
                                        class = astral95$class,
                                         fold = astral95$fold,
                                         superfamily = astral95$superfamily,
                                         family = astral95$family,
                                         matrix(NA, ncol = num_columns, nrow = length(astral95$pdbID)))
colnames(E210.astral95) = c("pdbID","chainID","length","start","end","class","fold","superfamily","family",  
                                         pair_inter)
E210.astral95_SS <- E210.astral95
for (NP in 1:Nprotein) {
  print(NP)
  NP = 1318
#---------------
# Your input string
input_string <- "4:,2:"
   input_string <- "B:1-38,B:152-190"
   input_string <- "4:,2:"
   input_string <- "A:,B:"
   input_string <- "L:,H:1-7,H:32-41"
   input_string <- "A:"
   input_string <- "A:1-98"
   input_string <- "A:0-34,A:515-583,A:678-808"
   input_string <- "L:1-107A"
   input_string <- "L:,C:,E:,F:"
   input_string <- "C:488-585,D:603-875,E:885-1007"

# Split the string by comma
split_strings <- unlist(strsplit(input_string, ","))

# Initialize vectors to store the data
chains <- character(0)
starts <- character(0)
ends <- character(0)

# Process each part of the string
for (part in split_strings) {
  parts <- strsplit(part, ":")[[1]]
  chain <- parts[1]
  if (length(parts) > 1) {
    range <- unlist(strsplit(parts[2], "-"))
    start <- ifelse(length(range) == 1, NA, as.numeric(range[1]))
    end <- ifelse(length(range) == 1, NA, as.numeric(range[2]))
  } else {
    start <- NA
    end <- NA
  }

  chains <- c(chains, chain)
  starts <- c(starts, start)
  ends <- c(ends, end)
}
# Create a dataframe
df <- data.frame(chain = chains, start = starts, end = ends)
# Replace NA values with "NA"
df[is.na(df)] <- "NA"
# Print the resulting dataframe
print(df)
#--------------------
  pdbname <- astral95$pdbID[NP]
  ch <- astral95$chainID[NP]
  tryCatch({
    pdb0 <- read.pdb(pdbname)
    flag <- grep("[,]",astral95$Position[NP])
    nums <- unlist(str_extract_all(astral95$Position[NP], "\\d+"))
    ch <- unlist(str_split(astral95$Position[NP], "[:,]"))
    ch<-ch[ch!='']
    (length(flag) == 0)
    ((!length(flag) == 0) & (!length(nums)==0))
    ((!length(flag) == 0) & (length(nums)==0))
    if (length(flag) == 0) {
      sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
      sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
      sele4 <- atom.select(pdb0, "noh")
      sele <- combine.select(sele1, sele2,sele4, operator="AND")
      if (!is.na(astral95$start[NP])){
         st <- astral95$start[NP]
         en <- astral95$end[NP]
         sele3 <- atom.select(pdb0,  resno=st:en, verbose=TRUE)
         sele <- combine.select(sele1, sele2,sele3,sele4, operator="AND")
      }
    }else if ((!length(flag) == 0) & (!length(nums)==0)){
      ch<-substr(astral95$Position[NP],1,1)
      pos<-unlist(str_extract_all(astral95$Position[NP], "\\d+"))
      pos<-c(pos[1]:pos[2],pos[3]:pos[4])
      sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
      sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
      sele4 <- atom.select(pdb0, "noh")
      sele3 <- atom.select(pdb0,  resno=pos, verbose=TRUE)
      sele <- combine.select(sele1, sele2,sele3,sele4, operator="AND")
    }else if ((!length(flag) == 0) & (length(nums)==0)){
      ch<-unlist(str_split(astral95$Position[NP], "[:,]"))
      ch<-ch[ch!='']
      sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
      sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
      sele4 <- atom.select(pdb0, "noh")
      sele <- combine.select(sele1, sele2,sele4, operator="AND")
    }
    pdbX <- pdb0$atom[sele$atom,]
    E210.astral95$length[NP] = length(unique(pdbX$resno))
    E210.astral95_SS$length[NP] = length(unique(pdbX$resno))
    Net_Frame <- NetworkFrame(PDB = pdb0, AtomSelect = sele)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210.astral95[NP,10:219] <- energy210
    # ****************keep res helix & sheet************************
    indhs <- inedx_HelixSheet(PDB = pdb0)
    Net_Frame_hs <- Net_Frame[Net_Frame$resnoif %in% indhs & Net_Frame$resnojf %in% indhs,]
    if(dim(Net_Frame_hs)[1]!=0){
        Net_Frame <- Net_Frame_hs
    }
    energy210_ss <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210.astral95_SS[NP,10:219] <- energy210_ss
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}
saveRDS(E210.astral95,file = "Data/rds/E210.astral95.rds")
saveRDS(E210.astral95_SS,file = "Data/rds/E210.astral95_SS.rds")

```




## Scope 2.08_40
```{r}
astrals_40 <- read_table("Data/csv/astral-scopedom-seqres-gd-sel-gs-bib-40-2.08.fa", col_names = FALSE)
astrals_40_seq <- readFASTA("Data/csv/astral-scopedom-seqres-gd-sel-gs-bib-40-2.08.fa")

rows_keep <- grep(">", astrals_40$X1)
astrals_40 <- astrals_40[rows_keep,c(1,2,3)]
colnames(astrals_40) <- c("pdbID","scopeID","chainID")
astral_40 <- data.frame(
  scope = character(dim(astrals_40)[1]),
    pdbID = character(dim(astrals_40)[1]),
    chainID = character(dim(astrals_40)[1]),
    position = integer(dim(astrals_40)[1]),
    start = integer(dim(astrals_40)[1]),
    end = integer(dim(astrals_40)[1]),
  seq = t(data.frame(astrals_40_seq)))
astral_40$scope <- substr(astrals_40$pdbID, start = 2, stop = nchar(astrals_40$pdbID))
astral_40$pdbID <- substr(astrals_40$pdbID, start = 3, stop = 6)
#----------Split scopeID into class, fold, super and family---------
split_values <- strsplit(astrals_40$scopeID, "\\.")
matrix_values <- matrix("", nrow = nrow(astrals_40), ncol = 4)
for (i in 1:4) {
  matrix_values[, i] <- sapply(split_values, function(x) paste(x[1:i], collapse = "."))
}
df_split <- as.data.frame(matrix_values)
colnames(df_split) <- c("class", "fold", "superfamily", "family")
#-------------------------------------------------------------------
# ------------split chainID, start and end--------------------------
astral_40$chainID <- gsub("[()]", "", astrals_40$chainID)
astral_40$position <- substr(astral_40$chainID, start = regexpr(":", astral_40$chainID) + 1, stop = nchar(astral_40$chainID))
astral_40$chainID <- substr(astral_40$chainID,start = 1, stop = 1)
astral_40$start <- NA
astral_40$end <- NA
split_values <- strsplit(astral_40$position, "-")
for (i in seq_along(split_values)) {
  if (length(split_values[[i]]) == 2) {
    astral_40$start[i] <- as.numeric(split_values[[i]][1])
    astral_40$end[i] <- as.numeric(split_values[[i]][2])
  }
}
#-----------------------------------------------------------------
astral_40 <- cbind(astral_40,df_split)
astral_40$length <- apply(data.frame(astral_40$seq),1,nchar)
saveRDS(astral_40, file = "Data/rds/astral_40.rds")

#ind_rem <- is.na(astral_40$start)
#underscore_count <- str_count(astral_40$position, "-")
#astral40.1domain <- astral_40[underscore_count < 1 & !is.na(astral_40$class),]
#saveRDS(astral40.1domain, file = "Data/rds/astral40.1domain.rds")
```


## Energy_Calculator_40

```{r}
energy_dell_dunbrack <- read.csv("Data/csv/energy.csv",header = FALSE,sep ="" )
energy_dell_dunbrack <- energy_dell_dunbrack[,5:33]
source("Functions/amacid2num.R")
source("Functions/atomtype2num167.R")
source("Functions/Energy_PC210.R")
source("Functions/NetworkFrame.R")
source("Functions/inedx_HelixSheet.R")
astral40 <- readRDS("Data/rds/astral_40.rds")
Nprotein <- nrow(astral40)
# ------Pairwise Amino Acid names-------------
AA <- c("PHE","LEU","ILE","VAL","TRP","TYR","MET","CYS","HIS","THR",
        "ARG","ALA","ASN","GLN","PRO","SER","ASP","GLY","LYS","GLU")
letters_list <- aa321(AA)
pair_inter <- character(0)
for (i in 1:20) {
  for (j in i:20) {
    pair_inter <- c(pair_inter, paste0(letters_list[i], letters_list[j]))
  }
} 
#----------------------------------------------
num_columns <- 210
E210.astral40 <- data.frame(pdbID = astral40$pdbID,chainID = astral40$chainID,length = rep(NA,Nprotein),
                            start = astral40$start,end = astral40$end,
                                        class = astral40$class,
                                         fold = astral40$fold,
                                         superfamily = astral40$superfamily,
                                         family = astral40$family,
                                         matrix(NA, ncol = num_columns, nrow = length(astral40$pdbID)))
colnames(E210.astral40) = c("pdbID","chainID","length","start","end","class","fold","superfamily","family",  
                                         pair_inter)
E210.astral40_SS <- E210.astral40
for (NP in 1:Nprotein) {
  print(NP)
  pdbname <- astral40$pdbID[NP]
  ch <- astral40$chainID[NP]
  tryCatch({
    pdb0 <- read.pdb(pdbname)
    sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
    sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
    sele4 <- atom.select(pdb0, "noh")
    sele <- combine.select(sele1, sele2,sele4, operator="AND")
    if (!is.na(astral40$start[NP])){
      st <- astral40$start[NP]
      en <- astral40$end[NP]
      sele3 <- atom.select(pdb0,  resno=st:en, verbose=TRUE)
      sele <- combine.select(sele1, sele2,sele3,sele4, operator="AND")
    }
    pdbX <- pdb0$atom[sele$atom,]
    E210.astral40$length[NP] = length(unique(pdbX$resno))
    E210.astral40_SS$length[NP] = length(unique(pdbX$resno))
    Net_Frame <- NetworkFrame(PDB = pdb0, AtomSelect = sele)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210.astral40[NP,10:219] <- energy210
    # ****************keep res helix & sheet************************
    indhs <- inedx_HelixSheet(PDB = pdb0)
    Net_Frame_hs <- Net_Frame[Net_Frame$resnoif %in% indhs & Net_Frame$resnojf %in% indhs,]
    if(dim(Net_Frame_hs)[1]!=0){
        Net_Frame <- Net_Frame_hs
    }
    energy210_ss <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210.astral40_SS[NP,10:219] <- energy210_ss
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}
saveRDS(E210.astral40,file = "Data/rds/E210.astral40.rds")
saveRDS(E210.astral40_SS,file = "Data/rds/E210.astral40_SS.rds")
```

## Ferritin-like

```{r}
ferritin <- read.csv("Data/csv/Ferritin_Like.csv",sep = ";")
energy_dell_dunbrack <- read.csv("Data/csv/energy.csv",header = FALSE,sep ="" )
energy_dell_dunbrack <- energy_dell_dunbrack[,5:33]
source("Functions/amacid2num.R")
source("Functions/atomtype2num167.R")
source("Functions/Energy_PC210.R")
source("Functions/NetworkFrame.R")
source("Functions/inedx_HelixSheet.R")
Nprotein <- nrow(ferritin)
# ------Pairwise Amino Acid names-------------
AA <- c("PHE","LEU","ILE","VAL","TRP","TYR","MET","CYS","HIS","THR",
        "ARG","ALA","ASN","GLN","PRO","SER","ASP","GLY","LYS","GLU")
letters_list <- aa321(AA)
pair_inter <- character(0)
for (i in 1:20) {
  for (j in i:20) {
    pair_inter <- c(pair_inter, paste0(letters_list[i], letters_list[j]))
  }
} 
#----------------------------------------------
num_columns <- 210
E210_ferritin <- data.frame(pdbID = ferritin$pdbID,length = rep(NA,Nprotein),group = ferritin$group,
                                         matrix(NA, ncol = num_columns, nrow = length(ferritin$pdbID)))
colnames(E210_ferritin) = c("pdbID","length","group",pair_inter)
E210_ferritin_SS <- E210_ferritin
for (NP in 1:Nprotein) {
  print(NP)
  pdbname <- ferritin$pdbID[NP]
  E210_ferritin_SS$pdbID[NP] <- pdbname
  ch <- ferritin$chainID[NP]
  E210_ferritin$pdbID[NP] <- paste0(pdbname,ch)
  E210_ferritin_SS$pdbID[NP] <- paste0(pdbname,ch)
  tryCatch({
    pdb0 <- read.pdb(pdbname)
    sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
    sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
    sele4 <- atom.select(pdb0, "noh")
    sele <- combine.select(sele1, sele2,sele4, operator="AND")
    if (!is.na(ferritin$start[NP])){
      st <- ferritin$start[NP]
      en <- ferritin$end[NP]
      sele3 <- atom.select(pdb0,  resno=st:en, verbose=TRUE)
      sele <- combine.select(sele1, sele2,sele3,sele4, operator="AND")
    }
    pdbX <- pdb0$atom[sele$atom,]
    E210_ferritin$length[NP] = length(unique(pdbX$resno))
    E210_ferritin_SS$length[NP] = length(unique(pdbX$resno))
    Net_Frame <- NetworkFrame(PDB = pdb0, AtomSelect = sele)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210_ferritin[NP,4:213] <- energy210
    # ****************keep res helix & sheet************************
    indhs <- inedx_HelixSheet(PDB = pdb0)
    Net_Frame_hs <- Net_Frame[Net_Frame$resnoif %in% indhs & Net_Frame$resnojf %in% indhs,]
    if(dim(Net_Frame_hs)[1]!=0){
        Net_Frame <- Net_Frame_hs
    }
    energy210_ss <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210_ferritin_SS[NP,4:213] <- energy210_ss
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}

saveRDS(E210_ferritin,file = "Data/rds/E210_ferritin.rds")
saveRDS(E210_ferritin_SS,file = "Data/rds/E210_ferritin_SS.rds")
```

## Phylogenetic Tree_Ferritin

```{r}
E210_ferritin <- readRDS("Data/rds/E210_ferritin.rds")
df <- E210_ferritin
df <- df[with(df, order(df$group)), ]
row.names(df) <- df[,1]
rn <- df[,1]
dis <- proxy::dist(df[,-c(1,3)], method = manhat_norm)
dis[is.nan(dis)]<-0
tree_data <- ape::bionj(dis)
grp <- list(Bacteri     = rn[1:5],
            BMM_alpha = rn[6:7],
            BMM_beta  = rn[8:10],
            Dps  = rn[11:25],
            Fatty_acid  = rn[26:29],
            Ferritins  = rn[30:40],
            NAA  = rn[41:45],
            RNRR2  = rn[46:53])

p_data <- ggtree(tree_data, layout = 'circular',
                 branch.length='none',size = 1.5) 

t = groupOTU(p_data, grp, "groups") + aes(color=groups) +
  theme(legend.position="right",
        legend.title = element_blank(),
        legend.text = element_text(size = 15, colour = "black"))+
  geom_tiplab2(aes(label=label,angle=angle),
               hjust=-.1,size=5,fontface="bold",align=TRUE)
#------------------tamam--------

df[,-c(1,2,3)] = df[,-c(1,2,3)]
df$total = rowSums(df[,-c(1,2,3)])
ggplot(df,aes(group,total))+geom_boxplot()
df$family[df$group %in% c("Bac","Fer","Dps","Manganese cataleses")] = "a.25.1.1"
df$family[! df$group %in% c("Bac","Fer","Dps")] = "a.25.1.2"
t.test(df$total[df$family == "a.25.1.0"],df$total[df$family == "a.25.1.2"],alternative = "greater")
wilcox.test(df$total[df$family == "a.25.1.0"],df$total[df$family == "a.25.1.2"],alternative = "greater")
ggplot(df,aes(family,total))+geom_boxplot()

dis <- dist(df[,-c(1,2,3)], method="manhattan")
dis <- sqrt(dis)
#------------------

#----------------------------------------
tot<-data.frame(total=df$total)
rownames(tot)<-rownames(df)
dis <- dist(tot)



x = df[,-c(1:3)]
x  <- as.matrix(x) 
hc <- x %>% dist(method = 'manhattan') %>% hclust(method = "ward.D")
dend <- hc %>% as.dendrogram

labcol <- c( "blue", "green","red","black","yellow","pink","magenta","purple")
ord_sml <- df[,][order.dendrogram(dend),3]
group_cols <- labcol[as.factor(ord_sml)]

dend2 <- dend %>%
  #color_branches(k=3, col = c("red","blue","green3")) %>%
  color_labels(col =   group_cols) %>%
  set("labels_cex", 0.5) 
plot(dend2, ylab = "Height")
legend("topright", legend = as.character(unique(ord_sml)),
       pch = 15, pt.cex = 2.3, cex = 1.4, bty = 'n',
       inset = c(-.15, -.1), # place outside
       col = unique(group_cols),horiz = FALSE,xpd = TRUE)







nnet <- neighborNet(dis)
plot(nnet)
ggsplitnet(nnet) + geom_tiplab2(aes(label=label),
                                hjust=-.1,size=5,fontface="bold")



tree_data <- ape::bionj(dis)
grp <- list(Bacteri     = rn[1:5],
            BMM_alpha = rn[6:7],
            BMM_beta  = rn[8:10],
            Dps  = rn[11:25],
            Fatty_acid  = rn[26:29],
            Ferritins  = rn[30:40],
            NAA  = rn[41:45],
            RNRR2  = rn[46:53])

p_data <- ggtree(tree_data, layout = 'circular',
                 branch.length='none',size = 1.5) 

t = groupOTU(p_data, grp, "groups") + aes(color=groups) +
  theme(legend.position="right",
        legend.title = element_blank(),
        legend.text = element_text(size = 15, colour = "black"))+
  geom_tiplab2(aes(label=label,angle=angle),
               hjust=-.1,size=5,fontface="bold",align=TRUE)




df0 <- E210_ferritin
df0 <- df0[with(df0, order(df0$group)), ]
pc <- prcomp(df0[,-c(1,2,3)],scale. = F,center = T) 
Rot_E_abcd <- as.matrix(df0[,-c(1,2,3)]) %*% pc$rotation[,1:3] 
df <- data.frame(class = df0$group
                 ,Rot_E_abcd) 
rn <- df0[,1]
row.names(df) <- rn




# ------------- UMAP ------------------------
ump.E_PC_abcd <- umap(E_PC_abcd[,-c(1)], n_neighbors = 10,
            min_dist = 0.001, random_state = 123)
#-------- UMAMP Visualization ---------------
umap.df.E_PC_abcd <- data.frame(
  sf = E_PC_abcd$class,
  UMAP1 = ump.E_PC_abcd$layout[, 1],
  UMAP2 = ump.E_PC_abcd$layout[, 2])
um.E_PC_abcd <- ggplot(umap.df.E_PC_abcd,aes(UMAP1, UMAP2, color= sf))+
  geom_point(alpha=.7) 




```
## Covid_Spike200

```{r}
#spike <- read.delim("Data/csv/spike_structures.tsv")
spike <- read.csv("Data/csv/spike200.csv", sep=";")
energy_dell_dunbrack <- read.csv("Data/csv/energy.csv",header = FALSE,sep ="" )
energy_dell_dunbrack <- energy_dell_dunbrack[,5:33]
source("Functions/amacid2num.R")
source("Functions/atomtype2num167.R")
source("Functions/Energy_PC210.R")
source("Functions/NetworkFrame.R")
source("Functions/inedx_HelixSheet.R")
aaenergy <- read.csv("Data/csv/Pij.csv", header = F,sep = ";")
aaenergy <- (aaenergy + t(aaenergy))/2
colnames(aaenergy) = rownames(aaenergy)<- 
  c("F","L","I","V","W","Y","M","C","H","T","R","A","N","Q","P","S","D","G","K","E")
#row.names(aaenergy) <- aa321(aaenergy[,1])
#aaenergy <- aaenergy[,-1]
#row.names(aaenergy) <- colnames(aaenergy)
letters_list <- tolower(row.names(aaenergy))
aa210<-aaenergy[lower.tri(aaenergy,diag = T)]

Nprotein <- nrow(spike)
# ------Pairwise Amino Acid names-------------
AA <- c("PHE","LEU","ILE","VAL","TRP","TYR","MET","CYS","HIS","THR",
        "ARG","ALA","ASN","GLN","PRO","SER","ASP","GLY","LYS","GLU")
letters_list <- aa321(AA)
pair_inter <- character(0)
for (i in 1:20) {
  for (j in i:20) {
    pair_inter <- c(pair_inter, paste0(letters_list[i], letters_list[j]))
  }
} 
#----------------------------------------------
num_columns <- 210
E210_spike200 <- data.frame(pdbID = spike$pdbID,length = rep(NA,Nprotein),virus = spike$VIRUS,
                         #description = spike$DESCRIPTION,
                         conformation = spike$Conformation,
                         matrix(NA, ncol = num_columns, nrow = length(spike$pdbID)))
colnames(E210_spike200) = c("pdbID","length","virus", "conformation",pair_inter)
E210_spike200_SS <- E210_spike200
E210.seq_spike200 <- data.frame(pdbID = spike$pdbID,length = rep(NA,Nprotein),
                             virus = spike$VIRUS,
                             conformation = spike$Conformation,
                                         matrix(NA, ncol = num_columns, 
                                                nrow = length(spike$pdbID)))

for (NP in 1:Nprotein) {
  print(NP)
  ch <- substr(spike$pdbID[NP],5,5)
  pdbname <- substr(spike$pdbID[NP],1,4)
  E210_spike200$pdbID[NP] <- spike$pdbID[NP]
  E210_spike200_SS$pdbID[NP] <- spike$pdbID[NP]
  E210.seq_spike200$pdbID[NP] <- spike$pdbID[NP]
  tryCatch({
    pdb0 <- read.pdb(tolower(pdbname))
    sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
    sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
    sele4 <- atom.select(pdb0, "noh")
    sele <- combine.select(sele1, sele2,sele4, operator="AND")
    pdbX <- pdb0$atom[sele$atom,]
    pdbXca <- pdbX[pdbX$elety == "CA",]
    seq = aa321(pdbXca$resid)
    E210.seq_spike200$length[NP] = length(unique(pdbX$resno))
    freq = c()
    for (j in 1:20){
        freq[j] <-  length(grep(letters_list[j],seq))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
    E210.seq_spike200[NP,5:214] <- aa210_p
    E210_spike200$length[NP] = length(unique(pdbX$resno))
    E210_spike200_SS$length[NP] = length(unique(pdbX$resno))
    Net_Frame <- NetworkFrame(PDB = pdb0, AtomSelect = sele)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210_spike200[NP,5:214] <- energy210
    # ****************keep res helix & sheet************************
    indhs <- inedx_HelixSheet(PDB = pdb0)
    Net_Frame_hs <- Net_Frame[Net_Frame$resnoif %in% indhs & Net_Frame$resnojf %in% indhs,]
    if(dim(Net_Frame_hs)[1]!=0){
        Net_Frame <- Net_Frame_hs
    }
    energy210_ss <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210_spike200_SS[NP,5:214] <- energy210_ss
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}

saveRDS(E210_spike200,file = "Data/rds/E210_spike200.rds")
saveRDS(E210_spike200_SS,file = "Data/rds/E210_spike200_SS.rds")
saveRDS(E210.seq_spike200,file = "Data/rds/E210.seq_spike200.rds")

```

## UMAP covid

```{r}
E210_spike200 <- readRDS("Data/rds/E210_spike200.rds")
E210_spike200_SS <- readRDS("Data/rds/E210_spike200_SS.rds")
E210.seq_spike200 <- readRDS("Data/rds/E210.seq_spike200.rds")

#spike200 <- read.csv("Data/csv/spike200.csv", sep = ";")
#spike200$chain = substr(spike200$pdbID,5,5)
#spike200$pdbID = paste0(tolower(substr(spike200$pdbID,1,4)),spike200$chain)
#df = left_join(spike200,E210_spike,by = "pdbID")
df <- E210.seq_spike200
df = df[!is.na(df$length),]
df0 = df[,-seq(1:4)]/df$length
#df = df[df$length>100 & df$description == " Spike glycoprotein",]
pc <- prcomp(df0,scale. = F,center = T) 
#%*% pc$rotation[,1:10] 
Rot_E_abcd <- as.matrix(df0) 
E_PC_abcd <- data.frame(class = df$virus,conf = df$conformation,Rot_E_abcd) 
# ------------- UMAP ------------------------
ump.E_PC_abcd <- umap(E_PC_abcd[,-c(1,2)], n_neighbors = 40,
            min_dist = 0.1, random_state = 123)
#-------- UMAMP Visualization ---------------
umap.df.E_PC_abcd <- data.frame(shape = E_PC_abcd$conf,
  sf = E_PC_abcd$class,
  UMAP1 = ump.E_PC_abcd$layout[, 1],
  UMAP2 = ump.E_PC_abcd$layout[, 2])
um.E_PC_abcd <- ggplot(umap.df.E_PC_abcd,aes(UMAP1, UMAP2, color= sf, shape = shape))+
  geom_point(alpha=.7) +
  mythem

```

## tree covid

```{r}
E210.seq_spike200 <- readRDS("Data/rds/E210.seq_spike200.rds")
E210_spike200 <- readRDS("Data/rds/E210_spike200.rds")
#-----------close
df = E210_spike200
df = df[!is.na(df$length),]
df = df[df$conformation == "Close",]

rownames(df) = df$pdbID
x = df[,-c(1,3:4)]
x  <- as.matrix(x) 
hc <- x %>% proxy::dist(method = manhat_norm)
hc[is.nan(hc)] <- 0
hc <- hc%>% hclust(method = "ward.D2")
dend <- hc %>% as.dendrogram

labcol <- c( "blue", "green","red")
ord_sml <- df[,][order.dendrogram(dend),3]
group_cols <- labcol[as.factor(ord_sml)]

dend2 <- dend %>%
  #color_branches(k=3, col = c("red","blue","green3")) %>%
  color_labels(col =   group_cols) %>%
  set("labels_cex", 0.5) 
plot(dend2, ylab = "Height")
legend("topright", legend = as.character(unique(ord_sml)),
       pch = 15, pt.cex = 2.3, cex = 1.4, bty = 'n',
       inset = c(-.15, -.1), # place outside
       col = unique(group_cols),horiz = FALSE,xpd = TRUE)
#---------------tamam----------------------
df$total<-rowSums(df[,-c(1:4)]/df$length)
ggplot(df,aes(virus,total))+geom_boxplot()

#----------------------open
df = left_join(spike200,E210_spike,by = "pdbID")
df = df[!is.na(df$length),]
df = df[df$Conformation == "Open",]
rownames(df) = df$pdbID
len = as.numeric(df$length)
  x = df[,-c(1:7)]/len
x  <- as.matrix(x) 
hc <- x %>% dist(method = 'manhattan') %>% hclust(method = "ward.D2")
dend <- hc %>% as.dendrogram

labcol <- c( "blue", "green","red")
ord_sml <- df[,][order.dendrogram(dend),2]
group_cols <- labcol[as.factor(ord_sml)]

dend2 <- dend %>%
  #color_branches(k=3, col = c("red","blue","green3")) %>%
  color_labels(col =   group_cols) %>%
  set("labels_cex", 0.5) 
plot(dend2, ylab = "Height")
legend("topright", legend = as.character(unique(ord_sml)),
       pch = 15, pt.cex = 2.3, cex = 1.4, bty = 'n',
       inset = c(-.15, -.1), # place outside
       col = unique(group_cols),horiz = FALSE,xpd = TRUE)
```


## DD dataset

```{r}
dd_train0 <- read_table("Data/csv/DDtrain.txt", col_names = FALSE)
rows_keep <- grep(">", dd_train0$X1)
dd_train0 <- dd_train0[rows_keep,c(1)]
colnames(dd_train0) <- c("pdbID")
dd_train <- data.frame(
    pdbID = (character(dim(dd_train0)[1])),
    chainID = character(dim(dd_train0)[1]),
    pdbcID = character(dim(dd_train0)[1]))
dd_train$pdbID <- tolower(substr(dd_train0$pdbID, start = 2, stop = 5))
dd_train$chainID <- substr(dd_train0$pdbID, start = 6, stop = 6)
dd_train$pdbcID <- paste0(dd_train$pdbID,dd_train$chainID)
astral95.1domain <- readRDS("Data/rds/astral95.1domain.rds")
astral95.1domain$pdbcID = paste0(astral95.1domain$pdbID,astral95.1domain$chainID)
dd_tr <- left_join(dd_train,astral95.1domain, by = "pdbcID")
```



## Energy_Calculator_DD
```{r}
energy_dell_dunbrack <- read.csv("Data/csv/energy.csv",header = FALSE,sep ="" )
energy_dell_dunbrack <- energy_dell_dunbrack[,5:33]
source("Functions/amacid2num.R")
source("Functions/atomtype2num167.R")
source("Functions/Energy_PC210.R")
source("Functions/NetworkFrame.R")
source("Functions/inedx_HelixSheet.R")
DD_train <- readRDS("Data/rds/astral_95.rds")
Nprotein <- nrow(DD_train)
# ------Pairwise Amino Acid names-------------
AA <- c("PHE","LEU","ILE","VAL","TRP","TYR","MET","CYS","HIS","THR",
        "ARG","ALA","ASN","GLN","PRO","SER","ASP","GLY","LYS","GLU")
letters_list <- aa321(AA)
pair_inter <- character(0)
for (i in 1:20) {
  for (j in i:20) {
    pair_inter <- c(pair_inter, paste0(letters_list[i], letters_list[j]))
  }
} 
aaenergy <- read.csv("Data/csv/Pij.csv", header = F,sep = ";")
aaenergy <- (aaenergy + t(aaenergy))/2
colnames(aaenergy) = rownames(aaenergy)<- 
  c("F","L","I","V","W","Y","M","C","H","T","R","A","N","Q","P","S","D","G","K","E")
letters_list <- tolower(row.names(aaenergy))
aa210<-aaenergy[lower.tri(aaenergy,diag = T)]
#----------------------------------------------
num_columns <- 210
E210.DD_train <- data.frame(pdbID = DD_train$pdbID,length = rep(NA,Nprotein),
                            fold = rep(NA,Nprotein),
                            fold_name = DD_train$fold_name,
                            scopeID = DD_train$scopeID, 
                            matrix(NA, ncol = num_columns, nrow = length(DD_train$pdbID)))
colnames(E210.DD_train) = c("pdbID","length","fold","fold_name","scopeID",pair_inter)
E210.DD_train_SS <-E210.seq_DD_train <-  E210.DD_train
for (NP in 1:Nprotein) {
  print(NP)
  pdbname <- DD_train$pdbID[NP]
  E210.DD_train$pdbID[NP] <- pdbname
  E210.DD_train_SS$pdbID[NP] <- pdbname
  E210.seq_DD_train$pdbID[NP] <- pdbname
  ch <- DD_train$chainID[NP]
  split_values <- strsplit(DD_train$scopeID[NP], "\\.")
  E210.DD_train_SS$fold[NP] <- sapply(split_values, function(x) paste(x[1:2], collapse = "."))
  E210.DD_train$fold[NP] <- E210.DD_train_SS$fold[NP]
  E210.seq_DD_train$fold[NP] <- E210.DD_train_SS$fold[NP]
  tryCatch({
    pdb0 <- read.pdb(pdbname)
    flag <- grep("[,]",DD_train$Position[NP])
    nums <- unlist(str_extract_all(DD_train$Position[NP], "\\d+"))
    if (length(flag) == 0) {
      sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
      sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
      sele4 <- atom.select(pdb0, "noh")
      sele <- combine.select(sele1, sele2,sele4, operator="AND")
      if (!is.na(DD_train$start[NP])){
         st <- DD_train$start[NP]
         en <- DD_train$end[NP]
         sele3 <- atom.select(pdb0,  resno=st:en, verbose=TRUE)
         sele <- combine.select(sele1, sele2,sele3,sele4, operator="AND")
      }
    }else if ((!length(flag) == 0) & (!length(nums)==0)){
      ch<-substr(DD_train$Position[NP],1,1)
      pos<-unlist(str_extract_all(DD_train$Position[NP], "\\d+"))
      pos<-c(pos[1]:pos[2],pos[3]:pos[4])
      sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
      sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
      sele4 <- atom.select(pdb0, "noh")
      sele3 <- atom.select(pdb0,  resno=pos, verbose=TRUE)
      sele <- combine.select(sele1, sele2,sele3,sele4, operator="AND")
    }else if ((!length(flag) == 0) & (length(nums)==0)){
      ch<-unlist(str_split(DD_train$Position[NP], "[:,]"))
      ch<-ch[ch!='']
      sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
      sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
      sele4 <- atom.select(pdb0, "noh")
      sele <- combine.select(sele1, sele2,sele4, operator="AND")
    }
    pdbX <- pdb0$atom[sele$atom,]
    E210.DD_train$length[NP] = length(unique(pdbX$resno))
    E210.DD_train_SS$length[NP] = length(unique(pdbX$resno))
    Net_Frame <- NetworkFrame(PDB = pdb0, AtomSelect = sele)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210.DD_train[NP,6:215] <- energy210
    # ****************keep res helix & sheet************************
    indhs <- inedx_HelixSheet(PDB = pdb0)
    Net_Frame_hs <- Net_Frame[Net_Frame$resnoif %in% indhs & Net_Frame$resnojf %in% indhs,]
    if(dim(Net_Frame_hs)[1]!=0){
        Net_Frame <- Net_Frame_hs
    }
    energy210_ss <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210.DD_train_SS[NP,6:215] <- energy210_ss
    #-----------sequence enregy--------
    pdbXca <- pdbX[pdbX$elety == "CA",]
    seq = aa321(pdbXca$resid)
    E210.seq_DD_train$length[NP] = length(unique(pdbX$resno))
    freq = c()
    for (j in 1:20){
        freq[j] <-  length(grep(letters_list[j],tolower(seq)))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
    E210.seq_DD_train[NP,6:215] <- aa210_p

    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}
saveRDS(E210.DD_train,file = "Data/rds/E210.EDD.rds")
saveRDS(E210.DD_train_SS,file = "Data/rds/E210.EDD_SS.rds")
saveRDS(E210.seq_DD_train,file = "Data/rds/E210.seq_EDD.rds")

```

## UMAP DD_train

```{r}
df <- E210.DD_train_SS
pc <- prcomp(df[,-seq(1:5)]/df$length,scale. = F,center = T) 
Rot_E_abcd <- as.matrix(df[,-seq(1:5)]) %*% pc$rotation[,1:7] 
E_PC_abcd <- data.frame(class = df$fold,Rot_E_abcd) 
# ------------- UMAP ------------------------
ump.E_PC_abcd <- umap(E_PC_abcd[,-c(1)], n_neighbors = 15,
            min_dist = 0.01, random_state = 123)
#-------- UMAMP Visualization ---------------
umap.df.E_PC_abcd <- data.frame(
  sf = E_PC_abcd$class,
  UMAP1 = ump.E_PC_abcd$layout[, 1],
  UMAP2 = ump.E_PC_abcd$layout[, 2])
um.E_PC_abcd <- ggplot(umap.df.E_PC_abcd,aes(UMAP1, UMAP2, color= sf))+
  geom_point(alpha=.7) +
  mythem
```

## DD_Test

```{r}

scope_2.08 <- read.delim("~/Documents/Projects/Git/ProteinEnergyProfileSimilarity/Data/csv/dir.cla.scope.2.08-stable.txt", header=FALSE, comment.char="#")
colnames(scope_2.08) <- c("scope","pdb","chain","scopeID","description")
dd_test <- read.csv("Data/csv/DDtest.fasta", header=FALSE, sep=";")
pdbid <- fold_name <- character(0)
for (line in dd_test$V1) {
  if (grepl("^FOLD", line)) {
    # Extract fold name
    current_fold <- line
  } else if (grepl("^>", line)) {
    # Extract pdbid and sequence
    parts <- strsplit(line, "\\s+")
    current_pdbid <- gsub(">", "", parts[[1]][1])  # Remove ">" symbol
    # Append to the vectors
    pdbid <- c(pdbid, current_pdbid)
    fold_name <- c(fold_name, current_fold)
  }
}
df <- data.frame(pdbid, fold_name,scope=character(length(fold_name)))
for (i in 1:dim(df)[1]){
    df$scope[i] <- paste0("d", tolower(df$pdbid[i]))
}

dx = scope_2.08[scope_2.08$scope %in% df$scope,]
DD <- merge(df,scope_2.08,"scope")
chain=start = end = c()
for (i in 1:dim(DD)[1]){
  parts <- unlist(strsplit(DD$chain[i], "[:-]"))
  chain[i] <- parts[1]  # B
  start[i] <- as.integer(parts[2])  # 141
  end[i] <- as.integer(parts[length(parts)])  # 267
}
DD$chainID = chain
DD$start = start
DD$end = end
DD$pdbID =DD$pdb
saveRDS(DD,"Data/rds/DD_test.rds")

```

## Random Forest_DD

```{r}
E210.DD_train <- readRDS("Data/rds/E210.seq_DD_train.rds")
E210.DD_test <- readRDS("Data/rds/E210.seq_DD_test.rds")

df = E210.DD_train
df = df[!is.na(df$FF),]
data<- cbind(class = df$fold,df[,-seq(1:8)])
data$class[data$class == "b.18"] = "b.1"
data$class[data$class == "c.93"] = "c.94"
data$class <- as.factor(data$class)
#---------------------------
rf_models <- list()
unique_classes <- unique(data$class)
for (class in unique_classes) {
  # Create a binary response variable for the OvA classification
  data$BinaryClass <- as.factor(ifelse(data$class == class, "Positive", "Negative"))
  data0 = data[,-1]
  # Train a Random Forest model for the current class
  traintask <- makeClassifTask(data = data0, target = "BinaryClass")
  rf <- makeLearner("classif.randomForest",
                   predict.type = "response",
                   par.vals = list(ntree= floor(0.1*nrow(data)),
                                   mtry = floor((ncol(data)-1)/3)))

  rf$par.vals <- list(importance = TRUE)
  # Tune the parameters
  # Grid search to find parapmeters
  rf_param <- makeParamSet(
  makeIntegerParam("ntree",lower = 10, upper = floor(1*nrow(data))),
  makeIntegerParam("mtry", lower = floor((ncol(data)-1)/3), upper = ncol(data)))
  #Random search for 10 iterations
  rancontrol <- makeTuneControlRandom(maxit = 10L)
  set_cv <- makeResampleDesc("CV",iters= 3L)
  rf_tune <- tuneParams(learner = rf,
                      resampling = set_cv,
                      task = traintask,
                      par.set = rf_param,
                      control = rancontrol,
                      measures = acc)
  RF_model <- randomForest(formula = as.factor(BinaryClass) ~ .,
                         data = data0,
                         ntree = rf_tune$x$ntree,
                         mtry = rf_tune$x$mtry, 
                         replace = TRUE)

  # Store the model in the list
  rf_models[[class]] <- RF_model
}


#data <- data[data$class %in% c("a.1.1", "a.104.1"),]
# one versus all classification#
rf_models <- list()
unique_classes <- unique(data$class)
for (class in unique_classes) {
  # Create a binary response variable for the OvA classification
  data$BinaryClass <- as.factor(ifelse(data$class == class, "Positive", "Negative"))
  data0 = data[,-1]
  # Train a Random Forest model for the current class
  rf_model <- tuneRF(data0, data0$BinaryClass, stepFactor = 1.1, improve = 0.01, trace=T, 
                   plot= T,doBest = T)
  # Store the model in the list
  rf_models[[class]] <- rf_model
}

k = 1
test <- cbind(class = E210.DD_test$fold,E210.DD_test[,-seq(1:8)])
test$class[test$class == "c.93"] = "c.94"
test$class[test$class == "b.18"] = "b.1"

test = test[!is.na(test$FF),]
table(test$class)

class_probabilities <- matrix(0,nrow = nrow(test),ncol = length(unique_classes))
for (i in unique_classes) {
  rf_model <- rf_models[[i]]
  test$BinaryClass <- as.factor(ifelse(test$class == i, "Positive", "Negative"))
  test0 = test[,-c(1)]
  prediction <- predict(rf_model, test0, type = "prob")
  class_probabilities[,k] <- prediction[,2]  # Probability of the positive class
  k = k+1
}
colnames(class_probabilities) <- unique_classes
result = data.frame(class = test$class,class_probabilities)
predicted_class = data.frame(class = test$class, predict = character(nrow(result)))
for (j in 1:nrow(result)){
  predicted_class$predict[j] <- as.character(unique_classes[which.max(as.vector(result[j,-1]))])
}

different_rows <- predicted_class$class != predicted_class$predict
predicted_class[different_rows,]

df_seq_str<-data.frame(seq = rowSums(E210.seq_DD_test[,-c(1:5)]),
                       str = rowSums(E210.DD_test[,-c(1:5)]),E210.DD_test[,c(1:5)])



library(mlr)
library(randomForest)
```

## Scope 2.08_95_original
```{r}
astrals_95 <- read_table("Data/csv/astral-scopedom-seqres-gd-sel-gs-bib-95-2.08.fa", col_names = FALSE)
astrals_95_seq <- readFASTA("Data/csv/astral-scopedom-seqres-gd-sel-gs-bib-95-2.08.fa")

rows_keep <- grep(">", astrals_95$X1)
astrals_95 <- astrals_95[rows_keep,c(1,2,3)]
colnames(astrals_95) <- c("pdbID","scopeID","chainID")
astral_95 <- data.frame(
  scope = character(dim(astrals_95)[1]),
    pdbID = character(dim(astrals_95)[1]),
    chainID = character(dim(astrals_95)[1]),
    Position = character(dim(astrals_95)[1]),
    start_end = character(dim(astrals_95)[1]),
    start = integer(dim(astrals_95)[1]),
    end = integer(dim(astrals_95)[1]),
  seq = t(data.frame(astrals_95_seq)))
row.names(astral_95) <- c(1:dim(astral_95)[1])
astral_95$scope <- substr(astrals_95$pdbID, start = 2, stop = nchar(astrals_95$pdbID))
astral_95$pdbID <- substr(astrals_95$pdbID, start = 3, stop = 6)
#----------Split scopeID into class, fold, super and family---------
split_values <- strsplit(astrals_95$scopeID, "\\.")
matrix_values <- matrix("", nrow = nrow(astrals_95), ncol = 4)
for (i in 1:4) {
  matrix_values[, i] <- sapply(split_values, function(x) paste(x[1:i], collapse = "."))
}
df_split <- as.data.frame(matrix_values)
colnames(df_split) <- c("class", "fold", "superfamily", "family")
#-------------------------------------------------------------------
# ------------split chainID, start and end--------------------------
astral_95$Position <- gsub("[()]", "", astrals_95$chainID)
astral_95$start_end <- substr(astral_95$Position, start = regexpr(":", astral_95$Position) + 1, stop = nchar(astral_95$Position))
astral_95$chainID <- substr(astral_95$Position,start = 1, stop = 1)
astral_95$start <- NA
astral_95$end <- NA
split_values <- strsplit(astral_95$start_end, "-")
for (i in seq_along(split_values)) {
  if (length(split_values[[i]]) == 2) {
    astral_95$start[i] <- as.numeric(split_values[[i]][1])
    astral_95$end[i] <- as.numeric(split_values[[i]][2])
  }
}
#-----------------------------------------------------------------
astral_95 <- cbind(astral_95,df_split)
astral_95$length <- apply(data.frame(astral_95$seq),1,nchar)
saveRDS(astral_95, file = "Data/rds/astral_95.rds")

#ind_rem <- is.na(astral_95$start)
#underscore_count <- str_count(astral_95$position, "-")
#astral95.1domain <- astral_95[underscore_count < 1 & !is.na(astral_95$class),]
#saveRDS(astral95.1domain, file = "Data/rds/astral95.1domain.rds")
```


## TG_original 

```{r}
scope_2.08 <- read.delim("Data/csv/dir.cla.scope.2.08-stable.txt", 
                         header=FALSE, comment.char="#")
colnames(scope_2.08) <- c("scope","pdb","chain","scopeID","description")
TG0 <- read.csv("Data/csv/TG.fasta", header=FALSE, sep=";")
TG_seq <- readFASTA("Data/csv/TG.fasta")

scope <-  character(0)
for (line in TG0$V1) {
  if (grepl("^>", line)) {
    parts <- strsplit(line, "\\s+")
    scp <-  gsub("__", "a_", parts[[1]][2])
    scope <- c(scope,scp)
  }
}

TG_all <- data.frame(scope)
row.names(TG_all) <- scope

TG1 <- merge(TG_all,scope_2.08,"scope")
rem_tg <- TG_all[!(TG_all$scope %in% scope_2.08$scope),]
scope <-  character(0)
scope_or <- character(0)
for (i in 1:length(rem_tg)){
  sci <- rem_tg[i]
  scope <- c(scope,gsub("_(1|2|3)$", "a\\1", sci))
  scope_or <- c(scope_or,sci)
}
TG_1 <- data.frame(scope)
TG2 <- merge(TG_1,scope_2.08,"scope")

rem_tg <- TG_1[! (TG_1$scope %in% scope_2.08$scope),]
scope <-  character(0)
scope_or <-  character(0)
for (i in 1:length(rem_tg)){
  sci <- rem_tg[i]
  scope <- c(scope,gsub("a_$", "a1", sci))
  scope_or <- c(scope_or,sci)
}
TG_2 <- data.frame(scope)
TG3 <- merge(TG_2,scope_2.08,"scope")

rem_tg <- TG_2[! (TG_2$scope %in% scope_2.08$scope),]
  
TG <- rbind(TG1,TG2,TG3)
TGsss <- left_join(TG,TG_seq,by = "scope")

chain=start = end = c()
for (i in 1:nrow(TG)){
  parts <- unlist(strsplit(TG$chain[i], "[:-]"))
  chain[i] <- parts[1]  # B
  start[i] <- as.integer(parts[2])  # 141
  end[i] <- as.integer(parts[length(parts)])  
  split_values <- strsplit(TG$scopeID[i], "\\.")
  TG$fold[i] <- sapply(split_values, function(x) paste(x[1:2], collapse = "."))
}
TG$pdbID <- TG$pdb
TG$chainID = chain
TG$start = start
TG$end = end

TG <- TG[,c("pdbID","chainID","start","end","fold","scopeID")]
#saveRDS(TG,"Data/rds/TG.rds")
```



## Energy_Calculator_EDD
```{r}
energy_dell_dunbrack <- read.csv("Data/csv/energy.csv",header = FALSE,sep ="" )
energy_dell_dunbrack <- energy_dell_dunbrack[,5:33]
source("Functions/amacid2num.R")
source("Functions/atomtype2num167.R")
source("Functions/Energy_PC210.R")
source("Functions/NetworkFrame.R")
source("Functions/inedx_HelixSheet.R")
DD_train <- readRDS("Data/rds/EDD.rds")
Nprotein <- nrow(DD_train)
# ------Pairwise Amino Acid names-------------
AA <- c("PHE","LEU","ILE","VAL","TRP","TYR","MET","CYS","HIS","THR",
        "ARG","ALA","ASN","GLN","PRO","SER","ASP","GLY","LYS","GLU")
letters_list <- aa321(AA)
pair_inter <- character(0)
for (i in 1:20) {
  for (j in i:20) {
    pair_inter <- c(pair_inter, paste0(letters_list[i], letters_list[j]))
  }
} 
#----------------------------------------------
num_columns <- 210
E210.DD_train <- data.frame(pdbID = DD_train$pdbID,length = rep(NA,Nprotein),
                            fold = rep(NA,Nprotein),fold_name = DD_train$fold_name,scopeID = DD_train$scopeID, 
                                         matrix(NA, ncol = num_columns, nrow = length(DD_train$pdbID)))
colnames(E210.DD_train) = c("pdbID","length","fold","fold_name","scopeID",pair_inter)
E210.DD_train_SS <-E210_seq_DD_train <-  E210.DD_train

for (NP in 1:Nprotein) {
  print(NP)
  pdbname <- DD_train$pdbID[NP]
  E210.DD_train$pdbID[NP] <- pdbname
  E210.DD_train_SS$pdbID[NP] <- pdbname
  E210_seq_DD_train$pdbID[NP] <- pdbname
  ch <- DD_train$chainID[NP]
  split_values <- strsplit(DD_train$scopeID[NP], "\\.")
  E210.DD_train_SS$fold[NP] <- sapply(split_values, function(x) paste(x[1:2], collapse = "."))
  E210.DD_train$fold[NP] <- E210.DD_train_SS$fold[NP]
  E210_seq_DD_train$fold[NP] <- E210.DD_train_SS$fold[NP]
  tryCatch({
    pdb0 <- read.pdb(pdbname)
    sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
    sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
    sele4 <- atom.select(pdb0, "noh")
    sele <- combine.select(sele1, sele2,sele4, operator="AND")
    if (!is.na(astral95.1domain_abcd$start[NP])){
      st <- astral95.1domain_abcd$start[NP]
      en <- astral95.1domain_abcd$end[NP]
      sele3 <- atom.select(pdb0,  resno=st:en, verbose=TRUE)
      sele <- combine.select(sele1, sele2,sele3,sele4, operator="AND")
    }
    pdbX <- pdb0$atom[sele$atom,]
    E210.DD_train$length[NP] = length(unique(pdbX$resno))
    E210.DD_train_SS$length[NP] = length(unique(pdbX$resno))
    E210_seq_DD_train$length[NP] = length(unique(pdbX$resno))
    Net_Frame <- NetworkFrame(PDB = pdb0, AtomSelect = sele)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210.DD_train[NP,6:215] <- energy210
    # ****************keep res helix & sheet************************
    indhs <- inedx_HelixSheet(PDB = pdb0)
    Net_Frame_hs <- Net_Frame[Net_Frame$resnoif %in% indhs & Net_Frame$resnojf %in% indhs,]
    if(dim(Net_Frame_hs)[1]!=0){
        Net_Frame <- Net_Frame_hs
    }
    energy210_ss <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210.DD_train_SS[NP,6:215] <- energy210_ss
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}
saveRDS(E210.DD_train,file = "Data/rds/E210.EDD.rds")
saveRDS(E210.DD_train_SS,file = "Data/rds/E210.EDD_SS.rds")

```


## Random Forest_EDD

```{r}
E210.EDD_SS <- readRDS("Data/rds/E210.EDD_SS.rds")
E210.EDD <- readRDS("Data/rds/E210.EDD.rds")

df = E210.EDD[!is.na(E210.EDD$FF),]
data<- cbind(class = df$fold,df[,-seq(1:5)]/df$length)
data$class <- as.factor(data$class)
table(data$class)
ind <- sample(2, nrow(data), replace = TRUE, prob = c(0.8, 0.2))
train <- data[ind==1,]
test <- data[ind==2,]
table(train$class)
table(test$class)
# one versus all classification#
rf_models <- list()
unique_classes <- unique(train$class)
for (class in unique_classes) {
  # Create a binary response variable for the OvA classification
  train$BinaryClass <- as.factor(ifelse(train$class == class, "Positive", "Negative"))
  data0 = train[,-1]
  # Train a Random Forest model for the current class
  rf_model <- tuneRF(data0, data0$BinaryClass, stepFactor = 1, improve = 0.1, trace=T, 
                   plot= T,doBest = T)
  # Store the model in the list
  rf_models[[class]] <- rf_model
}

class_probabilities <- matrix(0,nrow = nrow(test),ncol = length(unique_classes))
k = 1
for (i in unique_classes) {
  rf_model <- rf_models[[i]]
  test$BinaryClass <- as.factor(ifelse(test$class == i, "Positive", "Negative"))
  test0 = test[,-1]
  prediction <- predict(rf_model, test0, type = "prob")
  class_probabilities[,k] <- prediction[,2]  # Probability of the positive class
  k = k+1
}
colnames(class_probabilities) <- unique_classes
result = data.frame(class = test$class,class_probabilities)
predicted_class = data.frame(class = test$class, predict = character(nrow(result)))
for (j in 1:nrow(result)){
  predicted_class$predict[j] <- as.character(unique_classes[which.max(as.vector(result[j,-1]))])
}

different_rows <- predicted_class$class != predicted_class$predict
predicted_class[different_rows,]

```


## Energy_Calculator_TG
```{r}
energy_dell_dunbrack <- read.csv("Data/csv/energy.csv",header = FALSE,sep ="" )
energy_dell_dunbrack <- energy_dell_dunbrack[,5:33]
source("Functions/amacid2num.R")
source("Functions/atomtype2num167.R")
source("Functions/Energy_PC210.R")
source("Functions/NetworkFrame.R")
source("Functions/inedx_HelixSheet.R")
TG <- readRDS("Data/rds/TG.rds")
Nprotein <- nrow(TG)
# ------Pairwise Amino Acid names-------------
AA <- c("PHE","LEU","ILE","VAL","TRP","TYR","MET","CYS","HIS","THR",
        "ARG","ALA","ASN","GLN","PRO","SER","ASP","GLY","LYS","GLU")
letters_list <- aa321(AA)
pair_inter <- character(0)
for (i in 1:20) {
  for (j in i:20) {
    pair_inter <- c(pair_inter, paste0(letters_list[i], letters_list[j]))
  }
} 
#----------------------------------------------
num_columns <- 210
E210.seq_TG <- data.frame(pdbID = TG$pdbID,length = rep(NA,Nprotein),
                            fold = rep(NA,Nprotein),
                            scopeID = TG$scopeID, 
                                         matrix(NA, ncol = num_columns, nrow = length(TG$pdbID)))
colnames(E210.seq_TG) = c("pdbID","length","fold","scopeID",pair_inter)
E210.TG <- data.frame(pdbID = TG$pdbID,length = rep(NA,Nprotein),
                            fold = rep(NA,Nprotein),scopeID = TG$scopeID, 
                                         matrix(NA, ncol = num_columns, nrow = length(TG$pdbID)))
colnames(E210.TG) = c("pdbID","length","fold","scopeID",pair_inter)
E210.TG_SS <- E210.TG
for (NP in 1:Nprotein) {
  print(NP)
  pdbname <- TG$pdbID[NP]
  E210.TG$pdbID[NP] <- pdbname
  E210.seq_TG$pdbID[NP] <- pdbname
  E210.TG_SS$pdbID[NP] <- pdbname
  ch <- TG$chainID[NP]
  split_values <- strsplit(TG$scopeID[NP], "\\.")
  E210.TG_SS$fold[NP] <- sapply(split_values, function(x) paste(x[1:2], collapse = "."))
  E210.TG$fold[NP] <- E210.TG_SS$fold[NP]
  E210.seq_TG$fold[NP] <- E210.TG_SS$fold[NP]
  tryCatch({
    pdb0 <- read.pdb(pdbname)
    sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
    sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
    sele4 <- atom.select(pdb0, "noh")
    sele <- combine.select(sele1, sele2,sele4, operator="AND")
    if (!is.na(astral95.1domain_abcd$start[NP])){
      st <- astral95.1domain_abcd$start[NP]
      en <- astral95.1domain_abcd$end[NP]
      sele3 <- atom.select(pdb0,  resno=st:en, verbose=TRUE)
      sele <- combine.select(sele1, sele2,sele3,sele4, operator="AND")
    }
    pdbX <- pdb0$atom[sele$atom,]
    
    pdbXca <- pdbX[pdbX$elety == "CA",]
    seq = aa321(pdbXca$resid)
    E210.seq_TG$length[NP] = length(unique(pdbX$resno))
    freq = c()
    for (j in 1:20){
        freq[j] <-  length(grep(letters_list[j],seq))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
    E210.seq_TG[NP,6:215] <- aa210_p
    
    E210.TG$length[NP] = length(unique(pdbX$resno))
    E210.TG_SS$length[NP] = length(unique(pdbX$resno))
    E210.seq_TG$length[NP] = length(unique(pdbX$resno))
    Net_Frame <- NetworkFrame(PDB = pdb0, AtomSelect = sele)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210.TG[NP,6:215] <- energy210
    # ****************keep res helix & sheet************************
    indhs <- inedx_HelixSheet(PDB = pdb0)
    Net_Frame_hs <- Net_Frame[Net_Frame$resnoif %in% indhs & Net_Frame$resnojf %in% indhs,]
    if(dim(Net_Frame_hs)[1]!=0){
        Net_Frame <- Net_Frame_hs
    }
    energy210_ss <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210.TG_SS[NP,6:215] <- energy210_ss
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}
saveRDS(E210.TG,file = "Data/rds/E210.TG.rds")
saveRDS(E210.TG_SS,file = "Data/rds/E210.TG_SS.rds")
saveRDS(E210.seq_TG,file = "Data/rds/E210.seq_TG.rds")

```

## TOP40

```{r}
astral_40 <- readRDS("Data/rds/astral_40.rds")
E210.astral40_SS <- readRDS("Data/rds/E210.astral40_SS.rds")
E210.astral40 <- readRDS("Data/rds/E210.astral40.rds")
E210.seq_astral40 <- readRDS("Data/rds/E210.seq_astral40.rds")

scope_top40<-c('b.1','c.1','d.58','a.4','c.2','c.37','c.23','c.47',
               'b.40','b.34','c.55','c.94','a.118','d.15','c.66','g.3',
               'b.82','c.69','c.67','d.17','d.144','c.108','d.108','c.26',
               'a.60','b.29','b.36','b.55','c.93','a.39','a.24','c.3','d.38',
               'g.37','a.25','b.18','b.121','g.39','c.56','c.14','d.129','a.29','d.110')

E210.astral40_SS_top <- E210.astral40_SS[E210.astral40_SS$fold %in% scope_top40,]
E210.astral40_top <- E210.astral40[E210.astral40$fold %in% scope_top40,]
E210.seq_astral40_top <- E210.seq_astral40[E210.seq_astral40$fold %in% scope_top40,]
saveRDS(E210.seq_astral40_top,file = "Data/rds/E210.seq_astral40_top.rds")
saveRDS(E210.astral40_top,file = "Data/rds/E210.astral40_top.rds")
saveRDS(E210.astral40_SS_top,file = "Data/rds/E210.astral40_SS_top.rds")
```


## Random Forest_TOP40

```{r}
E210.astral40_top <- readRDS("Data/rds/E210.astral40_top.rds")
df = E210.seq_astral40_top[!is.na(E210.seq_astral40_top$X1),]


data<- cbind(class = df$fold,df[,-seq(1:10)]/df$length)
data$class <- as.factor(data$class)
table(data$class)
ind <- sample(2, nrow(data), replace = TRUE, prob = c(0.8, 0.2))
train <- data[ind==1,]
test <- data[ind==2,]
table(train$class)
table(test$class)
# one versus all classification#
rf_models <- list()
unique_classes <- unique(train$class)
for (class in unique_classes) {
  # Create a binary response variable for the OvA classification
  train$BinaryClass <- as.factor(ifelse(train$class == class, "Positive", "Negative"))
  data0 = train[,-1]
  # Train a Random Forest model for the current class
  rf_model <- tuneRF(data0, data0$BinaryClass, stepFactor = 1.2, improve = 0.1, trace=T, 
                   plot= T,doBest = T)
  # Store the model in the list
  rf_models[[class]] <- rf_model
}

class_probabilities <- matrix(0,nrow = nrow(test),ncol = length(unique_classes))
k = 1
for (i in unique_classes) {
  rf_model <- rf_models[[i]]
  test$BinaryClass <- as.factor(ifelse(test$class == i, "Positive", "Negative"))
  test0 = test[,-1]
  prediction <- predict(rf_model, test0, type = "prob")
  class_probabilities[,k] <- prediction[,2]  # Probability of the positive class
  k = k+1
}
colnames(class_probabilities) <- unique_classes
result = data.frame(class = test$class,class_probabilities)
predicted_class = data.frame(class = test$class, predict = character(nrow(result)))
for (j in 1:nrow(result)){
  predicted_class$predict[j] <- as.character(unique_classes[which.max(as.vector(result[j,-1]))])
}

different_rows <- predicted_class$class != predicted_class$predict
predicted_class[different_rows,]

```

# Sequence Level

## Astral40

```{r}
astral <- read.fasta('Data/csv/astral-scopedom-seqres-gd-sel-gs-bib-40-2.08.fa')
aaenergy <- read.csv("Data/csv/Pij.csv", header = F,sep = ";")
astral40 <- readRDS("Data/rds/astral_40.rds")

aaenergy <- (aaenergy + t(aaenergy))/2
colnames(aaenergy) = rownames(aaenergy)<- 
  c("F","L","I","V","W","Y","M","C","H","T","R","A","N","Q","P","S","D","G","K","E")
#row.names(aaenergy) <- aa321(aaenergy[,1])
#aaenergy <- aaenergy[,-1]
#row.names(aaenergy) <- colnames(aaenergy)
letters_list <- tolower(row.names(aaenergy))
nprotein <- length(astral)
aa210<-aaenergy[lower.tri(aaenergy,diag = T)]
num_columns <- 210
E210.seq_astral40 <- data.frame(pdbID = astral40$pdbID,length = rep(NA,nprotein),
                                         matrix(NA, ncol = num_columns, 
                                                nrow = length(astral40$pdbID)))
for (i in 1:nprotein){
  print(i)
  #seq = unlist(strsplit(tolower(DTI_df__seq$seq[i]),""))
  seq = astral[[i]]
  freq = c()
  for (j in 1:20){
    freq[j] <-  length(grep(letters_list[j],seq))
  }
  freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
  en_fr <- aaenergy * freq_matrix
  en_fr <- en_fr/length(seq)
  pair_es <- diag(freq) %*% as.matrix(en_fr)
  aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
  E210.seq_astral40[i,3:212] <- aa210_p
  E210.seq_astral40[i,2] <- length(seq)
}

E210.seq_astral40 <- cbind(astral40,E210.seq_astral40[,-c(1)])
saveRDS(E210.seq_astral40,file = "Data/rds/E210.seq_astral40.rds")
# ---------------------------------structure---------------------------------
E210.astral40 <- readRDS("Data/rds/E210.astral40.rds")
df = E210.astral40
df <- df[df$length > 100 & !is.na(df$length) & df$class %in% c("a","b"),]
df1 <- df[,-seq(1:10)]/(df$length)
pc <- prcomp(df1,scale. = F,center = T) 
Rot_E_abcd <- as.matrix(df1) 
E_PC_abcd <- data.frame(class = df$class,Rot_E_abcd) 
# ------------- UMAP ------------------------
ump.E_PC_abcd <- umap(E_PC_abcd[,-c(1)], n_neighbors = 10,metric = "manhattan",
            min_dist = .0001, random_state = 123)
#-------- UMAMP Visualization ---------------
umap.df.E_PC_abcd <- data.frame(
  sf = E_PC_abcd$class,
  UMAP1 = ump.E_PC_abcd$layout[, 1],
  UMAP2 = ump.E_PC_abcd$layout[, 2])
um.E_PC_abcd <- ggplot(umap.df.E_PC_abcd,aes(UMAP1, UMAP2, color= sf))+
  geom_point(alpha=.7) +
  mythem
#------------tSNE--------------------------
E_PC_abcd<-E_PC_abcd[!duplicated(E_PC_abcd[,-1]),]
tsne<-Rtsne(E_PC_abcd[,-c(1)],pca=F,dims=2,perplexity=100,
            verbose=T, max_iter=1000,theta=0,
            learning_rate=300,normalize=F)

tsne.df <- data.frame(
  group = E_PC_abcd$class,
  tSNE1=tsne$Y[,1],
  tSNE2=tsne$Y[,2]
  )

g.tsne <- ggplot(tsne.df,aes(tSNE1, tSNE2, color= group))+
  geom_point(size=1.5,alpha=.7)+
  mythem

# ---------------------------------Sequence---------------------------------
e_seq_str <- data.frame(eseq = rowSums(E210.seq_astral40[,-c(1:10)]),
                     estr = rowSums(E210.astral40[,-c(1:6)]))

e_seq_str <- cbind(E210.seq_astral40[,c(1:10)],e_seq_str)
cor.test(e_seq_str$eseq,e_seq_str$estr)
# ---------------------------------Sequence95---------------------------------

e_seq_str <- data.frame(eseq = rowSums(E210.seq_astral95[,-c(1:10)]),
                     estr = rowSums(E210.astral95[,-c(1:9)]))

e_seq_str <- cbind(E210.seq_astral95[,c(1:10)],e_seq_str)
e_seq_str1 <- e_seq_str[-grep(',',e_seq_str$seq),]
ggplot(e_seq_str1,aes(estr,eseq,col=class))+geom_jitter()
cor.test(e_seq_str1$eseq,e_seq_str1$estr)
```

### Random forest

```{r}
E210.seq_astral40 <- readRDS("Data/rds/E210.seq_astral40.rds")
df = E210.seq_astral40
#df <- df[!is.na(df$FF) & is.na(df$start) & df$length > 100 & df$class %in% c("a","b","c","d"),]
df <- df[ !is.na(df$length) & df$length > 100 & df$class %in% c("a","b","c","d"),]
data<- cbind(class = df$class,df[,-seq(1:10)])
#data <- data[data$class %in% c("a.1.1", "a.104.1"),]
data$class <- as.factor(data$class)
table(data$class)
ind <- sample(2, nrow(data), replace = TRUE, prob = c(0.8, 0.2))
train <- data[ind==1,]
test <- data[ind==2,]
rf <- randomForest(class~., data=train, proximity=TRUE)
p1 <- predict(rf, train)
confusionMatrix(p1, train$class)
plot(rf)
p2 <- predict(rf, test)
confusionMatrix(p2, test$class)
#cf <- createFolds(data$class,k =10,returnTrain = T)
#cv = c()
#for (i in 1:10){
#  ind = cf[[i]]
#  test = data[-ind,]
#  train = data[ind,]
#  bestmtry <- tuneRF(train,train$class,stepFactor = 1.2, improve = 0.01, trace=T, 
#                   plot= T,doBest = T) 
#  p3 <- predict(bestmtry, test)
# cn = confusionMatrix(p3, test$class)
# cv[i] = cn$overall[1]
#}
bestmtry <- tuneRF(train,train$class,stepFactor = 1.2, improve = 0.01, trace=T, 
                   plot= T,doBest = T) 

print(bestmtry)
p3 <- predict(bestmtry, test)
con <- confusionMatrix(p3, test$class)
con$byClass
hist(treesize(rf),
     main = "No. of Nodes for the Trees",
     col = "green")
varImpPlot(rf,
           sort = T,
           n.var = 10,
           main = "Top 10 - Variable Importance")
importance(rf)

MDSplot(rf, train$class)
MDSplot(rf, test$class)

```

## Ferritin-like

```{r}
ferritin <- read.csv("Data/csv/Ferritin_Like.csv",sep = ";")
energy_dell_dunbrack <- read.csv("Data/csv/energy.csv",header = FALSE,sep ="" )
energy_dell_dunbrack <- energy_dell_dunbrack[,5:33]
source("Functions/amacid2num.R")
source("Functions/atomtype2num167.R")
source("Functions/Energy_PC210.R")
Nprotein <- nrow(ferritin)
# ------Pairwise Amino Acid names-------------
AA <- c("PHE","LEU","ILE","VAL","TRP","TYR","MET","CYS","HIS","THR",
        "ARG","ALA","ASN","GLN","PRO","SER","ASP","GLY","LYS","GLU")
letters_list <- aa321(AA)
pair_inter <- character(0)
for (i in 1:20) {
  for (j in i:20) {
    pair_inter <- c(pair_inter, paste0(letters_list[i], letters_list[j]))
  }
} 
#----------------------------------------------
num_columns <- 210
E210_seq_ferritin <- data.frame(pdbID = ferritin$pdbID,length = rep(NA,Nprotein),group = ferritin$group,
                                         matrix(NA, ncol = num_columns, nrow = length(ferritin$pdbID)))
colnames(E210_seq_ferritin) = c("pdbID","length","group",pair_inter)
for (NP in 1:Nprotein) {
  print(NP)
  pdbname <- ferritin$pdbID[NP]
  ch <- ferritin$chainID[NP]
  E210_seq_ferritin$pdbID[NP] <- paste0(pdbname,ch)
  tryCatch({
    pdb0 <- read.pdb(pdbname)
    sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
    sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
    sele <- combine.select(sele1, sele2, operator="AND")
    if (!is.na(ferritin$start[NP])){
      st <- ferritin$start[NP]
      en <- ferritin$end[NP]
      sele3 <- atom.select(pdb0,  resno=st:en, verbose=TRUE)
      sele <- combine.select(sele1, sele2,sele3, operator="AND")
    }
    pdbX <- pdb0$atom[sele$atom,]
    pdbXca <- pdbX[pdbX$elety == "CA",]
    seq = aa321(pdbXca$resid)
    E210_seq_ferritin$length[NP] = length(unique(pdbX$resno))
    freq = c()
    for (j in 1:20){
        freq[j] <-  length(grep(letters_list[j],seq))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
    E210_seq_ferritin[NP,4:213] <- aa210_p
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}

saveRDS(E210_seq_ferritin,file = "Data/rds/E210_seq_ferritin.rds")

```


## Phylogenetic Tree_Ferritin

```{r}
E210_seq_ferritin <- readRDS("Data/rds/E210_seq_ferritin.rds")

df <- E210_seq_ferritin
df <- df[with(df, order(df$group)), ]
row.names(df) <- df[,1]
# PCA
#pc <- prcomp(df[,-seq(1:3)],scale. = F,center = T) 
#Rot_ferr <- as.matrix(df[,-seq(1:3)]) %*% pc$rotation[,1:5]
#dis <- dist(Rot_ferr, method="manhattan")
#dis <- log2(dis)
#------------------
rn <- df[,1]
dis <- dist(df[,-c(1,2,3)], method="manhattan")
#--------custom_distance--------------------
tm_score <- function(x, y){
  xy = x - y
  d =sum(1/ (1 + (xy)^2))
  return(d)
}
dis <- as.matrix(dist(df[,-c(1,2,3)], method = tm_score))
dis <- dist(df[,-c(1,2,3)], method = tm_score)
#--------------------------------------------
dis <- sqrt(dis)
#------------------
tree_data <- ape::bionj(dis)
grp <- list(Bacteri     = rn[1:5],
            BMM_alpha = rn[6:7],
            BMM_beta  = rn[8:10],
            Dps  = rn[11:25],
            Fatty_acid  = rn[26:29],
            Ferritins  = rn[30:40],
            NAA  = rn[41:45],
            RNRR2  = rn[46:53])

p_data <- ggtree(tree_data, layout = 'circular',
                 branch.length='none',size = 1.5) 

t = groupOTU(p_data, grp, "groups") + aes(color=groups) +
  theme(legend.position="right",
        legend.title = element_blank(),
        legend.text = element_text(size = 15, colour = "black"))+
  geom_tiplab2(aes(label=label,angle=angle),
               hjust=-.1,size=5,fontface="bold",align=TRUE)
```

## DD
```{r}
source("Functions/amacid2num.R")
source("Functions/atomtype2num167.R")
source("Functions/Energy_PC210.R")
source("Functions/NetworkFrame.R")
source("Functions/inedx_HelixSheet.R")
DD_train <- readRDS("Data/rds/DD_train.rds")
Nprotein <- nrow(DD_train)
aaenergy <- read.csv("Data/csv/Pij.csv", header = F,sep = ";")
aaenergy <- (aaenergy + t(aaenergy))/2
colnames(aaenergy) = rownames(aaenergy)<- 
  c("F","L","I","V","W","Y","M","C","H","T","R","A","N","Q","P","S","D","G","K","E")
letters_list <- tolower(row.names(aaenergy))
# ------Pairwise Amino Acid names-------------
AA <- c("PHE","LEU","ILE","VAL","TRP","TYR","MET","CYS","HIS","THR",
        "ARG","ALA","ASN","GLN","PRO","SER","ASP","GLY","LYS","GLU")
letters_list <- aa321(AA)
pair_inter <- character(0)
for (i in 1:20) {
  for (j in i:20) {
    pair_inter <- c(pair_inter, paste0(letters_list[i], letters_list[j]))
  }
} 
#----------------------------------------------
num_columns <- 210
E210.seq_DD_train <- data.frame(pdbID = DD_train$pdbID,length = rep(NA,Nprotein),
                            fold = rep(NA,Nprotein),fold_name = DD_train$fold_name,
                            scopeID = DD_train$scopeID, 
                                         matrix(NA, ncol = num_columns, nrow = length(DD_train$pdbID)))
colnames(E210.seq_DD_train) = c("pdbID","length","fold","fold_name","scopeID",pair_inter)
for (NP in 1:Nprotein) {
  print(NP)
  pdbname <- DD_train$pdbID[NP]
  E210.seq_DD_train$pdbID[NP] <- pdbname
  ch <- DD_train$chainID[NP]
  split_values <- strsplit(DD_train$scopeID[NP], "\\.")
  E210.seq_DD_train$fold[NP] <- sapply(split_values, function(x) paste(x[1:2], collapse = "."))
  tryCatch({
    pdb0 <- read.pdb(pdbname)
    sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
    sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
    sele <- combine.select(sele1, sele2, operator="AND")
    if (!is.na(DD_train$start[NP])){
      st <- DD_train$start[NP]
      en <- DD_train$end[NP]
      sele3 <- atom.select(pdb0,  resno=st:en, verbose=TRUE)
      sele <- combine.select(sele1, sele2,sele3, operator="AND")
    }
    pdbX <- pdb0$atom[sele$atom,]
    pdbXca <- pdbX[pdbX$elety == "CA",]
    seq = aa321(pdbXca$resid)
    E210.seq_DD_train$length[NP] = length(unique(pdbX$resno))
    freq = c()
    for (j in 1:20){
        freq[j] <-  length(grep(letters_list[j],seq))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
    E210.seq_DD_train[NP,6:215] <- aa210_p
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}
saveRDS(E210.seq_DD_train,file = "Data/rds/E210.seq_DD_train.rds")

```

## Random Forest_DD

```{r}
E210.seq_DD_train <- readRDS("Data/rds/E210.DD_train.rds")
E210.seq_DD_test <- readRDS("Data/rds/E210.DD_test.rds")

df = E210.seq_DD_train
df = df[!is.na(df$FF),]
data<- cbind(class = df$fold,df[,-seq(1:8)]/df$length)
test <- cbind(class = E210.seq_DD_test$fold,E210.seq_DD_test[,-seq(1:8)]/E210.seq_DD_test$length)
#data <- data[data$class %in% c("a.1.1", "a.104.1"),]
data$class[data$class == "b.18"] = "b.1"
data$class[data$class == "c.93"] = "c.94"

data$class <- as.factor(data$class)
table(data$class)
# one versus all classification#
rf_models <- list()
unique_classes <- unique(data$class)
for (class in unique_classes) {
  # Create a binary response variable for the OvA classification
  data$BinaryClass <- as.factor(ifelse(data$class == class, "Positive", "Negative"))
  data0 = data[,-1]
  # Train a Random Forest model for the current class
  rf_model <- my_tuneRF(data0[,-211], data0$BinaryClass, stepFactor = 1.5, improve = 0.1, trace=T, 
                   plot= T,doBest = T)
  # Store the model in the list
  rf_models[[class]] <- rf_model
}
E210.seq_DD_test = E210.seq_DD_test[!is.na(E210.seq_DD_test$FF),]
test <- cbind(class = E210.seq_DD_test$fold,E210.seq_DD_test[,-seq(1:8)]/E210.seq_DD_test$length)
class_probabilities <- matrix(0,nrow = nrow(test),ncol = length(unique_classes))
k = 1
test$class[test$class == "c.93"] = "c.94"

for (i in unique_classes) {
  rf_model <- rf_models[[i]]
  test$BinaryClass <- as.factor(ifelse(test$class == i, "Positive", "Negative"))
  test0 = test[,-1]
  prediction <- predict(rf_model, test0, type = "prob")
  class_probabilities[,k] <- prediction[,2]  # Probability of the positive class
  k = k+1
}
colnames(class_probabilities) <- unique_classes
result = data.frame(class = test$class,class_probabilities)
predicted_class = data.frame(class = test$class, predict = character(nrow(result)))
for (j in 1:nrow(result)){
  predicted_class$predict[j] <- as.character(unique_classes[which.max(as.vector(result[j,-1]))])
}

different_rows <- predicted_class$class != predicted_class$predict
predicted_class[different_rows,]
```

## EDD
```{r}
source("Functions/amacid2num.R")
source("Functions/atomtype2num167.R")
source("Functions/Energy_PC210.R")
source("Functions/NetworkFrame.R")
source("Functions/inedx_HelixSheet.R")
DD_train <- readRDS("Data/rds/EDD.rds")
Nprotein <- nrow(DD_train)
aaenergy <- read.csv("Data/csv/Pij.csv", header = F,sep = ";")
aaenergy <- (aaenergy + t(aaenergy))/2
colnames(aaenergy) = rownames(aaenergy)<- 
  c("F","L","I","V","W","Y","M","C","H","T","R","A","N","Q","P","S","D","G","K","E")
letters_list <- tolower(row.names(aaenergy))
# ------Pairwise Amino Acid names-------------
AA <- c("PHE","LEU","ILE","VAL","TRP","TYR","MET","CYS","HIS","THR",
        "ARG","ALA","ASN","GLN","PRO","SER","ASP","GLY","LYS","GLU")
letters_list <- aa321(AA)
pair_inter <- character(0)
for (i in 1:20) {
  for (j in i:20) {
    pair_inter <- c(pair_inter, paste0(letters_list[i], letters_list[j]))
  }
} 
#----------------------------------------------
num_columns <- 210
E210.seq_DD_train <- data.frame(pdbID = DD_train$pdbID,length = rep(NA,Nprotein),
                            fold = rep(NA,Nprotein),fold_name = DD_train$fold_name,
                            scopeID = DD_train$scopeID, 
                                         matrix(NA, ncol = num_columns, nrow = length(DD_train$pdbID)))
colnames(E210.seq_DD_train) = c("pdbID","length","fold","fold_name","scopeID",pair_inter)
for (NP in 1:Nprotein) {
  print(NP)
  pdbname <- DD_train$pdbID[NP]
  E210.seq_DD_train$pdbID[NP] <- pdbname
  ch <- DD_train$chainID[NP]
  split_values <- strsplit(DD_train$scopeID[NP], "\\.")
  E210.seq_DD_train$fold[NP] <- sapply(split_values, function(x) paste(x[1:2], collapse = "."))
  tryCatch({
    pdb0 <- read.pdb(pdbname)
    sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
    sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
    sele <- combine.select(sele1, sele2, operator="AND")
    if (!is.na(DD_train$start[NP])){
      st <- DD_train$start[NP]
      en <- DD_train$end[NP]
      sele3 <- atom.select(pdb0,  resno=st:en, verbose=TRUE)
      sele <- combine.select(sele1, sele2,sele3, operator="AND")
    }
    pdbX <- pdb0$atom[sele$atom,]
    pdbXca <- pdbX[pdbX$elety == "CA",]
    seq = aa321(pdbXca$resid)
    E210.seq_DD_train$length[NP] = length(unique(pdbX$resno))
    freq = c()
    for (j in 1:20){
        freq[j] <-  length(grep(letters_list[j],seq))
    }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
    E210.seq_DD_train[NP,6:215] <- aa210_p
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}
saveRDS(E210.seq_DD_train,file = "Data/rds/E210.seq_EDD.rds")

```
## Random Forest_EDD

```{r}
E210.EDD <- readRDS("Data/rds/E210.seq_EDD.rds")

df = E210.EDD[!is.na(E210.EDD$FF),]
data<- cbind(class = df$fold,df[,-seq(1:8)])
data$class <- as.factor(data$class)
table(data$class)
ind <- sample(2, nrow(data), replace = TRUE, prob = c(0.8, 0.2))
train <- data[ind==1,]
test <- data[ind==2,]
table(train$class)
table(test$class)
# one versus all classification#
rf_models <- list()
unique_classes <- unique(train$class)
for (class in unique_classes) {
  # Create a binary response variable for the OvA classification
  train$BinaryClass <- as.factor(ifelse(train$class == class, "Positive", "Negative"))
  data0 = train[,-1]
  # Train a Random Forest model for the current class
  rf_model <- my_tuneRF(data0[,], data0$BinaryClass, stepFactor = 1.2, improve = 0.1, trace=T, 
                   plot= T,doBest = T)
  # Store the model in the list
  rf_models[[class]] <- rf_model
}

class_probabilities <- matrix(0,nrow = nrow(test),ncol = length(unique_classes))
k = 1
for (i in unique_classes) {
  rf_model <- rf_models[[i]]
  test$BinaryClass <- as.factor(ifelse(test$class == i, "Positive", "Negative"))
  test0 = test[,-1]
  prediction <- predict(rf_model, test0, type = "prob")
  class_probabilities[,k] <- prediction[,2]  # Probability of the positive class
  k = k+1
}
colnames(class_probabilities) <- unique_classes
result = data.frame(class = test$class,class_probabilities)
predicted_class = data.frame(class = test$class, predict = character(nrow(result)))
for (j in 1:nrow(result)){
  predicted_class$predict[j] <- as.character(unique_classes[which.max(as.vector(result[j,-1]))])
}

different_rows <- predicted_class$class != predicted_class$predict
predicted_class[different_rows,]

```

## Aspartase

```{r}
E210.astral95 <- readRDS("Data/rds/E210.astral95.rds")
E210.astral95_SS <- readRDS("Data/rds/E210.astral95_SS.rds")
E210.seq_astral95 <- readRDS("Data/rds/E210.seq_astral95.rds")

aspartaze <- c("a.127.1.0","a.127.1.1","a.127.1.2")
E210.seq_aspartaze <- E210.astral95[E210.astral95$family %in% aspartaze,]
E210.aspartaze <- E210.astral95[E210.astral95$family %in% aspartaze,]
E210.aspartaze_SS <- E210.astral95_SS[E210.astral95_SS$family %in% aspartaze,]


scope_2.08 <- read.delim("Data/csv/dir.cla.scope.2.08-stable.txt", header=FALSE, comment.char="#")
colnames(scope_2.08) <- c("scope","pdb","chain","scopeID","description")
scope_aspartaze <- scope_2.08[scope_2.08$scopeID %in% aspartaze,]


data<- cbind(class = df$family,df[,-seq(1:10)])
#data <- data[data$class %in% c("a.1.1", "a.104.1"),]
data$class <- as.factor(data$class)
table(data$class)
ind <- sample(2, nrow(data), replace = TRUE, prob = c(0.8, 0.2))
train <- data[ind==1,]
test <- data[ind==2,]
bestmtry <- tuneRF(train,train$class,stepFactor = 1.5, improve = 0.0001, trace=T, 
                   plot= T,doBest = T) 

print(bestmtry)
p3 <- predict(bestmtry, test)
con <- confusionMatrix(p3, test$class)
```

## Drug_target

```{r}
dti_barabasi <- read_excel("Data/csv/dti_barabasi.xlsx")
colnames(dti_barabasi) <- c("drug","entrez")
#dti_barabasi <- dti_barabasi[1:100,]
dti_barabasi$entrez <- as.character(dti_barabasi$entrez)
library(rentrez)
get_protids_from_geneids <- function(gene_ids) {
  
  # Get the protein elink.
  protein_elink <- rentrez::entrez_link(dbfrom = "gene", id = gene_ids, db = "protein")
  
  # Get the protein id (refseq database)
  protein_ids <- protein_elink$links$gene_protein_refseq
  protein_ids
}

make_aa_fasta <- function(prot_ids, nameFile) {
  # Make a multi-fasta file with each protein id
  sapply(prot_ids, function(x) {
    protein_esummary <- rentrez::entrez_summary(db = "protein", id = x)
    protein_fasta <- rentrez::entrez_fetch(db = "protein", id = protein_esummary$uid, rettype = "fasta")
    # save amino acid sequences into a FASTA file ("nameFile"")
    write(protein_fasta, file = paste(nameFile, ".fasta", sep = ""), append = TRUE)
    #return(protein_fasta)
  } )
}
#-------------------------------------
dti_seq <- readRDS("Data/rds/DTI_df_seq.rds")
sAB <- readRDS("Data/rds/sAB.rds")

aaenergy <- read.csv("Data/csv/Pij.csv", header = F,sep = ";")
aaenergy <- (aaenergy + t(aaenergy))/2
colnames(aaenergy) = rownames(aaenergy)<- 
  c("F","L","I","V","W","Y","M","C","H","T","R","A","N","Q","P","S","D","G","K","E")
letters_list <- tolower(row.names(aaenergy))
nprotein <- nrow(dti_seq)
aa210<-aaenergy[lower.tri(aaenergy,diag = T)]
num_columns <- 210
E210.seq_dti <- data.frame(pdbID = dti_seq$drug,length = rep(NA,nprotein),
                                         matrix(NA, ncol = num_columns, 
                                                nrow = nrow(dti_seq)))
for (i in 1:nprotein){
  print(i)
  #seq = unlist(strsplit(tolower(DTI_df__seq$seq[i]),""))
  seq = unlist(strsplit(tolower(dti_seq$seq[i]), ""))
  freq = c()
  for (j in 1:20){
    freq[j] <-  length(grep(letters_list[j],seq))
  }
  freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
  en_fr <- aaenergy * freq_matrix
  en_fr <- en_fr/length(seq)
  pair_es <- diag(freq) %*% as.matrix(en_fr)
  aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
  E210.seq_dti[i,3:212] <- aa210_p
  E210.seq_dti[i,2] <- length(seq)
}
df0 = E210.seq_dti
df = cbind(drug = df[,1],df[,-c(1:2)]/df$length)
df <- df %>%
  group_by(drug) %>%
 dplyr::summarise(across(X1:X210, mean))

x  <- as.matrix(df[,-c(1)]) 
row.names(x) = df$drug
dist_matrix <- as.matrix(dist(x, method = "manhattan"))
dist_df <- as.data.frame(as.table(dist_matrix))
colnames(dist_df) <- c("d1", "d2", "distance")
dist_df <- dist_df[dist_df$d1 != dist_df$d2, ]


```

## Bacteriocin

```{r}
bacteriocin <- readRDS("Data/rds/bacteriocin.rds")
aaenergy <- read.csv("Data/csv/Pij.csv", header = F,sep = ";")
aaenergy <- (aaenergy + t(aaenergy))/2
colnames(aaenergy) = rownames(aaenergy)<-c("F","L","I","V","W","Y","M","C","H","T","R","A","N","Q","P","S","D","G","K","E")
letters_list <- tolower(row.names(aaenergy))
nprotein <- nrow(bacteriocin)
aa210<-aaenergy[lower.tri(aaenergy,diag = T)]
num_columns <- 210
E210.seq_bacteriocin <- data.frame(class = bacteriocin$class,length = rep(NA,nprotein),
                                   matrix(NA, ncol = num_columns,nrow = nrow(bacteriocin)))
for (i in 1:nprotein){
  print(i)
  seq = unlist(strsplit(tolower(bacteriocin$seq[i]), ""))
  freq = c()
  for (j in 1:20){
  freq[j] <-  length(grep(letters_list[j],seq))
  }
freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
en_fr <- aaenergy * freq_matrix
en_fr <- en_fr/length(seq)
pair_es <- diag(freq) %*% as.matrix(en_fr)
aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
E210.seq_bacteriocin[i,3:212] <- aa210_p
E210.seq_bacteriocin[i,2] <- length(seq)
}

saveRDS(E210.seq_bacteriocin,'Data/rds/E210.seq_bacteriocin.rds')
E210.seq_bacteriocin = readRDS("Data/rds/E210.seq_bacteriocin.rds")
df = E210.seq_bacteriocin
df1 <- df[,-seq(1:1)]

dis <- as.matrix(proxy::dist(df1, method = manhat_norm))
ump <- umap(d = dis,
            n_neighbors = 40,
            min_dist = 0.01,
            input="dist")

ump_df <- data.frame(df[,1:2],
                     UMAP1 = ump$layout[, 1],
                     UMAP2 = ump$layout[, 2])
ggplot(ump_df,aes(UMAP1, UMAP2, color= class)) +
   geom_point(alpha=.7)



Rot_E_abcd <- as.matrix(df1)
E_PC_abcd <- data.frame(class = df$class,Rot_E_abcd)
# ------------- UMAP ------------------------
ump.E_PC_abcd <- umap(E_PC_abcd[,-c(1)], n_neighbors = 25,metric = "manhattan",
                      min_dist = 0.1, random_state = 123)
#-------- UMAMP Visualization ---------------
umap.df.E_PC_abcd <- data.frame(
sf = E_PC_abcd$class,
UMAP1 = ump.E_PC_abcd$layout[, 1],
UMAP2 = ump.E_PC_abcd$layout[, 2])
ggplot(umap.df.E_PC_abcd,aes(UMAP1, UMAP2, color= sf))+
geom_point(alpha=.7) +
mythem

#----------tree
df = E210.seq_bacteriocin
len = as.numeric(df$length)
x = df[,-c(1:2)]/df$length
x  <- as.matrix(x)
hc <- x %>% dist(method = 'manhattan') %>% hclust(method = "ward.D2")
dend <- hc %>% as.dendrogram
labcol <- c( "blue", "green","red")
ord_sml <- df[,][order.dendrogram(dend),1]
group_cols <- labcol[as.factor(ord_sml)]
dend2 <- dend %>%
#color_branches(k=3, col = c("red","blue","green3")) %>%
color_labels(col =   group_cols) %>%
set("labels_cex", 0.5)
plot(dend2, ylab = "Height")
legend("topright", legend = as.character(unique(ord_sml)),
pch = 15, pt.cex = 2.3, cex = 1.4, bty = 'n',
inset = c(-.15, -.1), # place outside
col = unique(group_cols),horiz = FALSE,xpd = TRUE)
#------------boxplot
df = E210.seq_bacteriocin
x = df[,-c(1:1)]
hc <- as.matrix(x %>% dist(method = manhat_norm))
class1 = hc[1:499,1:499]
class1=class1[lower.tri(class1)]
class2 = hc[500:728,500:728]
class2=class2[lower.tri(class2)]
class3 = hc[729:821,729:821]
class3=class3[lower.tri(class3)]
class12 = hc[500:728,1:499]
class12 = hc[500:728,1:499]
class12=class12[lower.tri(class12)]
class13 = hc[729:821,1:499]
class13=class13[lower.tri(class13)]
class23 = hc[729:821,500:728]
class23=class23[lower.tri(class23)]
df<-data.frame(comp=c(rep('same',sum(length(class1),length(class2),length(class3))),
rep('dif',sum(length(class12),length(class13),length(class23)))),
dis=c(class1,class2,class3,class12,class13,class23))
ggplot(df,aes(comp,dis))+geom_boxplot()
df$scale_dis<-1/(1+df$dis)


```
## Remote homology

```{r}
df <- readRDS("data/rds/E210.seq_astral40.rds")
df = df[!is.na(df$FL),]
fa1 = "a.25.1.0"
fa2 = "a.25.1.2"
df_fa = df[df$family %in% c(fa1,fa2),]
dis_fa <- as.matrix(df_fa[,-c(1:8)] %>% dist(method = 'manhattan'))
table(df_fa$family)
# *****
class1 = dis_fa[1:15,1:15]
class1=class1[lower.tri(class1)]
class2 = dis_fa[16:26,16:26]
class2=class2[lower.tri(class2)]
class12 = dis_fa[1:15,16:26]
class12=as.vector(class12)

df_family<-data.frame(gr=rep(c('within','between'),
                             c(length(class1)+length(class2),length(class12))),
                      distance=c(class1,class2,class12))

g1<-ggplot(df_family,aes(gr,distance))+geom_boxplot()+ggtitle('a.25.1.0, a.25.1.2')
# *****************
sf1<-'a.29.2'
sf2<-'a.29.3'
df_sf = df[df$superfamily %in% c(sf1,sf2),]
dis_sf <- as.matrix(df_sf[,-c(1:8)] %>% dist(method = 'manhattan'))
table(df_sf$superfamily)
# *****
class1 = dis_sf[1:31,1:31]
class1=class1[lower.tri(class1)]
class2 = dis_sf[32:50,32:50]
class2=class2[lower.tri(class2)]
class12 = dis_sf[1:31,32:50]
class12=as.vector(class12)

df_superfamily<-data.frame(gr=rep(c('within','between'),
                             c(length(class1)+length(class2),length(class12))),
                      distance=c(class1,class2,class12))

g2<-ggplot(df_superfamily,aes(gr,distance))+geom_boxplot()+ggtitle('a.29.2, a.29.3')
ggarrange(g1,g2)
```

## fold
```{r}
df <- readRDS("Data/rds/E210.astral40.rds")
df <- df[!is.na(df$FL) & df$class %in% c('a','b','c','d'), ]
folds <- unique(df$fold)

dis <- as.matrix(dist(df[,-c(1:8)],method = tm_score3))
dis <- data.frame(gr=df$fold, dis)
fo<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(fo)) {  
  print(i)   
  x <- fo[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}

median(within,na.rm = T)
median(between,na.rm = T)

bet<-between[sample(length(between),10000)]
wit<-within[sample(length(within),10000)]
t.test(wit,bet)

df_fold <- data.frame(gr=rep(c('within','between'),c(length(wit),length(bet))),
                    distance=c(wit,bet))
ggplot(df_fold,aes(gr,distance,fill=gr))+geom_boxplot(alpha=.3)


```

## cusomize distance

```{r}
library(umap)
data <- matrix(runif(100), ncol = 2)
custom_distance_matrix <- dist(data, method = "euclidean")
umap_result <- umap(d = as.matrix(custom_distance_matrix))
plot(umap_result$layout, 
     main = "UMAP with Custom Distance Matrix",
     n_neighbors = 10,
     min_dist = 0.1, 
     random_state = 123)

```

# Mulidup
```{r}
mulidup <- read.delim('Data/rds/dup/dup.txt')

dis<-rep(0, nrow(mulidup))
len <- data.frame(len1=rep(0,nrow(mulidup)), len2=rep(0,nrow(mulidup)))
for(i in 1:nrow(mulidup)){
  print(i)
  path <- paste0('~/Desktop/rds/dup/', mulidup$Pair.Name[i])
  pdb1 <- read.pdb(paste0(path ,'/',mulidup$Domain1[i],'.pdb'))
  pdb2 <- read.pdb(paste0(path ,'/',mulidup$Domain2[i],'.pdb'))
  ch1 <- unique(pdb1$atom$chain)
  ch2 <- unique(pdb2$atom$chain)
  
  sele1 <- atom.select(pdb1,"protein",type = "ATOM", chain=ch1)
  sele2 <- atom.select(pdb1, "protein", elety=c("H","OXT"),inverse=TRUE)
  sele4 <- atom.select(pdb1, "noh")
  sele <- combine.select(sele1, sele2,sele4, operator="AND")
  pdb1 <- trim.pdb(pdb1, sele,verbose = FALSE)
  
  sele1 <- atom.select(pdb2,"protein",type = "ATOM", chain=ch2)
  sele2 <- atom.select(pdb2, "protein", elety=c("H","OXT"),inverse=TRUE)
  sele4 <- atom.select(pdb2, "noh")
  sele <- combine.select(sele1, sele2,sele4, operator="AND")
  pdb2 <- trim.pdb(pdb2, sele,verbose = FALSE)
  
  len[i,1] <- length(unique(pdb1$atom$resno))
  len[i,2] <- length(unique(pdb2$atom$resno))
  
  E210_2pdb <- data_frame(p1=rep(0,210),p2=rep(0,210))
  
  Net_Frame <- NetworkFrame(PDB = pdb1)
  energy210 <- Energy_PC210(Net_Frame ,energy_dell_dunbrack)
  E210_2pdb$p1 <- energy210
  
  Net_Frame <- NetworkFrame(PDB = pdb2)
  energy210 <- Energy_PC210(Net_Frame, energy_dell_dunbrack)
  E210_2pdb$p2 <- energy210
  
  df <- data.frame(t(E210_2pdb))
  df <- df/c(len$len1[i], len$len2[i])
  dis[i] <- dist(df,method = tm_score3)
}

len$mean<-(len$len1+len$len2)/2
len$min <- apply(len[,1:2],1,min)
len$max <- apply(len[,1:2],1,max)
len$diff <- len$max-len$min


E210.astral40 <- readRDS("Data/rds/E210.seq_astral40.rds") 
df40 <- E210.astral40[sample(nrow(E210.astral40),1000),]
dis40 <- as.vector(dist(df40[,-c(1:8)]/df40$length,
                        method = tm_score3))


df <- data.frame(group=rep(c('rend40','mulidup'),c(241,241)),
                 distance=c(dis40[sample(1000,241)], dis))
ggplot(df,aes(group,distance,fill=group))+geom_boxplot()


tm_score3 <- function(x, y){
  xy = abs(x - y)
  xy = xy[xy != 0]
  d = sum(xy)/length(xy)
  d <- d/max(.5,d)
  return(d)
}
```

