---
title: "Energetic Profile-Based Protein Comparison: A New and Fast Approach for Structural,
  and Evolutionary analysis"
author: "Peyman & Mehdi"
date: "2024-02-19"
output:
  pdf_document: default
  html_document:
    fig_caption: true
fontsize: 11pt
geometry: margin=1in
graphics: true
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,warning=FALSE, fig.pos= "h")
```

# Library   
```{r libraries, include=FALSE}
options(knitr.table.format = "latex")
library(readr)
library(data.table)
library(MASS)
library(tidyverse)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(ggpointdensity)
library(ggpubr)
library(igraph)
library(fpc)
library(grid)
library(gridExtra)
library(reshape2)
library(visNetwork)
library(pheatmap)
library(umap)
library(Rtsne)
library(viridis)
library(viridisLite)
library(bio3d)
library(caret)
library(e1071)
library(randomForest)
library(seqinr)
library(geometry)
library(ape)
library(colorspace)
library(dendextend)
library(dynamicTreeCut)
library(caret)
library(ggtree)
library(phangorn)
library(tanggle)
library(stringr)
library(protr)
library(proxy)
library(class)
library(readxl)
library(ggh4x)
library(googledrive)
```


# ASTRAL 95/40 SPE_CPE

This chunk calculate the profile of energy for protein domains in ASTRAL95/40 based on sequence and structure.

```{r Astrals_CPE_SPE}
source('Functions.R')
# ----------------------------------------------------
# ----------- astral 95 ------
astral <- readRDS("Data/rds/astral_95.rds")
Nprotein <- nrow(astral)
num_columns <- 210
E210.astral_95 <- data.frame(pdbID = astral$pdbID,
                            scopeID= astral$scopeID,
                            length = rep(NA,Nprotein),
                            position = astral$Position,
                            class = astral$class,
                            fold = astral$fold,
                            superfamily = astral$superfamily,
                            family = astral$family,
                            matrix(NA, ncol = num_columns, nrow = length(astral$pdbID)))
colnames(E210.astral_95)[9:218] <- pair_inter
E210.seq_astral_95 <- E210.astral_95 
for (NP in 1:nrow(E210.seq_astral_95)) {
  if (NP %% 50 == 0) {print(NP)}
  chain_resno <- split_position(astral$Position[NP])
  pdbname <- astral$pdbID[NP]
  tryCatch({
    pdb0 <- read.pdb(pdbname,verbose = FALSE)
    pdblist = list()
    for (ind in 1:dim(chain_resno)[1]){
      sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=chain_resno$chain[ind])
      sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
      sele4 <- atom.select(pdb0, "noh")
      sele <- combine.select(sele1, sele2,sele4, operator="AND")
      if (!(chain_resno$start[ind]=="NA") & !(chain_resno$end[ind])=="NA"){
         st <- chain_resno$start[ind]
         en <- chain_resno$end[ind]
         sele3 <- atom.select(pdb0,  resno=st:en, verbose=FALSE)
         sele <- combine.select(sele1, sele2,sele3,sele4, operator="AND")
      }
      pdblist[[ind]] <- trim.pdb(pdb0, sele,verbose = FALSE)
    }
    new_pdb = pdblist[[1]]
    if (dim(chain_resno)[1] > 1) {
      for (ind in 2:dim(chain_resno)[1]){
      new_pdb <- cat.pdb(new_pdb, pdblist[[ind]], rechain=FALSE)
      }
    }
    pdbX <- new_pdb$atom
    E210.astral_95$length[NP] = length(unique(pdbX$resno))
    Net_Frame <- NetworkFrame(PDB = new_pdb)
    energy210 <- Energy_SPE210(Net_Frame,energy_dell_dunbrack)
    E210.astral_95[NP,-c(1:8)] <- energy210
# ----------- sequence enregy based on astral --------
    seq = astral$seq[NP]
    seq = gsub("X", "", seq)
    E210.seq_astral_95$length[NP] = nchar(seq)
    aa210_p <- Energy_CPE210(seq)
    E210.seq_astral_95[NP,9:218] <- aa210_p
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}
# -----------------------------------------------------------
saveRDS(E210.astral_95, file = "Data/rds/E210.astral_95.rds")
saveRDS(E210.seq_astral_95, file = "Data/rds/E210.seq_astral_95.rds")
# -----------------------------------------------------------
# ----------- astral 40 ------
astral <- readRDS("Data/rds/astral_40.rds") 
Nprotein <- nrow(astral)
num_columns <- 210
E210.astral_40 <- data.frame(pdbID = astral$pdbID,
                            scopeID= astral$scopeID,
                            length = rep(NA,Nprotein),
                            position = astral$Position,
                            class = astral$class,
                            fold = astral$fold,
                            superfamily = astral$superfamily,
                            family = astral$family,
                            matrix(NA, ncol = num_columns, nrow = length(astral$pdbID)))
colnames(E210.astral_40)[9:218] <- pair_inter
E210.seq_astral_40 <- E210.astral_40
for (NP in 1:nrow(E210.seq_astral_40)) {
  if (NP %% 50 == 0) {print(NP)}
  chain_resno <- split_position(astral$Position[NP])
  pdbname <- astral$pdbID[NP]
  tryCatch({
    pdb0 <- read.pdb(pdbname,verbose = FALSE)
    pdblist = list()
    for (ind in 1:dim(chain_resno)[1]){
      sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=chain_resno$chain[ind])
      sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
      sele4 <- atom.select(pdb0, "noh")
      sele <- combine.select(sele1, sele2,sele4, operator="AND",verbose = FALSE)
      if (!(chain_resno$start[ind]=="NA") & !(chain_resno$end[ind])=="NA"){
         st <- chain_resno$start[ind]
         en <- chain_resno$end[ind]
         sele3 <- atom.select(pdb0,  resno=st:en, verbose=FALSE)
         sele <- combine.select(sele1, sele2,sele3,sele4, operator="AND")
      }
      pdblist[[ind]] <- trim.pdb(pdb0, sele,verbose = FALSE)
    }
    new_pdb = pdblist[[1]]
    if (dim(chain_resno)[1] > 1) {
      for (ind in 2:dim(chain_resno)[1]){
      new_pdb <- cat.pdb(new_pdb, pdblist[[ind]], rechain=FALSE)
      }
    }
    pdbX <- new_pdb$atom
    E210.astral_40$length[NP] = length(unique(pdbX$resno))
    Net_Frame <- NetworkFrame(PDB = new_pdb)
    energy210 <- Energy_SPE210(Net_Frame,energy_dell_dunbrack)
    E210.astral_40[NP,-c(1:8)] <- energy210
    # ----------- sequence enregy based on astral --------
    seq = astral$seq[NP]
    seq = gsub("X", "", seq)
    E210.seq_astral_40$length[NP] = nchar(seq)
    aa210_p <- Energy_CPE210(seq)
    E210.seq_astral_40[NP,9:218] <- aa210_p
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}

saveRDS(E210.astral_40,file = "Data/rds/E210.astral_40.rds")
saveRDS(E210.seq_astral_40,file = "Data/rds/E210.seq_astral_40.rds")
```


# C-terminal & Homing CPE
This chunk calculates the CPE of proteins from two superfamilies: the C-terminal domain in the DNA helicase RuvA subunit and the Homing endonucleases.

```{r C-terminal_Homing_CPE}
source('Functions.R')
# ----------------------------------------------------
df <- read.csv('Data/csv/CT_Ho_cathID_filtered.csv')

E210.seq_CT_Ho <- data.frame(df,
                         matrix(NA, ncol = 210, nrow = nrow(df)))

colnames(E210.seq_CT_Ho)[-c(1:9)] <- pair_inter

tictoc::tic()
for(NP in 1:nrow(E210.seq_CT_Ho)){
  print(NP)
  tryCatch({
      seq <- E210.seq_CT_Ho$seq[NP]
      aa210_p <- Energy_CPE210(seq)
      E210.seq_CT_Ho[NP,10:219] <- aa210_p
      }, error = function(e) {
       cat("Error downloading file:", e$message, "\n")
        }
  )
}
df2 <- E210.seq_CT_Ho
df2 <- df2[!is.na(df2$FF),]
dis <- as.matrix(proxy::dist(df2[,-c(1:9)], method = manhat))
tictoc::toc()
saveRDS(E210.seq_CT_Ho,'Data/rds/E210.seq_CT_Ho.rds')
```


# Five superfamily CPE

This chunk calculates the energy profile (CPE) for five distinct SCOP superfamilies: winged helix (a.4.5), PH domain-like (b.55.1), NTF-like (d.17.4), Ubiquitin-like (d.15.1), and Immunoglobulins (b.1.1).

```{r Five superfamily_CPE}
source('Functions.R')
# ----------------------------------------------------
df5 <- read.csv('Data/csv/fiveSF.csv')
Nprotein <- nrow(df5)
#----------------------------------------------
num_columns <- 210
E210.seq_df5 <- data.frame(df5,
                       matrix(NA,ncol=num_columns,nrow=Nprotein))

colnames(E210.seq_df5)[-c(1:9)] <- pair_inter
tictoc::tic()
for (NP in 1:Nprotein) {
  print(NP)
  seq <- E210.seq_df5$seq[NP]
  aa210_p <- Energy_CPE210(seq)
  E210.seq_df5[NP,10:219] <- aa210_p
}
df2 <- E210.seq_df5
df2 <- df2[!is.na(df2$FF),]
dis <- as.matrix(proxy::dist(df2[,-c(1:9)], method = manhat))
tictoc::toc()
saveRDS(E210.seq_df5,file = "Data/rds/E210.seq_fiveSF.rds")
```

# Ferritin-like Superfamily SPE

This segment computes the SPE for proteins belonging to the Ferritin-like superfamily. The PDB IDs of these proteins are located in the `Data\csv\Ferritin_Like.csv` file.

```{r Ferritin-like Superfamily}
source('Functions.R')
# ----------------------------------------------------
ferritin <- read.csv("Data/csv/Ferritin_Like.csv",sep = ",")
Nprotein <- nrow(ferritin)
#----------------------------------------------
E210_ferritin <- data.frame(pdbID = ferritin$pdbID,
                            length = rep(NA,Nprotein),
                            group = ferritin$group,
                            matrix(NA, ncol = 210, 
                                   nrow = Nprotein))
colnames(E210_ferritin)[-c(1:3)] <- pair_inter


for (NP in 1:Nprotein) {
  print(NP)
  pdbname <- ferritin$pdbID[NP]
  ch <- ferritin$chainID[NP]
  E210_ferritin$pdbID[NP] <- paste0(pdbname,ch)
  tryCatch({
    pdb0 <- read.pdb(pdbname)
    sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
    sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
    sele4 <- atom.select(pdb0, "noh")
    sele <- combine.select(sele1, sele2,sele4, operator="AND")
    if (!is.na(ferritin$start[NP])){
      st <- ferritin$start[NP]
      en <- ferritin$end[NP]
      sele3 <- atom.select(pdb0,  resno=st:en, verbose=TRUE)
      sele <- combine.select(sele1, sele2,sele3,sele4, operator="AND")
    }
    new_pdb <- trim.pdb(pdb0, sele,verbose = FALSE)
    pdbX <- new_pdb$atom
    pdbX <- pdbX[pdbX$elety == "CA",]
    E210_ferritin$length[NP] <- nrow(pdbX)
    Net_Frame <- NetworkFrame(PDB = new_pdb)
    energy210 <- Energy_SPE210(Net_Frame,energy_dell_dunbrack)
    E210_ferritin[NP,4:213] <- energy210
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}
saveRDS(E210_ferritin,file = "Data/rds/E210_ferritin.rds")

```

# Covid Spike SPE_CPE

This segment computes the SPE and CPE for spike glycoprotein proteins. The PDB IDs of these proteins are located in the 'Data/csv/spike.csv' file. 

```{r Covid_Spike_Proteins}
source('Functions.R')
# ----------------------------------------------------
path <- 'Data/covidPDB/'
spike <- read.csv(paste0(path,"spike_seq.csv"))
spike$length <- nchar(spike$seq)
Nprotein <- nrow(spike)
#----------------------------------------------
E210_spike <- data.frame(spike,
                            matrix(NA,ncol=210,nrow=Nprotein))

colnames(E210_spike)[-c(1:4)] <- pair_inter

tictoc::tic()
for (NP in 1:Nprotein) {
  print(NP)
  ch <- substr(E210_spike$pdbID[NP],5,5)
  pdbname <- substr(E210_spike$pdbID[NP],1,4)
  tryCatch({
    pdb0 <- read.pdb(paste0(path,tolower(pdbname),'.pdb'),
                     rm.alt = T,verbose = FALSE)
    sele1 <- atom.select(pdb0,"protein",type = "ATOM", chain=ch)
    sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
    sele4 <- atom.select(pdb0, "noh")
    sele <- combine.select(sele1,sele2,sele4, operator="AND",verbose = FALSE)
    # -------------------
    new_pdb <- trim.pdb(pdb0, sele,verbose = FALSE)
    Net_Frame <- NetworkFrame(PDB = new_pdb)
    energy210 <- Energy_SPE210(Net_Frame,energy_dell_dunbrack)
    E210_spike[NP, -c(1:4)] <- energy210
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}
dis <- E210_spike[,-c(1:4)] %>% proxy::dist(method = manhat)
tictoc::toc()

E210.seq_spike <- data.frame(spike,
                            matrix(NA,ncol=210,nrow=Nprotein))
colnames(E210.seq_spike)[-c(1:4)] <- pair_inter

tictoc::tic()
for (NP in 1:Nprotein) {
  print(NP)
  seq <- E210.seq_spike$seq[NP]
  aa210_p <- Energy_CPE210(seq)
  E210.seq_spike[NP,5:214] <- aa210_p
}
dis <- E210.seq_spike[,-c(1:4)] %>% proxy::dist(method = manhat)
tictoc::toc()

saveRDS(E210_spike,file = "Data/rds/E210_spike.rds")
saveRDS(E210.seq_spike,file = "Data/rds/E210.seq_spike.rds")
#write.csv(E210.seq_spike[,1:3], "Data/covidPDB/spike_seq.csv",row.names = F)
```


# Bacteriocin CPE

This section calculates the CPE and pairwise distances between Bacteriocin proteins. The file 'Data/csv/TM_vec_bac.csv' was obtained by requesting it from the authors of the paper 'Protein remote homology detection and structural alignment using deep learning'. This file contains results from AlphaFold2, OmegaFold, ESMFold, and TM-Vec software tools for comparison purposes. The 'bacteriocin_tm_vec' file includes their results along with those generated using the CPE method.

```{r Bacteriocin}
source('Functions.R')
# ----------------------------------------------------
tm <- read.csv('Data/csv/TM_vec_bac.csv') 
# ---------------------------------------------------------
# ------- ENERGY
# ---------------------------------------
seqIDs <- unique(c(tm$ID_x, tm$ID_y))
idx <- match(seqIDs, tm$ID_x)
idy <- match(seqIDs[is.na(idx)], tm$ID_y)
tmx <- tm[idx[!is.na(idx)], c(1,7,8,9,14)]
tmy <- tm[idy, c(2,10,11,12,16)]
colnames(tmx)<-colnames(tmy)<-gsub('_','',str_match(names(tmx),'.*_'))
tm_unique <- rbind(tmx, tmy)
#write.csv(tm_unique, 'Data/csv/Bacteriocin.csv', row.names = FALSE)
# ------------------------------------
# add our tm-vec
disTM <- read.csv('Data/csv/disTM_vec_bac.csv')
diag(disTM) <- 1
colnames(disTM) <- rownames(disTM) <- tm_unique$ID
disTM <- melt(as.matrix(disTM))
colnames(disTM) <- c('ID_x','ID_y','tm_vec')
tm <- left_join(tm,disTM,c('ID_x','ID_y'))
# ---------------------------------
nprotein <- nrow(tm_unique)
tm_unique <- data.frame(tm_unique, 
                        length=nchar(tm_unique$Seq),
                        matrix(NA,nrow=nprotein,ncol=210))
colnames(tm_unique)[-c(1:6)] <- pair_inter

tictoc::tic()
for (i in 1:nprotein){
  print(i)
  seq <- tm_unique$Seq[i]
  aa210_p <- Energy_CPE210(seq)
  tm_unique[i,7:216] <- aa210_p
}
df <- tm_unique[,-c(1:6)]
rownames(df) <- tm_unique$ID
x <- as.matrix(df)
dis <- x %>% proxy::dist(method = manhat)
tictoc::toc()
dis <- as.matrix(dis)
dis <- melt(dis)
colnames(dis) <- c('ID_x','ID_y', 'CPE_dis')
tm <- left_join(tm,dis,c('ID_x','ID_y'))
tm$CPE_dis <- (tm$CPE_dis - min(tm$CPE_dis))/(max(tm$CPE_dis)-min(tm$CPE_dis))
tm$CPE_dis <- 1-(tm$CPE_dis)
tm$CPE_dis <- (tm$CPE_dis) - 0.36

tm$status <- if_else(tm$class_x==tm$class_y,'same class','different class')
tm$status2 <- if_else(tm$Subclass_x==tm$Subclass_y&
                        tm$class_x=='Class_1'&tm$class_y=='Class_1',
                      'same class(sub class1)',NA)
saveRDS(tm,'Data/rds/bacteriocin_tm_vec.rds')
saveRDS(tm_unique,'Data/rds/E210.seq_bacteriocin.rds')
```

# Drug Target CPE
This section retrieves the protein sequences (drug targets) utilized in the study:
"Cheng F, Kovács IA, Barabási A-L. Network-based prediction of drug combinations. Nature communications 10, 1197 (2019)". 
The drug-target information downloaded from its supplementary information (`41467_2019_9186_MOESM4_ESM.xlsx`, `41467_2019_9186_MOESM4_ESM.xlsx`)
This section calculates the CPE of protein targets, using the sequences retrieved in the "Protein_Sequence_Target" Chunk.


```{r Protein_Targets_CPE}
source('Functions.R')
# ----------------------------------------------------
df_seq <- readRDS('Data/rds/DTI_df_seq.rds')

E210.seq_dti <- data.frame(df_seq,matrix(NA,nrow(df_seq),210))
colnames(E210.seq_dti)[-c(1:5)] <- pair_inter
for (NP in 1:nrow(df_seq)) {
  if(NP%%10==0)print(NP)
  seq <- E210.seq_dti$seq_aa[NP]
  aa210_p <- Energy_CPE210(seq)
  E210.seq_dti[NP,-c(1:5)] <- aa210_p
}
#saveRDS(E210.seq_dti,'Data/rds/E210.seq_dti.rds')
```
