---
title: "Figure and table"
author: "peyman"
date: "2024-03-12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Library   
```{r libraries, include=FALSE}
#library(kableExtra)
options(knitr.table.format = "latex")
set.seed(123)
library(readr)
library(data.table)
library(MASS)
library(tidyverse)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(ggpointdensity)
library(ggpubr)
library(igraph)
library(fpc)
library(grid)
library(gridExtra)
library(reshape2)
library(visNetwork)
library(pheatmap)
library(umap)
library(Rtsne)
library(viridis)
library(viridisLite)
library(bio3d)
library(caret)
library(e1071)
library(randomForest)
library(seqinr)
library(geometry)
library(ape)
library(colorspace)
library(dendextend)
library(dynamicTreeCut)
library(caret)
library(ggtree)
library(phangorn)
library(tanggle)
library(stringr)
library(protr)
library(proxy)
library(class)
library(readxl)
library(ggh4x)
library(mclust)
library(googledrive)
library(Biostrings)
library(msa)
library(binom)
library(lsa)
library(xlsx)
library(openxlsx)
```

# Figures
## Fig 2

```{r Fig 2}
source('Functions.R')
# ----------------------------------------------------
# ************** total energy 40 and 95
df_tot <- readRDS('Fig_Manuscript/fig_data/df_total_energy.rds')
df_tot$data <- ifelse(df_tot$data=='Astral 40','ASTRAL40','ASTRAL95')

A <- ggplot(df_tot[df_tot$data=='ASTRAL40',], aes(seq/2000, str/1000))+
       geom_pointdensity(adjust = .1,alpha=.3,show.legend = T,size = .5)+
       geom_smooth(method = 'glm',col='red',linewidth = .5)+
       stat_cor(method = "pearson", size=1.7, color = "red", geom = "label") +
       scale_color_viridis()+
  facet_wrap(vars(data))+
       theme_bw(base_line_size = .3)+
       xlab('Total energy (sequence)')+ylab('Total energy (structure)')+
  mytheme()+
  guides(color = guide_colorbar(barwidth = .5, barheight = 2))+
  theme(axis.text.x = element_text(size=5, face = "bold"),
        axis.text.y = element_text(size=5, face = "bold"),
        legend.text = element_text(size = 5, colour = "black"),
        strip.text.x = element_text(size = 5),
        axis.title.x = element_text(size = 5),
        axis.title.y = element_text(size = 5))

B <- ggplot(df_tot[df_tot$data=='ASTRAL95',], aes(seq/2000, str/1000))+
       geom_pointdensity(adjust = .1,alpha=.3,show.legend = T,size = .5)+
       geom_smooth(method = 'glm',col='red',linewidth = .5)+
       stat_cor(method = "pearson", size=1.7, color = "red", geom = "label") +
       scale_color_viridis()+
  facet_wrap(vars(data))+
       theme_bw(base_line_size = .3)+
       xlab('Total energy (sequence)')+ylab('Total energy (structure)')+
  mytheme()+
  guides(color = guide_colorbar(barwidth = .5, barheight = 2))+
  theme(axis.text.x = element_text(size=5, face = "bold"),
        axis.text.y = element_text(size=5, face = "bold"),
        legend.text = element_text(size = 5, colour = "black"),
        strip.text.x = element_text(size = 5),
        axis.title.x = element_text(size = 5),
        axis.title.y = element_text(size = 5))

# ----------SourceData---------------
df_tot$cpe <- df_tot$seq/2000
df_tot$spe <- df_tot$str/1000
MYwriteData(data=df_tot[,-c(2:5)],
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 2 A-B",
            row.Names = TRUE)
# ************** distance 40 and 95
df_dis40<-readRDS('Fig_Manuscript/fig_data/dis_40_manhat.rds')
df_dis95<-readRDS('Fig_Manuscript/fig_data/dis_95_manhat.rds')
df_dis <- rbind(df_dis40, df_dis95)
df_dis$data <- ifelse(df_dis$data=='Astral 40','ASTRAL40','ASTRAL95')

C <- ggplot(df_dis[df_dis$data=='ASTRAL40',], aes(dis_seq,dis_str))+
       geom_pointdensity(adjust = .1,alpha=.3,show.legend = T,size = .5)+
       geom_smooth(method = 'glm',col='red',linewidth = .5)+
       stat_cor(method = "pearson", size=1.7, color = "red", geom = "label") +
       scale_color_viridis()+facet_wrap(vars(data))+
       theme_bw(base_line_size = .3)+
       xlab('Distance (CPE)')+ylab('Distance (SPE)')+
       mytheme()+
  guides(color = guide_colorbar(barwidth = .5, barheight = 2))+
  theme(axis.text.x = element_text(size=5, face = "bold"),
        axis.text.y = element_text(size=5, face = "bold"),
        legend.text = element_text(size = 5, colour = "black"),
        strip.text.x = element_text(size = 5),
        axis.title.x = element_text(size = 5),
        axis.title.y = element_text(size = 5))

D <- ggplot(df_dis[df_dis$data=='ASTRAL95',], aes(dis_seq,dis_str))+
       geom_pointdensity(adjust = .1,alpha=.3,show.legend = T,size = .5)+
       geom_smooth(method = 'glm',col='red',linewidth = .5)+
       stat_cor(method = "pearson", size=1.7, color = "red", geom = "label") +
       scale_color_viridis()+facet_wrap(vars(data))+
       theme_bw(base_line_size = .3)+
       xlab('Distance (CPE)')+ylab('Distance (SPE)')+
       mytheme()+
  guides(color = guide_colorbar(barwidth = .5, barheight = 2))+
  theme(axis.text.x = element_text(size=5, face = "bold"),
        axis.text.y = element_text(size=5, face = "bold"),
        legend.text = element_text(size = 5, colour = "black"),
        strip.text.x = element_text(size = 5),
        axis.title.x = element_text(size = 5),
        axis.title.y = element_text(size = 5))

# ----------SourceData---------------
colnames(df_dis)[-1] <- c('dis_spe','dis_cpe')
MYwriteData(data=df_dis,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 2 D-E",
            row.Names = TRUE)
# ----------------------------------------
res <- readRDS('Data/rds/Length_Vs_Energy.rds')
res <- res[!is.na(res$length),]
p1 <- ggplot(res, aes(length, disTot))+
  #geom_point(alpha=.3)+
  geom_pointdensity(adjust = .1,alpha=.3,show.legend = T,size = .5)+
  scale_color_viridis()+
  geom_smooth(method = 'glm',col='red',linewidth = .5)+
  stat_cor(method = "pearson", size=1.7,
           color = "red", geom = "label") +
  theme_bw(base_line_size = .3)+
  mytheme()+
  ylab('Distance of Totoal Energy')+
  xlab('Length')+
  #xlim(0,1000)+
  ylim(-.5,.5)+
  guides(color = guide_colorbar(barwidth = .5, barheight = 2))+
  theme(axis.text.x = element_text(size=5, face = "bold"),
        axis.text.y = element_text(size=5, face = "bold"),
        legend.text = element_text(size = 5, colour = "black"),
        strip.text.x = element_text(size = 5),
        axis.title.x = element_text(size = 5),
        axis.title.y = element_text(size = 5))
# ----------SourceData---------------
MYwriteData(data=res[res$disTot>-.5&res$disTot<.5,1:3],
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 2 C",
            row.Names = TRUE)
# -------------------------
length_var <- res$length
other_vars <- res[, 4:213]

cor_values <- numeric(ncol(other_vars))

for (i in 1:ncol(other_vars)) {
  test <- cor.test(length_var, other_vars[, i],
                   method = 'spearman')
  cor_values[i] <- test$estimate
}

cor_results <- data.frame(Correlation = cor_values)

p2 <- ggplot(cor_results, aes(x = Correlation)) +
  geom_histogram(aes(fill = ..density..),
                 binwidth = 0.03, color = "black", alpha = 0.7) +
  scale_fill_viridis()+
  labs(#title = "Histogram of Correlation Values",
       x = "Correlation Coefficient",
       y = "Count") +
  mytheme()+
  theme(legend.key.width = unit(.3, "cm"),
        legend.key.height = unit(0.2, "cm"),
        axis.text.x = element_text(size=5, face = "bold"),
        axis.text.y = element_text(size=5, face = "bold"),
        legend.text = element_text(size = 5, colour = "black"),
        strip.text.x = element_text(size = 5),
        axis.title.x = element_text(size = 5),
        axis.title.y = element_text(size = 5))
# ----------SourceData---------------
MYwriteData(data=cor_results,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 2 F",
            row.Names = TRUE)
# --------------------------
p<-ggarrange(NULL,NULL,NULL,NULL,NULL,
             A,NULL,B,NULL,p1,
             NULL,NULL,NULL,NULL,NULL,
             C,NULL,D,NULL,p2,
             ncol = 5, nrow = 4,
             heights = c(.09, 1, .09, 1),
             widths = c(1, .09, 1, .09, 1),
             labels = c('', '', '', '', '',
                        'A', '', 'B', '', 'C',
                        '', '', '', '', '',
                        'D', '', 'E', '', 'F'),
             font.label = list(size = 15),    # color = "red"
             hjust = -.5, vjust = 0
             )
#ggsave(filename = "Fig2.pdf", plot = p,
#       width = 180, height = 100, units = "mm")
# save FIG2.pdf 16x9
```



## Fig 3 A

```{r Fig 3 A}
source('Functions.R')
# ----------------------------------------------------
# ******** str40
df <- readRDS("Data/rds/E210.astral_40.rds")
df <- df[df$fold%in%c('a.100','a.104'),]
df <- df[!is.na(df$FF),]
df <- df[,c(6,3,9:218)]

dis <- as.matrix(proxy::dist(df[,-c(1,2)], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_str <- data.frame(data='Astral 40',type='SPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])

# ******** seq40
df <- readRDS("Data/rds/E210.seq_astral_40.rds")
df <- df[df$fold%in%c('a.100','a.104'),]
df <- df[!is.na(df$FF),]
df <- df[,c(6,3,9:218)]

dis <- as.matrix(proxy::dist(df[,-c(1,2)], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_seq <- data.frame(data='Astral 40',type='CPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])
z <- rbind(ump_df_str,ump_df_seq)
z$data <- 'ASTRAL40'
z$class <- ifelse(z$class=='a.100', 'a.100    ','a.104')
z40 <- z

p1 <- ggplot(z,aes(UMAP1, UMAP2, fill= class)) +
  geom_point(alpha=.7, size=3, stroke = 1, colour = "black",shape = 21) +
  facet_nested(~ data+type,independent = T,scales = T)+
  ylab('')+xlab('')+
  mytheme()+
  theme(axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 18))+
  guides(color = guide_legend(override.aes = list(size = 3)))+
  scale_fill_manual(values = c('green', 'purple'))
# ---------------------
# ******** str95
df <- readRDS("Data/rds/E210.astral_95.rds")
df <- df[df$fold%in%c('a.100','a.104'),]
df <- df[!is.na(df$FF),]
df <- df[,c(6,3,9:218)]

dis <- as.matrix(proxy::dist(df[,-c(1,2)], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_str <- data.frame(data='Astral 95',type='SPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])

# ******** seq95
df <- readRDS("Data/rds/E210.seq_astral_95.rds")
df <- df[df$fold%in%c('a.100','a.104'),]
df <- df[!is.na(df$FF),]
df <- df[,c(6,3,9:218)]

dis <- as.matrix(proxy::dist(df[,-c(1,2)], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_seq <- data.frame(data='Astral 95',type='CPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])
z <- rbind(ump_df_str,ump_df_seq)
z$data <- 'ASTRAL95'
z$class <- ifelse(z$class=='a.100', 'a.100    ','a.104')
z95 <- z

p2 <- ggplot(z,aes(UMAP1, UMAP2, fill= class)) +
  geom_point(alpha=.7, size=3, stroke = 1, colour = "black",shape = 21) +
  facet_nested(~ data+type,independent = T,scales = T)+
  ylab('')+xlab('')+
  mytheme()+
  theme(axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 18))+
  guides(color = guide_legend(override.aes = list(size = 3)))+
  scale_fill_manual(values = c('green', 'purple'))

A<-ggarrange(NULL,NULL,NULL,NULL,
             NULL,p1,NULL,p2,nrow = 2,ncol = 4,
          common.legend = T,legend = 'right',
          widths = c(.1,1, 0.0, 1),
          heights = c(.07,1),
          labels = c('','','','',
                     '',"A", "", ""),
  font.label = list(size = 30),
  hjust = .2,vjust = 0)
#saveRDS(A,'Fig_Manuscript/Old_Figures/Fig4A.rds')
saveRDS(A,'Fig_Manuscript/fig_data/Fig3A.rds')
# ----------SourceData---------------
ump <- rbind(z40,z95)
MYwriteData(data=ump,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 3 A",
            row.Names = TRUE)
```

## Fig 3 B

```{r Fig 3 B}
source('Functions.R')
# ----------------------------------------------------
# ******** str40
df <- readRDS("Data/rds/E210.astral_40.rds")
df <- df[df$superfamily%in%c('a.29.2','a.29.3'),]
df <- df[!is.na(df$FF),]
df <- df[,c(7,3,9:218)]

dis <- as.matrix(proxy::dist(df[,-c(1,2)], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_str <- data.frame(data='Astral 40',type='SPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])

# ******** seq40
df <- readRDS("Data/rds/E210.seq_astral_40.rds")
df <- df[df$superfamily%in%c('a.29.2','a.29.3'),]
df <- df[!is.na(df$FF),]
df <- df[,c(7,3,9:218)]

dis <- as.matrix(proxy::dist(df[,-c(1,2)], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_seq <- data.frame(data='Astral 40',type='CPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])
z <- rbind(ump_df_str,ump_df_seq)
z$data <- 'ASTRAL40'
z$class <- ifelse(z$class=='a.29.2', 'a.29.2   ','a.29.3')
z40 <- z

p1 <- ggplot(z,aes(UMAP1, UMAP2, fill= class)) +
  geom_point(alpha=.7, size=3, stroke = 1, colour = "black",shape = 21) +
    facet_nested(~ data+type,independent = T,scales = T)+
  ylab('')+xlab('')+
  mytheme()+
  theme(axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 18))+
  guides(color = guide_legend(override.aes = list(size = 3)))+
  scale_fill_manual(values = c('green', 'purple'))
# ---------------------
# ******** str95
df <- readRDS("Data/rds/E210.astral_95.rds")
df <- df[df$superfamily%in%c('a.29.2','a.29.3'),]
df <- df[!is.na(df$FF),]
df <- df[,c(7,3,9:218)]

dis <- as.matrix(proxy::dist(df[,-c(1,2)], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_str <- data.frame(data='Astral 95',type='SPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])

# ******** seq95
df <- readRDS("Data/rds/E210.seq_astral_95.rds")
df <- df[df$superfamily%in%c('a.29.2','a.29.3'),]
df <- df[!is.na(df$FF),]
df <- df[,c(7,3,9:218)]

dis <- as.matrix(proxy::dist(df[,-c(1,2)], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df_seq <- data.frame(data='Astral 95',type='CPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])
z <- rbind(ump_df_str,ump_df_seq)
z$data <- 'ASTRAL95'
z$class <- ifelse(z$class=='a.29.2', 'a.29.2   ','a.29.3')
z95 <- z

p2 <- ggplot(z,aes(UMAP1, UMAP2, fill= class)) +
  geom_point(alpha=.7, size=3, stroke = 1, colour = "black",shape = 21) +
    facet_nested(~ data+type,independent = T,scales = T)+
  ylab('')+xlab('')+
  mytheme()+
  theme(axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 18))+
  guides(color = guide_legend(override.aes = list(size = 3)))+
  scale_fill_manual(values = c('green', 'purple'))

B <- ggarrange(NULL,NULL,NULL,NULL,
             NULL,p1,NULL,p2,nrow = 2,ncol = 4,
          common.legend = T,legend = 'right',
          widths = c(.1,1, 0.0, 1),
          heights = c(.07,1),
          labels = c('','','','',
                     '',"B", "", ""),
  font.label = list(size = 30),
  hjust = .2,vjust = 0)
saveRDS(B,'Fig_Manuscript/fig_data/Fig3B.rds')
# ----------SourceData---------------
ump <- rbind(z40,z95)
MYwriteData(data=ump,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 3 B",
            row.Names = TRUE)
```

## Fig 3 C

```{r Fig 3 C}
source('Functions.R')
# ----------------------------------------------------
# ******** str40
df <- readRDS("Data/rds/E210.astral_40.rds")
df <- df[df$family%in%c('a.25.1.0','a.25.1.2'),]
df <- df[!is.na(df$FF),]
df <- df[,c(8,3,9:218)]

dis <- as.matrix(proxy::dist(df[,-1], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 10,
            min_dist = 0.1,
            input="dist")
ump_df_str <- data.frame(data='Astral 40',type='SPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])

# ******** seq40
df <- readRDS("Data/rds/E210.seq_astral_40.rds")
df <- df[df$family%in%c('a.25.1.0','a.25.1.2'),]
df <- df[!is.na(df$FF),]
df <- df[,c(8,3,9:218)]

dis <- as.matrix(proxy::dist(df[,-c(1,2)], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 10,
            min_dist = 0.1,
            input="dist")
ump_df_seq <- data.frame(data='Astral 40',type='CPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])
z <- rbind(ump_df_str,ump_df_seq)
z$data <- 'ASTRAL40'
z40 <- z

p1 <- ggplot(z,aes(UMAP1, UMAP2, fill= class)) +
  geom_point(alpha=.7, size=3, stroke = 1, colour = "black",shape = 21) +
    facet_nested(~ data+type,independent = T,scales = T)+
  ylab('')+xlab('')+
  mytheme()+
  theme(axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 18))+
  guides(color = guide_legend(override.aes = list(size = 3)))+
  scale_fill_manual(values = c('green', 'purple'))
# ---------------------
# ******** str95
df <- readRDS("Data/rds/E210.astral_95.rds")
df <- df[df$family%in%c('a.25.1.0','a.25.1.2'),]
df <- df[!is.na(df$FF),]
df <- df[,c(8,3,9:218)]

dis <- as.matrix(proxy::dist(df[,-c(1,2)], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 10,
            min_dist = 0.1,
            input="dist")
ump_df_str <- data.frame(data='Astral 95',type='SPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])

# ******** seq95
df <- readRDS("Data/rds/E210.seq_astral_95.rds")
df <- df[df$family%in%c('a.25.1.0','a.25.1.2'),]
df <- df[!is.na(df$FF),]
df <- df[,c(8,3,9:218)]

dis <- as.matrix(proxy::dist(df[,-c(1,2)], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 10,
            min_dist = 0.1,
            input="dist")
ump_df_seq <- data.frame(data='Astral 95',type='CPE',
                           class=df[,1],
                           UMAP1 = ump$layout[, 1],
                           UMAP2 = ump$layout[, 2])
z <- rbind(ump_df_str,ump_df_seq)
z$data <- 'ASTRAL95'
z95 <- z

p2 <- ggplot(z,aes(UMAP1, UMAP2, fill= class)) +
  geom_point(alpha=.7, size=3, stroke = 1, colour = "black",shape = 21) +
    facet_nested(~ data+type,independent = T,scales = T)+
  ylab('')+xlab('')+
  mytheme()+
  theme(axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        legend.title = element_blank(),
        legend.text = element_text(size = 18))+
  guides(color = guide_legend(override.aes = list(size = 3)))+
  scale_fill_manual(values = c('green', 'purple'))

C <- ggarrange(NULL,NULL,NULL,NULL,
             NULL,p1,NULL,p2,nrow = 2,ncol = 4,
          common.legend = T,legend = 'right',
          widths = c(.1,1, 0.0, 1),
          heights = c(.07,1),
          labels = c('','','','',
                     '',"C", "", ""),
  font.label = list(size = 30),
  hjust = .2,vjust = 0)
saveRDS(C,'Fig_Manuscript/fig_data/Fig3C.rds')
# ----------SourceData---------------
ump <- rbind(z40,z95)
MYwriteData(data=ump,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 3 C",
            row.Names = TRUE)
```

##Fig 3 ABC

```{r Fig 3 ABC}
rm(list = ls())
A <- readRDS('Fig_Manuscript/fig_data/Fig3A.rds')
B <- readRDS('Fig_Manuscript/fig_data/Fig3B.rds')
C <- readRDS('Fig_Manuscript/fig_data/Fig3C.rds')
# ------------------------
p<-ggarrange(NULL,A,B,C,ncol = 1,
          heights = c(0.07,1,1,1))

ggsave(filename = "Fig3.pdf", plot = p,
       width = 180, height = 200, units = "mm")
# save FIG3.pdf 10x10
```


## Fig 4 AB 

```{r Fig 4 AB CT_Ho}
source('Functions.R')
# -----------------------------------
# time_for_SPE  = 187 Sec
# accuracy_for_SPE = 99
# Extracted from 'SPE for Ct_Ho' chunk
# ----------------------------------------------------
df <- data.frame(method=c('GR-Align','RMSD','TM-Score',
                          'YH(10Rotation)','YH(2500Rotation)','TM-Vec','CPE',
                          'SPE'),
                 Accuracy=c(62.3, 59.2, 61.5, 70.8, 81.5, 100, 100,
                            99),
                 TimeS=c(120, 3600, 33600, 600, 18600, 67, 1,
                         187))
# ------------------------------------------------------------------------
# ---------   the values for  `GR-Align`, `RMSD`, `TM-Score`, `YH(10Rotation)`, `YH(2500Rotation)` 
# ----------      were previously reported in the `Yau.S et.al. Journal of Biomolecular Structure and Dynamics (2018)`.
# --------------------------------------------------------------------------
df$Time <- round((df$TimeS*100)/(12*3600),5)
df$ratio<-log10(df$Accuracy/df$Time+1)

df <- df[order(df$ratio, decreasing = T),]
df$method <- factor(df$method, 
                    levels = df$method)
mycol <-c('magenta','orange','cyan','skyblue3','purple','blue','yellow4','green')


p12h <- ggplot(df, aes(x = method, y = Time, fill= method)) +
  geom_bar(stat="identity")+
  coord_polar(theta="y") +
  theme_light(base_rect_size = 0,base_line_size = 1)+
  theme(axis.text = element_blank(),
        axis.ticks=element_blank(),
        axis.title = element_blank(),
        legend.title = element_blank(),
        #legend.position = 'none'
        )+
  scale_fill_manual(values = mycol)+
  ylim(0,100)

p80s <- p12h+xlim(c('CPE','TM-Vec'))+
  ylim(0,.186)+ #1.86*12*3.6 = 80sec
  theme(legend.text = element_text(face = 'bold',size = 15))

df2 <- df
df2$Time <- df2$TimeS+1

p7 <- ggplot(df2,aes(Accuracy,Time,col=method,label=method))+
  geom_point(size=13,show.legend = F)+    
  geom_text(show.legend = F,vjust=.7,hjust=1.2,
            col='black',size=13, fontface='bold')+
  mytheme()+ylab('Time (s)')+
  scale_color_manual(values = mycol)+
  scale_y_log10()+xlim(50,103)+
  theme(axis.text.x = element_text(face = 'bold',color = 'black',size = 30),
        axis.text.y = element_text(face = 'bold',color = 'black',size = 30),
        axis.title.x = element_text(face = 'bold',color = 'black',size = 40),
        axis.title.y = element_text(face = 'bold',color = 'black',size = 40))
# ---------------------------------------------------------------
# ----------SourceData---------------
MYwriteData(data=df2[,1:3],
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 4 A-B",
            row.Names = TRUE)
```

## Fig 4 C 

```{r Fig 4 C Globin}
source('Functions.R')
# ------------------------
load('Data/Globin/dis_CPE_SPE_time.RData')
df <- read.csv('Data/Globin/Globin.csv')
df$globin <- gsub('-.*','',df$globin)
# ---------------------
tm <- read.csv('Data/Globin/disTM_vec_Globin.csv')
colnames(tm)<-rownames(tm)<-df$pdbID
tm <- 1 - tm
diag(tm) <- NA
t_tm <- 4.91
# ------------------tm----------------
ump_tm <- umap(d = tm,
            n_neighbors = 13,
            min_dist = 0.1,
            input="dist")
ump_tm <- ump_tm$layout
ump_tm <- data.frame(pdbID=rownames(ump_tm), 
                  UMAP1=ump_tm[,1],
                  UMAP2=ump_tm[,2])
ump_tm <- left_join(df, ump_tm,'pdbID')
# ------------------cpe----------------
ump_cpe <- umap(d = dis_CPE,
            n_neighbors = 13,
            min_dist = 0.1,
            input="dist")
ump_cpe <- ump_cpe$layout
ump_cpe <- data.frame(pdbID=rownames(ump_cpe), 
                  UMAP1=ump_cpe[,1],
                  UMAP2=ump_cpe[,2])
ump_cpe <- left_join(df, ump_cpe,'pdbID')
# ------------------SPE----------------
ump_spe <- umap(d = dis_SPE,
            n_neighbors = 13,
            min_dist = 0.1,
            input="dist")
ump_spe <- ump_spe$layout
ump_spe <- data.frame(pdbID=rownames(ump_spe), 
                  UMAP1=ump_spe[,1],
                  UMAP2=ump_spe[,2])
ump_spe <- left_join(df, ump_spe,'pdbID')
# ---------------------
ump_df <- data.frame(Method = rep(c('SPE', 'CPE', 'TM-Vec'), each=nrow(ump_cpe)),
                     rbind(ump_spe, ump_cpe, ump_tm))

ump_df$Method <- factor(ump_df$Method,
                        levels = c('SPE', 'CPE', 'TM-Vec'))

p<-ggplot(ump_df, aes(UMAP1, UMAP2, color= globin)) +
  geom_point(alpha=.7,size=3) +
  facet_grid( ~ Method)+
  mytheme() +
  theme(strip.text.x = element_text(size = 13, face = 'bold'),
        axis.text.x = element_text(size = 10))+
  scale_color_manual(values = c('darkorange3', 'purple'))
# save Fig8_new.pdf 10x5
# ----------SourceData---------------
MYwriteData(data=ump_df[,-c(3,4)],
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 4 C",
            row.Names = TRUE)
```

## Fig 5 B,D,F 

```{r Fig 5 ferritin}
source('Functions.R')
# ----------------------------------------------------
E210_ferritin <- readRDS("Data/rds/E210_ferritin.rds")
df <- E210_ferritin
rownames(df) <- df$pdbID
dis <- proxy::dist(df[,-c(1:4)], method = manhat)

nnet <- neighborNet(dis)
#ggsplitnet(nnet) + geom_tiplab2(aes(label=label),
#                                hjust=-.1,size=5,
#                                color='blue')+
#  ggexpand(.1) + ggexpand(.1, direction=-1)
# write.nexus.networx(nnet,'Data/rds/E210_ferritin_manhat.nxs')
# ----------SourceData---------------
MYwriteData(data=as.matrix(dis),
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 5 B",
            row.Names = TRUE)
# --------------------------------------------------------------------
# ----   the 'E210_ferritin_manhat.nxs' file 
# -----      was load on SplitsTree4(v 4.19.2) to visualize
# --------------------------------------------------------------------
df <- read.csv('Data/csv/Ferritin_Like_seq.csv')
dis <- read.csv('Data/csv/disTM_vec_Ferritin_Like_seq.csv')
dis <- 1 - dis
colnames(dis) <- rownames(dis) <- df$pdbID
nnet <- neighborNet(dis)
# write.nexus.networx(nnet,'Data/rds/ferritin_TMvec.nxs')
# ----------SourceData---------------
MYwriteData(data=as.matrix(dis),
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 5 D",
            row.Names = TRUE)
# -----------------------------------------
E210.seq_ferritin <- readRDS("Data/rds/E210.seq_ferritin.rds")
df <- E210.seq_ferritin
rownames(df) <- df$pdbID
dis <- proxy::dist(df[,-c(1:4)], method = manhat)

nnet <- neighborNet(dis)
#write.nexus.networx(nnet,'Data/rds/E210.seq_ferritin.nxs')
# ----------SourceData---------------
MYwriteData(data=as.matrix(dis),
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 5 F",
            row.Names = TRUE)
```


## Fig 5 C,E,G 

```{r Fig 5C,E,G ferritin (mean distance)}
source('Functions.R')
# -----------------------
mean_pairwise_dis <- function(dis, gr, gr_idx){
  n <- length(unique(gr))
  calculate_mean_distance <- function(group1_indices,group2_indices,
                                      distance_matrix) {
    submatrix <- distance_matrix[group1_indices, group2_indices]
    return(mean(submatrix))
  }
  mean_dis <- matrix(NA, nrow = n, ncol = n,
                     dimnames = list(levels(gr),levels(gr)))
  for (i in 1:n) {
    for (j in 1:n) {
      mean_dis[i, j] <- calculate_mean_distance(gr_idx[[i]],gr_idx[[j]],dis)
    }
  }
  diag(mean_dis) <- 0
  minmax_dis <- mean_dis
  minmax_dis <- (minmax_dis - min(minmax_dis))/(max(minmax_dis) - min(minmax_dis))
  minmax_dis <- round(minmax_dis, 2)
  return(minmax_dis)
}
# ------------------------------------
# -------- SPE
E210_ferritin <- readRDS("Data/rds/E210_ferritin.rds")
df <- E210_ferritin
rownames(df) <- df$pdbID
dis <- as.matrix(proxy::dist(df[,-c(1:4)], method = manhat))
group <-as.factor(df$group)
group_indices <-  split(seq_len(length(group)), group)
PWdisGroup<-mean_pairwise_dis(as.matrix(dis),
                             gr = group, gr_idx = group_indices)
tree_data <- ape::bionj(as.dist(PWdisGroup))

t_SPE <- ggtree(tree_data, layout = 'unrooted',
                size=5,aes(color = label)) +
  geom_tiplab(aes(label=label, col=label),
              hjust=.5,vjust=-.9,size=4,fontface="bold",align=F)+
  scale_color_discrete(type =  c('red',"orange2","yellow4","green3",
                                 'skyblue3','blue',
                                 'magenta','purple2'
  ))+
  theme(legend.position = 'none')
# ----------SourceData---------------
MYwriteData(data=PWdisGroup,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 5 C",
            row.Names = TRUE)
# ------------------------------
# ---------- CPE
E210.seq_ferritin <- readRDS("Data/rds/E210.seq_ferritin.rds")
df <- E210.seq_ferritin
rownames(df) <- df$pdbID
dis <- as.matrix(proxy::dist(df[,-c(1:4)], method = manhat))
group <- as.factor(df$group)
group_indices <- split(seq_len(length(group)), group)
PWdisGroup <- mean_pairwise_dis(as.matrix(dis),
                                gr = group, gr_idx = group_indices)
tree_data <- ape::bionj(as.dist(PWdisGroup))

t_CPE <- ggtree(tree_data, layout = 'unrooted',
                size=5,aes(color = label)) +
  geom_tiplab(aes(label=label, col=label),
              hjust=.5,vjust=-.9,size=4,fontface="bold",align=F)+
  scale_color_discrete(type =  c('red',"orange2","yellow4","green3",
                                 'skyblue3','blue',
                                 'magenta','purple2'
  ))+
  theme(legend.position = 'none')
# ----------SourceData---------------
MYwriteData(data=PWdisGroup,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 5 G",
            row.Names = TRUE)
# ----------------------
# --------- TM
df <- read.csv('Data/csv/Ferritin_Like_seq.csv')
dis <- read.csv('Data/csv/disTM_vec_Ferritin_Like_seq.csv')
dis <- 1 - dis
colnames(dis) <- rownames(dis) <- df$pdbID
group <- as.factor(df$group)
group_indices <- split(seq_len(length(group)), group)
PWdisGroup <- mean_pairwise_dis(as.matrix(dis),
                                gr = group, gr_idx = group_indices)
tree_data <- ape::bionj(as.dist(PWdisGroup))

t_TM <- ggtree(tree_data, layout = 'unrooted',
               size=5,aes(color = label)) +
  geom_tiplab(aes(label=label, col=label),
              hjust=.5,vjust=-.9,size=4,fontface="bold",align=F)+
  scale_color_discrete(type =  c('red',"orange2","yellow4","green3",
                                 'skyblue3','blue',
                                 'magenta','purple2'
  ))+
  theme(legend.position = 'none')
# ----------SourceData---------------
MYwriteData(data=PWdisGroup,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 5 E",
            row.Names = TRUE)
```


## Fig 6 

```{r Fig 6 covid}
source('Functions.R')
bootFun <- function(e){
  as.phylo(hclust(proxy::dist(e,
                              method = manhat),
                  method = 'ward.D2'))
  }
# ----------------------------------------------------
# --------------- CPE -----------------
E210.seq_spike <- readRDS("Data/rds/E210.seq_spike.rds")
df <- E210.seq_spike

rownames(df) <- df$pdbID
x <- df[,-c(1:4)]/df$length
x  <- as.matrix(x) 
dis <- x %>% proxy::dist(method = manhat)
hc <- dis %>% hclust(method = "ward.D2")
dend <- hc %>% as.dendrogram

df <- df[with(df, order(df$virus)), ]
row.names(df) <- df$pdbID
rn <- df$pdbID
grp <- list(MERS_Cov     = rn[1:32],
            SARS_Cov  = rn[33:63],
            SARS_Cov2  = rn[64:143])

tree_data <- as.phylo(dend)

myBoots <- boot.phylo(tree_data, x, bootFun, B = 100)

boot_ci <- data.frame(bootValues=myBoots, ci=NA)
for (i in 1:nrow(boot_ci)) {
  if(!is.na(boot_ci$bootValues[i])){
    ci <- binom.confint(boot_ci$bootValues[i],
                        100, methods="exact")
    boot_ci$ci[i] <- paste0('(',round(ci$lower * 100, 1), " , ", round(ci$upper * 100, 1), ')')
  }
}

p_data <- ggtree(tree_data, layout = 'dend',
                 size = 2)

p_data$data$boots <- c(rep(NA,143),boot_ci$bootValues)
p_data$data$ci <- c(rep(NA,143),boot_ci$ci)
p_data$data$edge <- paste0(p_data$data$parent, '_', p_data$data$node)

t_CPE <- groupOTU(p_data, grp, "groups") + aes(color=groups) +
  theme(legend.position="none",
        plot.title = element_text(size = 50, face="bold",hjust = 0.5,vjust = 2))+
#  geom_tiplab2(aes(label=label,angle=angle),
#               hjust=-.1,size=1.5,fontface="bold",align=F)+
#  geom_tiplab(aes(label=label),
#               hjust=1,size=1.5,fontface="bold",align=F)+
  geom_text2(aes(subset=edge%in%c('144_144','144_145'), label=boots),
             col='black', hjust=-.2, vjust=-.5)+
  geom_text2(aes(subset=edge%in%c('144_144','144_145'), label=ci),
             col='black', hjust=-.5, vjust=-.5)+
  scale_color_discrete(breaks=c('MERS_Cov', 'SARS_Cov', 'SARS_Cov2'),
                      type =  c('gray45',"blue","gold3","red3" ))+
  ggtitle('CPE')
# ----------SourceData---------------
MYwriteData(data=as.matrix(dis),
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 6 B",
            row.Names = TRUE)
# -----------------------------------------------------
# ------------ SPE ----------------
E210_spike <- readRDS("Data/rds/E210_spike.rds")
df <- E210_spike

rownames(df) <- df$pdbID
x <- df[,-c(1:4)]/df$length
x  <- as.matrix(x) 
dis <- x %>% proxy::dist(method = manhat)
hc <- dis %>% hclust(method = "ward.D2")
dend <- hc %>% as.dendrogram

df <- df[with(df, order(df$virus)), ]
row.names(df) <- df$pdbID
rn <- df$pdbID
grp <- list(MERS_Cov     = rn[1:32],
            SARS_Cov  = rn[33:63],
            SARS_Cov2  = rn[64:143])

tree_data <- as.phylo(dend)
p_data <- ggtree(tree_data, layout = 'dendrogram',
                 #branch.length='none',
                 size = 2) 
myBoots <- boot.phylo(tree_data, x, bootFun, B = 100)

boot_ci <- data.frame(bootValues=myBoots, ci=NA)
for (i in 1:nrow(boot_ci)) {
  if(!is.na(boot_ci$bootValues[i])){
    ci <- binom.confint(boot_ci$bootValues[i],
                        100, methods="exact")
    boot_ci$ci[i] <- paste0('(',round(ci$lower * 100, 1), " , ", round(ci$upper * 100, 1), ')')
  }
}

p_data <- ggtree(tree_data, layout = 'dend',
                 size = 2)

p_data$data$boots <- c(rep(NA,143),boot_ci$bootValues)
p_data$data$ci <- c(rep(NA,143),boot_ci$ci)
p_data$data$edge <- paste0(p_data$data$parent, '_', p_data$data$node)

t_SPE <- groupOTU(p_data, grp, "groups") + aes(color=groups) +
  theme(legend.position="none",
        plot.title = element_text(size = 50, face="bold",hjust = 0.5,vjust = 2))+
  geom_text2(aes(subset=edge%in%c('144_144','144_146'), label=boots),
             col='black', hjust=-.2, vjust=-.5)+
  geom_text2(aes(subset=edge%in%c('144_144','144_146'), label=ci),
             col='black', hjust=-.5, vjust=-.5)+
  scale_color_discrete(breaks=c('MERS_Cov', 'SARS_Cov', 'SARS_Cov2'),
                      type =  c('gray45',"blue","gold3","red3"  ))+
  ggtitle('SPE')
# ----------SourceData---------------
MYwriteData(data=as.matrix(dis),
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 6 D",
            row.Names = TRUE)
# -----------------------------------------------
# ------------- RMSD --------------
df <- E210.seq_spike
rms <- data.frame(readRDS('Data/covidPDB/RMSD_spike.rds'))
diag(rms)<-0

dis <- as.dist(rms)
hc <- dis %>% hclust(method = "ward.D2")

df <- df[with(df, order(df$virus)), ]
rn <- df$pdbID
grp <- list(MERS_Cov  = rn[1:32],
            SARS_Cov  = rn[33:63],
            SARS_Cov2 = rn[64:143])

tree_data <- as.phylo(hc)
p_data <- ggtree(tree_data, layout = 'dendrogram',
                 size = 2) 


t_RMSD <- groupOTU(p_data, grp, "groups") + aes(color=groups) +
       theme(legend.position="none",
             plot.title = element_text(size = 50, face="bold",hjust = 0.5,vjust = 2))+
#  geom_text2(aes(subset=!isTip, label=boots),
#             col='black', hjust=-.2, vjust=-.5)+
       scale_color_discrete(breaks=c('MERS_Cov', 'SARS_Cov', 'SARS_Cov2'),
                            type =  c("blue","gold3","red3" ))+
  ggtitle('RMSD')
# ----------SourceData---------------
MYwriteData(data=as.matrix(dis),
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 6 E",
            row.Names = TRUE)
# --------------------------------------------
# ----------- TM-Vec -----------------
df <- E210.seq_spike
tm_ebm <- read.csv('Data/covidPDB/TM_vec_emb_spike.csv')
rownames(tm_ebm) <- df$pdbID
tm_vec <- read.csv('Data/covidPDB/disTM_vec_spike.csv')
colnames(tm_vec) <- rownames(tm_vec) <- df$pdbID
tm_vec <- 1 - tm_vec

dis <- as.dist(tm_vec)
hc <- dis %>% hclust(method = "ward.D2")

dend <- hc %>% as.dendrogram

df <- df[with(df, order(df$virus,decreasing = T)), ]
rn <- df$pdbID
grp <- list(MERS_Cov  = rn[112:143],
            SARS_Cov  = rn[81:111],
            SARS_Cov2 = rn[1:80])

tree_data <- as.phylo(dend)
p_data <- ggtree(tree_data, layout = 'dendrogram',
                 size = 2) 

# ggtree(tr) + geom_text(aes(label=node))
p_data <- flip(p_data, 152, 151)

myBoots <- boot.phylo(tree_data, tm_ebm, FUN = function(e){
  d <- 1 - (proxy::dist(e,method = cosine))
  as.phylo(hclust(d, method = 'ward.D2'))},
  B = 100)

boot_ci <- data.frame(bootValues=myBoots, ci=NA)
for (i in 1:nrow(boot_ci)) {
  if(!is.na(boot_ci$bootValues[i])){
    ci <- binom.confint(boot_ci$bootValues[i],
                        100, methods="exact")
    boot_ci$ci[i] <- paste0('(',round(ci$lower * 100, 1), " , ", round(ci$upper * 100, 1), ')')
  }
}


p_data$data$boots <- c(rep(NA,143),boot_ci$bootValues)
p_data$data$ci <- c(rep(NA,143),boot_ci$ci)
p_data$data$edge <- paste0(p_data$data$parent, '_', p_data$data$node)

t_tmVec <- groupOTU(p_data, grp, "groups") + aes(color=groups) +
       theme(legend.position="none",
             plot.title = element_text(size = 50, face = 'bold',hjust = 0.5,vjust = 2))+
  geom_text2(aes(subset=edge%in%c('146_148','148_150'), label=boots),
             col='black', hjust=-.2, vjust=-.5)+
  geom_text2(aes(subset=edge%in%c('146_148','148_150'), label=ci),
             col='black', hjust=-.5, vjust=-.5)+
       scale_color_discrete(breaks=c('MERS_Cov', 'SARS_Cov', 'SARS_Cov2'),
                            type =  c("blue","gold3","red3" ))+
  ggtitle('TM-Vec')
# ----------SourceData---------------
MYwriteData(data=as.matrix(dis),
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 6 C",
            row.Names = TRUE)
# --------------------------------------
# --------------TM score-------------
spike <- read.csv("Data/covidPDB/spike_seq.csv")
df <- spike
tm <- read.csv('Data/covidPDB/disTM_score_spike.csv')
rownames(tm) <- colnames(tm) <- df$pdbID

tm_max <- tm
for (i in 1:nrow(tm_max)) {
  print(i)
  for (j in 1:nrow(tm_max)) {
    tm_max[i,j] <- max(tm_max[i,j], tm_max[j,i])
    tm_max[j,i] <- tm_max[i,j]
  }
}
diag(tm_max) <- 1
dis <- 1 - tm_max
hc <- as.dist(dis)
hc <- hc%>% hclust(method = "ward.D2")

dend <- hc %>% as.dendrogram

df <- df[with(df, order(df$virus)), ]
rn <- df$pdbID
tree_data <- as.phylo(dend)
grp <- list(MERS_Cov  = rn[1:32],
            SARS_Cov  = rn[33:63],
            SARS_Cov2 = rn[64:143])

p_data <- ggtree(tree_data, layout = 'dendrogram',
                 size = 2) 

t_tmScore<-groupOTU(p_data, grp, "groups") + aes(color=groups) +
  theme(legend.position="none",
    plot.title = element_text(size = 50, face = 'bold',
                              hjust = 0.5,vjust = 2))+
  #geom_tiplab2(aes(label=label,angle=angle),
  #              hjust=-.1,size=3.5,fontface="bold",align=F)+
  scale_color_discrete(breaks=c('MERS_Cov', 'SARS_Cov', 'SARS_Cov2'),
                       type =  c("blue","gold3","red3" ))+
  ggtitle('TM-Score')
# ----------SourceData---------------
MYwriteData(data=as.matrix(dis),
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 6 F",
            row.Names = TRUE)
# -----------------------------------------
# Covid-SequenceBased-Phylogenetic tree
# ---------- MSA ClustalW
# -------------------
df <- read.csv("Data/covidPDB/spike_seq.csv")
rownames(df) <- df$pdbID
df <- df[with(df, order(df$virus,decreasing = T)), ]
row.names(df) <- df$pdbID
rn <- df$pdbID
grp <- list(SARS_Cov2  = rn[1:80],
            SARS_Cov  = rn[81:111],
            MERS_Cov  = rn[112:143])

## --- Make a fasta file including spike protein sequences
#  ampir::df_to_faa(df[,c(1,3)],'Data/covidPDB/Spike.fasta')
# --- Read fasta sequences-------------
sequences <- readAAStringSet("Data/covidPDB/Spike.fasta")
# ---- Multiple sequence alignment using ClustalW---------
tictoc::tic()
alignment <- msa(sequences, method = "ClustalW")
tictoc::toc()
# --------------
# t := 71.8 Sec
# ------------------
alignment2 <- msaConvert(alignment, type="seqinr::alignment")
d <- seqinr::dist.alignment(alignment2, "identity")
upgma <- phangorn::upgma(d)

dend <- upgma %>% as.dendrogram
tree_data <- as.phylo(dend)

fu <- function(seq){
  z <- apply(seq, 1, paste, collapse = " ")
  z <- data.frame(pdbID=names(z), seq=z)
  ampir::df_to_faa(z,'Data/temp.fasta')
  z <- readAAStringSet(file = "Data/temp.fasta", format = "fasta")
  align <- msa(z, method = "ClustalW")
  align2 <- msaConvert(align, type="seqinr::alignment")
  did <- seqinr::dist.alignment(align2, "identity")
  upg <- phangorn::upgma(did)
  den <- upg %>% as.dendrogram
  tr <- as.phylo(den)
  return(tr)
}
#myBoots <- boot.phylo(tree_data, alignment_matrix, FUN = fu, B = 100)
#saveRDS(myBoots,'Data/rds/bootStraps_ClustalW.rds')
myBoots <- readRDS('Data/rds/bootStraps_ClustalW.rds')
boot_ci <- data.frame(bootValues=myBoots, ci=NA)
for (i in 1:nrow(boot_ci)) {
  print(i)
  if(!is.na(boot_ci$bootValues[i])){
    ci <- binom.confint(boot_ci$bootValues[i],
                        100, methods="exact")
    boot_ci$ci[i] <- paste0('(',round(ci$lower * 100, 1), " , ", round(ci$upper * 100, 1), ')')
  }
}

p_data <- ggtree(tree_data, layout = 'dend',
                 size = 2)

p_data$data$boots <- c(rep(NA,143),boot_ci$bootValues)
p_data$data$ci <- c(rep(NA,143),boot_ci$ci)
p_data$data$edge <- paste0(p_data$data$parent, '_', p_data$data$node)

t_Sequence_upgma <- groupOTU(p_data, grp, "groups") + aes(color=groups) +
  theme(legend.position="none",
        plot.title = element_text(size = 50, 
                                  face="bold",hjust = 0.5,vjust = 2))+
    geom_text2(aes(subset=edge%in%c('144_146','146_148'), label=boots),
             col='black', hjust=-.2, vjust=-.5)+
  geom_text2(aes(subset=edge%in%c('144_146','146_148'), label=ci),
             col='black', hjust=-.5, vjust=-.5)+
  scale_color_discrete(breaks=c('MERS_Cov', 'SARS_Cov', 'SARS_Cov2'),
                      type =  c('gray45',"blue","gold3","red3" ))+
  ggtitle('MSA (ClustalW)')
# ----------SourceData---------------
dis_seqid <- as.matrix(d)
dis_seqid <- dis_seqid[df$pdbID, df$pdbID]
MYwriteData(data=dis_seqid,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 6 A",
            row.Names = TRUE)
# ------------------------------------------
mylegend <- get_legend(t_SPE+
                       theme(legend.position = 'bottom',
                             legend.title = element_blank(),
                             legend.text = element_text(size = 20)))


AB<-ggarrange(NULL,NULL,
    t_Sequence_upgma,t_SPE, NULL,NULL,
    t_CPE,t_tmVec, NULL,NULL,
    t_RMSD, t_tmScore,
    ncol = 2,nrow = 6,
    heights = c(.15, 1.3, .2, 1.3),
    widths = c(1,1),
    labels = c('',"", "A", "B",'','','C','D', '', '', 'E', 'F'),
    font.label = list(size = 55), # color = "red"
    hjust = 0,vjust = .5
    )

p0 <- cowplot::plot_grid(AB, mylegend, ncol = 1, rel_heights = c(1,.05))
#annotate_figure(AB, top = title_grob)
# save FIG6.pdf 20x10
# ----------------------------------------------------------
# ----------------------------------------------------
df <- data.frame(method=c('RMSD','TM-Score','SPE',
                          'TM-Vec','CPE', 'MSA (ClustalW)'),
                 basedOn=rep(c('Structure based', 'Sequence based'),
                             c(3,3)),
                 CT_3=c(50, 50, 100, 16, 50, 49),
                 CT_4=c(50, 56, 95, 48, 95, 49),
                 TimeS=c(70*60, 9.7*3600, 3*60, 80, 0.9, 72))

df$Time <- round((df$TimeS*100)/(12*3600), 5)


df <- df[order(df$Time, decreasing = F),]
df$method <- factor(df$method, 
                    levels = df$method)
mycol <-c('magenta', 'purple','orange','cyan','blue','green')


p12h <- ggplot(df, aes(x = method, y = Time, fill= method)) +
  geom_bar(stat="identity")+
  coord_polar(theta="y") +
  theme_light(base_rect_size = 0,base_line_size = 1)+
  theme(axis.text = element_blank(),
        axis.ticks=element_blank(),
        axis.title = element_blank(),
        #legend.position = 'none',
        legend.title = element_blank())+
  scale_fill_manual(values = mycol)+
  ylim(0,100)

p80s <- p12h+xlim(c('CPE','MSA (ClustalW)','TM-Vec'))+
  ylim(0,.186)+ #1.86*12*3.6 = 80sec
    theme(legend.text = element_text(face = 'bold',size = 15))


df2 <- melt(df[,1:4])
df2$variable <- factor(ifelse(df2$variable=='CT_3',3,4))
p <- ggplot(df2, aes(variable,value, fill=method))+
  geom_bar(stat="identity", width = .5,
           position=position_dodge2(preserve = "single"),alpha=1,show.legend = F)+
  facet_grid( ~ basedOn)+xlab('Cut tree')+ylab('ARI')+
  scale_fill_manual(values = mycol)+
  mytheme()+
  theme(strip.text.x = element_text(size = 20, face = 'bold'),
        legend.text = element_text(face = 'bold',size=30),
        axis.text.x = element_text(size = 25, face = 'bold',color = 'black'),
        axis.text.y = element_text(size = 25, face = 'bold',color = 'black'),
        axis.title.x = element_text(size = 25, face = 'bold',color = 'black'),
        axis.title.y = element_text(size = 25, face = 'bold',color = 'black'))
# ----------SourceData---------------
MYwriteData(data=df,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 6 G-H",
            row.Names = TRUE)
# --------------------------------------------
p12h
p80s
p
```

## Fig 6i 

This chunk calculates the minimum pairwise distance among the SARS2, SARS, and MERS groups to assess their inter-similarity

```{r Fig 6i Spike Inter-Similarity}
source('Functions.R')
# --------------------SPE-----------
df <- readRDS("Data/rds/E210_spike.rds")
rownames(df) <- df$pdbID
x <- as.matrix(df[,-c(1:4)]/df$length)
dis_SPE <- as.matrix(x %>% proxy::dist(method = manhat))
# ----------------------------------
# -----------------CPE-------------------
df <- readRDS("Data/rds/E210.seq_spike.rds")
rownames(df) <- df$pdbID
x <- as.matrix(df[,-c(1:4)]/df$length)
dis_CPE <- as.matrix(x %>% proxy::dist(method = manhat))
# ------------------------------------------------
# -------------------Seq identity from MSA------------------------
sequences <- readAAStringSet("Data/covidPDB/Spike.fasta")
# ---- Multiple sequence alignment using ClustalW---------
alignment <- msa(sequences, method = "ClustalW")
alignment2 <- msaConvert(alignment,type="seqinr::alignment")
d <- seqinr::dist.alignment(alignment2, "identity")
dis_seqid <- as.matrix(d)
dis_seqid <- dis_seqid[df$pdbID, df$pdbID]
# -------------------------------------
# -------------RMSD---------------------
rms <- data.frame(readRDS('Data/covidPDB/RMSD_spike.rds'))
colnames(rms)<-rownames(rms)
diag(rms)<-0
rms <- as.matrix(rms)
# -------------------------------------
# -------------TM---------------------
tm <- read.csv('Data/covidPDB/disTM_vec_spike.csv')
colnames(tm) <- rownames(tm) <- df$pdbID
tm <- 1 - tm
tm_vec <- as.matrix(tm)
# -------------------------------------
# -------------TM score---------------------
spike <- read.csv("Data/covidPDB/spike_seq.csv")
df <- spike
tm <- read.csv('Data/covidPDB/disTM_score_spike.csv')
rownames(tm) <- colnames(tm) <- df$pdbID

tm_max <- tm
for (i in 1:nrow(tm_max)) {
  print(i)
  for (j in 1:nrow(tm_max)) {
    tm_max[i,j] <- max(tm_max[i,j], tm_max[j,i])
    tm_max[j,i] <- tm_max[i,j]
  }
}
diag(tm_max) <- 1
tm_max <- as.matrix(1 - tm_max)
# -------------Groups----------
group <- factor(df$virus)
group_indices <- split(seq_len(length(group)), group)
# -----------Function to calculate mean of pairwise distance between two groups
mean_pairwise_dis <- function(dis, gr, gr_idx){
  n <- length(unique(gr))
  calculate_mean_distance <- function(group1_indices,group2_indices,
                                     distance_matrix) {
  submatrix <- distance_matrix[group1_indices, group2_indices]
  return(mean(submatrix))
  }
mean_dis <- matrix(NA, nrow = n, ncol = n,
                   dimnames = list(levels(gr),levels(gr)))
for (i in 1:n) {
    for (j in 1:n) {
      mean_dis[i, j] <- calculate_mean_distance(gr_idx[[i]],gr_idx[[j]],dis)
    }
  }
diag(mean_dis)<-0
minmax_dis <- mean_dis
minmax_dis <- (minmax_dis - min(minmax_dis))/(max(minmax_dis) - min(minmax_dis))

minmax_dis <- round(minmax_dis,2)
return(minmax_dis)
}
# -----------------------------------------------------------
dists <- list(CPE=dis_CPE,SPE=dis_SPE,seqid=dis_seqid,
              TM_vec=tm_vec,RMSD=rms, TM_score=tm_max)
pair_dis_minPair <- lapply(dists, FUN = mean_pairwise_dis,
                           gr=group,gr_idx=group_indices)
lower_triangles <- lapply(pair_dis_minPair, function(x) x[lower.tri(x)])
mean_pair_dist <- do.call(rbind, lower_triangles)
colnames(mean_pair_dist) = c("MERS-CoV ~ SARS-CoV", "MERS-CoV ~ SARS-CoV-2", "SARS-CoV ~ SARS-CoV-2")
library(reshape2)
group_dis <- melt(mean_pair_dist)
colnames(group_dis) = c("Method","Comparison","Distance")
#saveRDS(group_dis, 'Data/rds/Spike_group_dis.rds')
# -------------------
group_dis <- readRDS('Data/rds/Spike_group_dis.rds')
# Plot the bar plot
p0<-ggplot(group_dis,
           aes(x = Comparison, y = Distance, fill = Method)) +
  geom_bar(stat = "identity", position = "dodge", width = 0.7,
           show.legend = F) +
  labs(x = "", y = "Distance") +
  mytheme() +
  theme(legend.title = element_blank(),
        axis.text.x = element_text(size = 15, face = 'bold',
                                   color = 'black', angle = 15, 
                                   hjust = .8, vjust = .8),
        axis.text.y = element_text(size = 20, face = 'bold',
                                   color = 'black'),
        axis.title.y = element_text(size = 25, face = 'bold',
                                   color = 'black'),)+
  scale_fill_manual(values = c('magenta', 'purple','cyan','orange','blue','green'))
# ----------SourceData---------------
MYwriteData(data=mean_pair_dist,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 6 I",
            row.Names = TRUE)
```


## Fig 7 AB

```{r Fig 7 AB Bacteriocin Tm-Vec}
source('Functions.R')
# ----------------------------------------------------
tm <- readRDS('Data/rds/bacteriocin_tm_vec.rds')
tm2 <- tm[!is.na(tm$status2),-24]
colnames(tm2)[24]<-'status'
tm2 <- rbind(tm[,-25],tm2)

tm2 <- tm2[,c(5,20,21,6,23,24)] # 22~6 DeepBlast
colnames(tm2)[-6] <- c('Alphafold2','Omegafold','ESMfold','TM_vec','CPE')
tm2$CPE[tm2$CPE<0] <- 0
tm3 <- melt(tm2,variable.name = 'method',value.name = 'score')
tm3$method <- factor(tm3$method,levels = c('TM_vec','Alphafold2','Omegafold','ESMfold','CPE'))

p1<-ggplot(tm3, aes(status,score,fill=method))+
  geom_boxplot(outlier.shape = NA,
               color='black',
               size=rep(c(rep(.2,4),.8),3),
               alpha=rep(c(rep(.5,4),1),3),
               width=.4
               )+
  coord_cartesian(ylim = c(0.1, 1))+
  xlab('')+
    geom_hline(yintercept = .5, linetype="dashed", 
             color = "red", size=.7)+
  mytheme()+
  theme(legend.title = element_blank(),
        axis.text.y = element_text(size=13),
        axis.text.x = element_text(color = 'black',size = 13, face = 'bold'),
        axis.title.y = element_text(size = 20,vjust = 2, face = 'bold'),
        legend.text = element_text(size = 15))+
  scale_fill_discrete(labels=c('TM_vec','Alphafold2',
                               'Omegafold','ESMfold',
                               expression(bold("CPE"))))+
  scale_fill_manual(values = c('orange', 'skyblue3',
                               'darkolivegreen','yellow3', 'magenta'))
# ----------SourceData---------------
MYwriteData(data=tm2,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 7 B",
            row.Names = TRUE)
# --------------------------------
# -----  UMAP
df <- readRDS('Data/rds/E210.seq_bacteriocin.rds')
dis <- as.matrix(proxy::dist(df[,-c(1:6)], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 100,
            min_dist = 0.1,
            input="dist")

ump_df <- data.frame(class=df$class,
                     UMAP1 = ump$layout[, 1],
                     UMAP2 = ump$layout[, 2])

p2<-ggplot(ump_df,
       aes(UMAP1, UMAP2, color= class)) +
  geom_point(alpha=.7, size=2) + 
  mytheme()+
  theme(legend.title = element_blank(),
        #legend.position = 'bottom',
        axis.title = element_text(size = 15),
        axis.title.y = element_text(size = 15,vjust = 0),
        legend.text = element_text(size = 15))+
  scale_color_manual(values = c('skyblue3','orange4','blue'))

# ----------SourceData---------------
MYwriteData(data=ump_df,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 7 A",
            row.Names = TRUE)


p1_2 <- ggarrange(p2,NULL,p1,ncol = 3,nrow = 1,
          widths = c(.5,.05,1))

# save FIG7AB.pdf 15x8

```

##Fig 7 D

```{r Fig 7 D DrugTarget}
source('Functions.R')
# ----------------------------------------------------
my_sab <- readRDS('Data/rds/my_sab.rds')
my_sab <- as.data.frame(as.table(my_sab)) 
colnames(my_sab) <- c("d1", "d2", "my_sAB") 
my_sab <- my_sab[!is.na(my_sab$my_sAB), ]
sab<-read_excel('Data/csv/41467_2019_9186_MOESM7_ESM.xlsx')
sab <- sab[-2081,-6]
colnames(sab)[1:2]<-c('d1','d2')

df <- left_join(sab,my_sab,c('d1','d2'))
p1 <- ggplot(df,aes(sAB,my_sAB))+
  geom_point(alpha=.5)+geom_smooth(method = 'lm')+
  stat_cor(col='red',size=8)+mytheme()+
    theme(axis.title.x = element_text(size=30),
          axis.title.y = element_text(size=30),
        axis.text.x = element_text(size = 17),
        axis.text.y = element_text(size = 17))+
  xlab('sAB (PPI)')+ylab('sAB (CPE)')
# save FIG7C.pdf 10x7

# ----------SourceData---------------
colnames(df)[5:6] <- c('sAB(PPI)','sAB(CPE)')
MYwriteData(data=df,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 7 D",
            row.Names = TRUE)
```

## Fig 7 C 

```{r Fig 7 C (PLPRO)}
source('Functions.R')
load('Data/Large_Scale_SARS2/SPE_CPE_time.RData')
df_CPE$pdbID <- gsub('.pdb','',df_CPE$pdbID)
df_SPE$pdbID <- gsub('.pdb','',df_SPE$pdbID)
# ----------- load TM from google drive ---------------
drive_deauth()
drive_user()
public_file <- drive_get(as_id("1QUAqdkypCPHraNESF11arxg29tDUp521"))
drive_download(public_file, overwrite = TRUE)
tm <- read.csv('disTM_vec_sars_proteom.csv')
file.remove('disTM_vec_sars_proteom.csv')
tm <- 1 - tm
# ------------------------------------
# -------------------
idx_PLpro <- grep('PLPro', df_CPE$sars_proteom)

df_CPE <- df_CPE[idx_PLpro,]
df_SPE <- df_SPE[idx_PLpro,]
rownames(df_CPE) <- df_CPE$pdbID
rownames(df_SPE) <- df_SPE$pdbID

tm <- tm[idx_PLpro, idx_PLpro]
colnames(tm) <- rownames(tm) <- df_CPE$pdbID

c1 <- read.delim('Data/Large_Scale_SARS2/NSP3_cd21732_betaCoV_PLPro/cluster1.txt', header = F)


c2 <- read.delim('Data/Large_Scale_SARS2/NSP3_cd21732_betaCoV_PLPro/cluster2.txt', header = F)

c3 <- read.delim('Data/Large_Scale_SARS2/NSP3_cd21732_betaCoV_PLPro/cluster3.txt', header = F)

c4 <- read.delim('Data/Large_Scale_SARS2/NSP3_cd21732_betaCoV_PLPro/cluster4.txt', header = F)


df_CPE <- data.frame(subClass=NA, df_CPE)
df_CPE$subClass[df_CPE$pdbID%in%c1$V1] <- 'Sarb'
df_CPE$subClass[df_CPE$pdbID%in%c2$V1] <- 'Nob'
df_CPE$subClass[df_CPE$pdbID%in%c3$V1] <- 'Merb'
df_CPE$subClass[df_CPE$pdbID%in%c4$V1] <- 'Emb'

df_SPE <- data.frame(subClass=NA, df_SPE)
df_SPE$subClass[df_SPE$pdbID%in%c1$V1] <- 'Sarb'
df_SPE$subClass[df_SPE$pdbID%in%c2$V1] <- 'Nob'
df_SPE$subClass[df_SPE$pdbID%in%c3$V1] <- 'Merb'
df_SPE$subClass[df_SPE$pdbID%in%c4$V1] <- 'Emb'

dff <- df_CPE[,-c(1:5)]/as.numeric(df_CPE$length)
dis_CPE <- as.matrix(proxy::dist(dff,
                                 method = manhat))
dff <- df_SPE[,-c(1:5)]/as.numeric(df_SPE$length)
dis_SPE <- as.matrix(proxy::dist(dff,
                                 method = manhat))

ump_tm <- umap(d = tm,n_neighbors = 13,input='dist',min_dist = 0.5)
ump_tm <- ump_tm$layout
ump_tm <- data.frame(pdbID=rownames(ump_tm), 
                  UMAP1=ump_tm[,1],
                  UMAP2=ump_tm[,2])
ump_tm <- left_join(df_CPE[,1:5], ump_tm, 'pdbID')
#----------------------CPE-----------
ump_CPE <- umap(d = dis_CPE,n_neighbors = 13,input='dist',min_dist = 0.5)
ump_CPE <- ump_CPE$layout
ump_CPE <- data.frame(pdbID=rownames(ump_CPE), 
                  UMAP1=ump_CPE[,1],
                  UMAP2=ump_CPE[,2])
ump_CPE <- left_join(df_CPE[,1:5], ump_CPE, 'pdbID')
#----------------SPE--------------------
ump_SPE <- umap(d = dis_SPE,n_neighbors = 13,input='dist',min_dist = 0.5)
ump_SPE <- ump_SPE$layout
ump_SPE <- data.frame(pdbID=rownames(ump_SPE), 
                  UMAP1=ump_SPE[,1],
                  UMAP2=ump_SPE[,2])
ump_SPE <- left_join(df_CPE[,1:5], ump_SPE, 'pdbID')
# ----------------
ump_df <- data.frame(Method = rep(c('SPE', 'CPE', 'TM-Vec'), each=nrow(ump_CPE)),
                     rbind(ump_SPE, ump_CPE, ump_tm))

ump_df$Method <- factor(ump_df$Method,
                        levels = c('SPE', 'CPE', 'TM-Vec'))

p<-ggplot(ump_df, aes(UMAP1, UMAP2, color= subClass)) +
  geom_point(alpha=.7,size=3) +
  facet_grid( ~ Method)+
  mytheme() +
  theme(strip.text.x = element_text(size = 13, face = 'bold'))+
  scale_color_manual(values = c('purple', 'skyblue3', 'yellow3', 'red4'))
# save Fig7D.pdf 10x5
# ----------SourceData---------------
MYwriteData(data=ump_df[,-c(5,6)],
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 7 C",
            row.Names = TRUE)
```

## Fig 7 E  

```{r Fig 7 E (Scalability)}
source('Functions.R')
# ---------------
cpe <- read.csv('Data/csv/CPE_prfc_astral95.csv')
tm <- read.csv('Data/csv/TM_prfc_astral95.csv')
cpe$method <- 'CPE'
tm$method <- 'TM-Vec'

df <- rbind(cpe, tm)

library(scales)
p0<-ggplot(df, aes(n_protein, time/sumAA, col=method)) +
  geom_point(size=5, alpha=1) +
  geom_smooth(method = "lm", se = FALSE, alpha=1,linewidth = 1) +  
#  annotation_custom(
#    grob = textGrob(bquote(paste("T/AA = 1.04 × 10"^"-8",
#                                 " N + 1.36 × 10"^"-5")),
#                    gp = gpar(col = "black", fontsize = 15)),
#    xmin = 10000, xmax = 10000, ymin = 0.0003, ymax = 0.0003)+
#  annotation_custom(
#    grob = textGrob(bquote(paste("T/AA = 2.33 × 10"^"-8",
#                                 " N + 4.71 × 10"^"-4")),
#                    gp = gpar(col = "black", fontsize = 15)),
#    xmin = 10000, xmax = 10000, ymin = 0.001, ymax = 0.001)+
  xlab('N Protein')+
  ylab('Time/AminoAcid (Sec/AA)') +
  scale_color_manual(values = c('magenta', 'orange2'))+
  mytheme()+
  theme(legend.position = c(.85,.09),
        legend.text = element_text(face = 'bold'),
        axis.title.y = element_text(size = 13, vjust = 2),
        axis.title.x = element_text(size = 13, vjust = -.5))+
  scale_x_continuous(breaks = unique(df$n_protein),
                     labels = label_number(scale = 1e-3))+
  scale_y_continuous(breaks = seq(0,1.2, .2)/1000,
                     labels = label_number(scale = 1e3))+
#  annotation_custom(
#  grob = textGrob(bquote("×10"^-3),
#                  gp = gpar(col = "black", fontsize = 15),rot=90),
#  xmin = -3100, xmax = -3100, ymin = 0.0011, ymax = 0.0011) +
#  annotation_custom(
#  grob = textGrob(bquote("×10"^3),
#                  gp = gpar(col = "black", fontsize = 15)),
#  xmin = 19000, xmax = 19000, ymin = -.00015, ymax = -.00015) +
  coord_cartesian(clip = "off") 

p1<-ggarrange(NULL,p0,ncol = 2,widths = c(.01,1))
# save Fig7E.pdf 7x4.5

# ----------SourceData---------------
MYwriteData(data=df,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure 7 E",
            row.Names = TRUE)
```

## Fig S1 

```{r Fig S1 210Cor}
source('Functions.R')
res <- readRDS('Data/rds/Length_Vs_Energy.rds')
res <- res[!is.na(res$length),-c(1,3:13)]
res <- res[-4383,]
# ----------SourceData---------------
MYwriteData(data=res,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure S1",
            row.Names = TRUE)
# -- save plots
#   st <- seq(1,210,35)
#   en <- seq(35,210,35)
#   
#   for (i in 1:6) {
#     print(i)
#     s <- st[i]+1
#     e <- en[i]+1
#     df <- res[,c(1, s:e)]
#     df2 <- df %>%
#       gather(key = "Int_i", value = "Value", -length)
#     p <- ggplot(df2, aes(length, Value))+
#       geom_point()+
#       stat_cor(aes(label = ..r.label..),
#                method = 'spearman', col='red', geom = "label",
#                label.x = 0, label.y = range(df2$Value,na.rm = #   T)[2]+.05)+
#       facet_wrap( ~ Int_i, ncol = 5)+
#       ylab('CPE - SPE') +
#       ylim(range(df2$Value,na.rm = T)[1]-.1, range(df2$Value,na.rm = #   T)[2]+.1)+
#       theme_bw(base_line_size = .1)+
#       theme(strip.text.x = element_text(size = 13),
#             axis.title.x = element_text(size = 20),
#             axis.title.y = element_text(size = 20),
#             panel.grid.major = element_line(color = "black"),
#       panel.grid.minor = element_line(color = "black"))
#     ggsave(filename = paste0('Fig_Manuscript/Revised/supp_Fig/210Cor', #   i, '.png'), 
#            plot = p, width = 9, height = 13, 
#            dpi = 600, device = "png")
#   }
```

## Fig S2

```{r Fig S2}
source('Functions.R')
# ----------------------------------------------------
df_tot <- readRDS('Fig_Manuscript/fig_data/df_total_energy.rds')
df_tot <- df_tot[df_tot$class%in%c('a','b','c','d'),]
df_tot$class[df_tot$class=='a']<-'All_alpha'
df_tot$class[df_tot$class=='b']<-'All_beta'
df_tot$class[df_tot$class=='c']<-'alpha/beta'
df_tot$class[df_tot$class=='d']<-'alpha+beta'
df_tot$data <- ifelse(df_tot$data=='Astral 40','ASTRAL40','ASTRAL95')

AB <- ggplot(df_tot,aes(class,str/length,fill=class))+
  geom_boxplot(outlier.alpha = .01)+
  facet_wrap(vars(data))+
  xlab('')+ylab('Total energy (structure)')+
  coord_cartesian(ylim = c(-15, -4))+
  mytheme()+theme(legend.title = element_blank(),
               axis.text.x = element_blank(),
               strip.text = element_text(size = 23))+
  scale_fill_manual(values = c('green', 'purple', 'gold3', 'magenta'))

CD <- ggplot(df_tot,aes(class,seq/(2*length),fill=class))+
  geom_boxplot(outlier.alpha = .1)+
  facet_wrap(vars(data))+
  xlab('')+ylab('Total energy (sequence)')+
  coord_cartesian(ylim = c(-15, -6))+
  mytheme()+theme(legend.title = element_blank(),
               axis.text.x = element_blank(),
               strip.text = element_text(size = 23))+
  scale_fill_manual(values = c('green', 'purple', 'gold3', 'magenta'))


p <- ggarrange(NULL,
              AB, NULL,CD, ncol = 1,
              heights = c(0.09, 1, 0.09, 1),
              labels = c('',"A", "", "B"),
              common.legend = T, legend = 'bottom',
              font.label = list(size = 30),    # color = "red"
              hjust = -.5,vjust = 0
              )
# save FIG3.pdf 10x9
# ----------SourceData---------------
df_tot$seq <- df_tot$seq/(2*df_tot$length)
df_tot$str <- df_tot$str/(df_tot$length)
MYwriteData(data=df_tot[,-3],
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure S2",
            row.Names = TRUE)
```

## Fig S3

```{r Fig S3}
source('Functions.R')
# ----------------------------------------------------
load('Fig_Manuscript/fig_data/UMPs.RData')
ump_df_seq40$class <- gsub('-',' ', ump_df_seq40$class)
ump_df_str40$class <- gsub('-',' ', ump_df_str40$class)
ump_df_seq95$class <- gsub('-',' ', ump_df_seq95$class)
ump_df_str95$class <- gsub('-',' ', ump_df_str95$class)

p1<-ggplot(ump_df_seq40,aes(UMAP1, UMAP2, color= class)) +
  geom_point(alpha=.7,size=.5) + 
  facet_grid(~type)+xlab('')+
  mytheme()+scale_color_manual(values = c(rgb(244/255, 200/255, 0/255),'purple'))+
  theme(legend.title = element_blank(),
        legend.text = element_text(size = 18),
        strip.text = element_text(size = 23))+
  guides(color = guide_legend(override.aes = list(size = 5)))

p2<-ggplot(ump_df_str40,aes(UMAP1, UMAP2, color= class)) +
  geom_point(alpha=.7,size=.5) +
  facet_grid(data ~ type)+xlab('')+ylab('')+
  mytheme()+scale_color_manual(values = c(rgb(244/255, 200/255, 0/255),'purple'))+
  theme(legend.title = element_blank(),
        legend.text = element_text(size = 18),
        strip.text = element_text(size = 23))+
  guides(color = guide_legend(override.aes = list(size = 5)))

p3<-ggplot(ump_df_seq95,aes(UMAP1, UMAP2, color= class)) +
  geom_point(alpha=.7,size=.5) +
  #facet_grid(cols = vars(type))+
  mytheme()+scale_color_manual(values = c(rgb(244/255, 200/255, 0/255),'purple'))+
  theme(legend.title = element_blank(),
        legend.text = element_text(size = 18),
        strip.text = element_text(size = 23))+
  guides(color = guide_legend(override.aes = list(size = 5)))


p4<-ggplot(ump_df_str95,aes(UMAP1, UMAP2, color= class)) +
  geom_point(alpha=.7,size=.5) + 
  facet_grid(rows = vars(data))+ylab('')+
  mytheme()+scale_color_manual(values = c(rgb(244/255, 200/255, 0/255),'purple'))+
  theme(legend.title = element_blank(),
        legend.text = element_text(size = 18),
        strip.text = element_text(size = 23))+
  guides(color = guide_legend(override.aes = list(size = 5)))
# ----------SourceData---------------
ump <- rbind(ump_df_seq40,ump_df_str40,
             ump_df_seq95, ump_df_str95)
MYwriteData(data=ump,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure S3",
            row.Names = TRUE)
# ********
p <- ggarrange(NULL,NULL,NULL,
          p1,NULL,p2,
          NULL,NULL,NULL,
          p3,NULL,p4,ncol = 3,nrow = 4,
          common.legend = T,
          labels = c('','','',
                     "", "", "",
                     '','','',
                     '','',''),
          font.label = list(size = 30),
          widths = c(1, 0.01, 1),
          heights = c(0, 1, 0.1, 1),
          hjust = -.5,vjust = 0,legend = 'bottom')


# save FIG4.pdf 10x10
```

## Fig S4 EF

```{r Fig S4 EF}
source('Functions.R')
# ----------------------------------------------------
# ******** a.29.3
# ****** structure40
df <- readRDS("Data/rds/E210.astral_40.rds") 
df = df[!is.na(df$FL),] 
idx<- df$superfamily%in% 'a.29.3'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$family[idx], dis)
fa<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(fa)) {  
  print(i)   
  x <- fa[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
df_family_str <- data.frame(data='ASTRAL40-superfamily a.29.3',type='SPE',
                            gr=rep(c('within','between'),
                                   c(length(within),length(between))),
                            distance=c(within,between)) 
# ****** sequence40
df <- readRDS("Data/rds/E210.seq_astral_40.rds") 
df = df[!is.na(df$FL),] 
idx<- df$superfamily%in% 'a.29.3'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$family[idx], dis)
fa<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(fa)) {  
  print(i)   
  x <- fa[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
df_family_seq <- data.frame(data='ASTRAL40-superfamily a.29.3',type='CPE',
                            gr=rep(c('within','between'),
                                   c(length(within),length(between))),
                            distance=c(within,between)) 
# **********
df_family <- rbind(df_family_seq, df_family_str)
# ----------SourceData---------------
MYwriteData(data=df_family,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure S4 E",
            row.Names = TRUE)


E <- ggboxplot(df_family, x='gr',y='distance', 
          fill = 'gr',alpha=.5,outlier.shape = NA)+
  stat_summary(fun.y=mean, geom="point", shape=8,
               size=2, color="red", fill="red")+xlab('')+
  mytheme()+
  theme(legend.position = 'none',
        axis.text.x = element_blank(),
    strip.text = element_text(size = 15))+
  scale_fill_viridis(discrete = TRUE, alpha=0.6)+
  stat_compare_means(method = 't.test',size=5,label.y = 18,
                     label.x = 1.4,
                     aes(label = paste0("p = ", ..p.format..)))+
  facet_nested(~ data + type)+
  coord_cartesian(ylim = c(0, 20))

mylegend <- get_legend(E+
                           theme(legend.position = 'bottom',
                                 legend.title = element_blank(),
                                 legend.text = element_text(size = 20)))
saveRDS(mylegend,'Fig_Manuscript/fig_data/FigS4mylegend.rds')
# ******************************************
# ****** structure95
df <- readRDS("Data/rds/E210.astral_95.rds") 
df = df[!is.na(df$FL),] 
idx<- df$superfamily%in% 'a.29.3'
dis <- as.matrix(dist(df[idx,-c(1:8)], method = manhat))
dis <- data.frame(gr=df$family[idx], dis)
fa<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(fa)) {  
  print(i)   
  x <- fa[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
df_family_str <- data.frame(data='ASTRAL95-superfamily a.29.3',type='SPE',
                            gr=rep(c('within','between'),
                                   c(length(within),length(between))),
                            distance=c(within,between)) 
# ****** sequence95
df <- readRDS("Data/rds/E210.seq_astral_95.rds") 
df = df[!is.na(df$FL),] 
idx<- df$superfamily%in% 'a.29.3'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$family[idx], dis)
fa<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(fa)) {  
  print(i)   
  x <- fa[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
df_family_seq <- data.frame(data='ASTRAL95-superfamily a.29.3',type='CPE',
                            gr=rep(c('within','between'),
                                   c(length(within),length(between))),
                            distance=c(within,between)) 
# **********
df_family <- rbind(df_family_seq, df_family_str)
# ----------SourceData---------------
MYwriteData(data=df_family,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure S4 F",
            row.Names = TRUE)

FF <- ggboxplot(df_family, x='gr',y='distance', 
          fill = 'gr',alpha=.5,outlier.shape = NA)+
  stat_summary(fun.y=mean, geom="point", shape=8,
               size=2, color="red", fill="red")+xlab('')+
  mytheme()+
  theme(legend.position = 'none',
    axis.text.x = element_blank(),
    strip.text = element_text(size = 15))+
  scale_fill_viridis(discrete = TRUE, alpha=0.6)+
  stat_compare_means(method = 't.test',size=5,
                     label.y = 18,
                     label.x = 1.4,
                     aes(label = paste0("p = ", ..p.format..)))+
  facet_nested(~ data + type)+ylab('')+
  coord_cartesian(ylim = c(0, 20))
# ********
E<-ggarrange(
  NULL,E, ncol = 1,
  heights = c(0.07, 1),
  labels = c('', "E"),
  font.label = list(size = 30), # color = "red"
  hjust = -.5,vjust = 0
)
FF<-ggarrange(
  NULL,FF, ncol = 1,
  heights = c(0.07, 1),
  labels = c('', "F"),
  font.label = list(size = 30), # color = "red"
  hjust = -.5,vjust = 0
)
EF<-ggarrange(
    E, NULL,FF, nrow = 1,
    widths = c(1, 0.3, 1),
    labels = c("", "", "")
    )

#saveRDS(EF,'Fig_Manuscript/fig_data/FigS4EF.rds')
```

## Fig S4 CD

```{r Fig S4 CD}
source('Functions.R')
# ----------------------------------------------------
# ******** a.29
# ****** structure40
df <- readRDS("Data/rds/E210.astral_40.rds") 
df = df[!is.na(df$FL),] 
idx<- df$fold%in% 'a.29'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$superfamily[idx], dis)
sf<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(sf)) {  
  print(i)   
  x <- sf[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
df_superfamily_str <- data.frame(data='ASTRAL40-fold a.29',type='SPE',
                            gr=rep(c('within','between'),
                                   c(length(within),length(between))),
                            distance=c(within,between)) 
# ****** sequence40
df <- readRDS("Data/rds/E210.seq_astral_40.rds") 
df = df[!is.na(df$FL),] 
idx<- df$fold%in% 'a.29'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$superfamily[idx], dis)
sf<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(sf)) {  
  print(i)   
  x <- sf[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
df_superfamily_seq <- data.frame(data='ASTRAL40-fold a.29',type='CPE',
                            gr=rep(c('within','between'),
                                   c(length(within),length(between))),
                            distance=c(within,between)) 
# **********
df_superfamily <- rbind(df_superfamily_seq, df_superfamily_str)
# ----------SourceData---------------
MYwriteData(data=df_superfamily,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure S4 C",
            row.Names = TRUE)


my_comparisons<-list(c('within','between'))

C <- ggboxplot(df_superfamily, x='gr',y='distance', 
          fill = 'gr',alpha=.5,outlier.shape = NA)+
  stat_summary(fun.y=mean, geom="point", shape=8,
               size=2, color="red", fill="red")+xlab('')+
  mytheme()+
  theme(legend.position = 'none',
        axis.text.x = element_blank(),
    strip.text = element_text(size = 15))+
  scale_fill_viridis(discrete = TRUE, alpha=0.6)+
  stat_compare_means(comparisons = my_comparisons, method = 't.test',
                     size=5,label.y = 17,
                     label.x = 1.4,bracket.size = NA,tip.length = NA)+
  facet_nested(~ data + type)+
  coord_cartesian(ylim = c(0, 20))
# ******************************************
# ****** structure95
df <- readRDS("Data/rds/E210.astral_95.rds") 
df = df[!is.na(df$FL),] 
idx<- df$fold%in% 'a.29'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$superfamily[idx], dis)
sf<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(sf)) {  
  print(i)   
  x <- sf[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
df_superfamily_str <- data.frame(data='ASTRAL95-fold a.29',type='SPE',
                            gr=rep(c('within','between'),
                                   c(length(within),length(between))),
                            distance=c(within,between)) 
# ****** sequence95
df <- readRDS("Data/rds/E210.seq_astral_95.rds") 
df = df[!is.na(df$FL),] 
idx<- df$fold%in% 'a.29'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$superfamily[idx], dis)
sf <- unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(sf)) {  
  print(i)   
  x <- sf[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
df_superfamily_seq <- data.frame(data='ASTRAL95-fold a.29',type='CPE',
                            gr=rep(c('within','between'),
                                   c(length(within),length(between))),
                            distance=c(within,between)) 
# **********
df_superfamily <- rbind(df_superfamily_seq, df_superfamily_str)
# ----------SourceData---------------
MYwriteData(data=df_superfamily,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure S4 D",
            row.Names = TRUE)

D <- ggboxplot(df_superfamily, x='gr',y='distance', 
          fill = 'gr',alpha=.5,outlier.shape = NA)+
  stat_summary(fun.y=mean, geom="point", shape=8,
               size=2, color="red", fill="red")+xlab('')+
  mytheme()+
  theme(legend.position = 'none',
        axis.text.x = element_blank(),
    strip.text = element_text(size = 15))+
  scale_fill_viridis(discrete = TRUE, alpha=0.6)+
  stat_compare_means(comparisons = my_comparisons, method = 't.test',size=5,
                     label.x = 1.4,label.y = 17,bracket.size = NA,tip.length = NA)+
  facet_nested(~ data + type)+ylab('')+
  coord_cartesian(ylim = c(0, 20))
# ********
C<-ggarrange(
  NULL,C, ncol = 1,
  heights = c(0.07, 1),
  labels = c('', "C"),
  font.label = list(size = 30), # color = "red"
  hjust = -.5,vjust = 0
)
D<-ggarrange(
  NULL,D, ncol = 1,
  heights = c(0.07, 1),
  labels = c('', "D"),
  font.label = list(size = 30), # color = "red"
  hjust = -.5,vjust = 0
)
CD<-ggarrange(
    C, NULL,D, nrow = 1,
    widths = c(1, 0.3, 1),
    labels = c("", "", "")
    )
#saveRDS(CD,'Fig_Manuscript/fig_data/FigS4CD.rds')
```

## Fig S4 AB

```{r Fig S4 AB}
source('Functions.R')
# ----------------------------------------------------
# ******** a
# ****** structure40
df <- readRDS("Data/rds/E210.astral_40.rds") 
df = df[!is.na(df$FL),] 
idx<- df$class%in% 'a'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$fold[idx], dis)
fo<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(fo)) {  
  print(i)   
  x <- fo[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
wit <- within[sample(length(within),1000)] 
bet <- between[sample(length(between),1000)] 
df_fold_str <- data.frame(data='ASTRAL40-class alpha',type='SPE',
                            gr=rep(c('within','between'),c(1000,1000)),
                            distance=c(wit,bet)) 
# ****** sequence40
df <- readRDS("Data/rds/E210.seq_astral_40.rds") 
df = df[!is.na(df$FL),] 
idx<- df$class%in% 'a'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$fold[idx], dis)
fo <- unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(fo)) {  
  print(i)   
  x <- fo[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
wit <- within[sample(length(within),1000)] 
bet <- between[sample(length(between),1000)] 
df_fold_seq <- data.frame(data='ASTRAL40-class alpha',type='CPE',
                            gr=rep(c('within','between'),c(1000,1000)),
                            distance=c(wit,bet))
# **********
df_fold <- rbind(df_fold_seq, df_fold_str)
# ----------SourceData---------------
MYwriteData(data=df_fold,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure S4 A",
            row.Names = TRUE)

my_comparisons<-list(c('within','between'))

A <- ggboxplot(df_fold, x='gr',y='distance', 
          fill = 'gr',alpha=.5,outlier.shape = NA)+
  stat_summary(fun.y=mean, geom="point", shape=8,
               size=2, color="red", fill="red")+xlab('')+
  mytheme()+
  theme(legend.position = 'none',
        axis.text.x = element_blank(),
    strip.text = element_text(size = 15))+
  scale_fill_viridis(discrete = TRUE, alpha=0.6)+
  stat_compare_means(comparisons = my_comparisons, method = 't.test',size=5,
                     label.y = 30,bracket.size = NA,tip.length = NA)+
  facet_nested(~ data + type)+
  coord_cartesian(ylim = c(0, 40))
# ******************************************
# ****** structure95
df <- readRDS("Data/rds/E210.astral_95.rds") 
df = df[!is.na(df$FL),] 
idx<- df$class%in% 'a'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$fold[idx], dis)
fo<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(fo)) {  
  print(i)   
  x <- fo[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
wit <- within[sample(length(within),1000)]
bet <- between[sample(length(between),1000)] 
df_fold_str <- data.frame(data='ASTRAL95-class alpha',type='SPE',
                            gr=rep(c('within','between'),c(1000,1000)),
                            distance=c(wit,bet))
# ****** sequence95
df <- readRDS("Data/rds/E210.seq_astral_95.rds") 
df = df[!is.na(df$FL),] 
idx<- df$class%in% 'a'
dis <- as.matrix(dist(df[idx,-c(1:8)],method = manhat))
dis <- data.frame(gr=df$fold[idx], dis)
fo<-unique(dis$gr)

between<-c() 
within<-c()
for (i in 1:length(fo)) {  
  print(i)   
  x <- fo[i]  
  idx <- which(dis$gr==x)
  y <- dis[idx,(idx+1)]
  within <- c(within, y[lower.tri(y)])
  x <- dis[idx,-c(1,idx+1)]   
  x <- as.vector(as.matrix(x))  
  x <- x[!is.na(x)]   
  between<-c(between,x) 
}
wit <- within[sample(length(within),1000)] 
bet <- between[sample(length(between),1000)] 
df_fold_seq <- data.frame(data='ASTRAL95-class alpha',type='CPE',
                            gr=rep(c('within','between'),c(1000,1000)),
                            distance=c(wit,bet)) 
# **********
df_fold <- rbind(df_fold_seq, df_fold_str)
# ----------SourceData---------------
MYwriteData(data=df_fold,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure S4 B",
            row.Names = TRUE)

B <- ggboxplot(df_fold, x='gr',y='distance', 
          fill = 'gr',alpha=.5,outlier.shape = NA)+
  stat_summary(fun.y=mean, geom="point", shape=8,
               size=3, color="red", fill="red")+xlab('')+
  mytheme()+
  theme(legend.position = 'none',
        axis.text.x = element_blank(),
    strip.text = element_text(size = 15))+
  scale_fill_viridis(discrete = TRUE, alpha=0.6)+
  stat_compare_means(comparisons = my_comparisons, method = 't.test',size=5,
                     label.y = 30, bracket.size = NA,tip.length = NA)+
  facet_nested(~ data + type)+ylab('')+
  coord_cartesian(ylim = c(0, 40))
# ********
A<-ggarrange(
  NULL,A, ncol = 1,
  heights = c(0.07, 1),
  labels = c('', "A"),
  font.label = list(size = 30), # color = "red"
  hjust = -.5,vjust = 0
)
B<-ggarrange(
  NULL,B, ncol = 1,
  heights = c(0.07, 1),
  labels = c('', "B"),
  font.label = list(size = 30), # color = "red"
  hjust = -.5,vjust = 0
)
AB<-ggarrange(
    A, NULL,B, nrow = 1,
    widths = c(1, 0.3, 1),
    labels = c("", "", "")
    )
saveRDS(AB,'Fig_Manuscript/fig_data/FigS4AB.rds')
```

##Fig S4 

```{r Fig S4 All}
rm(list = ls())
EF <- readRDS('Fig_Manuscript/fig_data/FigS4EF.rds')
CD <- readRDS('Fig_Manuscript/fig_data/FigS4CD.rds')
AB <- readRDS('Fig_Manuscript/fig_data/FigS4AB.rds')
mylegend <-readRDS('Fig_Manuscript/fig_data/FigS4mylegend.rds')
# ---------------------------------------------
f6all <- ggarrange(NULL,AB,CD,EF,ncol = 1,
          heights = c(0.1,1,1,1))
p<-cowplot::plot_grid(f6all, mylegend, ncol = 1, rel_heights = c(1,.03))
# save pdf 10x9
```


## Fig S9-S11 

```{r Fig S9-S11 UMAP}
source('Functions.R')
load('Data/Large_Scale_SARS2/SPE_CPE_time.RData')
library(RColorBrewer)
palette_Dark2 <- colorRampPalette(brewer.pal(28, "Dark2"))
#---------------------------------------------
# ----------SPE
dis <- as.matrix(proxy::dist(df_SPE[,-c(1:4)], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 150,
            min_dist = 0.1,
            input="dist")
ump <- ump$layout
ump_df <- data.frame(family=df_SPE$sars_proteom, 
                  UMAP1=ump[,1],
                  UMAP2=ump[,2])
spe <- ump_df
spe$method <- 'SPE'

p_SPE <- ggplot(ump_df, aes(UMAP1, UMAP2, color= family)) +
  geom_point(size = 3,alpha=.7) +
  scale_fill_manual(values = palette_Dark2(14)) +
  mytheme()+
  ggtitle('SPE')+
  theme(legend.position = "bottom",
        legend.text = element_text(size=13,color = 'black'),
        axis.text.x = element_text(size = 25,face = 'bold',color = 'black'),
        axis.text.y = element_text(size = 25,face = 'bold',color = 'black'),
        axis.title.x = element_text(size =25,face = 'bold',color = 'black'),
        axis.title.y = element_text(size =25,face = 'bold',color = 'black'),
        title = element_text(size =25,face = 'bold',color = 'black'))
# save SARS_Proteom_SPE.pdf 13x17
# ------------------------------------
# ------------CPE
dis <- as.matrix(proxy::dist(df_CPE[,-c(1:4)], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 150,
            min_dist = 0.1,
            input="dist")
ump <- ump$layout
ump_df <- data.frame(family=df_CPE$sars_proteom, 
                  UMAP1=ump[,1],
                  UMAP2=ump[,2])

cpe <- ump_df
cpe$method <- 'CPE'

p_CPE <- ggplot(ump_df, aes(UMAP1, UMAP2, color= family)) +
  geom_point(size = 3,alpha=.7) +
  scale_fill_manual(values = palette_Dark2(14)) +
  mytheme()+
  ggtitle('CPE')+
  theme(legend.position = "bottom",
        legend.text = element_text(size=13,color = 'black'),
        axis.text.x = element_text(size = 25,face = 'bold',color = 'black'),
        axis.text.y = element_text(size = 25,face = 'bold',color = 'black'),
        axis.title.x = element_text(size =25,face = 'bold',color = 'black'),
        axis.title.y = element_text(size =25,face = 'bold',color = 'black'),
        title = element_text(size =25,face = 'bold',color = 'black'))
# save SARS_Proteom_CPE.pdf 13x17
# ------------------------------------
# ------------TM-vec
# ----------- load from google drive ---------------
drive_deauth()
drive_user()
public_file <- drive_get(as_id("1QUAqdkypCPHraNESF11arxg29tDUp521"))
drive_download(public_file, overwrite = TRUE)
tm <- read.csv('disTM_vec_sars_proteom.csv')
file.remove('disTM_vec_sars_proteom.csv')
tm <- 1 - tm
ump <- umap(d = tm,
            n_neighbors = 150,
            min_dist = 0.1,
            input="dist")
ump <- ump$layout
ump_df <- data.frame(family=df_CPE$sars_proteom, 
                  UMAP1=ump[,1],
                  UMAP2=ump[,2])

tm <- ump_df
tm$method <- 'TM-Vec'

p_tm <- ggplot(ump_df, aes(UMAP1, UMAP2, color= family)) +
  geom_point(size = 3,alpha=.7) +
  scale_fill_manual(values = palette_Dark2(14)) +
  mytheme()+
  ggtitle('TM-Vec')+
  theme(legend.position = "bottom",
        legend.text = element_text(size=13,color = 'black'),
        axis.text.x = element_text(size = 25,face = 'bold',color = 'black'),
        axis.text.y = element_text(size = 25,face = 'bold',color = 'black'),
        axis.title.x = element_text(size =25,face = 'bold',color = 'black'),
        axis.title.y = element_text(size =25,face = 'bold',color = 'black'),
        title = element_text(size =25,face = 'bold',color = 'black'))
# save SARS_Proteom_TM_Vec.pdf 13x17
# -------------------------
#ggarrange(p_SPE,p_CPE,p_tm,common.legend = T,ncol = 1, legend = 'bottom')

# ----------SourceData---------------
ump <- rbind(cpe, spe, tm)
MYwriteData(data=ump,
            file="Fig_Manuscript/SourceData.xlsx", 
            sheet="Figure S9-S11",
            row.Names = TRUE)
```


#Tables

## Table 1 

```{r Table 1 CT_Ho}
source('Functions.R')
# ----------------------------------------------------
# ---------------- SPE -------------------
df <- readRDS('Data/rds/E210_CT_Ho.rds')
dis <- as.matrix(df[,-c(1:9)] %>% proxy::dist(method = manhat))
diag(dis) <- NA
im <- apply(dis,1, FUN = which.min)
res <- data.frame(source=df$cathID, target=df$cathID[im])
## ----------
twoSF <- unique(df$cathID)
cm <- matrix(0,2,2)
colnames(cm) <- rownames(cm) <- twoSF
for (sf in twoSF) {
  x <- res[res$source%in%sf,]
  z <- table(x$target)
  cm[sf,names(z)]<-z
}
ac <- round(sum(diag(cm))/sum(cm),2)*100
res_final <- data.frame(Method='SPE', Accuracy=ac, Time='187 sec') 
# -------------------------------------------------
# ------------------------------------------
# ------------ CPE ---------------------------
# -----------------------------------
df <- readRDS('Data/rds/E210.seq_CT_Ho.rds')
df <- df[!is.na(df$FF),]
twoSF <- unique(df$cathID)
dis <- as.matrix(dist(df[,-c(1:9)], method = manhat))
diag(dis) <- NA
# ---------- 1NN --------------------
im <- apply(dis,1, FUN = which.min)
res <- data.frame(source=df$cathID, target=df$cathID[im])
## ----------
cm <- matrix(0,2,2)
colnames(cm) <- rownames(cm) <- twoSF
for (sf in twoSF) {
  x <- res[res$source%in%sf,]
  z <- table(x$target)
  cm[sf,names(z)]<-z
}
ac<-round(sum(diag(cm))/sum(cm),2)*100
# ---------------------------------------------
# -------- Note: --------------
# the time was obtaind from the `C-terminal_Homing_CPE` chunck in `Profile_Energy.Rmd` script.
# ---------------------------------------------------------------
res_final <- rbind(res_final, c('CPE', ac, '1 sec'))
# ------------------------------------------
# ------------- TM-vec -----------------------------
# ---------------------------------------------
dis <- read.csv('Data/csv/disTM_vec_CT_Ho.csv')
diag(dis) <- 1
dis <- 1 - dis
diag(dis) <- NA

# ---------- 1NN --------------------
im <- apply(dis,1, FUN = which.min)
res <- data.frame(source=df$cathID, target=df$cathID[im])
## ----------
cm <- matrix(0,2,2)
colnames(cm) <- rownames(cm) <- twoSF
for (sf in twoSF) {
  x <- res[res$source%in%sf,]
  z <- table(x$target)
  cm[sf,names(z)]<-z
}
ac <- round(sum(diag(cm))/sum(cm),2)*100

res_final <- rbind(res_final, c('TM-Vec', ac, '67 sec'))
res_final
```

## Table 2 

```{r Table 2 FiveSF}
source('Functions.R')
# -----------------------------------------
# ----------------------CPE------------------------------
# -------- 5 sf
win <- 'a.4.5'
imu <- 'b.1.1'
ph <- 'b.55.1'
ub <- 'd.15.1'
ntf <- 'd.17.4'

fiveSF <- c(win,imu,ph,ub,ntf)
#############
df5 <- readRDS('Data/rds/E210.seq_fiveSF.rds')
# ----------------------------------------
# ---------- 1NN ----------------
# ---------------------------------------
dis <- as.matrix(dist(df5[,-c(1:9)], method = manhat))
diag(dis) <- NA

im <- apply(dis, 1, FUN = which.min)
res <- data.frame(source=df5$superfamily, target=df5$superfamily[im])
cm <- matrix(0,5,5)
colnames(cm) <- rownames(cm) <- fiveSF
for (sf in fiveSF) {
  x <- res[res$source%in%sf,]
  z <- table(x$target)
  cm[sf,names(z)]<-z
}
Accuracy<-sum(diag(cm))/sum(cm)
Precision <- diag(cm)/colSums(cm)
Recall <- diag(cm)/rowSums(cm)
F1 <- (2 * Precision * Recall) /(Precision + Recall)

res1NN_CPE<-data.frame(t(round(c(Accuracy,F1),2)))
# ----------------------------------------
# -----------------
res_final <- res1NN_CPE
colnames(res_final)<-colnames(res)
# -------------------------------------------------
# ----------------------SPE------------------------------
df5 <- readRDS('Data/rds/E210_fiveSF.rds')
df5 <- df5[!is.na(df5$FF),]
# ----------------------------------------
# ---------- 1NN ----------------
# ---------------------------------------
dis <- as.matrix(proxy::dist(df5[,-c(1:9)], method = manhat))
diag(dis) <- NA

im <- apply(dis, 1, FUN = which.min)
res <- data.frame(source=df5$superfamily, target=df5$superfamily[im])
cm <- matrix(0,5,5)
colnames(cm) <- rownames(cm) <- fiveSF
for (sf in fiveSF) {
  x <- res[res$source%in%sf,]
  z <- table(x$target)
  cm[sf,names(z)]<-z
}
Accuracy<-sum(diag(cm))/sum(cm)
Precision <- diag(cm)/colSums(cm)
Recall <- diag(cm)/rowSums(cm)
F1 <- (2 * Precision * Recall) /(Precision + Recall)

res1NN_SPE<-data.frame(t(round(c(Accuracy,F1),2)))
# ----------------------------------------
# -----------------
res_final <- rbind(res_final, res1NN_SPE)
# ----------------------------------------------------------------
# ----------------------------------------
# ---------- TM-Vec ----------------
# -------------------------------------------------------
# ----------- load from google drive ---------------
drive_deauth()
drive_user()
public_file <- drive_get(as_id("1coQdrd5zbTT9F0_pBkLARgR-uTHUINiC"))
drive_download(public_file, overwrite = TRUE)
dis<-read.csv('disTM_vec_fiveSF.csv')
file.remove('disTM_vec_fiveSF.csv')
diag(dis) <- NA
# ---------- 1NN ----------------
im <- apply(dis,1, FUN = which.max)
rm(dis)
res <- data.frame(source=df5$superfamily, target=df5$superfamily[im])
cm <- matrix(0,5,5)
colnames(cm) <- rownames(cm) <- fiveSF
for (sf in fiveSF) {
  x <- res[res$source%in%sf,]
  z <- table(x$target)
  cm[sf,names(z)]<-z
}
Accuracy<-sum(diag(cm))/sum(cm)
Precision <- diag(cm)/colSums(cm)
Recall <- diag(cm)/rowSums(cm)
F1 <- (2 * Precision * Recall) /(Precision + Recall)

res1nn_TM_Vec<-data.frame(t(c(Accuracy,F1)))
colnames(res1nn_TM_Vec) <- colnames(res_final)
# ------------------------
res_final <- rbind(res_final, res1nn_TM_Vec)
# --------------------------------------
tCPE <- 103
tSPE <- 3524
tTM <- 955
ts <- c(tCPE,tSPE,tTM)
res_final <- data.frame(method=c('CPE (1NN)',
                                 'SPE (1NN)',
                                 'TM_Vec (1NN)'),
                        Time=ts,
                        res_final)
res_final
```

## Table 3 

```{r Table 3 Rho Spearman Rank}
rm(list=ls())
get_ranks <- function(order, reference_order) {
  ranks <- match(order, reference_order)
  ranks[is.na(ranks)] <- NA 
  return(ranks)
}
# ----------------------------
# ----  a.25.1.1 family
# Define tree orders
reference_order <- c("R", "F", "B", "D")
SPE_order <- c("R", "F", "B", "D")
CPE_order <- c("R", "F", "D", "B")
TM_order <- c("R", "D", "B", "F")

# Compute ranks
reference_ranks <- get_ranks(reference_order, reference_order)
SPE_ranks <- get_ranks(SPE_order, reference_order)
CPE_ranks <- get_ranks(CPE_order, reference_order)
TM_ranks <- get_ranks(TM_order, reference_order)

# Calculate Spearman's rank correlation coefficient
SPE <- cor(reference_ranks, SPE_ranks, 
           method = "spearman", use = "complete.obs")
CPE <- cor(reference_ranks, CPE_ranks, 
           method = "spearman", use = "complete.obs")
TM <- cor(reference_ranks, TM_ranks, 
          method = "spearman", use = "complete.obs")
a.25.1.1 <- c(SPE, CPE, TM)
# -------------------------------
# ----  a.25.1.2 family
# Define tree orders
reference_order <- c("F", "R", "A", "B")
SPE_order <- c("R", "F", "B", "A")
CPE_order <- c("R", "F", "B", "A")
TM_order <- c("B", "R", "F", "A")

reference_ranks <- get_ranks(reference_order, reference_order)
SPE_ranks <- get_ranks(SPE_order, reference_order)
CPE_ranks <- get_ranks(CPE_order, reference_order)
TM_ranks <- get_ranks(TM_order, reference_order)

# Calculate Spearman's rank correlation coefficient
SPE <- cor(reference_ranks, SPE_ranks, 
           method = "spearman", use = "complete.obs")
CPE <- cor(reference_ranks, CPE_ranks, 
           method = "spearman", use = "complete.obs")
TM <- cor(reference_ranks, TM_ranks, 
          method = "spearman", use = "complete.obs")
a.25.1.2 <- c(SPE, CPE, TM)

cbind(Method=c('SPE', 'CPE', 'TM'),
      a.25.1.1, a.25.1.2)
```


## Table 4,S1

```{r Table S1}
source('Functions.R')
# ----------------------------------------------------
# -----------------CPE------------------------
df <- readRDS("Data/rds/E210.seq_spike.rds")
x <- as.matrix(df[,-c(1:5)]/df$length)
hc <- x %>% proxy::dist(method = manhat)
hc <- hc%>% hclust(method = "ward.D2")
ARI_CPE <- c()
CE_CPE <- c()
for (i in 3:6) {
  ARI_CPE <- c(ARI_CPE, adjustedRandIndex(cutree(hc, i), df$virus))
  CE_CPE <- c(CE_CPE, classError(cutree(hc, i), df$virus)$errorRate)
}
# -----------------------------------
# -----------------SPE------------------------
df <- readRDS("Data/rds/E210_spike.rds")
x <- as.matrix(df[,-c(1:5)]/df$length)
hc <- x %>% proxy::dist(method = manhat)
hc <- hc %>% hclust(method = "ward.D2")
ARI_SPE <- c()
CE_SPE <- c()
for (i in 3:6) {
  ARI_SPE <- c(ARI_SPE, adjustedRandIndex(cutree(hc, i), df$virus))
  CE_SPE <- c(CE_SPE, classError(cutree(hc, i), df$virus)$errorRate)
}
# -----------------------------------
# -----------------RMSD------------------------
rms <- data.frame(readRDS('Data/covidPDB/RMSD_spike.rds'))
diag(rms)<-0
hc <- as.dist(rms)
hc <- hc %>% hclust(method = "ward.D2")
ARI_RMSD <- c()
CE_RMSD <- c()
for (i in 3:6) {
  ARI_RMSD <- c(ARI_RMSD, adjustedRandIndex(cutree(hc, i), df$virus))
  CE_RMSD <- c(CE_RMSD, classError(cutree(hc, i), df$virus)$errorRate)
}
# -----------------------------------
# -----------------TM-Vec------------------------
tm_vec <- read.csv('Data/covidPDB/disTM_vec_spike.csv')
colnames(tm_vec) <- rownames(tm_vec) <- df$pdbID
tm_vec <- 1 - tm_vec
hc <- as.dist(tm_vec)
hc <- hc %>% hclust(method = "ward.D2")
ARI_TM_Vec <- c()
CE_TM_Vec <- c()
for (i in 3:6) {
  ARI_TM_Vec <- c(ARI_TM_Vec, adjustedRandIndex(cutree(hc, i), df$virus))
  CE_TM_Vec <- c(CE_TM_Vec, classError(cutree(hc, i), df$virus)$errorRate)
}
# -----------------------------------
# -----------------TM-score------------------------
tm <- read.csv('Data/covidPDB/disTM_score_spike.csv')
tm_max <- tm
for (i in 1:nrow(tm_max)) {
  print(i)
  for (j in 1:nrow(tm_max)) {
    tm_max[i,j] <- max(tm_max[i,j], tm_max[j,i])
    tm_max[j,i] <- tm_max[i,j]
  }
}
tm_max <- 1 - tm_max
diag(tm_max) <- NA
hc <- as.dist(tm_max)
hc <- hc%>% hclust(method = "ward.D2")
ARI_TM_score <- c()
CE_TM_score <- c()
for (i in 3:6) {
  ARI_TM_score <- c(ARI_TM_score, adjustedRandIndex(cutree(hc, i), df$virus))
  CE_TM_score <- c(CE_TM_score, classError(cutree(hc, i), df$virus)$errorRate)
}
# --------------------------------
# ---------- MSA ClustalW
# -------------------
df <- read.csv("Data/covidPDB/spike_seq.csv")
rownames(df) <- df$pdbID
df <- df[with(df, order(df$virus,decreasing = T)), ]
# --- Read fasta sequences-------------
sequences <- readAAStringSet("Data/covidPDB/Spike.fasta")
# ---- Multiple sequence alignment using ClustalW---------
alignment <- msa(sequences, method = "ClustalW")
alignment2 <- msaConvert(alignment, type="seqinr::alignment")
d <- seqinr::dist.alignment(alignment2, "identity")
upgma <- phangorn::upgma(d)
ARI_MSA <- c()
CE_MSA <- c()
for (i in 3:6) {
  ARI_MSA <- c(ARI_MSA, adjustedRandIndex(cutree(upgma, i), df$virus))
  CE_MSA <- c(CE_MSA, classError(cutree(upgma, i), df$virus)$errorRate)
}
# ---------------------------------------------
ARI_final <- rbind(ARI_CPE, ARI_SPE, ARI_RMSD, ARI_TM_Vec, ARI_TM_score,ARI_MSA,
                   CE_CPE, CE_SPE, CE_RMSD, CE_TM_Vec, CE_TM_score, CE_MSA)
colnames(ARI_final) <- paste0('cut_tree_', 3:6)
ARI_final
```


## Table 5,S3-S5 

```{r Table 5 performance}
source('Functions.R')
load('Data/Large_Scale_SARS2/SPE_CPE_time.RData')
dff <- df_CPE[,-c(1:4)]
dis_CPE <- as.matrix(dff %>% proxy::dist(method = manhat))
dff = df_SPE[,-c(1:4)]
dis_SPE <- as.matrix(dff %>% proxy::dist(method = manhat))
# ----------- load from google drive ---------------
drive_deauth()
drive_user()
public_file <- drive_get(as_id("1QUAqdkypCPHraNESF11arxg29tDUp521"))
drive_download(public_file, overwrite = TRUE)
tm <- read.csv('disTM_vec_sars_proteom.csv')
file.remove('disTM_vec_sars_proteom.csv')
tm <- 1 - tm
# -----------------------
# ----- from run 'Data/Large_Scale_SARS2/tm_vec_sars_proteom.py'
t_tm <- 685 
# ------------------
summary_1NN <- function(dist, group){
  diag(dist) <- NA
  sprot <- unique(group)
  # ---------- 1NN --------------------
  im <- apply(dist,1, FUN = which.min)
  res <- data.frame(source=group, target=group[im])
  ## ----------
  cm <- matrix(0,28,28)
  colnames(cm) <- rownames(cm) <- sprot
  for (sf in sprot) {
    x <- res[res$source%in%sf,]
    z <- table(x$target)
    cm[sf, names(z)]<-z
  }
  tp <- diag(cm)
  fp <- colSums(cm)-tp
  fn <- rowSums(cm)-tp
  tn <- sum(cm)-(tp+fp+fn)
  # -------------
  AC <- round(sum(tp)/sum(cm), 4)*100
  ac <- round((tp+tn)/sum(cm)*100, 4)
  pr <- round(tp/(tp+fp) , 4)*100
  re <- round(tp/(tp+fn) , 4)*100
  f1 <- round((2*pr*re)/(pr+re))
  summ <- data.frame(t(rbind(AC,ac,pr,re,f1)))
  summ <- rbind(round(colMeans(summ), 4), summ)
  rownames(summ)[1]<-'Average'
  return(summ)
}

dis <- list(tm, dis_CPE, dis_SPE)

summ <- lapply(dis, summary_1NN, df_CPE$sars_proteom)
names(summ) <- c('TM', 'CPE', 'SPE')


cbind(Method=c('TM', 'CPE', 'SPE'),
      rbind(summ[['TM']][1,],
            summ[['CPE']][1,],
            summ[['SPE']][1,]),
      Time=c(t_tm, t_CPE, t_SPE))


#for (sheet_name in names(summ)) {
#  write.xlsx(summ[[sheet_name]][,-1], file='Table_S3_S5.xlsx',
#             sheetName=paste0(sheet_name, '_TotalAccuracy=', #summ[[sheet_name]][1,1]), 
#             append=TRUE, row.names=TRUE)
#}
```