---
title: "Reviewers Response"
author: "peyman"
date: "2024-08-16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load package

```{r}
options(knitr.table.format = "latex")
library(RColorBrewer)
library(scales)
library(readr)
library(data.table)
library(MASS)
library(tidyverse)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggpmisc)
library(ggrepel)
library(ggpubr)
library(ggpointdensity)
library(ggpubr)
library(igraph)
library(fpc)
library(grid)
library(cluster)
library(gridExtra)
library(reshape2)
library(visNetwork)
library(pheatmap)
library(umap)
library(Rtsne)
library(viridis)
library(viridisLite)
library(bio3d)
library(caret)
library(e1071)
library(randomForest)
library(seqinr)
library(geometry)
library(ape)
library(colorspace)
library(dendextend)
library(dynamicTreeCut)
library(caret)
library(ggtree)
library(phangorn)
library(tanggle)
library(stringr)
library(protr)
library(proxy)
library(class)
library(readxl)
library(ggh4x)
library(mclust)
library(googledrive)
library(Biostrings)
library(msa)
library(lsa)
library(TreeDist)
library(binom)
```

# Overlap PDBs

```{r}
test4 <- read.csv('Data/csv/fiveSF.csv')
test3 <- read.csv('Data/csv/spike.csv')
test2 <- read.csv('Data/csv/Ferritin_Like_seq.csv')
test1 <- read.csv('Data/csv/CT_Ho_cathID_filtered.csv')
train <- read.csv("Train_Energy/list_with_chainID.txt",header = F)
# -------
pdbs1 <- toupper(substr(test1$pdbID,1,5))
pdbs2 <- toupper(substr(test2$pdbID,1,5))
pdbs3 <- toupper(substr(test3$pdbID,1,5))
pdbs4 <- toupper(paste0(test4$pdbID, substr(test4$position,1,1)))
pdbs <- unique(c(pdbs1, pdbs2, pdbs3, pdbs4))
m <- match(train$V1, pdbs)
idx_rm <- which(!is.na(m))
idx_rm <- data.frame(V1=idx_rm)
write.csv(idx_rm,'Train_Energy/idx_rm.csv',row.names = F)

```

# Pij_old_new

```{r}
Pij <- read.csv("Data/csv/Pij.csv", header = F,sep = ";")
pij_old <- (Pij + t(Pij))/2

Pij_rm_Olaps <- read.csv("Data/csv/Pij_rm_Olaps.csv",
                         header = F,sep = ";")

Pij_new <- (Pij_rm_Olaps + t(Pij_rm_Olaps))/2

df <- data.frame(Pij_old = pij_old[upper.tri(pij_old,diag = T)],
                 Pij_new = Pij_new[upper.tri(Pij_new,diag = T)])

saveRDS(df, 'Data/rds/Pij_old_new.rds')
```

## --- Visual
```{r}
source('Functions.R')
df <- readRDS('Data/rds/Pij_old_new.rds')

p<-ggplot(df, aes(Pij_old, Pij_new))+
  geom_point(size=3, alpha=.5)+
  geom_smooth(method = 'glm',col='skyblue2',linewidth = .2)+
       stat_cor(method = "pearson", size=5,
                r.accuracy = 1e-6,
                color = "red", geom = "label") +
       scale_color_viridis()+
       theme_bw(base_line_size = .3)+
       xlab('Old predictor')+ylab('New predictor')+
  mytheme()

```

# SARS_Proteom Large Scale

```{r}
source('Functions.R')
E210.SPE <- function(pdb){
  tryCatch({
    pdb0 <- pdb
    sele1 <- atom.select(pdb0,"protein",type = "ATOM")
    sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
    sele4 <- atom.select(pdb0, "noh")
    sele <- combine.select(sele1, sele2,sele4, operator="AND", verbose = FALSE)
    new_pdb <- trim.pdb(pdb0, sele,verbose = FALSE)
    pdbx <- new_pdb$atom
    pdbx <- pdbx[!duplicated(pdbx$resno),]
    seq <- paste0(aa321(pdbx$resid),collapse = '')
    Net_Frame <- NetworkFrame(PDB = new_pdb)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    return(c(nchar(seq), seq, energy210))
  }, error = function(e) {
    cat("Error downloading file:", e$message, "\n")
  })
}
# ------------------------------------------
sars_proteom <- list.dirs("Data/Large_Scale_SARS2",full.names = TRUE)
sars_proteom <- sars_proteom[basename(sars_proteom)!='__pycache__']
sars_proteom <- sars_proteom[basename(sars_proteom)=='RenamedModels']

df <- c()
tictoc::tic()
for (i in 1:length(sars_proteom)) {
  pdbFiles <- list.files(sars_proteom[i],full.names = TRUE)
  pdbs <- purrr::map(pdbFiles, read.pdb)
  print(paste0(i, ': ', length(pdbs)))
  x <- lapply(pdbs, E210.SPE)
  x <- do.call(rbind, x)
  x <- data.frame(sars_proteom=strsplit(sars_proteom[i],'/')[[1]][3],
                  pdbID=basename(pdbFiles),
                  x)
  df <- rbind(df, x)
}
colnames(df)[-c(1,2)] <- c('length', 'seq', pair_inter)
dis_SPE <- as.matrix(proxy::dist(df[,-c(1:4)], method = manhat))
t_SPE <- tictoc::toc()


write.csv(df[,1:4],'Data/Large_Scale_SARS2/sars_proteom.csv',
          row.names = F)
# ---------------
annot <- read.csv('Data/Large_Scale_SARS2/sars_proteom.csv')
annot <- df[,1:4]
df <- data.frame(annot,
                 matrix(NA,ncol=210,nrow=nrow(annot)))
colnames(df)[-c(1:4)] <- pair_inter
tictoc::tic()
for (NP in 1:nrow(df)) {
  print(NP)
  seq <- unlist(strsplit(df$seq[NP],''))
  freq <- c()
  for (j in 1:20){
    freq[j] <-  length(grep(tolower(letters_list[j]),tolower(seq)))
  }
  freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
  en_fr <- aaenergy * freq_matrix
  en_fr <- en_fr/length(seq)
  pair_es <- diag(freq) %*% as.matrix(en_fr)
  aa210_p <- pair_es[lower.tri(pair_es,diag = T)]
  df[NP,5:214] <- aa210_p
}
dis_CPE <- as.matrix(df[,-c(1:4)] %>% proxy::dist(method = manhat))
t_CPE <- tictoc::toc()

save(df_SPE, df_CPE, t_SPE, t_CPE,
     file='Data/Large_Scale_SARS2/SPE_CPE_time.RData')

```

## --- Summary

```{r}
load('Data/Large_Scale_SARS2/SPE_CPE_time.RData')
dff <- df_CPE[,-c(1:4)]/as.numeric(df_CPE$length)
dis_CPE <- as.matrix(dff %>% proxy::dist(method = manhat))
dff = df_SPE[,-c(1:4)]/as.numeric(df_SPE$length)
dis_SPE <- as.matrix(dff %>% proxy::dist(method = manhat))
# ----------- load from google drive ---------------
drive_deauth()
drive_user()
public_file <- drive_get(as_id("1QUAqdkypCPHraNESF11arxg29tDUp521"))
drive_download(public_file, overwrite = TRUE)
tm <- read.csv('disTM_vec_sars_proteom.csv')
file.remove('disTM_vec_sars_proteom.csv')
tm <- 1 - tm
# -----------------------
# ----- from run 'Data/Large_Scale_SARS2/tm_vec_sars_proteom.py'
t_tm <- 685 
# ------------------
summary_1NN <- function(dist, group){
  diag(dist) <- NA
  sprot <- unique(group)
  # ---------- 1NN --------------------
  im <- apply(dist,1, FUN = which.min)
  res <- data.frame(source=group, target=group[im])
  ## ----------
  cm <- matrix(0,28,28)
  colnames(cm) <- rownames(cm) <- sprot
  for (sf in sprot) {
    x <- res[res$source%in%sf,]
    z <- table(x$target)
    cm[sf, names(z)]<-z
  }
  ac <- round(sum(diag(cm))/sum(cm), 4)*100
  pr <- round(diag(cm)/colSums(cm) , 4)*100
  re <- round(diag(cm)/rowSums(cm) , 4)*100
  f1 <- round((2*pr*re)/(pr+re))
  summ <- data.frame(t(rbind(ac,pr,re,f1)))
  summ <- rbind(colMeans(summ), summ)
  rownames(summ)[1]<-'Average'
  return(summ)
}

dis <- list(tm, dis_CPE, dis_SPE)
dis <- list(dis_CPE, dis_SPE)

summ <- lapply(dis, summary_1NN, df_CPE$sars_proteom)
names(summ) <- c('TM', 'CPE', 'SPE')


cbind(Method=c('TM', 'CPE', 'SPE'),
      rbind(summ[['TM']][1,],
            summ[['CPE']][1,],
            summ[['SPE']][1,]),
      Time=c(t_tm, t_CPE, t_SPE))
```

## --- UMAP

```{r}
source('Functions.R')
load('Data/Large_Scale_SARS2/SPE_CPE_time.RData')
palette_Dark2 <- colorRampPalette(brewer.pal(28, "Dark2"))
#---------------------------------------------
# ----------SPE
dis <- as.matrix(proxy::dist(df_SPE[,-c(1:4)], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 150,
            min_dist = 0.1,
            input="dist")
ump <- ump$layout
ump_df <- data.frame(family=df_SPE$sars_proteom, 
                  UMAP1=ump[,1],
                  UMAP2=ump[,2])

p_SPE <- ggplot(ump_df, aes(UMAP1, UMAP2, color= family)) +
  geom_point(size = 3,alpha=.7) +
  scale_fill_manual(values = palette_Dark2(14)) +
  mytheme()+
  ggtitle('SPE')+
  theme(legend.position = "bottom",
        legend.text = element_text(size=13,color = 'black'),
        axis.text.x = element_text(size = 25,face = 'bold',color = 'black'),
        axis.text.y = element_text(size = 25,face = 'bold',color = 'black'),
        axis.title.x = element_text(size =25,face = 'bold',color = 'black'),
        axis.title.y = element_text(size =25,face = 'bold',color = 'black'),
        title = element_text(size =25,face = 'bold',color = 'black'))
# save SARS_Proteom_SPE.pdf 13x17
# ------------------------------------
# ------------CPE
dis <- as.matrix(proxy::dist(df_CPE[,-c(1:4)], method = manhat))
ump <- umap(d = dis,
            n_neighbors = 150,
            min_dist = 0.1,
            input="dist")
ump <- ump$layout
ump_df <- data.frame(family=df_CPE$sars_proteom, 
                  UMAP1=ump[,1],
                  UMAP2=ump[,2])

p_CPE <- ggplot(ump_df, aes(UMAP1, UMAP2, color= family)) +
  geom_point(size = 3,alpha=.7) +
  scale_fill_manual(values = palette_Dark2(14)) +
  mytheme()+
  ggtitle('CPE')+
  theme(legend.position = "bottom",
        legend.text = element_text(size=13,color = 'black'),
        axis.text.x = element_text(size = 25,face = 'bold',color = 'black'),
        axis.text.y = element_text(size = 25,face = 'bold',color = 'black'),
        axis.title.x = element_text(size =25,face = 'bold',color = 'black'),
        axis.title.y = element_text(size =25,face = 'bold',color = 'black'),
        title = element_text(size =25,face = 'bold',color = 'black'))
# save SARS_Proteom_CPE.pdf 13x17
# ------------------------------------
# ------------TM-vec
# ----------- load from google drive ---------------
drive_deauth()
drive_user()
public_file <- drive_get(as_id("1QUAqdkypCPHraNESF11arxg29tDUp521"))
drive_download(public_file, overwrite = TRUE)
tm <- read.csv('disTM_vec_sars_proteom.csv')
file.remove('disTM_vec_sars_proteom.csv')
tm <- 1 - tm
ump <- umap(d = tm,
            n_neighbors = 150,
            min_dist = 0.1,
            input="dist")
ump <- ump$layout
ump_df <- data.frame(family=df_CPE$sars_proteom, 
                  UMAP1=ump[,1],
                  UMAP2=ump[,2])

p_tm <- ggplot(ump_df, aes(UMAP1, UMAP2, color= family)) +
  geom_point(size = 3,alpha=.7) +
  scale_fill_manual(values = palette_Dark2(14)) +
  mytheme()+
  ggtitle('TM-Vec')+
  theme(legend.position = "bottom",
        legend.text = element_text(size=13,color = 'black'),
        axis.text.x = element_text(size = 25,face = 'bold',color = 'black'),
        axis.text.y = element_text(size = 25,face = 'bold',color = 'black'),
        axis.title.x = element_text(size =25,face = 'bold',color = 'black'),
        axis.title.y = element_text(size =25,face = 'bold',color = 'black'),
        title = element_text(size =25,face = 'bold',color = 'black'))
# save SARS_Proteom_TM_Vec.pdf 13x17
# -------------------------
ggarrange(p_SPE,p_CPE,p_tm,common.legend = T,ncol = 1, legend = 'bottom')
```

## --- PLPRO

```{r}
source('Functions.R')
load('Data/Large_Scale_SARS2/SPE_CPE_time.RData')
df_CPE$pdbID <- gsub('.pdb','',df_CPE$pdbID)
df_SPE$pdbID <- gsub('.pdb','',df_SPE$pdbID)
# ----------- load TM from google drive ---------------
drive_deauth()
drive_user()
public_file <- drive_get(as_id("1QUAqdkypCPHraNESF11arxg29tDUp521"))
drive_download(public_file, overwrite = TRUE)
tm <- read.csv('disTM_vec_sars_proteom.csv')
file.remove('disTM_vec_sars_proteom.csv')
tm <- 1 - tm
# ------------------------------------
# -------------------
idx_PLpro <- grep('PLPro', df_CPE$sars_proteom)

df_CPE <- df_CPE[idx_PLpro,]
df_SPE <- df_SPE[idx_PLpro,]
rownames(df_CPE) <- df_CPE$pdbID
rownames(df_SPE) <- df_SPE$pdbID

tm <- tm[idx_PLpro, idx_PLpro]
colnames(tm) <- rownames(tm) <- df_CPE$pdbID

c1 <- read.delim('Data/Large_Scale_SARS2/NSP3_cd21732_betaCoV_PLPro/cluster1.txt', header = F)


c2 <- read.delim('Data/Large_Scale_SARS2/NSP3_cd21732_betaCoV_PLPro/cluster2.txt', header = F)

c3 <- read.delim('Data/Large_Scale_SARS2/NSP3_cd21732_betaCoV_PLPro/cluster3.txt', header = F)

c4 <- read.delim('Data/Large_Scale_SARS2/NSP3_cd21732_betaCoV_PLPro/cluster4.txt', header = F)


df_CPE <- data.frame(subClass=NA, df_CPE)
df_CPE$subClass[df_CPE$pdbID%in%c1$V1] <- 'Sarb'
df_CPE$subClass[df_CPE$pdbID%in%c2$V1] <- 'Nob'
df_CPE$subClass[df_CPE$pdbID%in%c3$V1] <- 'Merb'
df_CPE$subClass[df_CPE$pdbID%in%c4$V1] <- 'Emb'

df_SPE <- data.frame(subClass=NA, df_SPE)
df_SPE$subClass[df_SPE$pdbID%in%c1$V1] <- 'Sarb'
df_SPE$subClass[df_SPE$pdbID%in%c2$V1] <- 'Nob'
df_SPE$subClass[df_SPE$pdbID%in%c3$V1] <- 'Merb'
df_SPE$subClass[df_SPE$pdbID%in%c4$V1] <- 'Emb'

dff = df_CPE[,-c(1:5)]/as.numeric(df_CPE$length)
dis_CPE <- as.matrix(proxy::dist(dff,
                                 method = manhat))
dff = df_SPE[,-c(1:5)]/as.numeric(df_SPE$length)
dis_SPE <- as.matrix(proxy::dist(dff,
                                 method = manhat))

ump_tm <- umap(d = tm,n_neighbors = 13,input='dist',min_dist = 0.5)
ump_tm <- ump_tm$layout
ump_tm <- data.frame(pdbID=rownames(ump_tm), 
                  UMAP1=ump_tm[,1],
                  UMAP2=ump_tm[,2])
ump_tm <- left_join(df_CPE[,1:5], ump_tm, 'pdbID')

tm_p = ggplot(ump_tm, aes(UMAP1, UMAP2, color= subClass)) +
  geom_point(alpha=.7,size=3) +
  ggtitle("TM-Vec") +
  mytheme() 
#----------------------CPE-----------
ump_CPE <- umap(d = dis_CPE,n_neighbors = 13,input='dist',min_dist = 0.5)
ump_CPE <- ump_CPE$layout
ump_CPE <- data.frame(pdbID=rownames(ump_CPE), 
                  UMAP1=ump_CPE[,1],
                  UMAP2=ump_CPE[,2])
ump_CPE <- left_join(df_CPE[,1:5], ump_CPE, 'pdbID')

cpe_p = ggplot(ump_CPE, aes(UMAP1, UMAP2, color= subClass)) +
  geom_point(alpha=.7,size=3) +
  ggtitle("CPE") +
  mytheme() 
#----------------SPE--------------------
ump_SPE <- umap(d = dis_SPE,n_neighbors = 13,input='dist',min_dist = 0.5)
ump_SPE <- ump_SPE$layout
ump_SPE <- data.frame(pdbID=rownames(ump_SPE), 
                  UMAP1=ump_SPE[,1],
                  UMAP2=ump_SPE[,2])
ump_SPE <- left_join(df_CPE[,1:5], ump_SPE, 'pdbID')

spe_p = ggplot(ump_SPE, aes(UMAP1, UMAP2, color= subClass)) +
  geom_point(alpha=.7,size=3) +
  ggtitle("SPE") +
  mytheme() 

ggarrange(spe_p,cpe_p,tm_p,ncol = 3,
          common.legend = T,legend = 'bottom')

```

# Globin Alpha_Beta

```{r}
source('Functions.R')
E210.SPE <- function(pdb){
  tryCatch({
    pdb0 <- pdb
    sele1 <- atom.select(pdb0,"protein",type = "ATOM")
    sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
    sele4 <- atom.select(pdb0, "noh")
    sele <- combine.select(sele1, sele2,sele4, operator="AND", verbose = FALSE)
    new_pdb <- trim.pdb(pdb0, sele,verbose = FALSE)
    pdbx <- new_pdb$atom
    pdbx <- pdbx[!duplicated(pdbx$resno),]
    seq <- paste0(aa321(pdbx$resid),collapse = '')
    Net_Frame <- NetworkFrame(PDB = new_pdb)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    return(c(seq,energy210))
  }, error = function(e) {
    cat("Error downloading file:", e$message, "\n")
  })
}
# -----------------------------------------
globin <- list.dirs("Data/Globin",
                          full.names = TRUE)
globin <- globin[basename(globin)!='Globin']


df <- c()
tictoc::tic()
for (i in 1:length(globin)) {
  pdbFiles <- list.files(globin[i],full.names = TRUE)
  pdbs <- purrr::map(pdbFiles, read.pdb)
  print(paste0(i, ': ', length(pdbs)))
  x <- lapply(pdbs, E210.SPE)
  x <- do.call(rbind, x)
  x <- data.frame(globin=strsplit(globin[i],'/')[[1]][3],
                  pdbID=basename(pdbFiles),
                  x)
  df <- rbind(df, x)
}
colnames(df)[-c(1:2)] <- c('seq',pair_inter)
rownames(df) <- df$pdbID
df <- cbind(df[,1:3],apply(df[,-c(1:3)], 2, FUN = as.numeric))
dis_SPE <- as.matrix(proxy::dist(df[,-c(1:3)], method = manhat))
z <- tictoc::toc()
t_SPE <- z$toc - z$tic
write.csv(df[,1:3],'Data/Globin/Globin.csv',
          row.names = F)
# ---------------
annot <- read.csv('Data/Globin/Globin.csv')
df <- data.frame(annot,
                 matrix(NA,ncol=210,nrow=nrow(annot)))
colnames(df)[-c(1:3)] <- pair_inter
rownames(df) <- df$pdbID
tictoc::tic()
for (NP in 1:nrow(df)) {
  print(NP)
  seq <- unlist(strsplit(df$seq[NP],''))
  freq <- c()
  for (j in 1:20){
    freq[j] <-  length(grep(tolower(letters_list[j]),tolower(seq)))
  }
  freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
  en_fr <- aaenergy * freq_matrix
  en_fr <- en_fr/length(seq)
  pair_es <- diag(freq) %*% as.matrix(en_fr)
  aa210_p <- pair_es[lower.tri(pair_es,diag = T)]
  df[NP, -c(1:3)] <- aa210_p
}
dis_CPE <- as.matrix(df[,-c(1:3)] %>% proxy::dist(method = manhat))
z <- tictoc::toc()
t_CPE <- z$toc - z$tic

save(t_SPE, t_CPE, dis_CPE, dis_SPE,
     file = 'Data/Globin/dis_CPE_SPE_time.RData')
```

## --- Visual
```{r}
source('Functions.R')
# ------------------------
load('Data/Globin/dis_CPE_SPE_time.RData')
df <- read.csv('Data/Globin/Globin.csv')
df$globin <- gsub('-.*','',df$globin)
# ---------------------
tm <- read.csv('Data/Globin/disTM_vec_Globin.csv')
colnames(tm)<-rownames(tm)<-df$pdbID
tm <- 1 - tm
diag(tm) <- NA
t_tm <- 4.91
# ------------------tm----------------
ump_tm <- umap(d = tm,
            n_neighbors = 13,
            min_dist = 0.1,
            input="dist")
ump_tm <- ump_tm$layout
ump_tm <- data.frame(pdbID=rownames(ump_tm), 
                  UMAP1=ump_tm[,1],
                  UMAP2=ump_tm[,2])
ump_df <- left_join(df, ump_tm,'pdbID')
gtm = ggplot(ump_df, aes(UMAP1, UMAP2, color= globin)) +
  geom_point(alpha=.7) +
  ggtitle("TM-Vec") +
  mytheme() 
# ------------------cpe----------------
ump_cpe <- umap(d = dis_CPE,
            n_neighbors = 13,
            min_dist = 0.1,
            input="dist")
ump_cpe <- ump_cpe$layout
ump_cpe <- data.frame(pdbID=rownames(ump_cpe), 
                  UMAP1=ump_cpe[,1],
                  UMAP2=ump_cpe[,2])
ump_df <- left_join(df, ump_cpe,'pdbID')
gcpe = ggplot(ump_df, aes(UMAP1, UMAP2, color= globin)) +
  geom_point(alpha=.7) +
  ggtitle("CPE") +
  mytheme() 
# ------------------SPE----------------
ump_spe <- umap(d = dis_SPE,
            n_neighbors = 13,
            min_dist = 0.1,
            input="dist")
ump_spe <- ump_spe$layout
ump_spe <- data.frame(pdbID=rownames(ump_spe), 
                  UMAP1=ump_spe[,1],
                  UMAP2=ump_spe[,2])
ump_df <- left_join(df, ump_spe,'pdbID')
gspe = ggplot(ump_df, aes(UMAP1, UMAP2, color= globin)) +
  geom_point(alpha=.7) +
  ggtitle("SPE") +
  mytheme() 
ggarrange(gspe,gcpe,gtm,ncol = 3, common.legend = T, legend = 'bottom')



diag(dis_CPE) <- NA
gl <- c('Alpha', 'Beta')
# ---------- 1NN --------------------
im <- apply(dis_CPE,1, FUN = which.min)
res <- data.frame(source=df$globin, target=df$globin[im])
cm <- matrix(0,2,2)
colnames(cm) <- rownames(cm) <- gl
for (sf in gl) {
  x <- res[res$source%in%sf,]
  z <- table(x$target)
  cm[sf, names(z)]<-z
}
ac <- round(sum(diag(cm))/sum(cm),2)*100
pr <- round(diag(cm)/colSums(cm) , 2)*100
re <- round(diag(cm)/rowSums(cm) , 2)*100
f1 <- round((2*pr*re)/(pr+re))
rbind(ac,pr,re,f1)
```



## Spike Inter-Similarity

This chunk calculates the minimum pairwise distance among the SARS2, SARS, and MERS groups to assess their inter-similarity

```{r}
source('Functions.R')
# --------------------SPE-----------
df <- readRDS("Data/rds/E210_spike.rds")
rownames(df) <- df$pdbID
x <- as.matrix(df[,-c(1:4)]/df$length)
dis_SPE <- as.matrix(x %>% proxy::dist(method = manhat))
# ----------------------------------
# -----------------CPE-------------------
df <- readRDS("Data/rds/E210.seq_spike.rds")
rownames(df) <- df$pdbID
x <- as.matrix(df[,-c(1:4)]/df$length)
dis_CPE <- as.matrix(x %>% proxy::dist(method = manhat))
# ------------------------------------------------
# -------------------Seq identity from MSA------------------------
sequences <- readAAStringSet("Data/covidPDB/Spike.fasta")
# ---- Multiple sequence alignment using ClustalW---------
alignment <- msa(sequences, method = "ClustalW")
alignment2 <- msaConvert(alignment,type="seqinr::alignment")
d <- seqinr::dist.alignment(alignment2, "identity")
#clustalo <- read_table("Data/covidPDB/clustalo.pim",
#                       col_names = FALSE, comment = "#")
#dis_seqid <- data.frame(clustalo[,-c(1,2)])
#colnames(dis_seqid) <- rownames(dis_seqid) <- clustalo$X2
#dis_seqid <- dis_seqid/100
#dis_seqid <- 1 - dis_seqid
#dis_seqid <- dis_seqid[df$pdbID, df$pdbID]
dis_seqid <- as.matrix(d)
dis_seqid <- dis_seqid[df$pdbID, df$pdbID]
# -------------------------------------
# -------------RMSD---------------------
rms <- data.frame(readRDS('Data/covidPDB/RMSD_spike.rds'))
colnames(rms)<-rownames(rms)
diag(rms)<-0
rms <- as.matrix(rms)
# -------------------------------------
# -------------TM---------------------
tm <- read.csv('Data/covidPDB/disTM_vec_spike.csv')
colnames(tm) <- rownames(tm) <- df$pdbID
tm <- 1 - tm
tm_vec <- as.matrix(tm)
# -------------------------------------
# -------------TM score---------------------
spike <- read.csv("Data/covidPDB/spike_seq.csv")
df <- spike
tm <- read.csv('Data/covidPDB/disTM_score_spike.csv')
rownames(tm) <- colnames(tm) <- df$pdbID

tm_max <- tm
for (i in 1:nrow(tm_max)) {
  print(i)
  for (j in 1:nrow(tm_max)) {
    tm_max[i,j] <- max(tm_max[i,j], tm_max[j,i])
    tm_max[j,i] <- tm_max[i,j]
  }
}
diag(tm_max) <- 1
tm_max <- as.matrix(1 - tm_max)
# -------------Groups----------
group <- factor(df$virus)
group_indices <- split(seq_len(length(group)), group)
# -----------Function to calculate mean of pairwise distance between two groups
mean_pairwise_dis <- function(dis, gr, gr_idx){
  n <- length(unique(gr))
  calculate_mean_distance <- function(group1_indices,group2_indices,
                                     distance_matrix) {
  submatrix <- distance_matrix[group1_indices, group2_indices]
  return(mean(submatrix))
  }
mean_dis <- matrix(NA, nrow = n, ncol = n,
                   dimnames = list(levels(gr),levels(gr)))
for (i in 1:n) {
    for (j in 1:n) {
      mean_dis[i, j] <- calculate_mean_distance(gr_idx[[i]],gr_idx[[j]],dis)
    }
  }
diag(mean_dis)<-0
minmax_dis <- mean_dis
minmax_dis <- (minmax_dis - min(minmax_dis))/(max(minmax_dis) - min(minmax_dis))

minmax_dis <- round(minmax_dis,2)
return(minmax_dis)
}
# -----------------------------------------------------------
dists <- list(CPE=dis_CPE,SPE=dis_SPE,seqid=dis_seqid,
              TM_vec=tm_vec,RMSD=rms, TM_score=tm_max)
pair_dis_minPair <- lapply(dists, FUN = mean_pairwise_dis,
                           gr=group,gr_idx=group_indices)
lower_triangles <- lapply(pair_dis_minPair, function(x) x[lower.tri(x)])
mean_pair_dist <- do.call(rbind, lower_triangles)
colnames(mean_pair_dist) = c("MERS-CoV ~ SARS-CoV", "MERS-CoV ~ SARS-CoV-2", "SARS-CoV ~ SARS-CoV-2")
library(reshape2)
group_dis <- melt(mean_pair_dist)
colnames(group_dis) = c("Method","Comparison","Distance")
```

## Visualization

```{r}
# Plot the bar plot
p0<-ggplot(group_dis, aes(x = Comparison, y = Distance, fill = Method)) +
  geom_bar(stat = "identity", position = "dodge", width = 0.7,
           show.legend = F) +
  labs(#title = "Comparison of Different Distance Metrics Between Virus Pairs",
       x = "", y = "Distance") +
  mytheme() +
  theme(legend.title = element_blank(),
        axis.text.x = element_text(size = 15, face = 'bold',
                                   color = 'black', angle = 15, 
                                   hjust = .8, vjust = .8),
        axis.text.y = element_text(size = 20, face = 'bold',
                                   color = 'black'),
        axis.title.y = element_text(size = 25, face = 'bold',
                                   color = 'black'),)+
  scale_fill_manual(values = c('magenta', 'blue','cyan','orange','red','green'))

```





# Ferritin
```{r}
source('Functions.R')
ferritin <- read.csv("Data/csv/Ferritin_Like_seq.csv",sep = ",")
Nprotein <- nrow(ferritin)
E210.seq_ferritin <- data.frame(ferritin,
                            matrix(NA, ncol = 210, 
                                   nrow = Nprotein))
colnames(E210.seq_ferritin)[-c(1:4)] <- pair_inter

for (NP in 1:Nprotein) {
  print(NP)
  seq <- unlist(strsplit(E210.seq_ferritin$seq[NP], ''))
  freq <- c()
    for (j in 1:20){
      freq[j] <-  length(grep(tolower(letters_list[j]),tolower(seq)))
      }
    freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
    en_fr <- aaenergy * freq_matrix
    en_fr <- en_fr/length(seq)
    pair_es <- diag(freq) %*% as.matrix(en_fr)
    aa210_p <- pair_es[lower.tri(pair_es,diag = T)]
    E210.seq_ferritin[NP,5:214] <- aa210_p
}

saveRDS(E210.seq_ferritin,file = "Data/rds/E210.seq_ferritin.rds")
```


## --- Visual NJ

```{r}
source('Functions.R')
E210_ferritin <- readRDS("Data/rds/E210_ferritin.rds")
df <- E210_ferritin
df$group[df$pdbID=='1otkA'] <- 'Pfam_1otkA'
df <- df[with(df, order(df$group)), ]
row.names(df) <- df$pdbID
rn <- df$pdbID
grp <- list(Bacteri     = rn[1:5],
            BMM_alpha = rn[6:7],
            BMM_beta  = rn[8:10],
            Dps  = rn[11:25],
            Fatty_acid  = rn[26:29],
            Ferritins  = rn[30:40],
            Pfam_1otkA = rn[41],
            RNRR2  = rn[42:49],
            Rubrerythrins = rn[50:53])

dis <- proxy::dist(df[,-c(1:3)], method = manhat)

tree_data <- ape::bionj(dis)
#newick_format <- ape::write.tree(tree_data)
#newick_no_values <- gsub(":[0-9\\.eE\\+-]+", "", newick_format)

p_data <- ggtree(tree_data, layout = 'unrooted',
                 branch.length='none',size = 1.5) 


t_SPE<-groupOTU(p_data, grp, "groups") + aes(color=groups) +
  theme(legend.position="right",
        legend.title = element_blank(),
        legend.text = element_text(size = 15, colour = "black"))+
  geom_tiplab(aes(label=label),
              hjust=0,size=3.5,fontface="bold",align=F)+
  scale_color_discrete(type =  c('red',"orange2","yellow4","green3",
                                 'skyblue3','blue','black',
                                 'magenta','purple2'
                                 ))
# ----------------------------
E210_ferritin <- readRDS("Data/rds/E210.seq_ferritin.rds")
df <- E210_ferritin
df$group[df$pdbID=='1otkA'] <- 'Pfam_1otkA'
df <- df[with(df, order(df$group)), ]
row.names(df) <- df$pdbID
rn <- df$pdbID
grp <- list(Bacteri     = rn[1:5],
            BMM_alpha = rn[6:7],
            BMM_beta  = rn[8:10],
            Dps  = rn[11:25],
            Fatty_acid  = rn[26:29],
            Ferritins  = rn[30:40],
            Pfam_1otkA = rn[41],
            RNRR2  = rn[42:49],
            Rubrerythrins = rn[50:53])

dis <- proxy::dist(df[,-c(1:4)], method = manhat)

tree_data <- ape::bionj(dis)


p_data <- ggtree(tree_data, layout = 'unrooted',
                 branch.length='none',size = 1.5) 


t_CPE<-groupOTU(p_data, grp, "groups") + aes(color=groups) +
  theme(legend.position="right",
        legend.title = element_blank(),
        legend.text = element_text(size = 15, colour = "black"))+
  geom_tiplab(aes(label=label),
              hjust=0,size=3.5,fontface="bold",align=F)+
  scale_color_discrete(type =  c('red',"orange2","yellow4","green3",
                                 'skyblue3','blue','black',
                                 'magenta','purple2'
                                 ))
# ----------------------------------------
ferritin <- read.csv('Data/csv/Ferritin_Like_seq.csv')
tm <- read.csv("Data/csv/disTM_vec_Ferritin_Like_seq.csv")
tm <- 1 - tm
colnames(tm) <- rownames(tm) <- ferritin$pdbID

tree_data <- ape::bionj(as.dist(tm))


p_data <- ggtree(tree_data, layout = 'unrooted',
                 branch.length='none',size = 1.5) 


t_tm<-groupOTU(p_data, grp, "groups") + aes(color=groups) +
  theme(legend.position="right",
        legend.title = element_blank(),
        legend.text = element_text(size = 15, colour = "black"))+
  geom_tiplab(aes(label=label),
              hjust=0,size=3.5,fontface="bold",align=F)+
  scale_color_discrete(type =  c('gray','red',"orange2","yellow4","green3",
                                 'skyblue3','blue','black',
                                 'magenta','purple2'
                                 ))
```

## --- mean distance

```{r}
source('Functions.R')
# -----------------------
mean_pairwise_dis <- function(dis, gr, gr_idx){
  n <- length(unique(gr))
  calculate_mean_distance <- function(group1_indices,group2_indices,
                                      distance_matrix) {
    submatrix <- distance_matrix[group1_indices, group2_indices]
    return(mean(submatrix))
  }
  mean_dis <- matrix(NA, nrow = n, ncol = n,
                     dimnames = list(levels(gr),levels(gr)))
  for (i in 1:n) {
    for (j in 1:n) {
      mean_dis[i, j] <- calculate_mean_distance(gr_idx[[i]],gr_idx[[j]],dis)
    }
  }
  diag(mean_dis) <- 0
  minmax_dis <- mean_dis
  minmax_dis <- (minmax_dis - min(minmax_dis))/(max(minmax_dis) - min(minmax_dis))
  minmax_dis <- round(minmax_dis, 2)
  return(minmax_dis)
}
# ------------------------------------
# -------- SPE
E210_ferritin <- readRDS("Data/rds/E210_ferritin.rds")
df <- E210_ferritin
rownames(df) <- df$pdbID
dis <- as.matrix(proxy::dist(df[,-c(1:4)], method = manhat))
group <-as.factor(df$group)
group_indices <-  split(seq_len(length(group)), group)
PWdisGroup<-mean_pairwise_dis(as.matrix(dis),
                             gr = group, gr_idx = group_indices)
tree_data <- ape::bionj(as.dist(PWdisGroup))

t_SPE <- ggtree(tree_data, layout = 'unrooted',
                size=5,aes(color = label)) +
  geom_tiplab(aes(label=label, col=label),
              hjust=.5,vjust=-.9,size=4,fontface="bold",align=F)+
  scale_color_discrete(type =  c('red',"orange2","yellow4","green3",
                                 'skyblue3','blue',
                                 'magenta','purple2'
  ))+
  theme(legend.position = 'none')
# ------------------------------
# ---------- CPE
E210.seq_ferritin <- readRDS("Data/rds/E210.seq_ferritin.rds")
df <- E210.seq_ferritin
rownames(df) <- df$pdbID
dis <- as.matrix(proxy::dist(df[,-c(1:4)], method = manhat))
group <- as.factor(df$group)
group_indices <- split(seq_len(length(group)), group)
PWdisGroup <- mean_pairwise_dis(as.matrix(dis),
                                gr = group, gr_idx = group_indices)
tree_data <- ape::bionj(as.dist(PWdisGroup))

t_CPE <- ggtree(tree_data, layout = 'unrooted',
                size=5,aes(color = label)) +
  geom_tiplab(aes(label=label, col=label),
              hjust=.5,vjust=-.9,size=4,fontface="bold",align=F)+
  scale_color_discrete(type =  c('red',"orange2","yellow4","green3",
                                 'skyblue3','blue',
                                 'magenta','purple2'
  ))+
  theme(legend.position = 'none')
# ----------------------
# --------- TM
df <- read.csv('Data/csv/Ferritin_Like_seq.csv')
dis <- read.csv('Data/csv/disTM_vec_Ferritin_Like_seq.csv')
dis <- 1 - dis
colnames(dis) <- rownames(dis) <- df$pdbID
group <- as.factor(df$group)
group_indices <- split(seq_len(length(group)), group)
PWdisGroup <- mean_pairwise_dis(as.matrix(dis),
                                gr = group, gr_idx = group_indices)
tree_data <- ape::bionj(as.dist(PWdisGroup))

t_TM <- ggtree(tree_data, layout = 'unrooted',
               size=5,aes(color = label)) +
  geom_tiplab(aes(label=label, col=label),
              hjust=.5,vjust=-.9,size=4,fontface="bold",align=F)+
  scale_color_discrete(type =  c('red',"orange2","yellow4","green3",
                                 'skyblue3','blue',
                                 'magenta','purple2'
  ))+
  theme(legend.position = 'none')
```

## --- Rho Spearman Rank

```{r}
rm(list=ls())
get_ranks <- function(order, reference_order) {
  ranks <- match(order, reference_order)
  ranks[is.na(ranks)] <- NA 
  return(ranks)
}
# ----------------------------
# ----  a.25.1.1 family
# Define tree orders
reference_order <- c("R", "F", "B", "D")
SPE_order <- c("R", "F", "B", "D")
CPE_order <- c("R", "F", "D", "B")
TM_order <- c("R", "D", "B", "F")

# Compute ranks
reference_ranks <- get_ranks(reference_order, reference_order)
SPE_ranks <- get_ranks(SPE_order, reference_order)
CPE_ranks <- get_ranks(CPE_order, reference_order)
TM_ranks <- get_ranks(TM_order, reference_order)

# Calculate Spearman's rank correlation coefficient
SPE <- cor(reference_ranks, SPE_ranks, 
           method = "spearman", use = "complete.obs")
CPE <- cor(reference_ranks, CPE_ranks, 
           method = "spearman", use = "complete.obs")
TM <- cor(reference_ranks, TM_ranks, 
          method = "spearman", use = "complete.obs")
a.25.1.1 <- c(SPE, CPE, TM)
# -------------------------------
# ----  a.25.1.2 family
# Define tree orders
reference_order <- c("F", "R", "A", "B")
SPE_order <- c("R", "F", "B", "A")
CPE_order <- c("R", "F", "B", "A")
TM_order <- c("B", "R", "F", "A")

reference_ranks <- get_ranks(reference_order, reference_order)
SPE_ranks <- get_ranks(SPE_order, reference_order)
CPE_ranks <- get_ranks(CPE_order, reference_order)
TM_ranks <- get_ranks(TM_order, reference_order)

# Calculate Spearman's rank correlation coefficient
SPE <- cor(reference_ranks, SPE_ranks, 
           method = "spearman", use = "complete.obs")
CPE <- cor(reference_ranks, CPE_ranks, 
           method = "spearman", use = "complete.obs")
TM <- cor(reference_ranks, TM_ranks, 
          method = "spearman", use = "complete.obs")
a.25.1.2 <- c(SPE, CPE, TM)

cbind(Method=c('SPE', 'CPE', 'TM'),
      a.25.1.1, a.25.1.2)
```

## --- compair

```{r}
library(TreeDist)
library(Quartet)
source('Functions.R')
set.seed(12345)
# -------------------------
# --------Schematic tree---------------------
tree_Ref <- read.tree(file = 'Data/Newick_Schematice.txt')
#-----------------------------------
E210_ferritin <- readRDS("Data/rds/E210_ferritin.rds")
df <- E210_ferritin
rownames(df) <- df$pdbID
dis <- proxy::dist(df[,-c(1:3)], method = manhat)
tree_data <- ape::bionj(dis)
newick_SPE <- ape::write.tree(tree_data)
newick_SPE <- gsub(":[0-9\\.eE\\+-]+", "", newick_SPE)
# ----------------------CPE---------------------------
E210_ferritin <- readRDS("Data/rds/E210.seq_ferritin.rds")
df <- E210_ferritin
rownames(df) <- df$pdbID
dis <- proxy::dist(df[,-c(1:4)], method = manhat)
tree_data <- ape::bionj(dis)
newick_CPE <- ape::write.tree(tree_data)
newick_CPE <- gsub(":[0-9\\.eE\\+-]+", "", newick_CPE)
# -----------------------TM-------------------
ferritin <- read.csv('Data/csv/Ferritin_Like_seq.csv')
tm <- read.csv("Data/csv/disTM_vec_Ferritin_Like_seq.csv")
tm <- 1 - tm
rownames(tm) <- colnames(tm) <- df$pdbID
tree_data <- ape::bionj(as.dist(tm))
newick_TM <- ape::write.tree(tree_data)
newick_TM <- gsub(":[0-9\\.eE\\+-]+", "", newick_TM)
# ----------------------------
# --- read trees---------------------
tree_SPE <- read.tree(text = newick_SPE)
tree_CPE <- read.tree(text = newick_CPE)
tree_TM <- read.tree(text = newick_TM)
# ----------------------------------------
RF_CPE <- treedist(tree_Ref, tree_CPE)
RF_SPE <- treedist(tree_Ref, tree_SPE)
RF_TM <- treedist(tree_Ref, tree_TM)

QD_CPE <- QuartetDivergence(QuartetStatus(tree_Ref, tree_CPE))
QD_SPE <- QuartetDivergence(QuartetStatus(tree_Ref, tree_SPE))
QD_TM <- QuartetDivergence(QuartetStatus(tree_Ref, tree_TM))

KC_CPE <- KendallColijn(tree_Ref, tree_CPE)
KC_SPE <- KendallColijn(tree_Ref, tree_SPE)
KC_TM <- KendallColijn(tree_Ref, tree_TM)

MS_CPE <- MatchingSplitDistance(tree_Ref, tree_CPE)
MS_SPE <- MatchingSplitDistance(tree_Ref, tree_SPE)
MS_TM <- MatchingSplitDistance(tree_Ref, tree_TM)
# -------------------------------------
# ------- generate random tree
resRtree <- data.frame(rQD_CPE=rep(NA,1000), rQD_SPE=NA, rQD_TM=NA,
                       rRF_CPE=NA, rRF_SPE=NA, rRF_TM=NA,
                       rPD_CPE=NA, rPD_SPE=NA, rPD_TM=NA,
                       rMS_CPE=NA, rMS_SPE=NA, rMS_TM=NA,
                       rKC_CPE=NA, rKC_SPE=NA, rKC_TM=NA,
                       maxPD_CPE=NA, maxPD_SPE=NA, maxPD_TM=NA,
                       maxMS_CPE=NA, maxMS_SPE=NA, maxMS_TM=NA)
# Generate a fully pectinate (caterpillar) tree
pectinate_tree <- stree(53, type = "left")
for (i in 1:1000) {
  print(i)
  # -------------CPE
  random_tree <- rcoal(length(tree_CPE$tip.label))
  random_tree$tip.label <- tree_CPE$tip.label
  # ---QD
  resRtree$rQD_CPE[i] <- QuartetDivergence(QuartetStatus(tree_Ref,random_tree))
  # ---RF_PD
  resRtree$rRF_CPE[i] <- treedist(tree_Ref, random_tree)[1]
  resRtree$rPD_CPE[i] <- treedist(tree_Ref, random_tree)[2]
  # ---MS
  resRtree$rMS_CPE[i] <- MatchingSplitDistance(tree_Ref, random_tree)
  # ---KC
  resRtree$rKC_CPE[i] <- KendallColijn(tree_Ref, random_tree)
  # ---Max PD_MS
  balanced_tree <- rtree(53) # Generate a fully balanced tree
  resRtree$maxPD_CPE[i] <- treedist(pectinate_tree, balanced_tree)[2]
  resRtree$maxMS_CPE[i] <- MatchingSplitDistance(pectinate_tree, balanced_tree)
  # -------------SPE
  random_tree <- rcoal(length(tree_SPE$tip.label))
  random_tree$tip.label <- tree_SPE$tip.label
  # ---QD
  resRtree$rQD_SPE[i] <- QuartetDivergence(QuartetStatus(tree_Ref,random_tree))
  # ---RF_PD
  resRtree$rRF_SPE[i] <- treedist(tree_Ref, random_tree)[1]
  resRtree$rPD_SPE[i] <- treedist(tree_Ref, random_tree)[2]
  # ---MS
  resRtree$rMS_SPE[i] <- MatchingSplitDistance(tree_Ref, random_tree)
  # ---KC
  resRtree$rKC_SPE[i] <- KendallColijn(tree_Ref, random_tree)
  # ---Max PD_MS
  balanced_tree <- rtree(53) # Generate a fully balanced tree
  resRtree$maxPD_SPE[i] <- treedist(pectinate_tree, balanced_tree)[2]
  resRtree$maxMS_SPE[i] <- MatchingSplitDistance(pectinate_tree, balanced_tree)
  # -------------TM
  random_tree <- rcoal(length(tree_TM$tip.label))
  random_tree$tip.label <- tree_TM$tip.label
  # ---QD
  resRtree$rQD_TM[i] <- QuartetDivergence(QuartetStatus(tree_Ref,random_tree))
  # ---RF_PD
  resRtree$rRF_TM[i] <- treedist(tree_Ref, random_tree)[1]
  resRtree$rPD_TM[i] <- treedist(tree_Ref, random_tree)[2]
  # ---MS
  resRtree$rMS_TM[i] <- MatchingSplitDistance(tree_Ref, random_tree)
  # ---KC
  resRtree$rKC_TM[i] <- KendallColijn(tree_Ref, random_tree)
  # ---Max PD_MS
  balanced_tree <- rtree(53) # Generate a fully balanced tree
  resRtree$maxPD_TM[i] <- treedist(pectinate_tree, balanced_tree)[2]
  resRtree$maxMS_TM[i] <- MatchingSplitDistance(pectinate_tree, balanced_tree)
}

# --- Pvalue Quartet Divergence-------------
pvl_QD_CPE <- mean(resRtree$rQD_CPE >= QD_CPE)
pvl_QD_SPE <- mean(resRtree$rQD_SPE >= QD_SPE)
pvl_QD_TM <- mean(resRtree$rQD_TM >= QD_TM)
# --- Pvalue Robinson Foulds-------------
pvl_RF_CPE <- mean(resRtree$rRF_CPE <= RF_CPE[1])
pvl_RF_SPE <- mean(resRtree$rRF_SPE <= RF_SPE[1])
pvl_RF_TM <- mean(resRtree$rRF_TM <= RF_TM[1])
# --- Pvalue path.difference-------------
pvl_PD_CPE <- mean(resRtree$rPD_CPE <= RF_CPE[2])
pvl_PD_SPE <- mean(resRtree$rPD_SPE <= RF_SPE[2])
pvl_PD_TM <- mean(resRtree$rPD_TM <= RF_TM[2])
# --- Pvalue KendallColijn-------------
pvl_KC_CPE <- mean(resRtree$rKC_CPE >= KC_CPE)
pvl_KC_SPE <- mean(resRtree$rKC_SPE >= KC_SPE)
pvl_KC_TM <- mean(resRtree$rKC_TM >= KC_TM)
# --- Pvalue MatchingSplitDistance-------------
pvl_MS_CPE <- mean(resRtree$rMS_CPE <= MS_CPE)
pvl_MS_SPE <- mean(resRtree$rMS_SPE <= MS_SPE)
pvl_MS_TM <- mean(resRtree$rMS_TM <= MS_TM)
# ----------------------------
RFs <- 1 - (c(RF_SPE[1], RF_CPE[1], RF_TM[1])/100)
KCs <- 1 - (c(KC_SPE, KC_CPE, KC_TM)/100)

PDs <- c(mean(resRtree$maxPD_SPE)-RF_SPE[2],
         mean(resRtree$maxPD_CPE)-RF_CPE[2],
         mean(resRtree$maxPD_TM)-RF_TM[2])
MaxPDs <- max(mean(resRtree$maxPD_SPE),mean(resRtree$maxPD_CPE),mean(resRtree$maxPD_TM))
pd <- c(0, MaxPDs, PDs)
pd01 <- (pd - min(pd))/(max(pd) - min(pd))

MSs <- c(mean(resRtree$maxMS_SPE)-MS_SPE,
         mean(resRtree$maxMS_CPE)-MS_CPE,
         mean(resRtree$maxMS_TM)-MS_TM)
MaxMSs <- max(mean(resRtree$maxMS_SPE),mean(resRtree$maxMS_CPE),mean(resRtree$maxMS_TM))
ms <- c(0, MaxMSs, MSs)
ms01 <- (ms - min(ms))/(max(ms) - min(ms))


df <- data.frame(metric = rep(c('Quartet.Divergence',
                                'Robinson.Foulds','Path.Difference',
                                'KendallColijn','MatchingSplit'),each=3),
                 method = c('SPE', 'CPE', 'TM'),
                 value = c(QD_SPE, QD_CPE, QD_TM,
                           RFs, pd01[3:5],
                           KCs, ms01[3:5]),
                 pvl = c(pvl_QD_SPE, pvl_QD_CPE, pvl_QD_TM,
                         pvl_RF_SPE, pvl_RF_CPE, pvl_RF_TM,
                         pvl_PD_SPE, pvl_PD_CPE, pvl_PD_TM,
                         pvl_KC_SPE, pvl_KC_CPE, pvl_KC_TM,
                         pvl_MS_SPE, pvl_MS_CPE, pvl_MS_TM))

p<-ggplot(df, aes(x = method, y = value, fill=method)) +
  geom_bar(stat = "identity", width = .2, alpha=.9) +
  facet_wrap(~ metric,nrow = 1)+
  mytheme()+
#  geom_text(aes(label = paste("p =", pvl), y = value + 0.05),
#            vjust = 0, size=5) +
  geom_text(aes(label = ifelse(pvl<0.05,'*',NA), y = value-0.01),
            vjust = 0, size=10) +
  scale_fill_manual(values = c('magenta','cyan','orange'))+
  xlab('')+ylab('')+
  theme(strip.text.x = element_text(size = 20),
        axis.text.x = element_blank(),
        axis.text.y = element_text(size = 20),
        legend.position = 'bottom')
```


# Human RAS superfamily
Nice data :/
```{r}
source('Functions.R')
# ------------ SPE ----------------------
E210.SPE <- function(pdb){
  tryCatch({
    pdb0 <- pdb
    sele1 <- atom.select(pdb0,"protein",type = "ATOM")
    sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
    sele4 <- atom.select(pdb0, "noh")
    sele <- combine.select(sele1, sele2,sele4, operator="AND", verbose = FALSE)
    new_pdb <- trim.pdb(pdb0, sele,verbose = FALSE)
    pdbx <- new_pdb$atom
    pdbx <- pdbx[!duplicated(pdbx$resno),]
    seq <- paste0(aa321(pdbx$resid),collapse = '')
    Net_Frame <- NetworkFrame(PDB = new_pdb)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    return(c(nchar(seq), seq, energy210))
  }, error = function(e) {
    cat("Error downloading file:", e$message, "\n")
  })
}
# ------------------------------------------
# ------------ ARF ------------
path <- 'Data/HumanRAS/'
fourSubFamily <- as.list(paste0(path,
                                c('ARF', 'RAB', 'RAS', 'RHO'),
                                '-RowData'))
names(fourSubFamily) <- c('ARF', 'RAB', 'RAS', 'RHO')
files <- lapply(fourSubFamily, list.files, full.names=TRUE)

df <- c()
tictoc::tic()
for (i in 1:length(files)) {
  pdbFiles <- files[[i]]
  pdbs <- purrr::map(pdbFiles, read.pdb)
  print(paste0(i, ': ', length(pdbs)))
  x <- lapply(pdbs, E210.SPE)
  x <- do.call(rbind, x)
  x <- data.frame(family=names(files)[i],
                  pdbID=gsub('.pdb','',basename(pdbFiles)),
                  x)
  df <- rbind(df, x)
}
colnames(df)[-c(1:2)] <- c('length','seq',pair_inter)
df <- data.frame(df[,c(1,2,4)], 
                 apply(df[-c(1,2,4)], 2, as.numeric))
rownames(df)<-df$pdbID
df[,-c(1:4)] = df[,-c(1:4)]/1000
dis_SPE <- as.matrix(proxy::dist(df[,-c(1:4)], method = manhat))
tictoc::toc()tictoc::toc()length
write.csv(df[,1:4], 'Data/HumanRAS/annot_HumanRAS.csv',
          row.names = FALSE)

dftot <- data.frame(family=df$family,
                    total=rowSums(df[,-c(1:4)])/df$length)
ggplot(dftot, aes(family, total))+
  geom_boxplot()


ump <- umap(d = dis_SPE,
            n_neighbors = 10,
            min_dist = 0.001,
            input="dist")
ump <- ump$layout
ump <- data.frame(pdbID=rownames(ump), 
                  UMAP1=ump[,1],
                  UMAP2=ump[,2])

ump_df <- left_join(df[,1:4], ump,'pdbID')

gg = ggplot(ump_df, aes(UMAP1, UMAP2, color= family)) +
  geom_point(alpha=.7)


diag(dis_SPE) <- NA
fourF <- names(files)
# ---------- 1NN --------------------
im <- apply(dis_SPE,1, FUN = which.min)
res <- data.frame(source=df$family, target=df$family[im])
## ----------
cm <- matrix(0,4,4)
colnames(cm) <- rownames(cm) <- fourF
for (sf in fourF) {
  x <- res[res$source%in%sf,]
  z <- table(x$target)
  cm[sf, names(z)]<-z
}
ac <- round(sum(diag(cm))/sum(cm),2)*100
pr <- round(diag(cm)/colSums(cm) , 2)*100
re <- round(diag(cm)/rowSums(cm) , 2)*100
f1 <- round((2*pr*re)/(pr+re))
rbind(ac,pr,re,f1)
# --------------------------------------
# -----------------CPE---------------
annot <- read.csv('Data/HumanRAS/annot_HumanRAS.csv')

df <- data.frame(annot, matrix(NA, nrow(annot), 210))
colnames(df)[-c(1:4)] <- pair_inter
rownames(df) <- df$pdbID

tictoc::tic()
for (NP in 1:nrow(df)) {
  print(NP)
  seq <- unlist(strsplit(unlist(df$seq[NP]),''))
  freq = c()
  for (j in 1:20){
    freq[j] <-  length(grep(tolower(letters_list[j]),tolower(seq)))
  }
  freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
  en_fr <- aaenergy * freq_matrix
  en_fr <- en_fr/length(seq)
  pair_es <- diag(freq) %*% as.matrix(en_fr)
  aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
  df[NP,-c(1:4)] <- aa210_p
}
dis_CPE <- as.matrix(proxy::dist(df[,-c(1:4)], method = manhat))
tictoc::toc()
dftot <- data.frame(family=df$family,
                    total=rowSums(df[,-c(1:5)])/df$length)
ggplot(dftot, aes(family, total))+
  geom_boxplot()


ump <- umap(d = dis_CPE,
            n_neighbors = 13,
            min_dist = 0.001,
            input="dist")

ump <- ump$layout
ump <- data.frame(pdbID=rownames(ump), 
                  UMAP1=ump[,1],
                  UMAP2=ump[,2])

ump_df <- left_join(df[,1:4], ump,'pdbID')

ggplot(ump_df, aes(UMAP1, UMAP2, color= family)) +
  geom_point(alpha=.7)

diag(dis_CPE) <- NA
# ---------- 1NN --------------------
im <- apply(dis_CPE,1, FUN = which.min)
res <- data.frame(source=df$family, target=df$family[im])
## ----------
cm <- matrix(0,4,4)
colnames(cm) <- rownames(cm) <- fourF
for (sf in fourF) {
  x <- res[res$source%in%sf,]
  z <- table(x$target)
  cm[sf, names(z)]<-z
}
ac <- round(sum(diag(cm))/sum(cm),2)*100
pr <- round(diag(cm)/colSums(cm) , 2)*100
re <- round(diag(cm)/rowSums(cm) , 2)*100
f1 <- round((2*pr*re)/(pr+re))
rbind(ac,pr,re,f1)
# --------------------------------
# --------tm-------------
tm <- read.csv('Data/HumanRAS/disTM_vec_HumanRAS.csv')
colnames(tm)<-rownames(tm)<-annot$pdbID
tm <- 1 - tm

ump <- umap(d = tm,
            n_neighbors = 13,
            min_dist = 0.001,
            input="dist")

ump <- ump$layout
ump <- data.frame(pdbID=rownames(ump), 
                  UMAP1=ump[,1],
                  UMAP2=ump[,2])

ump_df <- left_join(df[,1:4], ump,'pdbID')

ggplot(ump_df, aes(UMAP1, UMAP2, color= family)) +
  geom_point(alpha=.7)

diag(tm) <- NA
# ---------- 1NN --------------------
im <- apply(tm,1, FUN = which.min)
res <- data.frame(source=df$family, target=df$family[im])
## ----------
cm <- matrix(0,4,4)
colnames(cm) <- rownames(cm) <- fourF
for (sf in fourF) {
  x <- res[res$source%in%sf,]
  z <- table(x$target)
  cm[sf, names(z)]<-z
}
ac <- round(sum(diag(cm))/sum(cm),2)*100
pr <- round(diag(cm)/colSums(cm) , 2)*100
re <- round(diag(cm)/rowSums(cm) , 2)*100
f1 <- round((2*pr*re)/(pr+re))
rbind(ac,pr,re,f1)
```

# Length Vs Energy

```{r Astrals_CPE_SPE}
source('Functions.R')
# ----------------------------------------------------
df_seq <- readRDS("Data/rds/E210.seq_astral_40.rds") 
df_str <- readRDS("Data/rds/E210.astral_40.rds") 

result <- mapply(function(x, y) manhat(x, y),
                 split(df_seq[,-c(1:8)]/df_seq$length, 1:nrow(df_seq)), 
                 split(df_str[,-c(1:8)]/df_str$length, 1:nrow(df_str)))
res = data.frame(length = df_seq$length,dist = result)
cor(res$length,res$dist,use = "complete.obs")
res_60 = res[res$length < 60 , ]
cor(res_60$length,res_60$dist, use = "complete.obs")
res_100_200 = res[res$length < 100 & res$length > 60 ,]
cor(res_100_200$length,res_100_200$dist, use = "complete.obs")

result <- mapply(function(x, y) manhat(x,y),
                 split(df_seq[,-c(1:8)]/2, 1:nrow(df_seq)), 
                 split(df_str[,-c(1:8)], 1:nrow(df_str)))

res <- data.frame(scopeID=df_seq$scopeID,
                 length = df_seq$length,
                 disManhat = result
                 )
result <- mapply(function(x, y) cor(as.numeric(x), as.numeric(y)),
                 split(df_seq[,-c(1:8)]/2, 1:nrow(df_seq)), 
                 split(df_str[,-c(1:8)], 1:nrow(df_str)))
res$corr <- result
result <- mapply(function(x, y) cosine(as.numeric(x), as.numeric(y)),
                 split(df_seq[,-c(1:8)]/2, 1:nrow(df_seq)), 
                 split(df_str[,-c(1:8)], 1:nrow(df_str)))

res$cosine <- result
df_tot <- data.frame(scopeID=df_seq$scopeID,
                     CPE=rowSums(df_seq[,-c(1:8)]),
                     SPE=rowSums(df_str[,-c(1:8)]))

df_tot$disTot <- (df_tot$CPE/2000 - df_tot$SPE/1000)
bins <- c(seq(0, 500, by = 100), 1700)
res$length_bins <- cut(res$length, breaks = bins,
                       include.lowest = TRUE)
# ----------------------------------------------------
astral40 <- readRDS("Data/rds/astral_40.rds")
#-----
calculate_entropy <- function(protein_sequence) {
  aa_vector <- s2c(protein_sequence)
  aa_counts <- table(aa_vector)
  l <- sum(aa_counts) # length(aa_vector); nchar(protein_sequence)
  probs <- aa_counts / l
  p2 <- probs^2
  e1 <- 1 - (sum(p2)) # simpson diversity 
  e11 <- ((l * (sum(p2))) - 1)/(l - 1) # unbiased simpson
  e2 <- -sum(probs * log2(probs)) # shannon
  e22 <- e2 + ((length(aa_counts) - 1) / (2*l)) # unbiased shannon
  entropy_value <- c(e1, e2, e11, e22)
  names(entropy_value) <- c('simpson', 'shannon',
                            'simpsonAdjusted', 'shannonAhjusted')
  return(entropy_value)
}

entropyScore <- data.frame(scopeID=astral40$scopeID,
                        t(sapply(astral40$seq, calculate_entropy)),
                        row.names = NULL)

res <- left_join(res, df_tot, 'scopeID')
res <- left_join(res, entropyScore, 'scopeID')
# ------------------------------------
CPE_SPE_20 <- c()
for (i in 1:nrow(df_seq)) {
  print(i)
  x <- as.numeric(df_seq[i,-c(1:8)])
  m <- matrix(0, 20, 20)
  m[lower.tri(m, diag = T)] <- x
  m <- t(m)
  m[lower.tri(m, diag = T)] <- x
  cpe <- colSums(m)
  # --------
  x <- as.numeric(df_str[i,-c(1:8)])
  m <- matrix(0, 20, 20)
  m[lower.tri(m, diag = T)] <- x
  m <- t(m)
  m[lower.tri(m, diag = T)] <- x
  spe <- colSums(m)
  d <- c(cpe/2 , spe)
  CPE_SPE_20 <- rbind(CPE_SPE_20, d)
}
CPE_SPE_20 <- data.frame(CPE_SPE_20)
colnames(CPE_SPE_20) <- paste0(substr(colnames(df_seq[,-c(1:8)])[1:20], 2,2),
                             rep(c('_CPE', '_SPE'), each=20))
res <- data.frame(res, CPE_SPE_20)
# ---------------------------------
disDim <- df_seq[,-c(1:8)]/2 - df_str[,-c(1:8)]
colnames(disDim) <- colnames(df_seq)[-c(1:8)]
res <- data.frame(res, disDim)
saveRDS(res, 'Data/rds/Length_Vs_Energy.rds')
```

## --- Visual

```{r}
source('Functions.R')
# ---------------
res <- readRDS('Data/rds/Length_Vs_Energy.rds')
# ------------ save 210 cor plot
for (i in 1:210) {
  print(i)
  df <- res[,c(2,i+53)]
  aa <- colnames(df)[2]
  colnames(df)[2] <- 'dis'
  p <- ggplot(df, aes(length, dis))+
    geom_point()+
    ylab('CPE - SPE')
  file_name <- paste0('manuscript/Revised/sup_Fig/210cor/',
                      aa, ".pdf")
  ggsave(file_name, p, width = 9,height = 7)
}
```

# Scalability

```{r}
source('Functions.R')
E210.seq <- function(seq){
  seq <- unlist(strsplit(unlist(seq),''))
  freq = c()
  for (j in 1:20){
    freq[j] <-  length(grep(tolower(letters_list[j]),tolower(seq)))
  }
  freq_matrix <- matrix(rep(freq, each = 20), nrow = 20)
  en_fr <- aaenergy * freq_matrix
  en_fr <- en_fr/length(seq)
  pair_es <- diag(freq) %*% as.matrix(en_fr)
  aa210_p<-pair_es[lower.tri(pair_es,diag = T)]
  return(aa210_p)
}

path <- 'Data/csv/'
astral <- paste0(path,
                 'Astral95_subset_',
                 c(1,seq(5,30,5)),'k.csv')


df <- data.frame(subset=1:7,
                 n_protein=c(1,seq(5,30,5))*1000,
                 sumAA=NA,
                 time=NA)


for (i in 1:nrow(df)) {
  print(i)
  df2 <- read.csv(astral[i])
  tictoc::tic()
  x <- lapply(as.list(df2$seq), E210.seq)
  x <- do.call(rbind, x)
  dis <- x %>% proxy::dist(method = manhat)
  z <- tictoc::toc()
  df$time[i] <- z$toc - z$tic
  df$sumAA[i] <- sum(nchar(df2$seq))
  rm(df2, dis)
}

write.csv(df, 'Data/csv/CPE_prfc_astral95.csv', row.names = F)
```

## --- Visual 
```{r}
source('Functions.R')
# ---------------
cpe <- read.csv('Data/csv/CPE_prfc_astral95.csv')
tm <- read.csv('Data/csv/TM_prfc_astral95.csv')
cpe$method <- 'CPE'
tm$method <- 'TM'

df <- rbind(cpe, tm)

p<-ggplot(df, aes(n_protein, time/sumAA, col=method)) +
  geom_point(size=3, alpha=1) +
  geom_smooth(method = "lm", se = FALSE, alpha=1,linewidth = .3) +  
  stat_poly_eq(aes(label = paste(..eq.label..)),
               formula = y ~ x, parse = TRUE, label.x.npc = "right",
               label.y.npc = "top", size = 5)+
  xlab('N Protein')+
  ylab('Time/AminoAcid (Sec/AA)') +
  scale_color_manual(values = c('magenta', 'orange'))+
  mytheme()+
  theme(legend.position = c(.9,.15),
        legend.text = element_text(face = 'bold'),
        axis.title.y = element_text(size = 13, vjust = 2),
        axis.title.x = element_text(size = 13, vjust = -.5))+
  scale_x_continuous(breaks = unique(df$n_protein),
                     labels = label_number(scale = 1e-3))+
  scale_y_continuous(breaks = seq(0,1.2, .2)/1000,
                     labels = label_number(scale = 1e3))+
  annotation_custom(
  grob = textGrob(bquote("10"^-3),
                  gp = gpar(col = "blue", fontsize = 15),rot=90),
  xmin = -3100, xmax = -3100, ymin = 0.00096, ymax = 0.00096) +
  annotation_custom(
  grob = textGrob(bquote("10"^3),
                  gp = gpar(col = "blue", fontsize = 15)),
  xmin = 19000, xmax = 19000, ymin = -.00012, ymax = -.00012) +
  coord_cartesian(clip = "off") 

ggarrange(NULL,p,ncol = 2,widths = c(.01,1))
# save Fig13_new.pdf 7x6
```



# SPE for Ct_Ho
this chunk has run on win 10 with 2.4 GHz cpu and 8 GB RAM.
```{r}
source('Functions.R')
# ----------------------------------------------------
df <- read.csv('Data/csv/CT_Ho_cathID_filtered.csv')

E210_CT_Ho <- data.frame(df,
                         matrix(NA, ncol = 210, nrow = nrow(df)))

colnames(E210_CT_Ho)[-c(1:9)] <- pair_inter

tictoc::tic()
for (NP in 1:nrow(E210_CT_Ho)) {
  print(NP)
  pdbname <- substr(E210_CT_Ho$pdbID[NP],1,4)
  tryCatch({
    pdb0 <- read.pdb(pdbname, verbose = FALSE)
    sele1 <- atom.select(pdb0,"protein",type = "ATOM",
                         chain=substr(E210_CT_Ho$pdbID[NP],5,5))
    pdb0 <- trim.pdb(pdb0, sele1, verbose = FALSE)
    sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
    sele4 <- atom.select(pdb0, "noh")
    # ------------
    r <- unique(pdb0$atom$resno)
    pos <- as.numeric(unlist(strsplit(E210_CT_Ho$position[NP],'-')))
    st <- pos[1]
    en <- ifelse(pos[2]>length(r), length(r), pos[2])
    sele3 <- atom.select(pdb0,  resno=r[st]:r[en], verbose=FALSE)
    sele <- combine.select(sele2, sele4, operator="AND")
    new_pdb <- trim.pdb(pdb0, sele,verbose = FALSE)
    Net_Frame <- NetworkFrame(PDB = new_pdb)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210_CT_Ho[NP,-c(1:9)] <- energy210
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}
dis <- as.matrix(E210_CT_Ho[,-c(1:9)] %>% proxy::dist(method = manhat))
rt <- tictoc::toc()


saveRDS(E210_CT_Ho,'Data/rds/E210_CT_Ho.rds')

diag(dis) <- NA
im <- apply(dis,1, FUN = which.min)
res <- data.frame(source=df$cathID, target=df$cathID[im])
## ----------
twoSF <- unique(E210_CT_Ho$cathID)
cm <- matrix(0,2,2)
colnames(cm) <- rownames(cm) <- twoSF
for (sf in twoSF) {
  x <- res[res$source%in%sf,]
  z <- table(x$target)
  cm[sf,names(z)]<-z
}
ac <- round(sum(diag(cm))/sum(cm),2)*100

print(paste0('Accuracy = ', ac, '; RunningTime = ', rt$toc - rt$tic))
# time  = 187 Sec
```

# regenerat Fig 6

```{r}
source('Functions.R')
# -----------------------------------
# time_for_SPE  = 187 Sec
# accuracy_for_SPE = 99
# Extracted from 'SPE for Ct_Ho' chunk
# ----------------------------------------------------
df <- data.frame(method=c('GR-Align','RMSD','TM-Score',
                          'YH(10Rotation)','YH(2500Rotation)','TM-Vec','CPE',
                          'SPE'),
                 Accuracy=c(62.3, 59.2, 61.5, 70.8, 81.5, 100, 100,
                            99),
                 TimeS=c(120, 3600, 33600, 600, 18600, 67, 1,
                         187))
# ------------------------------------------------------------------------
# ---------   the values for  `GR-Align`, `RMSD`, `TM-Score`, `YH(10Rotation)`, `YH(2500Rotation)` 
# ----------      were previously reported in the `Yau.S et.al. Journal of Biomolecular Structure and Dynamics (2018)`.
# --------------------------------------------------------------------------
df$Time <- round((df$TimeS*100)/(12*3600),5)
df$ratio<-log10(df$Accuracy/df$Time+1)

df <- df[order(df$ratio, decreasing = T),]
df$method <- factor(df$method, 
                    levels = df$method)
mycol <-c('magenta','orange','cyan','skyblue3','purple','red','yellow4','green')


p12h <- ggplot(df, aes(x = method, y = Time, fill= method)) +
  geom_bar(stat="identity")+
  coord_polar(theta="y") +
  theme_light(base_rect_size = 0,base_line_size = 1)+
  theme(axis.text = element_blank(),
        axis.ticks=element_blank(),
        axis.title = element_blank(),
        legend.title = element_blank(),
        #legend.position = 'none'
        )+
  scale_fill_manual(values = mycol)+
  ylim(0,100)

p200s <- p12h+xlim(c('CPE','TM-Vec','GR-Align','SPE'))+
  ylim(0,.465)+ #3.0*12*3.6 = 200sec
  theme(legend.text = element_text(face = 'bold',size = 15))

df2 <- df
df2$Time <- df2$TimeS+1
p <- ggplot(df2,aes(Accuracy,Time,col=method,label=method))+
  geom_point(size=7,show.legend = F)+    
  geom_text(show.legend = F,vjust=.7,hjust=1.2,col='black',size=6, fontface='bold')+
  mytheme()+ylab('Time (s)')+
  scale_color_manual(values = mycol)+
  scale_y_log10()+xlim(50,103)+
  theme(axis.text.x = element_text(face = 'bold',color = 'black',size = 20),
        axis.text.y = element_text(face = 'bold',color = 'black',size = 20))
# --------------------------------------------------------------------------
```


# SPE for 5SF

```{r Five superfamily_CPE}
source('Functions.R')
# ----------------------------------------------------
df5 <- read.csv('Data/csv/fiveSF.csv')
Nprotein <- nrow(df5)
#----------------------------------------------
num_columns <- 210
E210_df5 <- data.frame(df5,
                       matrix(NA,ncol=num_columns,nrow=Nprotein))

colnames(E210_df5)[-c(1:9)] <- pair_inter
tictoc::tic()
for (NP in 1:Nprotein) {
  print(NP)
  chain_resno <- split_position(E210_df5$position[NP])
  pdbname <- E210_df5$pdbID[NP]
  tryCatch({
    pdb0 <- read.pdb(pdbname,verbose = FALSE)
    pdblist = list()
    for (ind in 1:dim(chain_resno)[1]){
      sele1 <- atom.select(pdb0,"protein",type = "ATOM",
                           chain=chain_resno$chain[ind])
      sele2 <- atom.select(pdb0, "protein", elety=c("H","OXT"),inverse=TRUE)
      sele4 <- atom.select(pdb0, "noh")
      sele <- combine.select(sele1, sele2,sele4, operator="AND",verbose = FALSE)
      if (!(chain_resno$start[ind]=="NA") & !(chain_resno$end[ind])=="NA"){
         st <- chain_resno$start[ind]
         en <- chain_resno$end[ind]
         sele3 <- atom.select(pdb0,  resno=st:en, verbose=FALSE)
         sele <- combine.select(sele1,sele2,sele3,sele4,operator="AND",verbose = FALSE)
      }
      pdblist[[ind]] <- trim.pdb(pdb0, sele,verbose = FALSE)
    }
    new_pdb = pdblist[[1]]
    if (dim(chain_resno)[1] > 1) {
      for (ind in 2:dim(chain_resno)[1]){
      new_pdb <- cat.pdb(new_pdb, pdblist[[ind]], rechain=FALSE)
      }
    }
    Net_Frame <- NetworkFrame(PDB = new_pdb)
    energy210 <- Energy_PC210(Net_Frame,energy_dell_dunbrack)
    E210_df5[NP,-c(1:9)] <- energy210
    }, error = function(e) {
     cat("Error downloading file:", e$message, "\n")
    })
}
df2 <- E210_df5
df2 <- df2[!is.na(df2$FF),]
dis <- as.matrix(proxy::dist(df2[,-c(1:9)], method = manhat))
tictoc::toc()
saveRDS(E210_df5, file = "Data/rds/E210_fiveSF.rds")
```


add SPE for fiveSF and CT_Ho
CPU
11th Gen Intel(R) Core(TM) i9-11950H @ 2.60GHz
Thread(s) per core:   2
Core(s) per socket:   8
Memory:        32GB


# MSA covid
## --- Visual

```{r}
df <- read.csv("Data/covidPDB/spike_seq.csv")
rownames(df) <- df$pdbID
df <- df[with(df, order(df$virus,decreasing = T)), ]
row.names(df) <- df$pdbID
rn <- df$pdbID
grp <- list(SARS_Cov2  = rn[1:80],
            SARS_Cov  = rn[81:111],
            MERS_Cov  = rn[112:143]
)
# ----------------------------------- 
#    write sequences in FASTA format 
#      to run MSA by Clustal Omega method as online on
#      https://www.ebi.ac.uk/jdispatcher/msa/clustalo
#   default parameters were employed.
# -----------------------------------
ampir::df_to_faa(df[,c(1,3)],'Data/covidPDB/Spike.fasta')
#  the output saved at 'Data/covidPDB/Spike.fasta.tree'
# -------------------
msa <- read.tree('Data/covidPDB/Spike.fasta.tree')
tree_data <- as.phylo(msa)
p_data <- ggtree(tree_data, layout = 'dendrogram',
                 size = 1,branch.length = 'none')

t_msa<-groupOTU(p_data, grp, "groups") + aes(color=groups) +
  theme(legend.position="top",
        legend.title=element_blank(),
        legend.text=element_text(size=20),
        plot.title = element_text(size = 50, face="bold",hjust = 0.5,vjust = 0))+
  scale_color_discrete(breaks=c('MERS_Cov', 'SARS_Cov', 'SARS_Cov2'),
                       type =  c('black',"blue","green","red" ))+
  ggtitle('MSA')
```


# Corplot210

```{r}
res <- readRDS('Data/rds/Length_Vs_Energy.rds')
# ----------
df <- res[,-c(1,3)]
length_var <- df[, 1]
other_vars <- df[, -1]

cor_values <- numeric(ncol(other_vars))
p_values <- numeric(ncol(other_vars))

for (i in 1:ncol(other_vars)) {
  test <- cor.test(length_var, other_vars[, i])
  cor_values[i] <- test$estimate
  p_values[i] <- test$p.value
}

# Create a data frame with the results
cor_results <- data.frame(Variable = names(other_vars), Correlation = cor_values, P_Value = p_values)

# Plotting the absolute correlation values
library(ggplot2)

# Plotting the p-values
p1<-ggplot(zzz[-1,], aes(x = Estimate, y = reorder(Variable, Estimate))) +
  geom_segment(aes(xend = 0, yend = Variable), color = "gray") +
  geom_point(aes(#size = ifelse(P_Value<0.05, -log10(P_Value)/100,NA),
                 color = ifelse(P_Value<0.05, P_Value,NA))) +  # Point size by p-value and color by p-value
  scale_color_gradient(low = "red", high = "blue") +  # Gradient color to indicate p-values
  labs(title = "Correlation of Length Variable with Other Variables",
       x = "Correlation Coefficient", 
       y = "Variable", 
       size = "-log10(p-value)", 
       color = "P-Value") +
  theme_minimal()


p2<-ggplot(zzz[-1,], aes(x = Estimate, fill = P_Value < 0.05)) +
  geom_histogram(binwidth = 0.01, color = "black", alpha = 0.3) +
  scale_fill_manual(values = c("red", "blue"),
                    labels = c("p >= 0.05", "p < 0.05")) +
  labs(title = "Histogram of Correlation Values",
       x = "Correlation Coefficient",
       y = "Count",
       fill = "Significance (p-value)") +
  theme_minimal()
# ------------------------
# -- save plots
for (i in 1:210) {
  print(i)
  df <- res[,c(2,i+53)]#,i+33
  aa <- colnames(df)[2]
  colnames(df)[2] <- 'dis'
  #df$dis <- df[,2]/2 - df[,3]
  p<- ggplot(df, aes(length, dis))+
    geom_point()
  file_name <- paste0('~/Desktop/20AA/', aa, ".pdf")
  ggsave(file_name, p, width = 9,height = 7)
}
```

# UmapDis

```{r}
df1 <- readRDS("Data/rds/E210.seq_astral_40.rds")
df1 <- df1[!is.na(df1$FF),]
df1[,-c(1:8)] <- df1[,-c(1:8)]/2
df3 <- readRDS("Data/rds/E210.astral_40.rds")
df3 <- df3[!is.na(df3$FF),]

df1$type <- 'CPE'
df3$type <- 'SPE'

m <- match(df3$scopeID, df1$scopeID)
df3 <- df3[!is.na(m),]
df <- rbind(df1[df1$scopeID%in%df3$scopeID,], df3)

dis <- as.matrix(proxy::dist(df[,-c(1:8, 219)], method = manhat))

ump <- umap(d = dis,
            n_neighbors = 30,
            min_dist = 0.1,
            input="dist")
ump_df <- data.frame(type=df$type,
                     length=df$length,
                     scopeID=df$scopeID,
                     UMAP1 = ump$layout[, 1],
                     UMAP2 = ump$layout[, 2])

ggplot(ump_df,aes(UMAP1, UMAP2, color= type)) +
  geom_point(alpha=.5, size=.7) + 
  mytheme()+
  theme(legend.title = element_blank(),
        legend.text = element_text(size = 18))+
  guides(color = guide_legend(override.aes = list(size = 5)))


scopeID <- unique(ump_df$scopeID)
df2 <- data.frame(scopeID=scopeID,length=NA,disUMAP=NA)
for (i in 1:length(scopeID)) {
  print(i)
  x <- ump_df[ump_df$scopeID%in%scopeID[i],]
  d <- dist(x[,c(4:5)])
  df2$disUMAP[i] <- as.numeric(d)
  df2$length[i] <- x$length[x$type=='CPE']
}
bins <- c(0,seq(150, 600, by = 100), 1700)
df2$length_bins <- cut(df2$length, breaks = bins,
                       include.lowest = TRUE)

ggplot(df2, aes(length_bins, disUMAP))+
  geom_boxplot()

ggplot(df2, aes(length, disUMAP))+
  geom_point()+
  ylim(0,7)+
  xlim(0,600)+
  stat_cor(method = "pearson", size=5,
           color = "red", geom = "label")

```



